<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en" data-bs-theme="auto"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><title>Intro to R and RStudio for Genomics: R Basics continued - factors and data frames</title><meta name="viewport" content="width=device-width, initial-scale=1"><script src="assets/themetoggle.js"></script><link rel="stylesheet" type="text/css" href="assets/styles.css"><script src="assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="favicons/dc/apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="favicons/dc/favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="favicons/dc/favicon-16x16.png"><link rel="manifest" href="favicons/dc/site.webmanifest"><link rel="mask-icon" href="favicons/dc/safari-pinned-tab.svg" color="#5bbad5"><meta name="msapplication-TileColor" content="#da532c"><meta name="theme-color" media="(prefers-color-scheme: light)" content="white"><meta name="theme-color" media="(prefers-color-scheme: dark)" content="black"></head><body>
    <header id="top" class="navbar navbar-expand-md top-nav data"><svg xmlns="http://www.w3.org/2000/svg" class="d-none"><symbol id="check2" viewbox="0 0 16 16"><path d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0z"></path></symbol><symbol id="circle-half" viewbox="0 0 16 16"><path d="M8 15A7 7 0 1 0 8 1v14zm0 1A8 8 0 1 1 8 0a8 8 0 0 1 0 16z"></path></symbol><symbol id="moon-stars-fill" viewbox="0 0 16 16"><path d="M6 .278a.768.768 0 0 1 .08.858 7.208 7.208 0 0 0-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 0 1 .81.316.733.733 0 0 1-.031.893A8.349 8.349 0 0 1 8.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 0 1 6 .278z"></path><path d="M10.794 3.148a.217.217 0 0 1 .412 0l.387 1.162c.173.518.579.924 1.097 1.097l1.162.387a.217.217 0 0 1 0 .412l-1.162.387a1.734 1.734 0 0 0-1.097 1.097l-.387 1.162a.217.217 0 0 1-.412 0l-.387-1.162A1.734 1.734 0 0 0 9.31 6.593l-1.162-.387a.217.217 0 0 1 0-.412l1.162-.387a1.734 1.734 0 0 0 1.097-1.097l.387-1.162zM13.863.099a.145.145 0 0 1 .274 0l.258.774c.115.346.386.617.732.732l.774.258a.145.145 0 0 1 0 .274l-.774.258a1.156 1.156 0 0 0-.732.732l-.258.774a.145.145 0 0 1-.274 0l-.258-.774a1.156 1.156 0 0 0-.732-.732l-.774-.258a.145.145 0 0 1 0-.274l.774-.258c.346-.115.617-.386.732-.732L13.863.1z"></path></symbol><symbol id="sun-fill" viewbox="0 0 16 16"><path d="M8 12a4 4 0 1 0 0-8 4 4 0 0 0 0 8zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"></path></symbol></svg><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-8">
      <div class="large-logo">
        <img id="dc-logo" alt="Data Carpentry" src="assets/images/data-logo.svg"><span class="badge text-bg-info">
          <abbr title="This lesson is in the beta phase, which means that it is ready for teaching by instructors outside of the original author team.">
            <a href="https://docs.carpentries.org/resources/curriculum/lesson-life-cycle.html" class="external-link alert-link">
              <i aria-hidden="true" class="icon" data-feather="alert-circle" style="border-radius: 5px"></i>
              Beta
            </a>
            <span class="visually-hidden">This lesson is in the beta phase, which means that it is ready for teaching by instructors outside of the original author team.</span>
          </abbr>
        </span>

      </div>
    </div>
    <div class="selector-container">
      <div id="theme-selector">
        <li class="nav-item dropdown" id="theme-button-list">
          <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
            <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="bd-theme-text"><li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                Light
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                Dark
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                Auto
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
          </ul></li>
      </div>

      <div class="dropdown" id="instructor-dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Learner View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1"><li><button class="dropdown-item" type="button" onclick="window.location.href='instructor/03-basics-factors-dataframes.html';">Instructor View</button></li>
        </ul></div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl bottom-nav data" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Data Carpentry" src="assets/images/data-logo-sm.svg"></div>
    <div class="lesson-title-md">
      Intro to R and RStudio for Genomics
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0"><li class="nav-item">
          <span class="lesson-title">
            Intro to R and RStudio for Genomics
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="reference.html#glossary">Glossary</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="profiles.html">Learner Profiles</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"><li><a class="dropdown-item" href="discuss.html">Discussion</a></li><li><a class="dropdown-item" href="reference.html">Glossary</a></li>
          </ul></li>
      </ul></div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a id="search-button" class="btn btn-primary" href="aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div><!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Intro to R and RStudio for Genomics
</div>

<aside class="col-md-12 lesson-progress"><div style="width: 35%" class="percentage">
    35%
  </div>
  <div class="progress data">
    <div class="progress-bar data" role="progressbar" style="width: 35%" aria-valuenow="35" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row" id="theme-row-mobile">
            <div class="col" id="theme-selector">
              <li class="nav-item dropdown" id="theme-button-list">
                <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
                  <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><span class="d-lg-none ms-1" id="bd-theme-text">Toggle Theme</span>
                </button>
                <ul class="dropdown-menu dropdown-menu-right" aria-labelledby="bd-theme-text"><li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                      Light
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                      Dark
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                      <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                      Auto
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                </ul></li>
            </div>
          </div>
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Learner View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="instructor/03-basics-factors-dataframes.html">Instructor View</a>
                      </div>
                    </div>
                  </div><!--/div.accordion-item-->
                </div><!--/div.accordion-flush-->
              </div><!--div.sidenav-view-selector -->
            </div><!--/div.col -->

            <hr></div><!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Setup</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="00-introduction.html">1. Introducing R and RStudio IDE</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="01-r-basics.html">2. R Basics</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="02-data-prelude.html">3. Introduction to the example dataset and file type</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlushcurrent">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-headingcurrent">
      <button class="accordion-button" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapsecurrent" aria-expanded="true" aria-controls="flush-collapsecurrent">
        <span class="visually-hidden">Current Chapter</span>
        <span class="current-chapter">
        4. R Basics continued - factors and data frames
        </span>
      </button>
    </div><!--/div.accordion-header-->

    <div id="flush-collapsecurrent" class="accordion-collapse collapse show" aria-labelledby="flush-headingcurrent" data-bs-parent="#accordionFlushcurrent">
      <div class="accordion-body">
        <ul><li><a href="#working-with-spreadsheets-tabular-data">Working with spreadsheets (tabular data)</a></li>
<li><a href="#importing-tabular-data-into-r">Importing tabular data into R</a></li>
<li><a href="#summarizing-subsetting-and-determining-the-structure-of-a-data-frame.">Summarizing, subsetting, and determining the structure of a data
frame.</a></li>
<li><a href="#introducing-factors">Introducing Factors</a></li>
<li><a href="#plotting-and-ordering-factors">Plotting and ordering factors</a></li>
<li><a href="#subsetting-data-frames">Subsetting data frames</a></li>
<li><a href="#coercing-values-in-data-frames">Coercing values in data frames</a></li>
<li><a href="#stringsasfactors">StringsAsFactors = ?</a></li>
<li><a href="#data-frame-bonus-material-math-sorting-renaming">Data frame bonus material: math, sorting, renaming</a></li>
<li><a href="#saving-your-data-frame-to-a-file">Saving your data frame to a file</a></li>
<li><a href="#importing-data-from-excel">Importing data from Excel</a></li>
        </ul></div><!--/div.accordion-body-->
    </div><!--/div.accordion-collapse-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="04-bioconductor-vcfr.html">5. Using packages from Bioconductor</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="05-dplyr.html">6. Data Wrangling and Analyses with Tidyverse</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush8">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading8">
        <a href="06-data-visualization.html">7. Data Visualization with ggplot2</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush9">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading9">
        <a href="07-r-help.html">8. Getting help with R</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width"><div class="accordion accordion-flush lesson-resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="lesson-resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul><li>
                        <a href="key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="reference.html#glossary">Glossary</a>
                      </li>
                      <li>
                        <a href="profiles.html">Learner Profiles</a>
                      </li>
                      <li><a href="discuss.html">Discussion</a></li><li><a href="reference.html">Glossary</a></li>
                    </ul></div>
                </div>
              </div>
            </div>
            <hr class="half-width lesson-resources"><a href="aio.html">See all in one page</a>


            <hr class="d-none d-sm-block d-md-none"><div class="d-grid gap-1">

            </div>
          </div><!-- /div.accordion -->
        </div><!-- /div.sidebar-inner -->
      </nav></div><!-- /div.sidebar -->
  </div><!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-instructor.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <nav class="lesson-content mx-md-4" aria-label="Previous and Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="02-data-prelude.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="04-bioconductor-vcfr.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="02-data-prelude.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Introduction to the
        </a>
        <a class="chapter-link float-end" href="04-bioconductor-vcfr.html" rel="next">
          Next: Using packages from...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
      <hr></nav><main id="main-content" class="main-content"><div class="container lesson-content">
        <h1>R Basics continued - factors and data frames</h1>
        <p>Last updated on 2025-10-27 |

        <a href="https://github.com/datacarpentry/genomics-r-intro/edit/main/episodes/03-basics-factors-dataframes.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>



        <div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>

        

<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul><li>How do I get started with tabular data (e.g. spreadsheets) in
R?</li>
<li>What are some best practices for reading data into R?</li>
<li>How do I save tabular data generated in R?</li>
</ul></div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul><li>Explain the basic principle of tidy datasets</li>
<li>Be able to load a tabular dataset using base R functions</li>
<li>Be able to determine the structure of a data frame including its
dimensions and the datatypes of variables</li>
<li>Be able to subset/retrieve values from a data frame</li>
<li>Understand how R may coerce data into different modes</li>
<li>Be able to change the mode of an object</li>
<li>Understand that R uses factors to store and manipulate categorical
data</li>
<li>Be able to manipulate a factor, including subsetting and
reordering</li>
<li>Be able to apply an arithmetic function to a data frame</li>
<li>Be able to coerce the class of an object (including variables in a
data frame)</li>
<li>Be able to import data from Excel</li>
<li>Be able to save a data frame as a delimited file</li>
</ul></div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="working-with-spreadsheets-tabular-data">Working with spreadsheets (tabular data)<a class="anchor" aria-label="anchor" href="#working-with-spreadsheets-tabular-data"></a></h2>
<hr class="half-width"><p>A substantial amount of the data we work with in genomics will be
tabular data, this is data arranged in rows and columns - also known as
spreadsheets. We could write a whole lesson on how to work with
spreadsheets effectively (<a href="https://datacarpentry.org/organization-genomics/" class="external-link">actually we
did</a>). For our purposes, we want to remind you of a few principles
before we work with our first set of example data:</p>
<p><strong>1) Keep raw data separate from analyzed data</strong></p>
<p>This is principle number one because if you can’t tell which files
are the original raw data, you risk making some serious mistakes
(e.g. drawing conclusion from data which have been manipulated in some
unknown way).</p>
<p><strong>2) Keep spreadsheet data Tidy</strong></p>
<p>The simplest principle of <strong>Tidy data</strong> is that we have
one row in our spreadsheet for each observation or sample, and one
column for every variable that we measure or report on. As simple as
this sounds, it’s very easily violated. Most data scientists agree that
significant amounts of their time is spent tidying data for analysis.
Read more about data organization in <a href="https://datacarpentry.org/organization-genomics/" class="external-link">our lesson</a>
and in <a href="https://www.jstatsoft.org/article/view/v059i10" class="external-link">this
paper</a>.</p>
<p><strong>3) Trust but verify</strong></p>
<p>Finally, while you don’t need to be paranoid about data, you should
have a plan for how you will prepare it for analysis. <strong>This a
focus of this lesson.</strong> You probably already have a lot of
intuition, expectations, assumptions about your data - the range of
values you expect, how many values should have been recorded, etc. Of
course, as the data get larger our human ability to keep track will
start to fail (and yes, it can fail for small data sets too). R will
help you to examine your data so that you can have greater confidence in
your analysis, and its reproducibility.</p>
<div id="tip-keep-your-raw-data-separate" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="tip-keep-your-raw-data-separate" class="callout-inner">
<h3 class="callout-title">Tip: Keep your raw data separate</h3>
<div class="callout-content">
<p>When you work with data in R, you are not changing the original file
you loaded that data from. This is different than (for example) working
with a spreadsheet program where changing the value of the cell leaves
you one “save”-click away from overwriting the original file. You have
to purposely use a writing function (e.g. <code>write.csv()</code>) to
save data loaded into R. In that case, be sure to save the manipulated
data into a new file. More on this later in the lesson.</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="importing-tabular-data-into-r">Importing tabular data into R<a class="anchor" aria-label="anchor" href="#importing-tabular-data-into-r"></a></h2>
<hr class="half-width"><p>There are several ways to import data into R. For our purpose here,
we will focus on using the tools every R installation comes with (so
called “base” R) to import a comma-delimited file containing the results
of our variant calling workflow. We will need to load the sheet using a
function called <code>read.csv()</code>.</p>
<div id="exercise-review-the-arguments-of-the-read.csv-function" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-review-the-arguments-of-the-read.csv-function" class="callout-inner">
<h3 class="callout-title">Exercise: Review the arguments of the
<code>read.csv()</code> function</h3>
<div class="callout-content">
<p><strong>Before using the <code>read.csv()</code> function, use R’s
help feature to answer the following questions</strong>.</p>
<p><em>Hint</em>: Entering ‘?’ before the function name and then running
that line will bring up the help documentation. Also, when reading this
particular help be careful to pay attention to the ‘read.csv’ expression
under the ‘Usage’ heading. Other answers will be in the ‘Arguments’
heading.</p>
<ol style="list-style-type: upper-alpha"><li><p>What is the default parameter for ‘header’ in the
<code>read.csv()</code> function?</p></li>
<li><p>What argument would you have to change to read a file that was
delimited by semicolons (;) rather than commas?</p></li>
<li><p>What argument would you have to change to read file in which
numbers used commas for decimal separation (i.e. 1,00)?</p></li>
<li><p>What argument would you have to change to read in only the first
10,000 rows of a very large file?</p></li>
</ol></div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<ol style="list-style-type: upper-alpha"><li><p>The <code>read.csv()</code> function has the argument ‘header’
set to TRUE by default, this means the function always assumes the first
row is header information, (i.e. column names)</p></li>
<li><p>The <code>read.csv()</code> function has the argument ‘sep’ set
to “,”. This means the function assumes commas are used as delimiters,
as you would expect. Changing this parameter (e.g. <code>sep=";"</code>)
would now interpret semicolons as delimiters.</p></li>
<li><p>Although it is not listed in the <code>read.csv()</code> usage,
<code>read.csv()</code> is a “version” of the function
<code>read.table()</code> and accepts all its arguments. If you set
<code>dec=","</code> you could change the decimal operator. We’d
probably assume the delimiter is some other character.</p></li>
<li><p>You can set <code>nrow</code> to a numeric value
(e.g. <code>nrow=10000</code>) to choose how many rows of a file you
read in. This may be useful for very large files where not all the data
is needed to test some data cleaning steps you are applying.</p></li>
</ol><p>Hopefully, this exercise gets you thinking about using the provided
help documentation in R. There are many arguments that exist, but which
we wont have time to cover. Look here to get familiar with functions you
use frequently, you may be surprised at what you find they can do.</p>
<div id="tip-why-does-read.csv-open-the-documentations-to-read.table" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="tip-why-does-read.csv-open-the-documentations-to-read.table" class="callout-inner">
<h3 class="callout-title">Tip: Why does ?read.csv open the
documentations to read.table?</h3>
<div class="callout-content">
<p>The reason for this is because <code>read.csv</code> is actually a
short cut for <code>read.table("file.csv", sep = ",")</code>. You can
see in the help documentation that there are several additional
variations of <code>read.table</code>, such as <code>read.csv2</code> to
read tables separated by <code>;</code> and <code>read.delim</code> to
read in tables separated by <code>\t</code> (tabs). If you know how your
table is separated, you can use one of the provided short cuts, but case
you run into an unconventional separator you are now equipped with the
knowledge to define it in the <code>sep =</code> argument of
<code>read.table</code>!</p>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
<p>Now, let’s read in the file <code>combined_tidy_vcf.csv</code> which
will be located in <code>/home/dcuser/r_data/</code>. Call this data
<code>variants</code>. The first argument to pass to our
<code>read.csv()</code> function is the file path for our data. The file
path must be in quotes and now is a good time to remember to use tab
autocompletion. <strong>If you use tab autocompletion you avoid typos
and errors in file paths.</strong> Use it!</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## read in a CSV file and save it as 'variants'</span></span>
<span></span>
<span><span class="va">variants</span> <span class="op">&lt;-</span> <span class="fu">read.csv</span><span class="op">(</span><span class="st">"/home/dcuser/r_data/combined_tidy_vcf.csv"</span><span class="op">)</span></span></code></pre>
</div>
<p>One of the first things you should notice is that in the Environment
window, you have the <code>variants</code> object, listed as 801 obs.
(observations/rows) of 29 variables (columns). Double-clicking on the
name of the object will open a view of the data in a new tab.</p>
<figure><img src="fig/rstudio_dataframeview.png" alt="RStudio data frame view" class="figure mx-auto d-block"><div class="figcaption">RStudio data frame view</div>
</figure><p>We can also quickly query the dimensions of the variable using
<code>dim()</code>. You’ll see that the first number <code>801</code>
shows the number of rows, then <code>29</code> the number of columns</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## get summary statistics on a data frame</span></span>
<span></span>
<span><span class="fu">dim</span><span class="op">(</span><span class="va">variants</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 801  29</code></pre>
</div>
</section><section><h2 class="section-heading" id="summarizing-subsetting-and-determining-the-structure-of-a-data-frame-">Summarizing, subsetting, and determining the structure of a data
frame.<a class="anchor" aria-label="anchor" href="#summarizing-subsetting-and-determining-the-structure-of-a-data-frame-"></a></h2>
<hr class="half-width"><p>A <strong>data frame is the standard way in R to store tabular
data</strong>. A data fame could also be thought of as a collection of
vectors, all of which have the same length. Using only two functions, we
can learn a lot about out data frame including some summary statistics
as well as well as the “structure” of the data frame. Let’s examine what
each of these functions can tell us:</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## get summary statistics on a data frame</span></span>
<span></span>
<span><span class="fu">summary</span><span class="op">(</span><span class="va">variants</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>  sample_id            CHROM                POS             ID
 Length:801         Length:801         Min.   :   1521   Mode:logical
 Class :character   Class :character   1st Qu.:1115970   NA's:801
 Mode  :character   Mode  :character   Median :2290361
                                       Mean   :2243682
                                       3rd Qu.:3317082
                                       Max.   :4629225

     REF                ALT                 QUAL          FILTER
 Length:801         Length:801         Min.   :  4.385   Mode:logical
 Class :character   Class :character   1st Qu.:139.000   NA's:801
 Mode  :character   Mode  :character   Median :195.000
                                       Mean   :172.276
                                       3rd Qu.:225.000
                                       Max.   :228.000

   INDEL              IDV              IMF               DP
 Mode :logical   Min.   : 2.000   Min.   :0.5714   Min.   : 2.00
 FALSE:700       1st Qu.: 7.000   1st Qu.:0.8823   1st Qu.: 7.00
 TRUE :101       Median : 9.000   Median :1.0000   Median :10.00
                 Mean   : 9.396   Mean   :0.9219   Mean   :10.57
                 3rd Qu.:11.000   3rd Qu.:1.0000   3rd Qu.:13.00
                 Max.   :20.000   Max.   :1.0000   Max.   :79.00
                 NA's   :700      NA's   :700
      VDB                 RPB              MQB              BQB
 Min.   :0.0005387   Min.   :0.0000   Min.   :0.0000   Min.   :0.1153
 1st Qu.:0.2180410   1st Qu.:0.3776   1st Qu.:0.1070   1st Qu.:0.6963
 Median :0.4827410   Median :0.8663   Median :0.2872   Median :0.8615
 Mean   :0.4926291   Mean   :0.6970   Mean   :0.5330   Mean   :0.7784
 3rd Qu.:0.7598940   3rd Qu.:1.0000   3rd Qu.:1.0000   3rd Qu.:1.0000
 Max.   :0.9997130   Max.   :1.0000   Max.   :1.0000   Max.   :1.0000
                     NA's   :773      NA's   :773      NA's   :773
      MQSB              SGB               MQ0F           ICB
 Min.   :0.01348   Min.   :-0.6931   Min.   :0.00000   Mode:logical
 1st Qu.:0.95494   1st Qu.:-0.6762   1st Qu.:0.00000   NA's:801
 Median :1.00000   Median :-0.6620   Median :0.00000
 Mean   :0.96428   Mean   :-0.6444   Mean   :0.01127
 3rd Qu.:1.00000   3rd Qu.:-0.6364   3rd Qu.:0.00000
 Max.   :1.01283   Max.   :-0.4536   Max.   :0.66667
 NA's   :48
   HOB                AC          AN        DP4                  MQ
 Mode:logical   Min.   :1   Min.   :1   Length:801         Min.   :10.00
 NA's:801       1st Qu.:1   1st Qu.:1   Class :character   1st Qu.:60.00
                Median :1   Median :1   Mode  :character   Median :60.00
                Mean   :1   Mean   :1                      Mean   :58.19
                3rd Qu.:1   3rd Qu.:1                      3rd Qu.:60.00
                Max.   :1   Max.   :1                      Max.   :60.00

    Indiv              gt_PL               gt_GT   gt_GT_alleles
 Length:801         Length:801         Min.   :1   Length:801
 Class :character   Class :character   1st Qu.:1   Class :character
 Mode  :character   Mode  :character   Median :1   Mode  :character
                                       Mean   :1
                                       3rd Qu.:1
                                       Max.   :1
                                                                     </code></pre>
</div>
<p>Our data frame had 29 variables, so we get 29 fields that summarize
the data. The <code>QUAL</code>, <code>IMF</code>, and <code>VDB</code>
variables (and several others) are numerical data and so you get summary
statistics on the min and max values for these columns, as well as mean,
median, and interquartile ranges. Many of the other variables
(e.g. <code>sample_id</code>) are treated as characters data (more on
this in a bit).</p>
<p>There is a lot to work with, so we will subset the columns into a new
data frame using the <code>data.frame()</code> function. To subset/index
a two dimensional variable, we need to define them on the appropriate
side of the brackets. The left hand side of the comma indicates the rows
you want to subset, and the right is the column position (e.g. [“row
index”, “column index”]).</p>
<p>Let’s put the columns 1, 2, 3, and 6 into a new data frame called
subset:</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Notice that we are wrapping the numbers in a c() function, to indicate a vector</span></span>
<span><span class="co">## in the right hand side of the comma. </span></span>
<span><span class="va">subset</span> <span class="op">&lt;-</span> <span class="fu">data.frame</span><span class="op">(</span><span class="va">variants</span><span class="op">[</span>, <span class="fu">c</span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="fl">6</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span></code></pre>
</div>
<p>Now, let’s use the <code>str()</code> (structure) function to look a
little more closely at how data frames work:</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## get the structure of a data frame</span></span>
<span></span>
<span><span class="fu">str</span><span class="op">(</span><span class="va">subset</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>'data.frame':	801 obs. of  4 variables:
 $ sample_id: chr  "SRR2584863" "SRR2584863" "SRR2584863" "SRR2584863" ...
 $ CHROM    : chr  "CP000819.1" "CP000819.1" "CP000819.1" "CP000819.1" ...
 $ POS      : int  9972 263235 281923 433359 473901 648692 1331794 1733343 2103887 2333538 ...
 $ ALT      : chr  "G" "T" "T" "CTTTTTTTT" ...</code></pre>
</div>
<p>Ok, thats a lot up unpack! Some things to notice.</p>
<ul><li><p>The object type <code>data.frame</code> is displayed in the first
row along with its dimensions, in this case 801 observations (rows) and
4 variables (columns)</p></li>
<li><p>Each variable (column) has a name (e.g. <code>sample_id</code>).
Notice that before each variable name there is a <code>$</code> - this
will be important later.</p></li>
<li>
<p>Each variable name is followed by the data type it contains
(e.g. chr, int, etc.). The <code>int</code> type shows an integer, which
is a type of numerical data, where it can only store whole numbers
(i.e. no decimal points ).</p>
<div id="exercise-revisiting-modes-and-classess" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div class="callout-inner">
<h3 class="callout-title">Exercise: Revisiting modes and classess</h3>
<div class="callout-content">
<p>Remeber when we said mode and class are sometimes different? If you
do, here is a chance to check. What happens when you try the
following?</p>
<ol style="list-style-type: decimal"><li><code>mode(variants)</code></li>
<li><code>class(variants)</code></li>
</ol></div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">mode</span><span class="op">(</span><span class="va">variants</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "list"</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">class</span><span class="op">(</span><span class="va">variants</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "data.frame"</code></pre>
</div>
<p>This result makes sense because <code>mode()</code> (which deals with
how an object is stored) tells us that <code>variants</code> is treated
as a <strong>list</strong> in R. A data frame is in some sense a “fancy”
list. However, data fames do have some specific properties beyond that
of a basic list, so they have their own class
(<strong>data.frame</strong>), which is important for functions (and
programmers) to know.</p>
</div>
</div>
</div>
</div>
</li>
</ul></section><section><h2 class="section-heading" id="introducing-factors">Introducing Factors<a class="anchor" aria-label="anchor" href="#introducing-factors"></a></h2>
<hr class="half-width"><p>Factors are the final major data structure we will introduce in our R
genomics lessons. Factors can be thought of as vectors which are
specialized for categorical data. Given R’s specialization for
statistics, this make sense since categorical and continuous variables
are usually treated differently. Sometimes you may want to have data
treated as a factor, but in other cases, this may be undesirable.</p>
<p>Let’s explore the value of treating some vectors that are categorical
in nature as factors. To do this we’ll take a look at just the alternate
alleles. We can use the <code>$</code> operator to access or extract a
column by its name in data frames (or to extract objects within named
lists).</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## extract the "ALT" column to a new object</span></span>
<span></span>
<span><span class="va">alt_alleles</span> <span class="op">&lt;-</span> <span class="va">subset</span><span class="op">$</span><span class="va">ALT</span></span></code></pre>
</div>
<p>Let’s look at the first few items in our factor using
<code>head()</code>:</p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="va">alt_alleles</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "G"         "T"         "T"         "CTTTTTTTT" "CCGCGC"    "T"        </code></pre>
</div>
<p>There are 801 alleles (one for each row). To simplify, lets look at
just the single-nucleotide alleles (SNPs).</p>
<p>Let’s review some of the vector indexing skills from the last episode
that can help:</p>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># This will find all matching alleles with the single nucleotide "A" and provide a TRUE/FASE vector</span></span>
<span><span class="va">alt_alleles</span> <span class="op">==</span> <span class="st">"A"</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>  [1] FALSE FALSE FALSE FALSE FALSE FALSE  TRUE  TRUE FALSE FALSE FALSE FALSE
 [13] FALSE FALSE  TRUE FALSE  TRUE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
 [25] FALSE FALSE FALSE FALSE FALSE  TRUE FALSE FALSE FALSE FALSE FALSE FALSE
 [37]  TRUE FALSE FALSE FALSE FALSE FALSE FALSE  TRUE  TRUE FALSE FALSE FALSE
 [49] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
 [61] FALSE FALSE FALSE FALSE FALSE FALSE FALSE  TRUE  TRUE FALSE FALSE FALSE
 [73] FALSE  TRUE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
 [85] FALSE  TRUE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
 [97] FALSE FALSE FALSE FALSE  TRUE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
[109] FALSE FALSE FALSE  TRUE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
[121] FALSE FALSE FALSE  TRUE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
[133] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
[145] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
[157] FALSE FALSE  TRUE FALSE FALSE FALSE  TRUE FALSE FALSE FALSE FALSE FALSE
[169] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
[181] FALSE FALSE FALSE FALSE FALSE FALSE  TRUE FALSE FALSE FALSE FALSE FALSE
[193]  TRUE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
[205] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE  TRUE FALSE FALSE
[217] FALSE FALSE  TRUE FALSE FALSE FALSE  TRUE FALSE FALSE FALSE FALSE FALSE
[229] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE  TRUE FALSE
[241] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
[253] FALSE FALSE FALSE FALSE FALSE  TRUE FALSE  TRUE  TRUE FALSE FALSE FALSE
[265] FALSE  TRUE  TRUE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
[277]  TRUE FALSE FALSE FALSE FALSE FALSE FALSE  TRUE FALSE FALSE FALSE  TRUE
[289] FALSE FALSE FALSE FALSE FALSE FALSE  TRUE FALSE  TRUE  TRUE FALSE FALSE
[301] FALSE  TRUE FALSE  TRUE FALSE FALSE FALSE FALSE FALSE  TRUE FALSE  TRUE
[313] FALSE FALSE FALSE FALSE  TRUE FALSE  TRUE  TRUE FALSE FALSE  TRUE FALSE
[325] FALSE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE FALSE  TRUE FALSE FALSE  TRUE
[337] FALSE FALSE  TRUE FALSE FALSE  TRUE FALSE FALSE FALSE FALSE FALSE FALSE
[349] FALSE  TRUE  TRUE  TRUE FALSE  TRUE FALSE FALSE FALSE  TRUE FALSE FALSE
[361] FALSE FALSE FALSE  TRUE  TRUE  TRUE  TRUE FALSE  TRUE FALSE FALSE FALSE
[373] FALSE FALSE FALSE  TRUE FALSE  TRUE FALSE FALSE  TRUE FALSE FALSE  TRUE
[385] FALSE FALSE FALSE  TRUE FALSE  TRUE FALSE  TRUE FALSE FALSE FALSE  TRUE
[397]  TRUE FALSE  TRUE  TRUE  TRUE FALSE FALSE  TRUE  TRUE FALSE FALSE  TRUE
[409] FALSE FALSE FALSE FALSE FALSE  TRUE FALSE FALSE FALSE  TRUE  TRUE FALSE
[421] FALSE  TRUE FALSE  TRUE  TRUE FALSE  TRUE FALSE  TRUE FALSE FALSE FALSE
[433]  TRUE FALSE FALSE FALSE FALSE FALSE FALSE  TRUE  TRUE FALSE FALSE FALSE
[445]  TRUE  TRUE  TRUE  TRUE FALSE  TRUE  TRUE  TRUE FALSE  TRUE FALSE  TRUE
[457]  TRUE  TRUE FALSE  TRUE FALSE  TRUE FALSE FALSE  TRUE  TRUE  TRUE  TRUE
[469]  TRUE  TRUE FALSE FALSE  TRUE  TRUE  TRUE  TRUE FALSE FALSE FALSE  TRUE
[481] FALSE FALSE  TRUE FALSE  TRUE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
[493]  TRUE FALSE FALSE FALSE  TRUE  TRUE FALSE FALSE FALSE FALSE  TRUE  TRUE
[505]  TRUE  TRUE  TRUE  TRUE FALSE FALSE FALSE  TRUE FALSE  TRUE FALSE FALSE
[517] FALSE FALSE FALSE  TRUE  TRUE FALSE FALSE FALSE FALSE FALSE  TRUE FALSE
[529] FALSE FALSE  TRUE FALSE FALSE  TRUE FALSE FALSE  TRUE  TRUE  TRUE FALSE
[541]  TRUE FALSE FALSE  TRUE FALSE  TRUE  TRUE  TRUE FALSE FALSE FALSE FALSE
[553] FALSE FALSE FALSE FALSE  TRUE  TRUE FALSE FALSE FALSE  TRUE  TRUE  TRUE
[565] FALSE  TRUE FALSE  TRUE  TRUE  TRUE FALSE  TRUE  TRUE FALSE FALSE  TRUE
[577] FALSE FALSE  TRUE  TRUE FALSE  TRUE FALSE FALSE FALSE FALSE FALSE FALSE
[589] FALSE FALSE FALSE FALSE FALSE FALSE FALSE  TRUE FALSE FALSE FALSE  TRUE
[601] FALSE FALSE FALSE FALSE FALSE FALSE  TRUE FALSE FALSE FALSE FALSE FALSE
[613] FALSE FALSE FALSE FALSE FALSE  TRUE FALSE  TRUE FALSE  TRUE  TRUE FALSE
[625]  TRUE  TRUE FALSE FALSE FALSE FALSE FALSE  TRUE  TRUE FALSE FALSE FALSE
[637] FALSE FALSE FALSE  TRUE  TRUE  TRUE FALSE  TRUE FALSE FALSE FALSE FALSE
[649]  TRUE  TRUE  TRUE FALSE FALSE FALSE  TRUE FALSE FALSE FALSE FALSE FALSE
[661] FALSE FALSE FALSE  TRUE  TRUE  TRUE  TRUE FALSE FALSE  TRUE FALSE FALSE
[673]  TRUE FALSE  TRUE FALSE  TRUE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
[685] FALSE  TRUE  TRUE FALSE  TRUE FALSE  TRUE  TRUE FALSE FALSE  TRUE  TRUE
[697]  TRUE FALSE FALSE  TRUE  TRUE  TRUE FALSE FALSE FALSE FALSE FALSE FALSE
[709] FALSE FALSE FALSE FALSE  TRUE  TRUE FALSE FALSE FALSE FALSE FALSE FALSE
[721] FALSE FALSE FALSE FALSE  TRUE FALSE FALSE  TRUE FALSE FALSE FALSE FALSE
[733]  TRUE FALSE  TRUE FALSE FALSE  TRUE  TRUE FALSE  TRUE FALSE  TRUE FALSE
[745] FALSE FALSE FALSE FALSE  TRUE FALSE FALSE FALSE FALSE FALSE  TRUE  TRUE
[757] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE  TRUE FALSE FALSE
[769] FALSE  TRUE FALSE FALSE FALSE FALSE  TRUE FALSE FALSE FALSE FALSE  TRUE
[781] FALSE FALSE FALSE FALSE FALSE  TRUE FALSE FALSE FALSE FALSE FALSE FALSE
[793]  TRUE  TRUE FALSE FALSE FALSE FALSE FALSE FALSE FALSE</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Then, we wrap them into an index to pull all the positions that match this. </span></span>
<span><span class="va">alt_alleles</span><span class="op">[</span><span class="va">alt_alleles</span> <span class="op">==</span> <span class="st">"A"</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>  [1] "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A"
 [19] "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A"
 [37] "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A"
 [55] "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A"
 [73] "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A"
 [91] "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A"
[109] "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A"
[127] "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A"
[145] "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A"
[163] "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A"
[181] "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A"
[199] "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A"</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># If we repeat this for each nucleotide A, T, G, and C, and connect them using `c()`,</span></span>
<span><span class="co"># we can index all the single nucleotide changes.</span></span>
<span><span class="va">snps</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="va">alt_alleles</span><span class="op">[</span><span class="va">alt_alleles</span> <span class="op">==</span> <span class="st">"A"</span><span class="op">]</span>,</span>
<span>  <span class="va">alt_alleles</span><span class="op">[</span><span class="va">alt_alleles</span><span class="op">==</span><span class="st">"T"</span><span class="op">]</span>,</span>
<span>  <span class="va">alt_alleles</span><span class="op">[</span><span class="va">alt_alleles</span><span class="op">==</span><span class="st">"G"</span><span class="op">]</span>,</span>
<span>  <span class="va">alt_alleles</span><span class="op">[</span><span class="va">alt_alleles</span><span class="op">==</span><span class="st">"C"</span><span class="op">]</span><span class="op">)</span></span></code></pre>
</div>
<p>This leaves us with a vector of the 701 alternative alleles which
were single nucleotides. Right now, they are being treated a characters,
but we could treat them as categories of SNP. Doing this will enable
some nice features. For example, we can try to generate a plot of this
character vector as it is right now:</p>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot</span><span class="op">(</span><span class="va">snps</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">WARNING<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="warning" tabindex="0"><code>Warning in xy.coords(x, y, xlabel, ylabel, log): NAs introduced by coercion</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">WARNING<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="warning" tabindex="0"><code>Warning in min(x): no non-missing arguments to min; returning Inf</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">WARNING<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="warning" tabindex="0"><code>Warning in max(x): no non-missing arguments to max; returning -Inf</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in plot.window(...): need finite 'ylim' values</code></pre>
</div>
<p>Whoops! Though the <code>plot()</code> function will do its best to
give us a quick plot, it is unable to do so here. Let’s use
<code>str()</code> to see why this might be:</p>
<div class="codewrapper sourceCode" id="cb26">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">str</span><span class="op">(</span><span class="va">snps</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> chr [1:707] "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" "A" ...</code></pre>
</div>
<p>R may not know how to plot a character vector! One way to fix this it
to tell R to treat the SNPs as categories (i.e. a factor vector); we
will create a new object to avoid confusion using the
<code>factor()</code> function:</p>
<div class="codewrapper sourceCode" id="cb28">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">factor_snps</span> <span class="op">&lt;-</span> <span class="fu">factor</span><span class="op">(</span><span class="va">snps</span><span class="op">)</span></span></code></pre>
</div>
<p>Let’s learn a little more about this new type of vector:</p>
<div class="codewrapper sourceCode" id="cb29">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">str</span><span class="op">(</span><span class="va">factor_snps</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> Factor w/ 4 levels "A","C","G","T": 1 1 1 1 1 1 1 1 1 1 ...</code></pre>
</div>
<p>What we get back are the categories (“A”,“C”,“G”,“T”) in our factor;
these are called “Levels”. <strong>Levels are the different categories
contained in a factor</strong>. By default, R will organize the levels
in a factor in alphabetical order. So the first level in this factor is
“A”.</p>
<p>For the sake of efficiency, R stores the content of a factor as a
vector of integers, which an integer is assigned to each of the possible
levels. Recall levels are assigned in alphabetical order. In this case,
the first item in our <code>factor_snps</code> object is “A”, which
happens to be the 1st level of our factor, ordered alphabetically. This
explains the sequence of “1”s (“Factor w/ 4 levels”A”,“C”,“G”,“T”: 1 1 1
1 1 1 1 1 1 1 …“), since”A” is the first level, and the first few items
in our factor are all “A”s.</p>
<p>We can see how many items in our vector fall into each category:</p>
<div class="codewrapper sourceCode" id="cb31">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">summary</span><span class="op">(</span><span class="va">factor_snps</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>  A   C   G   T
211 139 154 203 </code></pre>
</div>
<div class="codewrapper sourceCode" id="cb33">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Compare the character vector </span></span>
<span><span class="fu">summary</span><span class="op">(</span><span class="va">snps</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>   Length     Class      Mode
      707 character character </code></pre>
</div>
<p>As you can imagine, factors are already useful when you want to
generate a tally.</p>
<div id="tip-treating-objects-as-categories-without-changing-their-mode" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="tip-treating-objects-as-categories-without-changing-their-mode" class="callout-inner">
<h3 class="callout-title">Tip: treating objects as categories without
changing their mode</h3>
<div class="callout-content">
<p>You don’t have to make an object a factor to get the benefits of
treating an object as a factor. See what happens when you use the
<code>as.factor()</code> function on <code>factor_snps</code>. To
generate a tally, you can sometimes also use the <code>table()</code>
function; though sometimes you may need to combine both (i.e.
<code>table(as.factor(object))</code>)</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="plotting-and-ordering-factors">Plotting and ordering factors<a class="anchor" aria-label="anchor" href="#plotting-and-ordering-factors"></a></h2>
<hr class="half-width"><p>One of the most common uses for factors will be when you plot
categorical values. For example, suppose we want to know how many of our
variants had each possible SNP we could generate a plot:</p>
<div class="codewrapper sourceCode" id="cb35">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot</span><span class="op">(</span><span class="va">factor_snps</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/03-basics-factors-dataframes-rendered-unnamed-chunk-18-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>This isn’t a particularly pretty example of a plot but it works.
We’ll be learning much more about creating nice, publication-quality
graphics later in this lesson.</p>
<p>If you recall, factors are ordered alphabetically. That might make
sense, but categories (e.g., “red”, “blue”, “green”) often do not have
an intrinsic order. What if we wanted to order our plot according to the
numerical value (i.e., in descending order of SNP frequency)? We can
enforce an order on our factors:</p>
<div class="codewrapper sourceCode" id="cb36">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ordered_factor_snps</span> <span class="op">&lt;-</span> <span class="fu">factor</span><span class="op">(</span><span class="va">factor_snps</span>, levels <span class="op">=</span> <span class="fu">names</span><span class="op">(</span><span class="fu">sort</span><span class="op">(</span><span class="fu">table</span><span class="op">(</span><span class="va">factor_snps</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<p>Let’s deconstruct this from the inside out (you can try each of these
commands to see why this works):</p>
<ol style="list-style-type: decimal"><li>We create a table of <code>factor_snps</code> to get the frequency
of each SNP: <code>table(factor_snps)</code>
</li>
<li>We sort this table: <code>sort(table(factor_snps))</code>; use the
<code>decreasing =</code> parameter for this function if you wanted to
change from the default of FALSE</li>
<li>Using the <code>names</code> function gives us just the character
names of the table sorted by
frequencies:<code>names(sort(table(factor_snps)))</code>
</li>
<li>The <code>factor</code> function is what allows us to create a
factor. We give it the <code>factor_snps</code> object as input, and use
the <code>levels=</code> parameter to enforce the ordering of the
levels.</li>
</ol><p>Now we see our plot has be reordered:</p>
<div class="codewrapper sourceCode" id="cb37">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot</span><span class="op">(</span><span class="va">ordered_factor_snps</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/03-basics-factors-dataframes-rendered-unnamed-chunk-20-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Factors come in handy in many places when using R. Even using more
sophisticated plotting packages such as ggplot2 will sometimes require
you to understand how to manipulate factors.</p>
<div id="tip-packages-in-r-what-are-they-and-why-do-we-use-them" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="tip-packages-in-r-what-are-they-and-why-do-we-use-them" class="callout-inner">
<h3 class="callout-title">Tip: Packages in R – what are they and why do
we use them?</h3>
<div class="callout-content">
<p>Packages are simply collections of functions and/or data that can be
used to extend the capabilities of R beyond the core functionality that
comes with it by default. The default set of functions and packages that
come ‘in the box’ when you install R for the first time on a given
computer are called ‘base R’. However, one of the major benefits of
using an open source programming language is that there are thousands of
useful R packages freely available that span all types of statistical
analysis, data visualization, and more. The main place that these
additional R packages are made available is from a website called the
Comprehensive R Archive Network (<a href="https://cran.r-project.org/" class="external-link">CRAN</a>). When you use the built-in
R function <code>install.packages()</code>, it will look on CRAN for the
package and install it on your computer. So, for example, to install
packages such as <code>dplyr</code> and <code>ggplot2</code> (which
you’ll do in the next few lessons), you would use the following
command:</p>
<div class="codewrapper sourceCode" id="cb38">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install a package from CRAN</span></span>
<span><span class="fu">install.packages</span><span class="op">(</span><span class="st">"ggplot2"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>The following package(s) will be installed:
- ggplot2 [4.0.0]
These packages will be installed into "~/work/genomics-r-intro/genomics-r-intro/renv/profiles/lesson-requirements/renv/library/linux-ubuntu-jammy/R-4.5/x86_64-pc-linux-gnu".

# Installing packages --------------------------------------------------------
- Installing ggplot2 ...                        OK [linked from cache]
Successfully installed 1 package in 5.3 milliseconds.</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb40">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">install.packages</span><span class="op">(</span><span class="st">"dplyr"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>The following package(s) will be installed:
- dplyr [1.1.4]
These packages will be installed into "~/work/genomics-r-intro/genomics-r-intro/renv/profiles/lesson-requirements/renv/library/linux-ubuntu-jammy/R-4.5/x86_64-pc-linux-gnu".

# Installing packages --------------------------------------------------------
- Installing dplyr ...                          OK [linked from cache]
Successfully installed 1 package in 8.9 milliseconds.</code></pre>
</div>
<p>These two packages are among the most popular add on packages used in
R, and they are part of a large set of very useful packages called the
<a href="https://www.tidyverse.org" class="external-link">tidyverse</a>. Packages in the
tidyverse are designed to work well together and are made to work with
tidy data (which we described earlier in this lesson).</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="subsetting-data-frames">Subsetting data frames<a class="anchor" aria-label="anchor" href="#subsetting-data-frames"></a></h2>
<hr class="half-width"><p>Next, we are going to talk about how you can get specific values from
data frames, and where necessary, change the mode of a column of
values.</p>
<p>The first thing to remember is that a data frame is two-dimensional
(rows and columns). Therefore, to select a specific value we will will
once again use <code>[]</code> (bracket) notation, but we will specify
more than one value (except in some cases where we are taking a
range).</p>
<div id="exercise-subsetting-a-data-frame" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-subsetting-a-data-frame" class="callout-inner">
<h3 class="callout-title">Exercise: Subsetting a data frame</h3>
<div class="callout-content">
<p><strong>Try the following indices and functions and try to figure out
what they return</strong></p>
<ol style="list-style-type: lower-alpha"><li><p><code>variants[1,1]</code></p></li>
<li><p><code>variants[2,4]</code></p></li>
<li><p><code>variants[801,29]</code></p></li>
<li><p><code>variants[2, ]</code></p></li>
<li><p><code>variants[-1, ]</code></p></li>
<li><p><code>variants[1:4,1]</code></p></li>
<li><p><code>variants[1:10,c("REF","ALT")]</code></p></li>
<li><p><code>variants[,c("sample_id")]</code></p></li>
<li><p><code>head(variants)</code></p></li>
<li><p><code>tail(variants)</code></p></li>
<li><p><code>variants$sample_id</code></p></li>
<li><p><code>variants[variants$REF == "A",]</code></p></li>
</ol></div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3"> Show me the solution </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" aria-labelledby="headingSolution3" data-bs-parent="#accordionSolution3">
<div class="accordion-body">
<ol style="list-style-type: lower-alpha"><li>
</li></ol><div class="codewrapper sourceCode" id="cb42">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">variants</span><span class="op">[</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "SRR2584863"</code></pre>
</div>
<ol start="2" style="list-style-type: lower-alpha"><li>
</li></ol><div class="codewrapper sourceCode" id="cb44">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">variants</span><span class="op">[</span><span class="fl">2</span>, <span class="fl">4</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] NA</code></pre>
</div>
<ol start="3" style="list-style-type: lower-alpha"><li>
</li></ol><div class="codewrapper sourceCode" id="cb46">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">variants</span><span class="op">[</span><span class="fl">801</span>, <span class="fl">29</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "T"</code></pre>
</div>
<ol start="4" style="list-style-type: lower-alpha"><li>
</li></ol><div class="codewrapper sourceCode" id="cb48">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">variants</span><span class="op">[</span><span class="fl">2</span>, <span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>   sample_id      CHROM    POS ID REF ALT QUAL FILTER INDEL IDV IMF DP      VDB
2 SRR2584863 CP000819.1 263235 NA   G   T   85     NA FALSE  NA  NA  6 0.096133
  RPB MQB BQB MQSB       SGB     MQ0F ICB HOB AC AN     DP4 MQ
2   1   1   1   NA -0.590765 0.166667  NA  NA  1  1 0,1,0,5 33
                                                               Indiv gt_PL
2 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam 112,0
  gt_GT gt_GT_alleles
2     1             T</code></pre>
</div>
<ol start="5" style="list-style-type: lower-alpha"><li>
</li></ol><div class="codewrapper sourceCode" id="cb50">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">variants</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span>, <span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>   sample_id      CHROM     POS ID      REF       ALT QUAL FILTER INDEL IDV IMF
2 SRR2584863 CP000819.1  263235 NA        G         T   85     NA FALSE  NA  NA
3 SRR2584863 CP000819.1  281923 NA        G         T  217     NA FALSE  NA  NA
4 SRR2584863 CP000819.1  433359 NA CTTTTTTT CTTTTTTTT   64     NA  TRUE  12 1.0
5 SRR2584863 CP000819.1  473901 NA     CCGC    CCGCGC  228     NA  TRUE   9 0.9
6 SRR2584863 CP000819.1  648692 NA        C         T  210     NA FALSE  NA  NA
7 SRR2584863 CP000819.1 1331794 NA        C         A  178     NA FALSE  NA  NA
  DP      VDB RPB MQB BQB     MQSB       SGB     MQ0F ICB HOB AC AN     DP4 MQ
2  6 0.096133   1   1   1       NA -0.590765 0.166667  NA  NA  1  1 0,1,0,5 33
3 10 0.774083  NA  NA  NA 0.974597 -0.662043 0.000000  NA  NA  1  1 0,0,4,5 60
4 12 0.477704  NA  NA  NA 1.000000 -0.676189 0.000000  NA  NA  1  1 0,1,3,8 60
5 10 0.659505  NA  NA  NA 0.916482 -0.662043 0.000000  NA  NA  1  1 1,0,2,7 60
6 10 0.268014  NA  NA  NA 0.916482 -0.670168 0.000000  NA  NA  1  1 0,0,7,3 60
7  8 0.624078  NA  NA  NA 0.900802 -0.651104 0.000000  NA  NA  1  1 0,0,3,5 60
                                                               Indiv gt_PL
2 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam 112,0
3 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam 247,0
4 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam  91,0
5 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam 255,0
6 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam 240,0
7 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam 208,0
  gt_GT gt_GT_alleles
2     1             T
3     1             T
4     1     CTTTTTTTT
5     1        CCGCGC
6     1             T
7     1             A</code></pre>
</div>
<ol start="6" style="list-style-type: lower-alpha"><li>
</li></ol><div class="codewrapper sourceCode" id="cb52">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">variants</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span>, <span class="fl">1</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "SRR2584863" "SRR2584863" "SRR2584863" "SRR2584863"</code></pre>
</div>
<ol start="7" style="list-style-type: lower-alpha"><li>
</li></ol><div class="codewrapper sourceCode" id="cb54">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">variants</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="fu">c</span><span class="op">(</span><span class="st">"REF"</span>, <span class="st">"ALT"</span><span class="op">)</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>                                REF
1                                 T
2                                 G
3                                 G
4                          CTTTTTTT
5                              CCGC
6                                 C
7                                 C
8                                 G
9  ACAGCCAGCCAGCCAGCCAGCCAGCCAGCCAG
10                               AT
                                                        ALT
1                                                         G
2                                                         T
3                                                         T
4                                                 CTTTTTTTT
5                                                    CCGCGC
6                                                         T
7                                                         A
8                                                         A
9  ACAGCCAGCCAGCCAGCCAGCCAGCCAGCCAGCCAGCCAGCCAGCCAGCCAGCCAG
10                                                      ATT</code></pre>
</div>
<ol start="8" style="list-style-type: lower-alpha"><li>
</li></ol><div class="codewrapper sourceCode" id="cb56">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">variants</span><span class="op">[</span>, <span class="fu">c</span><span class="op">(</span><span class="st">"sample_id"</span><span class="op">)</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "SRR2584863" "SRR2584863" "SRR2584863" "SRR2584863" "SRR2584863"
[6] "SRR2584863"</code></pre>
</div>
<ol style="list-style-type: lower-roman"><li>
</li></ol><div class="codewrapper sourceCode" id="cb58">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="va">variants</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>   sample_id      CHROM    POS ID      REF       ALT QUAL FILTER INDEL IDV IMF
1 SRR2584863 CP000819.1   9972 NA        T         G   91     NA FALSE  NA  NA
2 SRR2584863 CP000819.1 263235 NA        G         T   85     NA FALSE  NA  NA
3 SRR2584863 CP000819.1 281923 NA        G         T  217     NA FALSE  NA  NA
4 SRR2584863 CP000819.1 433359 NA CTTTTTTT CTTTTTTTT   64     NA  TRUE  12 1.0
5 SRR2584863 CP000819.1 473901 NA     CCGC    CCGCGC  228     NA  TRUE   9 0.9
6 SRR2584863 CP000819.1 648692 NA        C         T  210     NA FALSE  NA  NA
  DP       VDB RPB MQB BQB     MQSB       SGB     MQ0F ICB HOB AC AN     DP4 MQ
1  4 0.0257451  NA  NA  NA       NA -0.556411 0.000000  NA  NA  1  1 0,0,0,4 60
2  6 0.0961330   1   1   1       NA -0.590765 0.166667  NA  NA  1  1 0,1,0,5 33
3 10 0.7740830  NA  NA  NA 0.974597 -0.662043 0.000000  NA  NA  1  1 0,0,4,5 60
4 12 0.4777040  NA  NA  NA 1.000000 -0.676189 0.000000  NA  NA  1  1 0,1,3,8 60
5 10 0.6595050  NA  NA  NA 0.916482 -0.662043 0.000000  NA  NA  1  1 1,0,2,7 60
6 10 0.2680140  NA  NA  NA 0.916482 -0.670168 0.000000  NA  NA  1  1 0,0,7,3 60
                                                               Indiv gt_PL
1 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam 121,0
2 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam 112,0
3 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam 247,0
4 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam  91,0
5 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam 255,0
6 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam 240,0
  gt_GT gt_GT_alleles
1     1             G
2     1             T
3     1             T
4     1     CTTTTTTTT
5     1        CCGCGC
6     1             T</code></pre>
</div>
<ol start="10" style="list-style-type: lower-alpha"><li>
</li></ol><div class="codewrapper sourceCode" id="cb60">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">tail</span><span class="op">(</span><span class="va">variants</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>     sample_id      CHROM     POS ID REF ALT QUAL FILTER INDEL IDV IMF DP
796 SRR2589044 CP000819.1 3444175 NA   G   T  184     NA FALSE  NA  NA  9
797 SRR2589044 CP000819.1 3481820 NA   A   G  225     NA FALSE  NA  NA 12
798 SRR2589044 CP000819.1 3893550 NA  AG AGG  101     NA  TRUE   4   1  4
799 SRR2589044 CP000819.1 3901455 NA   A  AC   70     NA  TRUE   3   1  3
800 SRR2589044 CP000819.1 4100183 NA   A   G  177     NA FALSE  NA  NA  8
801 SRR2589044 CP000819.1 4431393 NA TGG   T  225     NA  TRUE  10   1 10
          VDB RPB MQB BQB     MQSB       SGB MQ0F ICB HOB AC AN     DP4 MQ
796 0.4714620  NA  NA  NA 0.992367 -0.651104    0  NA  NA  1  1 0,0,4,4 60
797 0.8707240  NA  NA  NA 1.000000 -0.680642    0  NA  NA  1  1 0,0,4,8 60
798 0.9182970  NA  NA  NA 1.000000 -0.556411    0  NA  NA  1  1 0,0,3,1 52
799 0.0221621  NA  NA  NA       NA -0.511536    0  NA  NA  1  1 0,0,3,0 60
800 0.9272700  NA  NA  NA 0.900802 -0.651104    0  NA  NA  1  1 0,0,3,5 60
801 0.7488140  NA  NA  NA 1.007750 -0.670168    0  NA  NA  1  1 0,0,4,6 60
                                                                 Indiv gt_PL
796 /home/dcuser/dc_workshop/results/bam/SRR2589044.aligned.sorted.bam 214,0
797 /home/dcuser/dc_workshop/results/bam/SRR2589044.aligned.sorted.bam 255,0
798 /home/dcuser/dc_workshop/results/bam/SRR2589044.aligned.sorted.bam 131,0
799 /home/dcuser/dc_workshop/results/bam/SRR2589044.aligned.sorted.bam 100,0
800 /home/dcuser/dc_workshop/results/bam/SRR2589044.aligned.sorted.bam 207,0
801 /home/dcuser/dc_workshop/results/bam/SRR2589044.aligned.sorted.bam 255,0
    gt_GT gt_GT_alleles
796     1             T
797     1             G
798     1           AGG
799     1            AC
800     1             G
801     1             T</code></pre>
</div>
<ol start="11" style="list-style-type: lower-alpha"><li>
</li></ol><div class="codewrapper sourceCode" id="cb62">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">variants</span><span class="op">$</span><span class="va">sample_id</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "SRR2584863" "SRR2584863" "SRR2584863" "SRR2584863" "SRR2584863"
[6] "SRR2584863"</code></pre>
</div>
<ol start="12" style="list-style-type: lower-alpha"><li>
</li></ol><div class="codewrapper sourceCode" id="cb64">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">variants</span><span class="op">[</span><span class="va">variants</span><span class="op">$</span><span class="va">REF</span> <span class="op">==</span> <span class="st">"A"</span>, <span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>    sample_id      CHROM     POS ID REF ALT QUAL FILTER INDEL IDV IMF DP
11 SRR2584863 CP000819.1 2407766 NA   A   C  104     NA FALSE  NA  NA  9
12 SRR2584863 CP000819.1 2446984 NA   A   C  225     NA FALSE  NA  NA 20
14 SRR2584863 CP000819.1 2665639 NA   A   T  225     NA FALSE  NA  NA 19
16 SRR2584863 CP000819.1 3339313 NA   A   C  211     NA FALSE  NA  NA 10
18 SRR2584863 CP000819.1 3481820 NA   A   G  200     NA FALSE  NA  NA  9
19 SRR2584863 CP000819.1 3488669 NA   A   C  225     NA FALSE  NA  NA 13
         VDB      RPB      MQB      BQB     MQSB       SGB     MQ0F ICB HOB AC
11 0.0230738 0.900802 0.150134 0.750668 0.500000 -0.590765 0.333333  NA  NA  1
12 0.0714027       NA       NA       NA 1.000000 -0.689466 0.000000  NA  NA  1
14 0.9960390       NA       NA       NA 1.000000 -0.690438 0.000000  NA  NA  1
16 0.4059360       NA       NA       NA 1.007750 -0.670168 0.000000  NA  NA  1
18 0.1070810       NA       NA       NA 0.974597 -0.662043 0.000000  NA  NA  1
19 0.0162706       NA       NA       NA 1.000000 -0.680642 0.000000  NA  NA  1
   AN      DP4 MQ
11  1  3,0,3,2 25
12  1 0,0,10,6 60
14  1 0,0,12,5 60
16  1  0,0,4,6 60
18  1  0,0,4,5 60
19  1  0,0,8,4 60
                                                                Indiv gt_PL
11 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam 131,0
12 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam 255,0
14 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam 255,0
16 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam 241,0
18 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam 230,0
19 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam 255,0
   gt_GT gt_GT_alleles
11     1             C
12     1             C
14     1             T
16     1             C
18     1             G
19     1             C</code></pre>
</div>
</div>
</div>
</div>
</div>
<p>The subsetting notation is very similar to what we learned for
vectors. The key differences include:</p>
<ul><li>Typically provide two values separated by commas: data.frame[row,
column]</li>
<li>In cases where you are taking a continuous range of numbers use a
colon between the numbers (start:stop, inclusive)</li>
<li>For a non continuous set of numbers, pass a vector using
<code>c()</code>
</li>
<li>Index using the name of a column(s) by passing them as vectors using
<code>c()</code>
</li>
</ul><p>Finally, in all of the subsetting exercises above, we printed values
to the screen. You can create a new data frame object by assigning them
to a new object name:</p>
<div class="codewrapper sourceCode" id="cb66">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># create a new data frame containing only observations from SRR2584863</span></span>
<span></span>
<span><span class="va">SRR2584863_variants</span> <span class="op">&lt;-</span> <span class="va">variants</span><span class="op">[</span><span class="va">variants</span><span class="op">$</span><span class="va">sample_id</span> <span class="op">==</span> <span class="st">"SRR2584863"</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># check the dimension of the data frame</span></span>
<span></span>
<span><span class="fu">dim</span><span class="op">(</span><span class="va">SRR2584863_variants</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 25 29</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb68">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># get a summary of the data frame</span></span>
<span></span>
<span><span class="fu">summary</span><span class="op">(</span><span class="va">SRR2584863_variants</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>  sample_id            CHROM                POS             ID
 Length:25          Length:25          Min.   :   9972   Mode:logical
 Class :character   Class :character   1st Qu.:1331794   NA's:25
 Mode  :character   Mode  :character   Median :2618472
                                       Mean   :2464989
                                       3rd Qu.:3488669
                                       Max.   :4616538

     REF                ALT                 QUAL         FILTER
 Length:25          Length:25          Min.   : 31.89   Mode:logical
 Class :character   Class :character   1st Qu.:104.00   NA's:25
 Mode  :character   Mode  :character   Median :211.00
                                       Mean   :172.97
                                       3rd Qu.:225.00
                                       Max.   :228.00

   INDEL              IDV             IMF               DP
 Mode :logical   Min.   : 2.00   Min.   :0.6667   Min.   : 2.0
 FALSE:19        1st Qu.: 3.25   1st Qu.:0.9250   1st Qu.: 9.0
 TRUE :6         Median : 8.00   Median :1.0000   Median :10.0
                 Mean   : 7.00   Mean   :0.9278   Mean   :10.4
                 3rd Qu.: 9.75   3rd Qu.:1.0000   3rd Qu.:12.0
                 Max.   :12.00   Max.   :1.0000   Max.   :20.0
                 NA's   :19      NA's   :19
      VDB               RPB              MQB               BQB
 Min.   :0.01627   Min.   :0.9008   Min.   :0.04979   Min.   :0.7507
 1st Qu.:0.07140   1st Qu.:0.9275   1st Qu.:0.09996   1st Qu.:0.7627
 Median :0.37674   Median :0.9542   Median :0.15013   Median :0.7748
 Mean   :0.40429   Mean   :0.9517   Mean   :0.39997   Mean   :0.8418
 3rd Qu.:0.65951   3rd Qu.:0.9771   3rd Qu.:0.57507   3rd Qu.:0.8874
 Max.   :0.99604   Max.   :1.0000   Max.   :1.00000   Max.   :1.0000
                   NA's   :22       NA's   :22        NA's   :22
      MQSB             SGB               MQ0F           ICB
 Min.   :0.5000   Min.   :-0.6904   Min.   :0.00000   Mode:logical
 1st Qu.:0.9599   1st Qu.:-0.6762   1st Qu.:0.00000   NA's:25
 Median :0.9962   Median :-0.6620   Median :0.00000
 Mean   :0.9442   Mean   :-0.6341   Mean   :0.04667
 3rd Qu.:1.0000   3rd Qu.:-0.6168   3rd Qu.:0.00000
 Max.   :1.0128   Max.   :-0.4536   Max.   :0.66667
 NA's   :3
   HOB                AC          AN        DP4                  MQ
 Mode:logical   Min.   :1   Min.   :1   Length:25          Min.   :10.00
 NA's:25        1st Qu.:1   1st Qu.:1   Class :character   1st Qu.:60.00
                Median :1   Median :1   Mode  :character   Median :60.00
                Mean   :1   Mean   :1                      Mean   :55.52
                3rd Qu.:1   3rd Qu.:1                      3rd Qu.:60.00
                Max.   :1   Max.   :1                      Max.   :60.00

    Indiv              gt_PL               gt_GT   gt_GT_alleles
 Length:25          Length:25          Min.   :1   Length:25
 Class :character   Class :character   1st Qu.:1   Class :character
 Mode  :character   Mode  :character   Median :1   Mode  :character
                                       Mean   :1
                                       3rd Qu.:1
                                       Max.   :1
                                                                     </code></pre>
</div>
</section><section><h2 class="section-heading" id="coercing-values-in-data-frames">Coercing values in data frames<a class="anchor" aria-label="anchor" href="#coercing-values-in-data-frames"></a></h2>
<hr class="half-width"><div id="tip-coercion-isnt-limited-to-data-frames" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="tip-coercion-isnt-limited-to-data-frames" class="callout-inner">
<h3 class="callout-title">Tip: coercion isn’t limited to data
frames</h3>
<div class="callout-content">
<p>While we are going to address coercion in the context of data frames
most of these methods apply to other data structures, such as
vectors</p>
</div>
</div>
</div>
<p>Sometimes, it is possible that R will misinterpret the type of data
represented in a data frame, or store that data in a mode which prevents
you from operating on the data the way you wish. For example, a long
list of gene names isn’t usually thought of as a categorical variable,
the way that your experimental condition (e.g. control, treatment) might
be. More importantly, some R packages you use to analyze your data may
expect characters as input, not factors. At other times (such as
plotting or some statistical analyses) a factor may be more appropriate.
Ultimately, you should know how to change the mode of an object.</p>
<p>First, its very important to recognize that coercion happens in R all
the time. This can be a good thing when R gets it right, or a bad thing
when the result is not what you expect. Consider:</p>
<div class="codewrapper sourceCode" id="cb70">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">snp_chromosomes</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="st">'3'</span>, <span class="st">'11'</span>, <span class="st">'X'</span>, <span class="st">'6'</span><span class="op">)</span></span>
<span><span class="fu">typeof</span><span class="op">(</span><span class="va">snp_chromosomes</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "character"</code></pre>
</div>
<p>Although there are several numbers in our vector, they are all in
quotes, so we have explicitly told R to consider them as characters.
However, even if we removed the quotes from the numbers, R would coerce
everything into a character:</p>
<div class="codewrapper sourceCode" id="cb72">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">snp_chromosomes_2</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="fl">3</span>, <span class="fl">11</span>, <span class="st">'X'</span>, <span class="fl">6</span><span class="op">)</span></span>
<span><span class="fu">typeof</span><span class="op">(</span><span class="va">snp_chromosomes_2</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "character"</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb74">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">snp_chromosomes_2</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "3"</code></pre>
</div>
<p>We can use the <code>as.</code> functions to explicitly coerce values
from one form into another. Consider the following vector of characters,
which all happen to be valid numbers:</p>
<div class="codewrapper sourceCode" id="cb76">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">snp_positions_2</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="st">"8762685"</span>, <span class="st">"66560624"</span>, <span class="st">"67545785"</span>, <span class="st">"154039662"</span><span class="op">)</span></span>
<span><span class="fu">typeof</span><span class="op">(</span><span class="va">snp_positions_2</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "character"</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb78">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">snp_positions_2</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "8762685"</code></pre>
</div>
<p>Now we can coerce <code>snp_positions_2</code> into a numeric type
using <code>as.numeric()</code>:</p>
<div class="codewrapper sourceCode" id="cb80">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">snp_positions_2</span> <span class="op">&lt;-</span> <span class="fu">as.numeric</span><span class="op">(</span><span class="va">snp_positions_2</span><span class="op">)</span></span>
<span><span class="fu">typeof</span><span class="op">(</span><span class="va">snp_positions_2</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "double"</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb82">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">snp_positions_2</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 8762685</code></pre>
</div>
<p>Sometimes coercion is straight forward, but what would happen if we
tried using <code>as.numeric()</code> on
<code>snp_chromosomes_2</code></p>
<div class="codewrapper sourceCode" id="cb84">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">snp_chromosomes_2</span> <span class="op">&lt;-</span> <span class="fu">as.numeric</span><span class="op">(</span><span class="va">snp_chromosomes_2</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">WARNING<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="warning" tabindex="0"><code>Warning: NAs introduced by coercion</code></pre>
</div>
<p>If we check, we will see that an <code>NA</code> value (R’s default
value for missing data) has been introduced.</p>
<div class="codewrapper sourceCode" id="cb86">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">snp_chromosomes_2</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1]  3 11 NA  6</code></pre>
</div>
<p>Trouble can really start when we try to coerce a factor. For example,
when we try to coerce the <code>factor_snps</code> vector into a numeric
mode look at the result:</p>
<div class="codewrapper sourceCode" id="cb88">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">as.numeric</span><span class="op">(</span><span class="va">factor_snps</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>  [1] 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 [38] 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 [75] 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
[112] 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
[149] 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
[186] 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 4 4 4 4 4 4 4 4 4 4 4
[223] 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4
[260] 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4
[297] 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4
[334] 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4
[371] 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4
[408] 4 4 4 4 4 4 4 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3
[445] 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3
[482] 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3
[519] 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3
[556] 3 3 3 3 3 3 3 3 3 3 3 3 3 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
[593] 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
[630] 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
[667] 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
[704] 2 2 2 2</code></pre>
</div>
<p>Strangely, it works! Almost. Instead of giving an error message, R
returns numeric values, which in this case are the integers assigned to
the levels in this factor. This kind of behavior can lead to
hard-to-find bugs, for example when we do have numbers in a factor, and
we get numbers from a coercion. If we don’t look carefully, we may not
notice a problem.</p>
<p>If you need to coerce an entire column you can overwrite it using an
expression like this one:</p>
<div class="codewrapper sourceCode" id="cb90">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># make the 'REF' column a character type column</span></span>
<span></span>
<span><span class="va">variants</span><span class="op">$</span><span class="va">REF</span> <span class="op">&lt;-</span> <span class="fu">as.character</span><span class="op">(</span><span class="va">variants</span><span class="op">$</span><span class="va">REF</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># check the type of the column</span></span>
<span><span class="fu">typeof</span><span class="op">(</span><span class="va">variants</span><span class="op">$</span><span class="va">REF</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "character"</code></pre>
</div>
</section><section><h2 class="section-heading" id="stringsasfactors">StringsAsFactors = ?<a class="anchor" aria-label="anchor" href="#stringsasfactors"></a></h2>
<hr class="half-width"><p>Lets summarize this section on coercion with a few take home
messages.</p>
<ul><li>When you explicitly coerce one data type into another (this is known
as <strong>explicit coercion</strong>), be careful to check the result.
Ideally, you should try to see if its possible to avoid steps in your
analysis that force you to coerce.</li>
<li>R will sometimes coerce without you asking for it. This is called
(appropriately) <strong>implicit coercion</strong>. For example when we
tried to create a vector with multiple data types, R chose one type
through implicit coercion.</li>
<li>Check the structure (<code>str()</code>) of your data frames before
working with them!</li>
</ul><div id="tip-coercion-isnt-limited-to-data-frames-1" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="tip-coercion-isnt-limited-to-data-frames-1" class="callout-inner">
<h3 class="callout-title">Tip: coercion isn’t limited to data
frames</h3>
<div class="callout-content">
<p>Prior to R 4.0 when importing a data frame using any one of the
<code>read.table()</code> functions such as <code>read.csv()</code> ,
the argument <code>StringsAsFactors</code> was by default set to true
TRUE. Setting it to FALSE will treat any non-numeric column to a
character type. <code>read.csv()</code> documentation, you will also see
you can explicitly type your columns using the <code>colClasses</code>
argument. Other R packages (such as the Tidyverse “readr”) don’t have
this particular conversion issue, but many packages will still try to
guess a data type.</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="data-frame-bonus-material-math-sorting-renaming">Data frame bonus material: math, sorting, renaming<a class="anchor" aria-label="anchor" href="#data-frame-bonus-material-math-sorting-renaming"></a></h2>
<hr class="half-width"><p>Here are a few operations that don’t need much explanation, but which
are good to know.</p>
<p>There are lots of arithmetic functions you may want to apply to your
data frame, covering those would be a course in itself (there is some
starting material in <a href="https://swcarpentry.github.io/r-novice-inflammation/15-supp-loops-in-depth.html" class="external-link">the
<em>Loops in R</em> episode of one of the Software Carpentry
lessons</a>). Our lessons will cover some additional summary statistical
functions in a subsequent lesson, but overall we will focus on data
cleaning and visualization.</p>
<p>You can use functions like <code>mean()</code>, <code>min()</code>,
<code>max()</code> on an individual column. Let’s look at the “DP” or
filtered depth. This value shows the number of filtered reads that
support each of the reported variants.</p>
<div class="codewrapper sourceCode" id="cb92">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">max</span><span class="op">(</span><span class="va">variants</span><span class="op">$</span><span class="va">DP</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 79</code></pre>
</div>
<p>You can sort a data frame using the <code>order()</code>
function:</p>
<div class="codewrapper sourceCode" id="cb94">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sorted_by_DP</span> <span class="op">&lt;-</span> <span class="va">variants</span><span class="op">[</span><span class="fu">order</span><span class="op">(</span><span class="va">variants</span><span class="op">$</span><span class="va">DP</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">sorted_by_DP</span><span class="op">$</span><span class="va">DP</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 2 2 2 2 2 2</code></pre>
</div>
<div id="exercise" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise" class="callout-inner">
<h3 class="callout-title">Exercise</h3>
<div class="callout-content">
<p>The <code>order()</code> function lists values in increasing order by
default. Look at the documentation for this function and change
<code>sorted_by_DP</code> to start with variants with the greatest
filtered depth (“DP”).</p>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4"> Show me the solution </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" aria-labelledby="headingSolution4" data-bs-parent="#accordionSolution4">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb96">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span>   <span class="va">sorted_by_DP</span> <span class="op">&lt;-</span> <span class="va">variants</span><span class="op">[</span><span class="fu">order</span><span class="op">(</span><span class="va">variants</span><span class="op">$</span><span class="va">DP</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, <span class="op">]</span></span>
<span>   <span class="fu">head</span><span class="op">(</span><span class="va">sorted_by_DP</span><span class="op">$</span><span class="va">DP</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 79 46 41 29 29 27</code></pre>
</div>
</div>
</div>
</div>
</div>
<p>You can rename columns:</p>
<div class="codewrapper sourceCode" id="cb98">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">colnames</span><span class="op">(</span><span class="va">variants</span><span class="op">)</span><span class="op">[</span><span class="fu">colnames</span><span class="op">(</span><span class="va">variants</span><span class="op">)</span> <span class="op">==</span> <span class="st">"sample_id"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"strain"</span></span>
<span></span>
<span><span class="co"># check the column name (hint names are returned as a vector)</span></span>
<span><span class="fu">colnames</span><span class="op">(</span><span class="va">variants</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> [1] "strain"        "CHROM"         "POS"           "ID"
 [5] "REF"           "ALT"           "QUAL"          "FILTER"
 [9] "INDEL"         "IDV"           "IMF"           "DP"
[13] "VDB"           "RPB"           "MQB"           "BQB"
[17] "MQSB"          "SGB"           "MQ0F"          "ICB"
[21] "HOB"           "AC"            "AN"            "DP4"
[25] "MQ"            "Indiv"         "gt_PL"         "gt_GT"
[29] "gt_GT_alleles"</code></pre>
</div>
</section><section><h2 class="section-heading" id="saving-your-data-frame-to-a-file">Saving your data frame to a file<a class="anchor" aria-label="anchor" href="#saving-your-data-frame-to-a-file"></a></h2>
<hr class="half-width"><p>We can save data to a file. We will save our
<code>SRR2584863_variants</code> object to a .csv file using the
<code>write.csv()</code> function:</p>
<div class="codewrapper sourceCode" id="cb100">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">write.csv</span><span class="op">(</span><span class="va">SRR2584863_variants</span>, file <span class="op">=</span> <span class="st">"data/SRR2584863_variants.csv"</span><span class="op">)</span></span></code></pre>
</div>
<p>The <code>write.csv()</code> function has some additional arguments
listed in the help, but at a minimum you need to tell it what data frame
to write to file, and give a path to a file name in quotes (if you only
provide a file name, the file will be written in the current working
directory).</p>
</section><section><h2 class="section-heading" id="importing-data-from-excel">Importing data from Excel<a class="anchor" aria-label="anchor" href="#importing-data-from-excel"></a></h2>
<hr class="half-width"><p>Excel is one of the most common formats, so we need to discuss how to
make these files play nicely with R. The simplest way to import data
from Excel is to <strong>save your Excel file in .csv format</strong>*.
You can then import into R right away. Sometimes you may not be able to
do this (imagine you have data in 300 Excel files, are you going to open
and export all of them?).</p>
<p>One common R package (a set of code with features you can download
and add to your R installation) is the <a href="https://CRAN.R-project.org/package=readxl" class="external-link">readxl package</a>
which can open and import Excel files. Rather than addressing package
installation this second (we’ll discuss this soon!), we can take
advantage of RStudio’s import feature which integrates this package.
(Note: this feature is available only in the latest versions of RStudio
such as is installed on our cloud instance).</p>
<p>First, in the RStudio menu go to <strong>File</strong>, select
<strong>Import Dataset</strong>, and choose <strong>From Excel…</strong>
(notice there are several other options you can explore).</p>
<figure><img src="fig/rstudio_import_menu.png" alt="RStudio import menu" class="figure mx-auto d-block"><div class="figcaption">RStudio import menu</div>
</figure><p>Next, under <strong>File/Url:</strong> click the <kbd>Browse</kbd>
button and navigate to the <strong>Ecoli_metadata.xlsx</strong> file
located at <code>/home/dcuser/dc_sample_data/R</code>. You should now
see a preview of the data to be imported:</p>
<figure><img src="fig/rstudio_import_screen.png" alt="RStudio import screen" class="figure mx-auto d-block"><div class="figcaption">RStudio import screen</div>
</figure><p>Notice that you have the option to change the data type of each
variable by clicking arrow (drop-down menu) next to each column title.
Under <strong>Import Options</strong> you may also rename the data,
choose a different sheet to import, and choose how you will handle
headers and skipped rows. Under <strong>Code Preview</strong> you can
see the code that will be used to import this file. We could have
written this code and imported the Excel file without the RStudio import
function, but now you can choose your preference.</p>
<p>In this exercise, we will leave the title of the data frame as
<strong>Ecoli_metadata</strong>, and there are no other options we need
to adjust. Click the <kbd>Import</kbd> button to import the data.</p>
<p>Finally, let’s check the first few lines of the
<code>Ecoli_metadata</code> data frame:</p>
<div class="codewrapper sourceCode" id="cb101">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="va">Ecoli_metadata</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 6 × 7
  sample   generation clade   strain cit     run       genome_size
  &lt;chr&gt;         &lt;dbl&gt; &lt;chr&gt;   &lt;chr&gt;  &lt;chr&gt;   &lt;chr&gt;           &lt;dbl&gt;
1 REL606            0 NA      REL606 unknown &lt;NA&gt;             4.62
2 REL1166A       2000 unknown REL606 unknown SRR098028        4.63
3 ZDB409         5000 unknown REL606 unknown SRR098281        4.6
4 ZDB429        10000 UC      REL606 unknown SRR098282        4.59
5 ZDB446        15000 UC      REL606 unknown SRR098283        4.66
6 ZDB458        20000 (C1,C2) REL606 unknown SRR098284        4.63</code></pre>
</div>
<p>The type of this object is ‘tibble’, a type of data frame we will
talk more about in the ‘dplyr’ section. If you needed a true R data
frame you could coerce with <code>as.data.frame()</code>.</p>
<div id="exercise-putting-it-all-together---data-frames" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-putting-it-all-together---data-frames" class="callout-inner">
<h3 class="callout-title">Exercise: Putting it all together - data
frames</h3>
<div class="callout-content">
<p><strong>Using the <code>Ecoli_metadata</code> data frame created
above, answer the following questions</strong></p>
<ol style="list-style-type: upper-alpha"><li><p>What are the dimensions (# rows, # columns) of the data
frame?</p></li>
<li><p>What are categories are there in the <code>cit</code> column?
<em>hint</em>: treat column as factor</p></li>
<li><p>How many of each of the <code>cit</code> categories are
there?</p></li>
<li><p>What is the genome size for the 7th observation in this data
set?</p></li>
<li><p>What is the median value of the variable
<code>genome_size</code></p></li>
<li><p>Rename the column <code>sample</code> to
<code>sample_id</code></p></li>
<li><p>Create a new column (name genome_size_bp) and set it equal to the
genome_size multiplied by 1,000,000</p></li>
<li><p>Save the edited Ecoli_metadata data frame as
“exercise_solution.csv” in your current working directory.</p></li>
</ol></div>
</div>
</div>
<div id="accordionSolution5" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution5" aria-expanded="false" aria-controls="collapseSolution5">
  <h4 class="accordion-header" id="headingSolution5"> Show me the solution </h4>
</button>
<div id="collapseSolution5" class="accordion-collapse collapse" aria-labelledby="headingSolution5" data-bs-parent="#accordionSolution5">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb103">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">dim</span><span class="op">(</span><span class="va">Ecoli_metadata</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 30  7</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb105">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">levels</span><span class="op">(</span><span class="fu">as.factor</span><span class="op">(</span><span class="va">Ecoli_metadata</span><span class="op">$</span><span class="va">cit</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "minus"   "plus"    "unknown"</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb107">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">table</span><span class="op">(</span><span class="fu">as.factor</span><span class="op">(</span><span class="va">Ecoli_metadata</span><span class="op">$</span><span class="va">cit</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
  minus    plus unknown
      9       9      12 </code></pre>
</div>
<div class="codewrapper sourceCode" id="cb109">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">Ecoli_metadata</span><span class="op">[</span><span class="fl">7</span>, <span class="fl">7</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 1 × 1
  genome_size
        &lt;dbl&gt;
1        4.62</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb111">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">median</span><span class="op">(</span><span class="va">Ecoli_metadata</span><span class="op">$</span><span class="va">genome_size</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 4.625</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb113">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">colnames</span><span class="op">(</span><span class="va">Ecoli_metadata</span><span class="op">)</span><span class="op">[</span><span class="fu">colnames</span><span class="op">(</span><span class="va">Ecoli_metadata</span><span class="op">)</span> <span class="op">==</span> <span class="st">"sample"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"sample_id"</span></span>
<span><span class="va">Ecoli_metadata</span><span class="op">$</span><span class="va">genome_size_bp</span> <span class="op">&lt;-</span> <span class="va">Ecoli_metadata</span><span class="op">$</span><span class="va">genome_size</span> <span class="op">*</span> <span class="fl">1000000</span></span>
<span><span class="fu">write.csv</span><span class="op">(</span><span class="va">Ecoli_metadata</span>, file <span class="op">=</span> <span class="st">"exercise_solution.csv"</span><span class="op">)</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul><li>It is easy to import data into R from tabular formats including
Excel. However, you still need to check that R has imported and
interpreted your data correctly</li>
<li>There are best practices for organizing your data (keeping it tidy)
and R is great for this</li>
<li>Base R has many useful functions for manipulating your data, but all
of R’s capabilities are greatly enhanced by software packages developed
by the community</li>
</ul></div>
</div>
</div>
</section></div> <!-- / div.lesson-content -->
    </main><!-- / main#main-content.main-content --><nav class="bottom-pagination mx-md-4" aria-label="Previous and Next Chapter"><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="02-data-prelude.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="04-bioconductor-vcfr.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="02-data-prelude.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Introduction to the
        </a>
        <a class="chapter-link float-end" href="04-bioconductor-vcfr.html" rel="next">
          Next: Using packages from...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
    </nav></div> <!-- / div.primary-content.col-xs-12 -->
<!-- END:   inst/pkgdown/templates/content-instructor.html-->

      </div><!--/div.row-->
      		<footer class="row footer mx-md-3"><hr><div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>

        <a href="https://github.com/datacarpentry/genomics-r-intro/edit/main/episodes/03-basics-factors-dataframes.Rmd" class="external-link">Edit on GitHub</a>

	
        | <a href="https://github.com/datacarpentry/genomics-r-intro/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/datacarpentry/genomics-r-intro/" class="external-link">Source</a></p>
				<p><a href="https://github.com/datacarpentry/genomics-r-intro/blob/main/CITATION" class="external-link">Cite</a> | <a href="mailto:team@carpentries.org">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">

        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>

        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.17.1" class="external-link">sandpaper (0.17.1)</a>, <a href="https://github.com/carpentries/pegboard/tree/0.7.9" class="external-link">pegboard (0.7.9)</a>, and <a href="https://github.com/carpentries/varnish/tree/1.0.7" class="external-link">varnish (1.0.7)</a></p>
			</div>
		</footer></div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "LearningResource",
  "@id": "https://datacarpentry.github.io/genomics-r-intro/03-basics-factors-dataframes.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/LearningResource/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, The Carpentries",
  "name": "R Basics continued - factors and data frames",
  "creativeWorkStatus": "active",
  "url": "https://datacarpentry.github.io/genomics-r-intro/03-basics-factors-dataframes.html",
  "identifier": "https://datacarpentry.github.io/genomics-r-intro/03-basics-factors-dataframes.html",
  "dateCreated": "2018-03-12",
  "dateModified": "2025-10-27",
  "datePublished": "2025-10-27"
}

  </script><script>
		feather.replace();
	</script><!-- Matomo --><script>
          var _paq = window._paq = window._paq || [];
          /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
          _paq.push(["setDocumentTitle", document.domain + "/" + document.title]);
          _paq.push(["setDomains", ["*.lessons.carpentries.org","*.datacarpentry.github.io","*.datacarpentry.org","*.librarycarpentry.github.io","*.librarycarpentry.org","*.swcarpentry.github.io", "*.carpentries.github.io"]]);
          _paq.push(["setDoNotTrack", true]);
          _paq.push(["disableCookies"]);
          _paq.push(["trackPageView"]);
          _paq.push(["enableLinkTracking"]);
          (function() {
              var u="https://matomo.carpentries.org/";
              _paq.push(["setTrackerUrl", u+"matomo.php"]);
              _paq.push(["setSiteId", "1"]);
              var d=document, g=d.createElement("script"), s=d.getElementsByTagName("script")[0];
              g.async=true; g.src="https://matomo.carpentries.org/matomo.js"; s.parentNode.insertBefore(g,s);
          })();
        </script><!-- End Matomo Code --></body></html><!-- END:   inst/pkgdown/templates/layout.html-->

