<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<title>Intro to R and RStudio for Genomics: All in One View</title>
<meta name="viewport" content="width=device-width, initial-scale=1">
<link rel="stylesheet" type="text/css" href="assets/styles.css">
<script src="assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [ ['$','$'], ['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="apple-touch-icon.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="manifest" href="site.webmanifest">
<link rel="mask-icon" href="safari-pinned-tab.svg" color="#5bbad5">
<meta name="msapplication-TileColor" content="#da532c">
<meta name="theme-color" content="#ffffff">
</head>
<body>
    <header id="top" class="navbar navbar-expand-md navbar-light bg-white top-nav data"><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-6">
      <div class="large-logo">
        <img alt="Data Carpentry" src="assets/images/data-logo.svg"><abbr class="icon" title="This lesson is in the alpha phase, which means that it has been taught once and lesson authors are iterating on feedback." style="text-decoration: unset">
           
          <a href="https://cdh.carpentries.org/the-lesson-life-cycle.html#field-testing-alpha-stage" class="external-link alert-link" style="color: #383838">Alpha
            <i aria-hidden="true" class="icon" data-feather="alert-triangle" style="background: #FFC700; border-radius: 5px"></i>
          </a>
          <span class="visually-hidden">This lesson is in the alpha phase, which means that it has been taught once and lesson authors are iterating on feedback.</span>
        </abbr>
        
      </div>
    </div>
    <div class="selector-container ">
      
      
      <div class="dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Learner View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1">
<li><button class="dropdown-item" type="button" onclick="window.location.href='instructor/aio.html';">Instructor View</button></li>
        </ul>
</div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl navbar-light bg-white bottom-nav data" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Data Carpentry" src="assets/images/data-logo-sm.svg">
</div>
    <div class="lesson-title-md">
      Intro to R and RStudio for Genomics
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
      <i role="img" aria-label="search button" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0">
<li class="nav-item">
          <span class="lesson-title">
            Intro to R and RStudio for Genomics
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="reference.html#glossary">Glossary</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="profiles.html">Learner Profiles</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown">
<li><a class="dropdown-item" href="discuss.html">Discussion</a></li>
<li><a class="dropdown-item" href="reference.html">Glossary</a></li>
          </ul>
</li>
      </ul>
</div>
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
<input class="form-control me-2 searchbox" type="search" placeholder="Search" aria-label="Search"><button class="btn btn-outline-success tablet-search-button" type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="search button"></i>
        </button>
      </fieldset>
</form>
  </div>
<!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Intro to R and RStudio for Genomics
</div>

<aside class="col-md-12 lesson-progress"><div style="width: %" class="percentage">
    %
  </div>
  <div class="progress data">
    <div class="progress-bar data" role="progressbar" style="width: %" aria-valuenow="" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
  <div id="sidebar-col" class="col-lg-4">
    <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle"><i class="search-icon" data-feather="x" role="img"></i></button>
        <div class="sidebar-inner">
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Learner View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="instructor/aio.html">Instructor View</a>
                      </div>
                    </div>
                  </div>
<!--/div.accordion-item-->
                </div>
<!--/div.accordion-flush-->
              </div>
<!--div.sidenav-view-selector -->
            </div>
<!--/div.col -->
      
            <hr>
</div>
<!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Setup</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="00-introduction.html">1. Introducing R and RStudio IDE</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="01-r-basics.html">2. R Basics</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="02-data-prelude.html">3. Introduction to the example dataset and file type</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="03-basics-factors-dataframes.html">4. R Basics continued - factors and data frames</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="04-bioconductor-vcfr.html">5. Using packages from Bioconductor</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="05-dplyr.html">6. Data Wrangling and Analyses with Tidyverse</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush8">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading8">
        <a href="06-data-visualization.html">7. Data Visualization with ggplot2</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush9">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading9">
        <a href="07-r-help.html">8. Getting help with R</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width">
<div class="accordion accordion-flush resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul>
<li>
                        <a href="key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="reference.html#glossary">Glossary</a>
                      </li>
                      <li>
                        <a href="profiles.html">Learner Profiles</a>
                      </li>
                      <li><a href="discuss.html">Discussion</a></li>
<li><a href="reference.html">Glossary</a></li>
                    </ul>
</div>
                </div>
              </div>
            </div>
            <hr class="half-width resources">
<a href="aio.html">See all in one page</a>

            <hr class="d-none d-sm-block d-md-none">
<div class="d-grid gap-1">
            
            </div>
          </div>
<!-- /div.accordion -->
        </div>
<!-- /div.sidebar-inner -->
      </nav>
</div>
<!-- /div.sidebar -->
  </div>
<!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-extra.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <main id="main-content" class="main-content"><div class="container lesson-content">
        
        
<section id="aio-00-introduction"><p>Content from <a href="00-introduction.html">Introducing R and RStudio IDE</a></p>
<hr>
<p> Last updated on 2023-05-01 | 
        
        <a href="https://github.com/datacarpentry/genomics-r-intro/edit/main/episodes/00-introduction.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>Why use R?</li>
<li>Why use RStudio and how does it differ from R?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Know advantages of analyzing data in R</li>
<li>Know advantages of using RStudio</li>
<li>Create an RStudio project, and know the benefits of working within a
project</li>
<li>Be able to customize the RStudio layout</li>
<li>Be able to locate and change the current working directory with
<code>getwd()</code> and <code>setwd()</code>
</li>
<li>Compose an R script file containing comments and commands</li>
<li>Understand what an R function is</li>
<li>Locate help for an R function using <code>?</code>, <code>??</code>,
and <code>args()</code>
</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section id="getting-ready-to-use-r-for-the-first-time"><h2 class="section-heading">Getting ready to use R for the first time<a class="anchor" aria-label="anchor" href="#getting-ready-to-use-r-for-the-first-time"></a>
</h2>
<hr class="half-width">
<p>In this lesson we will take you through the very first things you
need to get R working.</p>
<div id="tip-this-lesson-works-best-on-the-cloud" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="tip-this-lesson-works-best-on-the-cloud" class="callout-inner">
<h3 class="callout-title">Tip: This lesson works best on the cloud<a class="anchor" aria-label="anchor" href="#tip-this-lesson-works-best-on-the-cloud"></a>
</h3>
<div class="callout-content">
<p>Remember, these lessons assume we are using the pre-configured
virtual machine instances provided to you at a genomics workshop. Much
of this work could be done on your laptop, but we use instances to
simplify workshop setup requirements, and to get you familiar with using
the cloud (a common requirement for working with big data). Visit the <a href="https://www.datacarpentry.org/genomics-workshop/setup.html" class="external-link">Genomics
Workshop setup page</a> for details on getting this instance running on
your own, or for the info you need to do this on your own computer.</p>
</div>
</div>
</div>
</section><section id="a-brief-history-of-r"><h2 class="section-heading">A Brief History of R<a class="anchor" aria-label="anchor" href="#a-brief-history-of-r"></a>
</h2>
<hr class="half-width">
<p><a href="https://en.wikipedia.org/wiki/R_(programming_language)" class="external-link">R</a> has
been around since 1995, and was created by Ross Ihaka and Robert
Gentleman at the University of Auckland, New Zealand. R is based off the
S programming language developed at Bell Labs and was developed to teach
intro statistics. See this <a href="https://www.stat.auckland.ac.nz/~ihaka/downloads/Massey.pdf" class="external-link">slide
deck</a> by Ross Ihaka for more info on the subject.</p>
</section><section id="advantages-of-using-r"><h2 class="section-heading">Advantages of using R<a class="anchor" aria-label="anchor" href="#advantages-of-using-r"></a>
</h2>
<hr class="half-width">
<p>At more than 20 years old, R is fairly mature and <a href="https://www.tiobe.com/tiobe-index/r/" class="external-link">growing in popularity</a>.
However, programming isn’t a popularity contest. Here are key advantages
of analyzing data in R:</p>
<ul>
<li>
<strong>R is <a href="https://en.wikipedia.org/wiki/Open-source_software" class="external-link">open
source</a></strong>. This means R is free - an advantage if you are at
an institution where you have to pay for your own MATLAB or SAS license.
Open source, is important to your colleagues in parts of the world where
expensive software in inaccessible. It also means that R is actively
developed by a community (see <a href="https://www.r-project.org/" class="external-link">r-project.org</a>), and there are
regular updates.</li>
<li>
<strong>R is widely used</strong>. Ok, maybe programming is a
popularity contest. Because, R is used in many areas (not just
bioinformatics), you are more likely to find help online when you need
it. Chances are, almost any error message you run into, someone else has
already experienced.</li>
<li>
<strong>R is powerful</strong>. R runs on multiple platforms
(Windows/MacOS/Linux). It can work with much larger datasets than
popular spreadsheet programs like Microsoft Excel, and because of its
scripting capabilities is far more reproducible. Also, there are
thousands of available software packages for science, including genomics
and other areas of life science.</li>
</ul>
<div id="discussion-your-experience" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="discussion-your-experience" class="callout-inner">
<h3 class="callout-title">Discussion: Your experience<a class="anchor" aria-label="anchor" href="#discussion-your-experience"></a>
</h3>
<div class="callout-content">
<p>What has motivated you to learn R? Have you had a research question
for which spreadsheet programs such as Excel have proven difficult to
use, or where the size of the data set created issues?</p>
</div>
</div>
</div>
</section><section id="introducing-rstudio-server"><h2 class="section-heading">Introducing RStudio Server<a class="anchor" aria-label="anchor" href="#introducing-rstudio-server"></a>
</h2>
<hr class="half-width">
<p>In these lessons, we will be making use of a software called <a href="https://www.rstudio.com/products/RStudio/" class="external-link">RStudio</a>, an <a href="https://en.wikipedia.org/wiki/Integrated_development_environment" class="external-link">Integrated
Development Environment (IDE)</a>. RStudio, like most IDEs, provides a
graphical interface to R, making it more user-friendly, and providing
dozens of useful features. We will introduce additional benefits of
using RStudio as you cover the lessons. In this case, we are
specifically using <a href="https://www.rstudio.com/products/RStudio/#Server" class="external-link">RStudio
Server</a>, a version of RStudio that can be accessed in your web
browser. RStudio Server has the same features of the Desktop version of
RStudio you could download as standalone software.</p>
</section><section id="log-on-to-rstudio-server"><h2 class="section-heading">Log on to RStudio Server<a class="anchor" aria-label="anchor" href="#log-on-to-rstudio-server"></a>
</h2>
<hr class="half-width">
<p>Open a web browser and enter the URL you used to log in at the
terminal (provided by your instructors), followed by <code>:8787</code>.
For example, if your URL was ec2.12.2.45.678.compute-1.amazonaws.com,
you should enter:</p>
<pre><code>http://ec2.12.2.45.678.compute-1.amazonaws.com:8787</code></pre>
<div id="tip-make-sure-there-are-no-spaces-before-or-after-your-url." class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="tip-make-sure-there-are-no-spaces-before-or-after-your-url." class="callout-inner">
<h3 class="callout-title">Tip: Make sure there are no spaces before or
after your URL.<a class="anchor" aria-label="anchor" href="#tip-make-sure-there-are-no-spaces-before-or-after-your-url."></a>
</h3>
<div class="callout-content">
<p>If your URL has spaces, your web browser may interpret it as a search
query.</p>
</div>
</div>
</div>
<p>You should now be looking at a page that will allow you to login to
the RStudio server:</p>
<figure><img src="fig/rstudio_login_screen.png" alt="rstudio default session" style="width: 1000px;" class="figure mx-auto d-block"></figure><p>Enter your user credentials and click <kbd>Sign In</kbd>. The
credentials for the genomics Data Carpentry instances will be provided
by your instructors.</p>
<p>You should now see the RStudio interface:</p>
<figure><img src="fig/rstudio_session_default.png" alt="rstudio default session" style="width:1000px;" class="figure mx-auto d-block"></figure></section><section id="create-an-rstudio-project"><h2 class="section-heading">Create an RStudio project<a class="anchor" aria-label="anchor" href="#create-an-rstudio-project"></a>
</h2>
<hr class="half-width">
<p>One of the first benefits we will take advantage of in RStudio is
something called an <strong>RStudio Project</strong>. An RStudio project
allows you to more easily:</p>
<ul>
<li>Save data, files, variables, packages, etc. related to a specific
analysis project</li>
<li>Restart work where you left off</li>
<li>Collaborate, especially if you are using version control such as <a href="https://swcarpentry.github.io/git-novice/" class="external-link">git</a>.</li>
</ul>
<ol style="list-style-type: decimal">
<li>To create a project, go to the <kbd>File</kbd> menu, and click
<kbd>New Project…</kbd>.</li>
</ol>
<figure><img src="fig/new_project_window.png" alt="rstudio default session" style="width: 600px;" class="figure mx-auto d-block"></figure><ol start="2" style="list-style-type: decimal">
<li><p>In the window that opens select <strong>New Directory</strong>,
then <strong>New Project</strong>. For “Directory name:” enter
<strong>dc_genomics_r</strong>. For “Create project as subdirectory of”,
click <kbd>Browse…</kbd> and then click <kbd>Choose</kbd> which will
select your home directory “~”.</p></li>
<li><p>Finally click <kbd>Create Project</kbd>. In the “Files” tab of
your output pane (more about the RStudio layout in a moment), you should
see an RStudio project file, <strong>dc_genomics_r.Rproj</strong>. All
RStudio projects end with the “<strong>.Rproj</strong>” file
extension.</p></li>
</ol>
<div id="tip-make-your-project-more-reproducible-with-renv" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="tip-make-your-project-more-reproducible-with-renv" class="callout-inner">
<h3 class="callout-title">Tip: Make your project more reproducible with
renv<a class="anchor" aria-label="anchor" href="#tip-make-your-project-more-reproducible-with-renv"></a>
</h3>
<div class="callout-content">
<p>One of the most wonderful and also frustrating aspects of working
with R is managing packages. We will talk more about them, but packages
(e.g. ggplot2) are add-ons that extend what you can do with R.
Unfortunately it is very common that you may run into versions of R
and/or R packages that are not compatible. This may make it difficult
for someone to run your R script using their version of R or a given R
package, and/or make it more difficult to run their scripts on your
machine. <a href="https://rstudio.github.io/renv/" class="external-link">renv</a> is an
RStudio add-on that will associate your packages and project so that
your work is more portable and reproducible. To turn on renv click on
the <kbd>Tools</kbd> menu and select <kbd>Project Options</kbd>. Under
<strong>Enviornments</strong> check off “<strong>Use renv with this
project</strong>” and follow any installation instructions.</p>
</div>
</div>
</div>
</section><section id="creating-your-first-r-script"><h2 class="section-heading">Creating your first R script<a class="anchor" aria-label="anchor" href="#creating-your-first-r-script"></a>
</h2>
<hr class="half-width">
<p>Now that we are ready to start exploring R, we will want to keep a
record of the commands we are using. To do this we can create an R
script:</p>
<p>Click the <kbd>File</kbd> menu and select <kbd>New File</kbd> and
then <kbd>R Script</kbd>. Before we go any further, save your script by
clicking the save/disk icon that is in the bar above the first line in
the script editor, or click the <kbd>File</kbd> menu and select
<kbd>save</kbd>. In the “Save File” window that opens, name your file
<strong>“genomics_r_basics”</strong>. The new script
<strong>genomics_r_basics.R</strong> should appear under “files” in the
output pane. By convention, R scripts end with the file extension
<strong>.R</strong>.</p>
</section><section id="overview-and-customization-of-the-rstudio-layout"><h2 class="section-heading">Overview and customization of the RStudio layout<a class="anchor" aria-label="anchor" href="#overview-and-customization-of-the-rstudio-layout"></a>
</h2>
<hr class="half-width">
<p>Here are the major windows (or panes) of the RStudio environment:</p>
<figure><img src="fig/rstudio_session_4pane_layout.png" alt="rstudio default session" style="width: 1000px;" class="figure mx-auto d-block"></figure><ul>
<li>
<strong>Source</strong>: This pane is where you will write/view R
scripts. Some outputs (such as if you view a dataset using
<code>View()</code>) will appear as a tab here.</li>
<li>
<strong>Console/Terminal/Jobs</strong>: This is actually where you
see the execution of commands. This is the same display you would see if
you were using R at the command line without RStudio. You can work
interactively (i.e. enter R commands here), but for the most part we
will run a script (or lines in a script) in the source pane and watch
their execution and output here. The “Terminal” tab give you access to
the BASH terminal (the Linux operating system, unrelated to R). RStudio
also allows you to run jobs (analyses) in the background. This is useful
if some analysis will take a while to run. You can see the status of
those jobs in the background.</li>
<li>
<strong>Environment/History</strong>: Here, RStudio will show you
what datasets and objects (variables) you have created and which are
defined in memory. You can also see some properties of objects/datasets
such as their type and dimensions. The “History” tab contains a history
of the R commands you’ve executed R.</li>
<li>
<strong>Files/Plots/Packages/Help/Viewer</strong>: This multipurpose
pane will show you the contents of directories on your computer. You can
also use the “Files” tab to navigate and set the working directory. The
“Plots” tab will show the output of any plots generated. In “Packages”
you will see what packages are actively loaded, or you can attach
installed packages. “Help” will display help files for R functions and
packages. “Viewer” will allow you to view local web content (e.g. HTML
outputs).</li>
</ul>
<div id="tip-uploads-and-downloads-in-the-cloud" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="tip-uploads-and-downloads-in-the-cloud" class="callout-inner">
<h3 class="callout-title">Tip: Uploads and downloads in the cloud<a class="anchor" aria-label="anchor" href="#tip-uploads-and-downloads-in-the-cloud"></a>
</h3>
<div class="callout-content">
<p>In the “Files” tab you can select a file and download it from your
cloud instance (click the “more” button) to your local computer. Uploads
are also possible.</p>
</div>
</div>
</div>
<p>All of the panes in RStudio have configuration options. For example,
you can minimize/maximize a pane, or by moving your mouse in the space
between panes you can resize as needed. The most important customization
options for pane layout are in the <kbd>View</kbd> menu. Other options
such as font sizes, colors/themes, and more are in the <kbd>Tools</kbd>
menu under <kbd>Global Options</kbd>.</p>
<div id="you-are-working-with-r" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="you-are-working-with-r" class="callout-inner">
<h3 class="callout-title">You are working with R<a class="anchor" aria-label="anchor" href="#you-are-working-with-r"></a>
</h3>
<div class="callout-content">
<p>Although we won’t be working with R at the terminal, there are lots
of reasons to. For example, once you have written an RScript, you can
run it at any Linux or Windows terminal without the need to start up
RStudio. We don’t want you to get confused - RStudio runs R, but R is
not RStudio. For more on running an R Script at the terminal see this <a href="https://swcarpentry.github.io/r-novice-inflammation/05-cmdline/" class="external-link">Software
Carpentry lesson</a>.</p>
</div>
</div>
</div>
</section><section id="getting-to-work-with-r-navigating-directories"><h2 class="section-heading">Getting to work with R: navigating directories<a class="anchor" aria-label="anchor" href="#getting-to-work-with-r-navigating-directories"></a>
</h2>
<hr class="half-width">
<p>Now that we have covered the more aesthetic aspects of RStudio, we
can get to work using some commands. We will write, execute, and save
the commands we learn in our <strong>genomics_r_basics.R</strong> script
that is loaded in the Source pane. First, lets see what directory we are
in. To do so, type the following command into the script:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">getwd</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<p>To execute this command, make sure your cursor is on the same line
the command is written. Then click the <kbd>Run</kbd> button that is
just above the first line of your script in the header of the Source
pane.</p>
<p>In the console, we expect to see the following output*:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "/home/dcuser/dc_genomics_r"</code></pre>
</div>
<p>* Notice, at the Console, you will also see the instruction you
executed above the output in blue.</p>
<p>Since we will be learning several commands, we may already want to
keep some short notes in our script to explain the purpose of the
command. Entering a <code>#</code> before any line in an R script turns
that line into a comment, which R will not try to interpret as code.
Edit your script to include a comment on the purpose of commands you are
learning, e.g.:</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># this command shows the current working directory</span></span>
<span><span class="fu">getwd</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<div id="exercise-work-interactively-in-r" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-work-interactively-in-r" class="callout-inner">
<h3 class="callout-title">Exercise: Work interactively in R<a class="anchor" aria-label="anchor" href="#exercise-work-interactively-in-r"></a>
</h3>
<div class="callout-content">
<p>What happens when you try to enter the <code>getwd()</code> command
in the Console pane?</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<p>You will get the same output you did as when you ran
<code>getwd()</code> from the source. You can run any command in the
Console, however, executing it from the source script will make it
easier for us to record what we have done, and ultimately run an entire
script, instead of entering commands one-by-one.</p>
</div>
</div>
</div>
</div>
<p>For the purposes of this exercise we want you to be in the directory
<code>"/home/dcuser/R_data"</code>. What if you weren’t? You can set
your home directory using the <code>setwd()</code> command. Enter this
command in your script, but <em>don’t run</em> this yet.</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># This sets the working directory</span></span>
<span><span class="fu">setwd</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<p>You may have guessed, you need to tell the <code>setwd()</code>
command what directory you want to set as your working directory. To do
so, inside of the parentheses, open a set of quotes. Inside the quotes
enter a <code>/</code> which is the root directory for Linux. Next, use
the <kbd>Tab</kbd> key, to take advantage of RStudio’s
Tab-autocompletion method, to select <code>home</code>,
<code>dcuser</code>, and <code>dc_genomics_r</code> directory. The path
in your script should look like this:</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># This sets the working directory</span></span>
<span><span class="fu">setwd</span><span class="op">(</span><span class="st">"/home/dcuser/dc_genomics_r"</span><span class="op">)</span></span></code></pre>
</div>
<p>When you run this command, the console repeats the command, but gives
you no output. Instead, you see the blank R prompt: <code>&gt;</code>.
Congratulations! Although it seems small, knowing what your working
directory is and being able to set your working directory is the first
step to analyzing your data.</p>
<div id="tip-never-use-setwd" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="tip-never-use-setwd" class="callout-inner">
<h3 class="callout-title">Tip: Never use <code>setwd()</code><a class="anchor" aria-label="anchor" href="#tip-never-use-setwd"></a>
</h3>
<div class="callout-content">
<p>Wait, what was the last 2 minutes about? Well, setting your working
directory is something you need to do, you need to be very careful about
using this as a step in your script. For example, what if your script is
being on a computer that has a different directory structure? The
top-level path in a Unix file system is root <code>/</code>, but on
Windows it is likely <code>C:\</code>. This is one of several ways you
might cause a script to break because a file path is configured
differently than your script anticipates. R packages like <a href="https://cran.r-project.org/package=here" class="external-link">here</a> and <a href="https://www.rdocumentation.org/packages/base/versions/3.4.3/topics/file.path" class="external-link">file.path</a>
allow you to specify file paths is a way that is more operating system
independent. See Jenny Bryan’s <a href="https://www.tidyverse.org/articles/2017/12/workflow-vs-script/" class="external-link">blog
post</a> for this and other R tips.</p>
</div>
</div>
</div>
</section><section id="using-functions-in-r-without-needing-to-master-them"><h2 class="section-heading">Using functions in R, without needing to master them<a class="anchor" aria-label="anchor" href="#using-functions-in-r-without-needing-to-master-them"></a>
</h2>
<hr class="half-width">
<p>A function in R (or any computing language) is a short program that
takes some input and returns some output. Functions may seem like an
advanced topic (and they are), but you have already used at least one
function in R. <code>getwd()</code> is a function! The next sections
will help you understand what is happening in any R script.</p>
<div id="exercise-what-do-these-functions-do" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-what-do-these-functions-do" class="callout-inner">
<h3 class="callout-title">Exercise: What do these functions do?<a class="anchor" aria-label="anchor" href="#exercise-what-do-these-functions-do"></a>
</h3>
<div class="callout-content">
<p>Try the following functions by writing them in your script. See if
you can guess what they do, and make sure to add comments to your script
about your assumed purpose.</p>
<ul>
<li><code>dir()</code></li>
<li><code>sessionInfo()</code></li>
<li><code>date()</code></li>
<li><code>Sys.time()</code></li>
</ul>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<ul>
<li>
<code>dir()</code> # Lists files in the working directory</li>
<li>
<code>sessionInfo()</code> # Gives the version of R and additional
info including on attached packages</li>
<li>
<code>date()</code> # Gives the current date</li>
<li>
<code>Sys.time()</code> # Gives the current time</li>
</ul>
<p><em>Notice</em>: Commands are case sensitive!</p>
</div>
</div>
</div>
</div>
<p>You have hopefully noticed a pattern - an R function has three key
properties:</p>
<ul>
<li>Functions have a name (e.g. <code>dir</code>, <code>getwd</code>);
note that functions are case sensitive!</li>
<li>Following the name, functions have a pair of <code>()</code>
</li>
<li>Inside the parentheses, a function may take 0 or more arguments</li>
</ul>
<p>An argument may be a specific input for your function and/or may
modify the function’s behavior. For example the function
<code>round()</code> will round a number with a decimal:</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># This will round a number to the nearest integer</span></span>
<span><span class="fu">round</span><span class="op">(</span><span class="fl">3.14</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 3</code></pre>
</div>
</section><section id="getting-help-with-function-arguments"><h2 class="section-heading">Getting help with function arguments<a class="anchor" aria-label="anchor" href="#getting-help-with-function-arguments"></a>
</h2>
<hr class="half-width">
<p>What if you wanted to round to one significant digit?
<code>round()</code> can do this, but you may first need to read the
help to find out how. To see the help (In R sometimes also called a
“vignette”) enter a <code>?</code> in front of the function name:</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">?</span><span class="fu">round</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<p>The “Help” tab will show you information (often, too much
information). You will slowly learn how to read and make sense of help
files. Checking the “Usage” or “Examples” headings is often a good place
to look first. If you look under “Arguments,” we also see what arguments
we can pass to this function to modify its behavior. You can also see a
function’s argument using the <code>args()</code> function:</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">args</span><span class="op">(</span><span class="va">round</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code><span><span class="kw">function</span> <span class="op">(</span><span class="va">x</span>, <span class="va">digits</span> <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> </span>
<span><span class="cn">NULL</span></span></code></pre>
</div>
<p><code>round()</code> takes two arguments, <code>x</code>, which is
the number to be rounded, and a <code>digits</code> argument. The
<code>=</code> sign indicates that a default (in this case 0) is already
set. Since <code>x</code> is not set, <code>round()</code> requires we
provide it, in contrast to <code>digits</code> where R will use the
default value 0 unless you explicitly provide a different value. We can
explicitly set the digits parameter when we call the function:</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">round</span><span class="op">(</span><span class="fl">3.14159</span>, digits <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 3.14</code></pre>
</div>
<p>Or, R accepts what we call “positional arguments”, if you pass a
function arguments separated by commas, R assumes that they are in the
order you saw when we used <code>args()</code>. In the case below that
means that <code>x</code> is 3.14159 and digits is 2.</p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">round</span><span class="op">(</span><span class="fl">3.14159</span>, <span class="fl">2</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 3.14</code></pre>
</div>
<p>Finally, what if you are using <code>?</code> to get help for a
function in a package not installed on your system, such as when you are
running a script which has dependencies.</p>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">?</span><span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<p>will return an error:</p>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in .helpForCall(topicExpr, parent.frame()) :
   no methods for ‘geom_point' and no documentation for it as a function</code></pre>
</div>
<p>Use two question marks (i.e. <code>??geom_point()</code>) and R will
return results from a search of the documentation for packages you have
installed on your computer in the “Help” tab. Finally, if you think
there should be a function, for example a statistical test, but you
aren’t sure what it is called in R, or what functions may be available,
use the <code>help.search()</code> function.</p>
<div id="exercise-searching-for-r-functions" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-searching-for-r-functions" class="callout-inner">
<h3 class="callout-title">Exercise: Searching for R functions<a class="anchor" aria-label="anchor" href="#exercise-searching-for-r-functions"></a>
</h3>
<div class="callout-content">
<p>Use <code>help.search()</code> to find R functions for the following
statistical functions. Remember to put your search query in quotes
inside the function’s parentheses.</p>
<ul>
<li>Chi-Squared test</li>
<li>Student t-test</li>
<li>mixed linear model</li>
</ul>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" data-bs-parent="#accordionSolution3" aria-labelledby="headingSolution3">
<div class="accordion-body">
<p>While your search results may return several tests, we list a few you
might find:</p>
<ul>
<li>Chi-Squared test: <code>stats::Chisquare</code>
</li>
<li>Student t-test: <code>stats::t.test</code>
</li>
<li>mixed linear model: <code>stats::lm.glm</code>
</li>
</ul>
</div>
</div>
</div>
</div>
<p>We will discuss more on where to look for the libraries and packages
that contain functions you want to use. For now, be aware that two
important ones are <a href="https://cran.r-project.org/" class="external-link">CRAN</a> - the
main repository for R, and <a href="https://bioconductor.org/" class="external-link">Bioconductor</a> - a popular repository
for bioinformatics-related R packages.</p>
</section><section id="rstudio-contextual-help"><h2 class="section-heading">RStudio contextual help<a class="anchor" aria-label="anchor" href="#rstudio-contextual-help"></a>
</h2>
<hr class="half-width">
<p>Here is one last bonus we will mention about RStudio. It’s difficult
to remember all of the arguments and definitions associated with a given
function. When you start typing the name of a function and hit the
<kbd>Tab</kbd> key, RStudio will display functions and associated
help:</p>
<figure><img src="fig/studio_contexthelp1.png" alt="rstudio default session" style="width: 600px;" class="figure mx-auto d-block"></figure><p>Once you type a function, hitting the <kbd>Tab</kbd> inside the
parentheses will show you the function’s arguments and provide
additional help for each of these arguments.</p>
<figure><img src="fig/studio_contexthelp2.png" alt="rstudio default session" style="width: 600px;" class="figure mx-auto d-block"></figure><div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Keypoints<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>R is a powerful, popular open-source scripting language</li>
<li>You can customize the layout of RStudio, and use the project feature
to manage the files and packages used in your analysis</li>
<li>RStudio allows you to run R in an easy-to-use interface and makes it
easy to find help</li>
</ul>
</div>
</div>
</div>
</section></section><section id="aio-01-r-basics"><p>Content from <a href="01-r-basics.html">R Basics</a></p>
<hr>
<p> Last updated on 2023-05-01 | 
        
        <a href="https://github.com/datacarpentry/genomics-r-intro/edit/main/episodes/01-r-basics.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What will these lessons not cover?</li>
<li>What are the basic features of the R language?</li>
<li>What are the most common objects in R?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Be able to create the most common R objects including vectors</li>
<li>Understand that vectors have modes, which correspond to the type of
data they contain</li>
<li>Be able to use arithmetic operators on R objects</li>
<li>Be able to retrieve (subset), name, or replace, values from a
vector</li>
<li>Be able to use logical operators in a subsetting operation</li>
<li>Understand that lists can hold data of more than one mode and can be
indexed</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section id="the-fantastic-world-of-r-awaits-you-or-nobody-wants-to-learn-how-to-use-r"><h2 class="section-heading">“The fantastic world of R awaits you” OR “Nobody wants to learn how
to use R”<a class="anchor" aria-label="anchor" href="#the-fantastic-world-of-r-awaits-you-or-nobody-wants-to-learn-how-to-use-r"></a>
</h2>
<hr class="half-width">
<p>Before we begin this lesson, we want you to be clear on the goal of
the workshop and these lessons. We believe that every learner can
<strong>achieve competency with R</strong>. You have reached competency
when you find that you are able to <strong>use R to handle common
analysis challenges in a reasonable amount of time</strong> (which
includes time needed to look at learning materials, search for answers
online, and ask colleagues for help). As you spend more time using R
(there is no substitute for regular use and practice) you will find
yourself gaining competency and even expertise. The more familiar you
get, the more complex the analyses you will be able to carry out, with
less frustration, and in less time - the fantastic world of R awaits
you!</p>
</section><section id="what-these-lessons-will-not-teach-you"><h2 class="section-heading">What these lessons will not teach you<a class="anchor" aria-label="anchor" href="#what-these-lessons-will-not-teach-you"></a>
</h2>
<hr class="half-width">
<p>Nobody wants to learn how to use R. People want to learn how to use R
to analyze their own research questions! Ok, maybe some folks learn R
for R’s sake, but these lessons assume that you want to start analyzing
genomic data as soon as possible. Given this, there are many valuable
pieces of information about R that we simply won’t have time to cover.
Hopefully, we will clear the hurdle of giving you just enough knowledge
to be dangerous, which can be a high bar in R! We suggest you look into
the additional learning materials in the tip box below.</p>
<p><strong>Here are some R skills we will <em>not</em> cover in these
lessons</strong></p>
<ul>
<li>How to create and work with R matrices and R lists</li>
<li>How to create and work with loops and conditional statements, and
the “apply” family of functions (which are super useful, read more <a href="https://www.r-bloggers.com/r-tutorial-on-the-apply-family-of-functions/" class="external-link">here</a>)</li>
<li>How to do basic string manipulations (e.g. finding patterns in text
using grep, replacing text)</li>
<li>How to plot using the default R graphic tools (we <em>will</em>
cover plot creation, but will do so using the popular plotting package
<code>ggplot2</code>)</li>
<li>How to use advanced R statistical functions</li>
</ul>
<div id="tip-where-to-learn-more" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="tip-where-to-learn-more" class="callout-inner">
<h3 class="callout-title">Tip: Where to learn more<a class="anchor" aria-label="anchor" href="#tip-where-to-learn-more"></a>
</h3>
<div class="callout-content">
<p>The following are good resources for learning more about R. Some of
them can be quite technical, but if you are a regular R user you may
ultimately need this technical knowledge.</p>
<ul>
<li>
<a href="https://cran.r-project.org/doc/contrib/Paradis-rdebuts_en.pdf" class="external-link">R
for Beginners</a>: By Emmanuel Paradis and a great starting point</li>
<li>
<a href="https://cran.r-project.org/manuals.html" class="external-link">The R Manuals</a>:
Maintained by the R project</li>
<li>
<a href="https://cran.r-project.org/other-docs.html" class="external-link">R contributed
documentation</a>: Also linked to the R project; importantly there are
materials available in several languages</li>
<li>
<a href="https://r4ds.had.co.nz/" class="external-link">R for Data Science</a>: A
wonderful collection by noted R educators and developers Garrett
Grolemund and Hadley Wickham</li>
<li>
<a href="https://peerj.com/collections/50-practicaldatascistats/" class="external-link">Practical
Data Science for Stats</a>: Not exclusively about R usage, but a nice
collection of pre-prints on data science and applications for R</li>
<li>
<a href="https://software-carpentry.org/lessons/" class="external-link">Programming in R
Software Carpentry lesson</a>: There are several Software Carpentry
lessons in R to choose from</li>
</ul>
</div>
</div>
</div>
</section><section id="creating-objects-in-r"><h2 class="section-heading">Creating objects in R<a class="anchor" aria-label="anchor" href="#creating-objects-in-r"></a>
</h2>
<hr class="half-width">
<div id="reminder" class="callout prereq">
<div class="callout-square">
<i class="callout-icon" data-feather="check"></i>
</div>
<div id="reminder" class="callout-inner">
<h3 class="callout-title">Reminder<a class="anchor" aria-label="anchor" href="#reminder"></a>
</h3>
<div class="callout-content">
<p>At this point you should be coding along in the
“<strong>genomics_r_basics.R</strong>” script we created in the last
episode. Writing your commands in the script (and commenting it) will
make it easier to record what you did and why.</p>
</div>
</div>
</div>
<p>What might be called a variable in many languages is called an
<strong>object</strong> in R.</p>
<p><strong>To create an object you need:</strong></p>
<ul>
<li>a name (e.g. ‘a’)</li>
<li>a value (e.g. ‘1’)</li>
<li>the assignment operator (‘&lt;-’)</li>
</ul>
<p>In your script, “<strong>genomics_r_basics.R</strong>”, using the R
assignment operator ‘&lt;-’, assign ‘1’ to the object ‘a’ as shown.
Remember to leave a comment in the line above (using the ‘#’) to explain
what you are doing:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># this line creates the object 'a' and assigns it the value '1'</span></span>
<span></span>
<span><span class="va">a</span> <span class="op">&lt;-</span> <span class="fl">1</span></span></code></pre>
</div>
<p>Next, run this line of code in your script. You can run a line of
code by hitting the <kbd>Run</kbd> button that is just above the first
line of your script in the header of the Source pane or you can use the
appropriate shortcut:</p>
<ul>
<li>Windows execution shortcut: <kbd>Ctrl</kbd>+<kbd>Enter</kbd>
</li>
<li>Mac execution shortcut: <kbd>Cmd(⌘)</kbd>+<kbd>Enter</kbd>
</li>
</ul>
<p>To run multiple lines of code, you can highlight all the line you
wish to run and then hit <kbd>Run</kbd> or use the shortcut key combo
listed above.</p>
<p>In the RStudio ‘Console’ you should see:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>a &lt;- 1
&gt;</code></pre>
</div>
<p>The ‘Console’ will display lines of code run from a script and any
outputs or status/warning/error messages (usually in red).</p>
<p>In the ‘Environment’ window you will also get a table:</p>
<table class="table">
<colgroup>
<col width="7%">
<col width="92%">
</colgroup>
<thead><tr class="header">
<th>Values</th>
<th></th>
</tr></thead>
<tbody><tr class="odd">
<td>a</td>
<td>1</td>
</tr></tbody>
</table>
<p>The ‘Environment’ window allows you to keep track of the objects you
have created in R.</p>
<div id="exercise-create-some-objects-in-r" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-create-some-objects-in-r" class="callout-inner">
<h3 class="callout-title">Exercise: Create some objects in R<a class="anchor" aria-label="anchor" href="#exercise-create-some-objects-in-r"></a>
</h3>
<div class="callout-content">
<p>Create the following objects; give each object an appropriate name
(your best guess at what name to use is fine):</p>
<ol style="list-style-type: decimal">
<li>Create an object that has the value of number of pairs of human
chromosomes</li>
<li>Create an object that has a value of your favorite gene name</li>
<li>Create an object that has this URL as its value: “<a href="ftp://ftp.ensemblgenomes.org/pub/bacteria/release-39/fasta/bacteria_5_collection/escherichia_coli_b_str_rel606/">ftp://ftp.ensemblgenomes.org/pub/bacteria/release-39/fasta/bacteria_5_collection/escherichia_coli_b_str_rel606/</a>”</li>
<li>Create an object that has the value of the number of chromosomes in
a diploid human cell</li>
</ol>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<p>Here as some possible answers to the challenge:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">human_chr_number</span> <span class="op">&lt;-</span> <span class="fl">23</span></span>
<span><span class="va">gene_name</span> <span class="op">&lt;-</span> <span class="st">'pten'</span></span>
<span><span class="va">ensemble_url</span> <span class="op">&lt;-</span> <span class="st">'ftp://ftp.ensemblgenomes.org/pub/bacteria/release-39/fasta/bacteria_5_collection/escherichia_coli_b_str_rel606/'</span></span>
<span><span class="va">human_diploid_chr_num</span> <span class="op">&lt;-</span>  <span class="fl">2</span> <span class="op">*</span> <span class="va">human_chr_number</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
</section><section id="naming-objects-in-r"><h2 class="section-heading">Naming objects in R<a class="anchor" aria-label="anchor" href="#naming-objects-in-r"></a>
</h2>
<hr class="half-width">
<p>Here are some important details about naming objects in R.</p>
<ul>
<li>
<strong>Avoid spaces and special characters</strong>: Object names
cannot contain spaces or the minus sign (<code>-</code>). You can use
‘_’ to make names more readable. You should avoid using special
characters in your object name (e.g. ! @ # . , etc.). Also, object names
cannot begin with a number.</li>
<li>
<strong>Use short, easy-to-understand names</strong>: You should
avoid naming your objects using single letters (e.g. ‘n’, ‘p’, etc.).
This is mostly to encourage you to use names that would make sense to
anyone reading your code (a colleague, or even yourself a year from
now). Also, avoiding excessively long names will make your code more
readable.</li>
<li>
<strong>Avoid commonly used names</strong>: There are several names
that may already have a definition in the R language (e.g. ‘mean’,
‘min’, ‘max’). One clue that a name already has meaning is that if you
start typing a name in RStudio and it gets a colored highlight or
RStudio gives you a suggested autocompletion you have chosen a name that
has a reserved meaning.</li>
<li>
<strong>Use the recommended assignment operator</strong>: In R, we
use ‘&lt;-’ as the preferred assignment operator. ‘=’ works too, but is
most commonly used in passing arguments to functions (more on functions
later). There is a shortcut for the R assignment operator:
<ul>
<li>Windows execution shortcut: <kbd>Alt</kbd>+<kbd>-</kbd>
</li>
<li>Mac execution shortcut: <kbd>Option</kbd>+<kbd>-</kbd>
</li>
</ul>
</li>
</ul>
<p>There are a few more suggestions about naming and style you may want
to learn more about as you write more R code. There are several “style
guides” that have advice, and one to start with is the <a href="https://style.tidyverse.org/index.html" class="external-link">tidyverse R style
guide</a>.</p>
<div id="tip-pay-attention-to-warnings-in-the-script-console" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="tip-pay-attention-to-warnings-in-the-script-console" class="callout-inner">
<h3 class="callout-title">Tip: Pay attention to warnings in the script
console<a class="anchor" aria-label="anchor" href="#tip-pay-attention-to-warnings-in-the-script-console"></a>
</h3>
<div class="callout-content">
<p>If you enter a line of code in your script that contains an error,
RStudio may give you an error message and underline this mistake.
Sometimes these messages are easy to understand, but often the messages
may need some figuring out. Paying attention to these warnings will help
you avoid mistakes. In the example below, our object name has a space,
which is not allowed in R. The error message does not say this directly,
but R is “not sure” about how to assign the name to “human_ chr_number”
when the object name we want is “human_chr_number”.</p>
<figure><img src="fig/rstudio_script_warning.png" alt="rstudio script warning" style="width: 600px;" class="figure mx-auto d-block"></figure>
</div>
</div>
</div>
</section><section id="reassigning-object-names-or-deleting-objects"><h2 class="section-heading">Reassigning object names or deleting objects<a class="anchor" aria-label="anchor" href="#reassigning-object-names-or-deleting-objects"></a>
</h2>
<hr class="half-width">
<p>Once an object has a value, you can change that value by overwriting
it. R will not give you a warning or error if you overwriting an object,
which may or may not be a good thing depending on how you look at
it.</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># gene_name has the value 'pten' or whatever value you used in the challenge.</span></span>
<span><span class="co"># We will now assign the new value 'tp53'</span></span>
<span><span class="va">gene_name</span> <span class="op">&lt;-</span> <span class="st">'tp53'</span></span></code></pre>
</div>
<p>You can also remove an object from R’s memory entirely. The
<code>rm()</code> function will delete the object.</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># delete the object 'gene_name'</span></span>
<span><span class="fu">rm</span><span class="op">(</span><span class="va">gene_name</span><span class="op">)</span></span></code></pre>
</div>
<p>If you run a line of code that has only an object name, R will
normally display the contents of that object. In this case, we are told
the object no longer exists.</p>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error: object 'gene_name' not found</code></pre>
</div>
</section><section id="understanding-object-data-types-modes"><h2 class="section-heading">Understanding object data types (modes)<a class="anchor" aria-label="anchor" href="#understanding-object-data-types-modes"></a>
</h2>
<hr class="half-width">
<p>In R, <strong>every object has two properties</strong>:</p>
<ul>
<li>
<strong>Length</strong>: How many distinct values are held in that
object</li>
<li>
<strong>Mode</strong>: What is the classification (type) of that
object.</li>
</ul>
<p>We will get to the “length” property later in the lesson. The
<strong>“mode” property</strong> <strong>corresponds to the type of data
an object represents</strong>. The most common modes you will encounter
in R are:</p>
<table class="table">
<colgroup>
<col width="7%">
<col width="92%">
</colgroup>
<thead><tr class="header">
<th>Mode (abbreviation)</th>
<th>Type of data</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Numeric (num)</td>
<td>Numbers such floating point/decimals (1.0, 0.5, 3.14), there are
also more specific numeric types (dbl - Double, int - Integer). These
differences are not relevant for most beginners and pertain to how these
values are stored in memory</td>
</tr>
<tr class="even">
<td>Character (chr)</td>
<td>A sequence of letters/numbers in single ’’ or double ” ” quotes</td>
</tr>
<tr class="odd">
<td>Logical</td>
<td>Boolean values - TRUE or FALSE</td>
</tr>
</tbody>
</table>
<p>There are a few other modes (i.e. “complex”, “raw” etc.) but these
are the three we will work with in this lesson.</p>
<p>Data types are familiar in many programming languages, but also in
natural language where we refer to them as the parts of speech,
e.g. nouns, verbs, adverbs, etc. Once you know if a word - perhaps an
unfamiliar one - is a noun, you can probably guess you can count it and
make it plural if there is more than one (e.g. 1 <a href="https://en.wikipedia.org/wiki/Tuatara" class="external-link">Tuatara</a>, or 2
Tuataras). If something is a adjective, you can usually change it into
an adverb by adding “-ly” (e.g. <a href="https://www.merriam-webster.com/dictionary/jejune" class="external-link">jejune</a> vs.
jejunely). Depending on the context, you may need to decide if a word is
in one category or another (e.g “cut” may be a noun when it’s on your
finger, or a verb when you are preparing vegetables). These concepts
have important analogies when working with R objects.</p>
<div id="exercise-create-objects-and-check-their-modes" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-create-objects-and-check-their-modes" class="callout-inner">
<h3 class="callout-title">Exercise: Create objects and check their
modes<a class="anchor" aria-label="anchor" href="#exercise-create-objects-and-check-their-modes"></a>
</h3>
<div class="callout-content">
<p>Create the following objects in R, then use the <code>mode()</code>
function to verify their modes. Try to guess what the mode will be
before you look at the solution</p>
<ol style="list-style-type: decimal">
<li><code>chromosome_name &lt;- 'chr02'</code></li>
<li><code>od_600_value &lt;- 0.47</code></li>
<li><code>chr_position &lt;- '1001701'</code></li>
<li><code>spock &lt;- TRUE</code></li>
<li><code>pilot &lt;- Earhart</code></li>
</ol>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'Earhart' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">mode</span><span class="op">(</span><span class="va">chromosome_name</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "character"</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">mode</span><span class="op">(</span><span class="va">od_600_value</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "numeric"</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">mode</span><span class="op">(</span><span class="va">chr_position</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "character"</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">mode</span><span class="op">(</span><span class="va">spock</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "logical"</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">mode</span><span class="op">(</span><span class="va">pilot</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'pilot' not found</code></pre>
</div>
</div>
</div>
</div>
</div>
<p>Notice from the solution that even if a series of numbers is given as
a value R will consider them to be in the “character” mode if they are
enclosed as single or double quotes. Also, notice that you cannot take a
string of alphanumeric characters (e.g. Earhart) and assign as a value
for an object. In this case, R looks for an object named
<code>Earhart</code> but since there is no object, no assignment can be
made. If <code>Earhart</code> did exist, then the mode of
<code>pilot</code> would be whatever the mode of <code>Earhart</code>
was originally. If we want to create an object called <code>pilot</code>
that was the <strong>name</strong> “Earhart”, we need to enclose
<code>Earhart</code> in quotation marks.</p>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pilot</span> <span class="op">&lt;-</span> <span class="st">"Earhart"</span></span>
<span><span class="fu">mode</span><span class="op">(</span><span class="va">pilot</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "character"</code></pre>
</div>
</section><section id="mathematical-and-functional-operations-on-objects"><h2 class="section-heading">Mathematical and functional operations on objects<a class="anchor" aria-label="anchor" href="#mathematical-and-functional-operations-on-objects"></a>
</h2>
<hr class="half-width">
<p>Once an object exists (which by definition also means it has a mode),
R can appropriately manipulate that object. For example, objects of the
numeric modes can be added, multiplied, divided, etc. R provides several
mathematical (arithmetic) operators including:</p>
<table class="table">
<colgroup>
<col width="7%">
<col width="92%">
</colgroup>
<thead><tr class="header">
<th>Operator</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>+</td>
<td>addition</td>
</tr>
<tr class="even">
<td>-</td>
<td>subtraction</td>
</tr>
<tr class="odd">
<td>*</td>
<td>multiplication</td>
</tr>
<tr class="even">
<td>/</td>
<td>division</td>
</tr>
<tr class="odd">
<td>^ or **</td>
<td>exponentiation</td>
</tr>
<tr class="even">
<td>a%/%b</td>
<td>integer division (division where the remainder is discarded)</td>
</tr>
<tr class="odd">
<td>a%%b</td>
<td>modulus (returns the remainder after division)</td>
</tr>
</tbody>
</table>
<p>These can be used with literal numbers:</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="fl">1</span> <span class="op">+</span> <span class="op">(</span><span class="fl">5</span> <span class="op">**</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span><span class="op">/</span><span class="fl">2</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 1.618034</code></pre>
</div>
<p>and importantly, can be used on any object that evaluates to
(i.e. interpreted by R) a numeric object:</p>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># multiply the object 'human_chr_number' by 2</span></span>
<span></span>
<span><span class="va">human_chr_number</span> <span class="op">*</span> <span class="fl">2</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 46</code></pre>
</div>
<div id="exercise-compute-the-golden-ratio" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-compute-the-golden-ratio" class="callout-inner">
<h3 class="callout-title">Exercise: Compute the golden ratio<a class="anchor" aria-label="anchor" href="#exercise-compute-the-golden-ratio"></a>
</h3>
<div class="callout-content">
<p>One approximation of the golden ratio (φ) can be found by taking the
sum of 1 and the square root of 5, and dividing by 2 as in the example
above. Compute the golden ratio to 3 digits of precision using the
<code>sqrt()</code> and <code>round()</code> functions. Hint: remember
the <code>round()</code> function can take 2 arguments.</p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" data-bs-parent="#accordionSolution3" aria-labelledby="headingSolution3">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb24">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">round</span><span class="op">(</span><span class="op">(</span><span class="fl">1</span> <span class="op">+</span> <span class="fu">sqrt</span><span class="op">(</span><span class="fl">5</span><span class="op">)</span><span class="op">)</span><span class="op">/</span><span class="fl">2</span>, digits <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 1.618</code></pre>
</div>
<p>Notice that you can place one function inside of another.</p>
</div>
</div>
</div>
</div>
</section><section id="vectors"><h2 class="section-heading">Vectors<a class="anchor" aria-label="anchor" href="#vectors"></a>
</h2>
<hr class="half-width">
<p>Vectors are probably the most used commonly used object type in R.
<strong>A vector is a collection of values that are all of the same type
(numbers, characters, etc.)</strong>. One of the most common ways to
create a vector is to use the <code>c()</code> function - the
“concatenate” or “combine” function. Inside the function you may enter
one or more values; for multiple values, separate each value with a
comma:</p>
<div class="codewrapper sourceCode" id="cb26">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create the SNP gene name vector</span></span>
<span></span>
<span><span class="va">snp_genes</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="st">"OXTR"</span>, <span class="st">"ACTN3"</span>, <span class="st">"AR"</span>, <span class="st">"OPRM1"</span><span class="op">)</span></span></code></pre>
</div>
<p>Vectors always have a <strong>mode</strong> and a
<strong>length</strong>. You can check these with the
<code>mode()</code> and <code>length()</code> functions respectively.
Another useful function that gives both of these pieces of information
is the <code>str()</code> (structure) function.</p>
<div class="codewrapper sourceCode" id="cb27">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Check the mode, length, and structure of 'snp_genes'</span></span>
<span><span class="fu">mode</span><span class="op">(</span><span class="va">snp_genes</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "character"</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb29">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">length</span><span class="op">(</span><span class="va">snp_genes</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 4</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb31">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">str</span><span class="op">(</span><span class="va">snp_genes</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> chr [1:4] "OXTR" "ACTN3" "AR" "OPRM1"</code></pre>
</div>
<p>Vectors are quite important in R. Another data type that we will work
with later in this lesson, data frames, are collections of vectors. What
we learn here about vectors will pay off even more when we start working
with data frames.</p>
</section><section id="creating-and-subsetting-vectors"><h2 class="section-heading">Creating and subsetting vectors<a class="anchor" aria-label="anchor" href="#creating-and-subsetting-vectors"></a>
</h2>
<hr class="half-width">
<p>Let’s create a few more vectors to play around with:</p>
<div class="codewrapper sourceCode" id="cb33">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Some interesting human SNPs</span></span>
<span><span class="co"># while accuracy is important, typos in the data won't hurt you here</span></span>
<span></span>
<span><span class="va">snps</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="st">"rs53576"</span>, <span class="st">"rs1815739"</span>, <span class="st">"rs6152"</span>, <span class="st">"rs1799971"</span><span class="op">)</span></span>
<span><span class="va">snp_chromosomes</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="st">"3"</span>, <span class="st">"11"</span>, <span class="st">"X"</span>, <span class="st">"6"</span><span class="op">)</span></span>
<span><span class="va">snp_positions</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="fl">8762685</span>, <span class="fl">66560624</span>, <span class="fl">67545785</span>, <span class="fl">154039662</span><span class="op">)</span></span></code></pre>
</div>
<p>Once we have vectors, one thing we may want to do is specifically
retrieve one or more values from our vector. To do so, we use
<strong>bracket notation</strong>. We type the name of the vector
followed by square brackets. In those square brackets we place the index
(e.g. a number) in that bracket as follows:</p>
<div class="codewrapper sourceCode" id="cb34">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># get the 3rd value in the snp vector</span></span>
<span><span class="va">snp</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'snp' not found</code></pre>
</div>
<p>In R, every item your vector is indexed, starting from the first item
(1) through to the final number of items in your vector. You can also
retrieve a range of numbers:</p>
<div class="codewrapper sourceCode" id="cb36">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># get the 1st through 3rd value in the snp vector</span></span>
<span></span>
<span><span class="va">snp</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'snp' not found</code></pre>
</div>
<p>If you want to retrieve several (but not necessarily sequential)
items from a vector, you pass a <strong>vector of indices</strong>; a
vector that has the numbered positions you wish to retrieve.</p>
<div class="codewrapper sourceCode" id="cb38">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># get the 1st, 3rd, and 4th value in the snp vector</span></span>
<span></span>
<span><span class="va">snp</span><span class="op">[</span><span class="fu">c</span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span>, <span class="fl">4</span><span class="op">)</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'snp' not found</code></pre>
</div>
<p>There are additional (and perhaps less commonly used) ways of
subsetting a vector (see <a href="https://thomasleeper.com/Rcourse/Tutorials/vectorindexing.html" class="external-link">these
examples</a>). Also, several of these subsetting expressions can be
combined:</p>
<div class="codewrapper sourceCode" id="cb40">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># get the 1st through the 3rd value, and 4th value in the snp vector</span></span>
<span><span class="co"># yes, this is a little silly in a vector of only 4 values.</span></span>
<span><span class="va">snp</span><span class="op">[</span><span class="fu">c</span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>,<span class="fl">4</span><span class="op">)</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'snp' not found</code></pre>
</div>
</section><section id="adding-to-removing-or-replacing-values-in-existing-vectors"><h2 class="section-heading">Adding to, removing, or replacing values in existing vectors<a class="anchor" aria-label="anchor" href="#adding-to-removing-or-replacing-values-in-existing-vectors"></a>
</h2>
<hr class="half-width">
<p>Once you have an existing vector, you may want to add a new item to
it. To do so, you can use the <code>c()</code> function again to add
your new value:</p>
<div class="codewrapper sourceCode" id="cb42">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># add the gene "CYP1A1" and "APOA5" to our list of snp genes</span></span>
<span><span class="co"># this overwrites our existing vector</span></span>
<span><span class="va">snp_genes</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="va">snp_genes</span>, <span class="st">"CYP1A1"</span>, <span class="st">"APOA5"</span><span class="op">)</span></span></code></pre>
</div>
<p>We can verify that “snp_genes” contains the new gene entry</p>
<div class="codewrapper sourceCode" id="cb43">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">snp_genes</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "OXTR"   "ACTN3"  "AR"     "OPRM1"  "CYP1A1" "APOA5" </code></pre>
</div>
<p>Using a negative index will return a version of a vector with that
index’s value removed:</p>
<div class="codewrapper sourceCode" id="cb45">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">snp_genes</span><span class="op">[</span><span class="op">-</span><span class="fl">6</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "OXTR"   "ACTN3"  "AR"     "OPRM1"  "CYP1A1"</code></pre>
</div>
<p>We can remove that value from our vector by overwriting it with this
expression:</p>
<div class="codewrapper sourceCode" id="cb47">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">snp_genes</span> <span class="op">&lt;-</span> <span class="va">snp_genes</span><span class="op">[</span><span class="op">-</span><span class="fl">6</span><span class="op">]</span></span>
<span><span class="va">snp_genes</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "OXTR"   "ACTN3"  "AR"     "OPRM1"  "CYP1A1"</code></pre>
</div>
<p>We can also explicitly rename or add a value to our index using
double bracket notation:</p>
<div class="codewrapper sourceCode" id="cb49">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">snp_genes</span><span class="op">[</span><span class="fl">6</span><span class="op">]</span><span class="op">&lt;-</span> <span class="st">"APOA5"</span></span>
<span><span class="va">snp_genes</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "OXTR"   "ACTN3"  "AR"     "OPRM1"  "CYP1A1" "APOA5" </code></pre>
</div>
<div id="exercise-examining-and-subsetting-vectors" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-examining-and-subsetting-vectors" class="callout-inner">
<h3 class="callout-title">Exercise: Examining and subsetting
vectors<a class="anchor" aria-label="anchor" href="#exercise-examining-and-subsetting-vectors"></a>
</h3>
<div class="callout-content">
<p>Answer the following questions to test your knowledge of vectors</p>
<p>Which of the following are true of vectors in R? A) All vectors have
a mode <strong>or</strong> a length<br>
B) All vectors have a mode <strong>and</strong> a length<br>
C) Vectors may have different lengths<br>
D) Items within a vector may be of different modes<br>
E) You can use the <code>c()</code> to add one or more items to an
existing vector<br>
F) You can use the <code>c()</code> to add a vector to an exiting
vector</p>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" data-bs-parent="#accordionSolution4" aria-labelledby="headingSolution4">
<div class="accordion-body">
<ol style="list-style-type: upper-alpha">
<li>False - Vectors have both of these properties<br>
</li>
<li>True<br>
</li>
<li>True<br>
</li>
<li>False - Vectors have only one mode (e.g. numeric, character); all
items in<br>
a vector must be of this mode.</li>
<li>True<br>
</li>
<li>True</li>
</ol>
</div>
</div>
</div>
</div>
</section><section id="logical-subsetting"><h2 class="section-heading">Logical Subsetting<a class="anchor" aria-label="anchor" href="#logical-subsetting"></a>
</h2>
<hr class="half-width">
<p>There is one last set of cool subsetting capabilities we want to
introduce. It is possible within R to retrieve items in a vector based
on a logical evaluation or numerical comparison. For example, let’s say
we wanted get all of the SNPs in our vector of SNP positions that were
greater than 100,000,000. We could index using the ‘&gt;’ (greater than)
logical operator:</p>
<div class="codewrapper sourceCode" id="cb51">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">snp_positions</span><span class="op">[</span><span class="va">snp_positions</span> <span class="op">&gt;</span> <span class="fl">100000000</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 154039662</code></pre>
</div>
<p>In the square brackets you place the name of the vector followed by
the comparison operator and (in this case) a numeric value. Some of the
most common logical operators you will use in R are:</p>
<table class="table">
<colgroup>
<col width="7%">
<col width="92%">
</colgroup>
<thead><tr class="header">
<th>Operator</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>&lt;</td>
<td>less than</td>
</tr>
<tr class="even">
<td>&lt;=</td>
<td>less than or equal to</td>
</tr>
<tr class="odd">
<td>&gt;</td>
<td>greater than</td>
</tr>
<tr class="even">
<td>&gt;=</td>
<td>greater than or equal to</td>
</tr>
<tr class="odd">
<td>==</td>
<td>exactly equal to</td>
</tr>
<tr class="even">
<td>!=</td>
<td>not equal to</td>
</tr>
<tr class="odd">
<td>!x</td>
<td>not x</td>
</tr>
<tr class="even">
<td>a</td>
<td>b</td>
</tr>
<tr class="odd">
<td>a &amp; b</td>
<td>a and b</td>
</tr>
</tbody>
</table>
<div id="the-magic-of-programming" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="the-magic-of-programming" class="callout-inner">
<h3 class="callout-title">The magic of programming<a class="anchor" aria-label="anchor" href="#the-magic-of-programming"></a>
</h3>
<div class="callout-content">
<p>The reason why the expression
<code>snp_positions[snp_positions &gt; 100000000]</code> works can be
better understood if you examine what the expression “snp_positions &gt;
100000000” evaluates to:</p>
<div class="codewrapper sourceCode" id="cb53">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">snp_positions</span> <span class="op">&gt;</span> <span class="fl">100000000</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] FALSE FALSE FALSE  TRUE</code></pre>
</div>
<p>The output above is a logical vector, the 4th element of which is
TRUE. When you pass a logical vector as an index, R will return the true
values:</p>
<div class="codewrapper sourceCode" id="cb55">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">snp_positions</span><span class="op">[</span><span class="fu">c</span><span class="op">(</span><span class="cn">FALSE</span>, <span class="cn">FALSE</span>, <span class="cn">FALSE</span>, <span class="cn">TRUE</span><span class="op">)</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 154039662</code></pre>
</div>
<p>If you have never coded before, this type of situation starts to
expose the “magic” of programming. We mentioned before that in the
bracket notation you take your named vector followed by brackets which
contain an index: <strong>named_vector[index]</strong>. The “magic” is
that the index needs to <em>evaluate to</em> a number. So, even if it
does not appear to be an integer (e.g. 1, 2, 3), as long as R can
evaluate it, we will get a result. That our expression
<code>snp_positions[snp_positions &gt; 100000000]</code> evaluates to a
number can be seen in the following situation. If you wanted to know
which <strong>index</strong> (1, 2, 3, or 4) in our vector of SNP
positions was the one that was greater than 100,000,000?</p>
<p>We can use the <code>which()</code> function to return the indices of
any item that evaluates as TRUE in our comparison:</p>
<div class="codewrapper sourceCode" id="cb57">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">which</span><span class="op">(</span><span class="va">snp_positions</span> <span class="op">&gt;</span> <span class="fl">100000000</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 4</code></pre>
</div>
<p><strong>Why this is important</strong></p>
<p>Often in programming we will not know what inputs and values will be
used when our code is executed. Rather than put in a pre-determined
value (e.g 100000000) we can use an object that can take on whatever
value we need. So for example:</p>
<div class="codewrapper sourceCode" id="cb59">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">snp_marker_cutoff</span> <span class="op">&lt;-</span> <span class="fl">100000000</span></span>
<span><span class="va">snp_positions</span><span class="op">[</span><span class="va">snp_positions</span> <span class="op">&gt;</span> <span class="va">snp_marker_cutoff</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 154039662</code></pre>
</div>
<p>Ultimately, it’s putting together flexible, reusable code like this
that gets at the “magic” of programming!</p>
</div>
</div>
</div>
</section><section id="a-few-final-vector-tricks"><h2 class="section-heading">A few final vector tricks<a class="anchor" aria-label="anchor" href="#a-few-final-vector-tricks"></a>
</h2>
<hr class="half-width">
<p>Finally, there are a few other common retrieve or replace operations
you may want to know about. First, you can check to see if any of the
values of your vector are missing (i.e. are <code>NA</code>, that stands
for <code>not avaliable</code>). Missing data will get a more detailed
treatment later, but the <code>is.NA()</code> function will return a
logical vector, with TRUE for any NA value:</p>
<div class="codewrapper sourceCode" id="cb61">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># current value of 'snp_genes': </span></span>
<span><span class="co"># chr [1:7] "OXTR" "ACTN3" "AR" "OPRM1" "CYP1A1" NA "APOA5"</span></span>
<span></span>
<span><span class="fu">is.na</span><span class="op">(</span><span class="va">snp_genes</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] FALSE FALSE FALSE FALSE FALSE FALSE</code></pre>
</div>
<p>Sometimes, you may wish to find out if a specific value (or several
values) is present a vector. You can do this using the comparison
operator <code>%in%</code>, which will return TRUE for any value in your
collection that is in the vector you are searching:</p>
<div class="codewrapper sourceCode" id="cb63">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># current value of 'snp_genes':</span></span>
<span><span class="co"># chr [1:7] "OXTR" "ACTN3" "AR" "OPRM1" "CYP1A1" NA "APOA5"</span></span>
<span></span>
<span><span class="co"># test to see if "ACTN3" or "APO5A" is in the snp_genes vector</span></span>
<span><span class="co"># if you are looking for more than one value, you must pass this as a vector</span></span>
<span></span>
<span><span class="fu">c</span><span class="op">(</span><span class="st">"ACTN3"</span>,<span class="st">"APOA5"</span><span class="op">)</span> <span class="op">%in%</span> <span class="va">snp_genes</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] TRUE TRUE</code></pre>
</div>
<div id="review-exercise-1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="review-exercise-1" class="callout-inner">
<h3 class="callout-title">Review Exercise 1<a class="anchor" aria-label="anchor" href="#review-exercise-1"></a>
</h3>
<div class="callout-content">
<p>What data modes are the following vectors? a. <code>snps</code><br>
b. <code>snp_chromosomes</code><br>
c. <code>snp_positions</code></p>
</div>
</div>
</div>
<div id="accordionSolution5" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution5" aria-expanded="false" aria-controls="collapseSolution5">
  <h4 class="accordion-header" id="headingSolution5">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution5" class="accordion-collapse collapse" data-bs-parent="#accordionSolution5" aria-labelledby="headingSolution5">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb65">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">mode</span><span class="op">(</span><span class="va">snps</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "character"</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb67">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">mode</span><span class="op">(</span><span class="va">snp_chromosomes</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "character"</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb69">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">mode</span><span class="op">(</span><span class="va">snp_positions</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "numeric"</code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="review-exercise-2" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="review-exercise-2" class="callout-inner">
<h3 class="callout-title">Review Exercise 2<a class="anchor" aria-label="anchor" href="#review-exercise-2"></a>
</h3>
<div class="callout-content">
<p>Add the following values to the specified vectors: a. To the
<code>snps</code> vector add: “rs662799”<br>
b. To the <code>snp_chromosomes</code> vector add: 11<br>
c. To the <code>snp_positions</code> vector add: 116792991</p>
</div>
</div>
</div>
<div id="accordionSolution6" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution6" aria-expanded="false" aria-controls="collapseSolution6">
  <h4 class="accordion-header" id="headingSolution6">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution6" class="accordion-collapse collapse" data-bs-parent="#accordionSolution6" aria-labelledby="headingSolution6">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb71">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">snps</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="va">snps</span>, <span class="st">"rs662799"</span><span class="op">)</span></span>
<span><span class="va">snps</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "rs53576"   "rs1815739" "rs6152"    "rs1799971" "rs662799" </code></pre>
</div>
<div class="codewrapper sourceCode" id="cb73">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">snp_chromosomes</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="va">snp_chromosomes</span>, <span class="st">"11"</span><span class="op">)</span> <span class="co"># did you use quotes?</span></span>
<span><span class="va">snp_chromosomes</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "3"  "11" "X"  "6"  "11"</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb75">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">snp_positions</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="va">snp_positions</span>, <span class="fl">116792991</span><span class="op">)</span></span>
<span><span class="va">snp_positions</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1]   8762685  66560624  67545785 154039662 116792991</code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="review-exercise-3" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="review-exercise-3" class="callout-inner">
<h3 class="callout-title">Review Exercise 3<a class="anchor" aria-label="anchor" href="#review-exercise-3"></a>
</h3>
<div class="callout-content">
<p>Make the following change to the <code>snp_genes</code> vector:</p>
<p>Hint: Your vector should look like this in ‘Environment’:
<code>chr [1:7] "OXTR" "ACTN3" "AR" "OPRM1" "CYP1A1" NA "APOA5"</code>.
If not recreate the vector by running this expression:
<code>snp_genes &lt;- c("OXTR", "ACTN3", "AR", "OPRM1", "CYP1A1", NA, "APOA5")</code></p>
<ol style="list-style-type: lower-alpha">
<li>Create a new version of <code>snp_genes</code> that does not contain
CYP1A1 and then<br>
</li>
<li>Add 2 NA values to the end of <code>snp_genes</code>
</li>
</ol>
</div>
</div>
</div>
<div id="accordionSolution7" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution7" aria-expanded="false" aria-controls="collapseSolution7">
  <h4 class="accordion-header" id="headingSolution7">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution7" class="accordion-collapse collapse" data-bs-parent="#accordionSolution7" aria-labelledby="headingSolution7">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb77">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">snp_genes</span> <span class="op">&lt;-</span> <span class="va">snp_genes</span><span class="op">[</span><span class="op">-</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="va">snp_genes</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="va">snp_genes</span>, <span class="cn">NA</span>, <span class="cn">NA</span><span class="op">)</span></span>
<span><span class="va">snp_genes</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "OXTR"  "ACTN3" "AR"    "OPRM1" "APOA5" NA      NA     </code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="review-exercise-4" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="review-exercise-4" class="callout-inner">
<h3 class="callout-title">Review Exercise 4<a class="anchor" aria-label="anchor" href="#review-exercise-4"></a>
</h3>
<div class="callout-content">
<p>Using indexing, create a new vector named <code>combined</code> that
contains:</p>
<ul>
<li>The the 1st value in <code>snp_genes</code>
</li>
<li>The 1st value in <code>snps</code>
</li>
<li>The 1st value in <code>snp_chromosomes</code>
</li>
<li>The 1st value in <code>snp_positions</code>
</li>
</ul>
</div>
</div>
</div>
<div id="accordionSolution8" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution8" aria-expanded="false" aria-controls="collapseSolution8">
  <h4 class="accordion-header" id="headingSolution8">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution8" class="accordion-collapse collapse" data-bs-parent="#accordionSolution8" aria-labelledby="headingSolution8">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb79">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">combined</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="va">snp_genes</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="va">snps</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="va">snp_chromosomes</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="va">snp_positions</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">combined</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "OXTR"    "rs53576" "3"       "8762685"</code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="review-exercise-5" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="review-exercise-5" class="callout-inner">
<h3 class="callout-title">Review Exercise 5<a class="anchor" aria-label="anchor" href="#review-exercise-5"></a>
</h3>
<div class="callout-content">
<p>What type of data is <code>combined</code>?</p>
</div>
</div>
</div>
<div id="accordionSolution9" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution9" aria-expanded="false" aria-controls="collapseSolution9">
  <h4 class="accordion-header" id="headingSolution9">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution9" class="accordion-collapse collapse" data-bs-parent="#accordionSolution9" aria-labelledby="headingSolution9">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb81">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">typeof</span><span class="op">(</span><span class="va">combined</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "character"</code></pre>
</div>
</div>
</div>
</div>
</div>
</section><section id="bonus-material-lists"><h2 class="section-heading">Bonus material: Lists<a class="anchor" aria-label="anchor" href="#bonus-material-lists"></a>
</h2>
<hr class="half-width">
<p>Lists are quite useful in R, but we won’t be using them in the
genomics lessons. That said, you may come across lists in the way that
some bioinformatics programs may store and/or return data to you. One of
the key attributes of a list is that, unlike a vector, a list may
contain data of more than one mode. Learn more about creating and using
lists using this <a href="https://r4ds.had.co.nz/vectors.html#lists" class="external-link">nice tutorial</a>. In
this one example, we will create a named list and show you how to
retrieve items from the list.</p>
<div class="codewrapper sourceCode" id="cb83">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create a named list using the 'list' function and our SNP examples</span></span>
<span><span class="co"># Note, for easy reading we have placed each item in the list on a separate line</span></span>
<span><span class="co"># Nothing special about this, you can do this for any multiline commands</span></span>
<span><span class="co"># To run this command, make sure the entire command (all 4 lines) are highlighted</span></span>
<span><span class="co"># before running</span></span>
<span><span class="co"># Note also, as we are doing all this inside the list() function use of the</span></span>
<span><span class="co"># '=' sign is good style</span></span>
<span><span class="va">snp_data</span> <span class="op">&lt;-</span> <span class="fu">list</span><span class="op">(</span>genes <span class="op">=</span> <span class="va">snp_genes</span>,</span>
<span>                 refference_snp <span class="op">=</span> <span class="va">snps</span>,</span>
<span>                 chromosome <span class="op">=</span> <span class="va">snp_chromosomes</span>,</span>
<span>                 position <span class="op">=</span> <span class="va">snp_positions</span><span class="op">)</span></span>
<span><span class="co"># Examine the structure of the list</span></span>
<span><span class="fu">str</span><span class="op">(</span><span class="va">snp_data</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>List of 4
 $ genes         : chr [1:7] "OXTR" "ACTN3" "AR" "OPRM1" ...
 $ refference_snp: chr [1:5] "rs53576" "rs1815739" "rs6152" "rs1799971" ...
 $ chromosome    : chr [1:5] "3" "11" "X" "6" ...
 $ position      : num [1:5] 8.76e+06 6.66e+07 6.75e+07 1.54e+08 1.17e+08</code></pre>
</div>
<p>To get all the values for the <code>position</code> object in the
list, we use the <code>$</code> notation:</p>
<div class="codewrapper sourceCode" id="cb85">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># return all the values of position object</span></span>
<span></span>
<span><span class="va">snp_data</span><span class="op">$</span><span class="va">position</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1]   8762685  66560624  67545785 154039662 116792991</code></pre>
</div>
<p>To get the first value in the <code>position</code> object, use the
<code>[]</code> notation to index:</p>
<div class="codewrapper sourceCode" id="cb87">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># return first value of the position object</span></span>
<span></span>
<span><span class="va">snp_data</span><span class="op">$</span><span class="va">position</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 8762685</code></pre>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Keypoints<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>Effectively using R is a journey of months or years. Still you don’t
have to be an expert to use R and you can start using and analyzing your
data with with about a day’s worth of training</li>
<li>It is important to understand how data are organized by R in a given
object type and how the mode of that type (e.g. numeric, character,
logical, etc.) will determine how R will operate on that data.</li>
<li>Working with vectors effectively prepares you for understanding how
data are organized in R.</li>
</ul>
</div>
</div>
</div>
</section></section><section id="aio-02-data-prelude"><p>Content from <a href="02-data-prelude.html">Introduction to the example dataset and file type</a></p>
<hr>
<p> Last updated on 2023-05-01 | 
        
        <a href="https://github.com/datacarpentry/genomics-r-intro/edit/main/episodes/02-data-prelude.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What data are we using in the lesson?</li>
<li>What are VCF files?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Know what the example dataset represents</li>
<li>Know the concepts of how VCF files are generated</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section id="preface"><h2 class="section-heading">Preface<a class="anchor" aria-label="anchor" href="#preface"></a>
</h2>
<hr class="half-width">
<p>The Intro to R and RStudio for Genomics is a part of the Genomics
Data Carpentry lessons. In this lesson we will learn the necessary skill
sets for R and RStudio and apply them directly to a real next-generation
sequencing (NGS) data in the variant calling format (VCF) file type.
Previous Genomics Data Carpentry lessons teach learners how to generate
a VCF file from FASTQ files downloaded from NCBI Sequence Read Archive
(SRA), so we won’t cover that here. Instead, in this episode we will
give a brief overview of the data and a what VCF file types are for
those who wish to teach the Intro to R and RStudio for Genomics lesson
independently of the Genomics Data Carpentry lessons.</p>
<p>This dataset was selected for several reasons, including:</p>
<ul>
<li>Simple, but iconic NGS-problem: Examine a population where we want
to characterize changes in sequence <em>a priori</em>
</li>
<li>Dataset publicly available - in this case through the NCBI SRA (<a href="https://www.ncbi.nlm.nih.gov/sra" class="external-link">http://www.ncbi.nlm.nih.gov/sra</a>)</li>
</ul></section><section id="introduction-to-the-dataset"><h2 class="section-heading">Introduction to the dataset<a class="anchor" aria-label="anchor" href="#introduction-to-the-dataset"></a>
</h2>
<hr class="half-width">
<p>Microbes are ideal organisms for exploring ‘Long-term Evolution
Experiments’ (LTEEs) - thousands of generations can be generated and
stored in a way that would be virtually impossible for more complex
eukaryotic systems. In <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4988878/" class="external-link">Tenaillon
et al 2016</a>, 12 populations of <em>Escherichia coli</em> were
propagated for more than 50,000 generations in a glucose-limited minimal
medium. This medium was supplemented with citrate which <em>E. coli</em>
cannot metabolize in the aerobic conditions of the experiment.
Sequencing of the populations at regular time points reveals that
spontaneous citrate-using mutants (Cit+) appeared in a population of
<em>E.coli</em> (designated Ara-3) at around 31,000 generations. It
should be noted that spontaneous Cit+ mutants are extraordinarily rare -
inability to metabolize citrate is one of the defining characters of the
<em>E. coli</em> species. Eventually, Cit+ mutants became the dominant
population as the experimental growth medium contained a high
concentration of citrate relative to glucose. Around the same time that
this mutation emerged, another phenotype become prominent in the Ara-3
population. Many <em>E. coli</em> began to develop excessive numbers of
mutations, meaning they became hypermutable.</p>
<p>Strains from generation 0 to generation 50,000 were sequenced,
including ones that were both Cit+ and Cit- and hypermutable in later
generations.</p>
<p>For the purposes of this workshop we’re going to be working with 3 of
the sequence reads from this experiment.</p>
<table class="table">
<colgroup>
<col width="3%">
<col width="82%">
<col width="2%">
<col width="1%">
<col width="3%">
<col width="2%">
<col width="4%">
</colgroup>
<thead><tr class="header">
<th>SRA Run Number</th>
<th>Clone</th>
<th>Generation</th>
<th>Cit</th>
<th>Hypermutable</th>
<th>Read Length</th>
<th>Sequencing Depth</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>SRR2589044</td>
<td>REL2181A</td>
<td>5,000</td>
<td>Unknown</td>
<td>None</td>
<td>150</td>
<td>60.2</td>
</tr>
<tr class="even">
<td>SRR2584863</td>
<td>REL7179B</td>
<td>15,000</td>
<td>Unknown</td>
<td>None</td>
<td>150</td>
<td>88</td>
</tr>
<tr class="odd">
<td>SRR2584866</td>
<td>REL11365</td>
<td>50,000</td>
<td>Cit+</td>
<td>plus</td>
<td>150</td>
<td>138.3</td>
</tr>
</tbody>
</table>
<p>We want to be able to look at differences in mutation rates between
hypermutable and non-hypermutable strains. We also want to analyze the
sequences to figure out what changes occurred in genomes to make the
strains Cit+. Ultimately, we will use R to answer the questions:</p>
<ul>
<li>How many base pair changes are there between the Cit+ and Cit-
strains?</li>
<li>What are the base pair changes between strains?</li>
</ul></section><section id="how-vcf-files-are-generated"><h2 class="section-heading">How VCF files are generated<a class="anchor" aria-label="anchor" href="#how-vcf-files-are-generated"></a>
</h2>
<hr class="half-width">
<p>Publicly accessible sequencing files in FASTQ formats can be
downloaded from NCBI SRA. However, at FASTQ files contain unaligned
sequences of varying quality, and requires clean up and alignment steps
for variants to be called from the reference genome.</p>
<p>Five steps are taken to transform FASTQ files to variant calls
contained in VCF files and at each step, specialized non-R based
bioinformatics tools that are used:</p>
<center>
<img src="fig/variant_calling_workflow.png" alt="variant calling workflow. Sequence reads (FASTQ files), Quality control (FASTQ files), Alignment to Genome (SAM/BAM files), Alignment cleanup (BAM file ready for variant calling), Variant Calling (VCF file)" style="width: 250px;" class="figure">
</center>
</section><section id="how-variant-calls-are-stored-in-vcf-files"><h2 class="section-heading">How variant calls are stored in VCF files<a class="anchor" aria-label="anchor" href="#how-variant-calls-are-stored-in-vcf-files"></a>
</h2>
<hr class="half-width">
<p>VCF files contain variants that were called against a reference
genome. These files are slightly more complicated than regular tables
you can open using programs like Excel and contain two sections: header
and records.</p>
<p>Below you will see the header (which describes the format), the time
and date the file was created, the version of bcftools that was used,
the command line parameters used, and some additional information:</p>
<pre><code><span><span class="co">##fileformat=VCFv4.2</span></span>
<span><span class="co">##FILTER=&lt;ID=PASS,Description="All filters passed"&gt;</span></span>
<span><span class="co">##bcftoolsVersion=1.8+htslib-1.8</span></span>
<span><span class="co">##bcftoolsCommand=mpileup -O b -o results/bcf/SRR2584866_raw.bcf -f data/ref_genome/ecoli_rel606.fasta results/bam/SRR2584866.aligned.sorted.bam</span></span>
<span><span class="co">##reference=file://data/ref_genome/ecoli_rel606.fasta</span></span>
<span><span class="co">##contig=&lt;ID=CP000819.1,length=4629812&gt;</span></span>
<span><span class="co">##ALT=&lt;ID=*,Description="Represents allele(s) other than observed."&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=INDEL,Number=0,Type=Flag,Description="Indicates that the variant is an INDEL."&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=IDV,Number=1,Type=Integer,Description="Maximum number of reads supporting an indel"&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=IMF,Number=1,Type=Float,Description="Maximum fraction of reads supporting an indel"&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=DP,Number=1,Type=Integer,Description="Raw read depth"&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=VDB,Number=1,Type=Float,Description="Variant Distance Bias for filtering splice-site artefacts in RNA-seq data (bigger is better)",Version=</span></span>
<span><span class="co">##INFO=&lt;ID=RPB,Number=1,Type=Float,Description="Mann-Whitney U test of Read Position Bias (bigger is better)"&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=MQB,Number=1,Type=Float,Description="Mann-Whitney U test of Mapping Quality Bias (bigger is better)"&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=BQB,Number=1,Type=Float,Description="Mann-Whitney U test of Base Quality Bias (bigger is better)"&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=MQSB,Number=1,Type=Float,Description="Mann-Whitney U test of Mapping Quality vs Strand Bias (bigger is better)"&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=SGB,Number=1,Type=Float,Description="Segregation based metric."&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=MQ0F,Number=1,Type=Float,Description="Fraction of MQ0 reads (smaller is better)"&gt;</span></span>
<span><span class="co">##FORMAT=&lt;ID=PL,Number=G,Type=Integer,Description="List of Phred-scaled genotype likelihoods"&gt;</span></span>
<span><span class="co">##FORMAT=&lt;ID=GT,Number=1,Type=String,Description="Genotype"&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=ICB,Number=1,Type=Float,Description="Inbreeding Coefficient Binomial test (bigger is better)"&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=HOB,Number=1,Type=Float,Description="Bias in the number of HOMs number (smaller is better)"&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=AC,Number=A,Type=Integer,Description="Allele count in genotypes for each ALT allele, in the same order as listed"&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=AN,Number=1,Type=Integer,Description="Total number of alleles in called genotypes"&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=DP4,Number=4,Type=Integer,Description="Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases"&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=MQ,Number=1,Type=Integer,Description="Average mapping quality"&gt;</span></span>
<span><span class="co">##bcftools_callVersion=1.8+htslib-1.8</span></span>
<span><span class="co">##bcftools_callCommand=call --ploidy 1 -m -v -o results/bcf/SRR2584866_variants.vcf results/bcf/SRR2584866_raw.bcf; Date=Tue Oct  9 18:48:10 2018</span></span></code></pre>
<p>Followed by information on each of the variations observed:</p>
<pre><code>#CHROM  POS     ID      REF     ALT     QUAL    FILTER  INFO    FORMAT  results/bam/SRR2584866.aligned.sorted.bam
CP000819.1      1521    .       C       T       207     .       DP=9;VDB=0.993024;SGB=-0.662043;MQSB=0.974597;MQ0F=0;AC=1;AN=1;DP4=0,0,4,5;MQ=60
CP000819.1      1612    .       A       G       225     .       DP=13;VDB=0.52194;SGB=-0.676189;MQSB=0.950952;MQ0F=0;AC=1;AN=1;DP4=0,0,6,5;MQ=60
CP000819.1      9092    .       A       G       225     .       DP=14;VDB=0.717543;SGB=-0.670168;MQSB=0.916482;MQ0F=0;AC=1;AN=1;DP4=0,0,7,3;MQ=60
CP000819.1      9972    .       T       G       214     .       DP=10;VDB=0.022095;SGB=-0.670168;MQSB=1;MQ0F=0;AC=1;AN=1;DP4=0,0,2,8;MQ=60      GT:PL
CP000819.1      10563   .       G       A       225     .       DP=11;VDB=0.958658;SGB=-0.670168;MQSB=0.952347;MQ0F=0;AC=1;AN=1;DP4=0,0,5,5;MQ=60
CP000819.1      22257   .       C       T       127     .       DP=5;VDB=0.0765947;SGB=-0.590765;MQSB=1;MQ0F=0;AC=1;AN=1;DP4=0,0,2,3;MQ=60      GT:PL
CP000819.1      38971   .       A       G       225     .       DP=14;VDB=0.872139;SGB=-0.680642;MQSB=1;MQ0F=0;AC=1;AN=1;DP4=0,0,4,8;MQ=60      GT:PL
CP000819.1      42306   .       A       G       225     .       DP=15;VDB=0.969686;SGB=-0.686358;MQSB=1;MQ0F=0;AC=1;AN=1;DP4=0,0,5,9;MQ=60      GT:PL
CP000819.1      45277   .       A       G       225     .       DP=15;VDB=0.470998;SGB=-0.680642;MQSB=0.95494;MQ0F=0;AC=1;AN=1;DP4=0,0,7,5;MQ=60
CP000819.1      56613   .       C       G       183     .       DP=12;VDB=0.879703;SGB=-0.676189;MQSB=1;MQ0F=0;AC=1;AN=1;DP4=0,0,8,3;MQ=60      GT:PL
CP000819.1      62118   .       A       G       225     .       DP=19;VDB=0.414981;SGB=-0.691153;MQSB=0.906029;MQ0F=0;AC=1;AN=1;DP4=0,0,8,10;MQ=59
CP000819.1      64042   .       G       A       225     .       DP=18;VDB=0.451328;SGB=-0.689466;MQSB=1;MQ0F=0;AC=1;AN=1;DP4=0,0,7,9;MQ=60      GT:PL</code></pre>
<p>The first few columns represent the information we have about a
predicted variation.</p>
<table class="table">
<colgroup>
<col width="4%">
<col width="95%">
</colgroup>
<thead><tr class="header">
<th>column</th>
<th>info</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>CHROM</td>
<td>contig location where the variation occurs</td>
</tr>
<tr class="even">
<td>POS</td>
<td>position within the contig where the variation occurs</td>
</tr>
<tr class="odd">
<td>ID</td>
<td>a <code>.</code> until we add annotation information</td>
</tr>
<tr class="even">
<td>REF</td>
<td>reference genotype (forward strand)</td>
</tr>
<tr class="odd">
<td>ALT</td>
<td>sample genotype (forward strand)</td>
</tr>
<tr class="even">
<td>QUAL</td>
<td>Phred-scaled probability that the observed variant exists at this
site (higher is better)</td>
</tr>
<tr class="odd">
<td>FILTER</td>
<td>a <code>.</code> if no quality filters have been applied, PASS if a
filter is passed, or the name of the filters this variant failed</td>
</tr>
</tbody>
</table>
<p>In an ideal world, the information in the <code>QUAL</code> column
would be all we needed to filter out bad variant calls. However, in
reality we need to filter on multiple other metrics.</p>
<p>The last two columns contain the genotypes and can be tricky to
decode.</p>
<table class="table">
<colgroup>
<col width="4%">
<col width="95%">
</colgroup>
<thead><tr class="header">
<th>column</th>
<th>info</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>FORMAT</td>
<td>lists in order the metrics presented in the final column</td>
</tr>
<tr class="even">
<td>results</td>
<td>lists the values associated with those metrics in order</td>
</tr>
</tbody>
</table>
<p>For our file, the metrics presented are GT:PL:GQ.</p>
<table class="table">
<colgroup>
<col width="4%">
<col width="95%">
</colgroup>
<thead><tr class="header">
<th>metric</th>
<th>definition</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>AD, DP</td>
<td>the depth per allele by sample and coverage</td>
</tr>
<tr class="even">
<td>GT</td>
<td>the genotype for the sample at this loci. For a diploid organism,
the GT field indicates the two alleles carried by the sample, encoded by
a 0 for the REF allele, 1 for the first ALT allele, 2 for the second ALT
allele, etc. A 0/0 means homozygous reference, 0/1 is heterozygous, and
1/1 is homozygous for the alternate allele.</td>
</tr>
<tr class="odd">
<td>PL</td>
<td>the likelihoods of the given genotypes</td>
</tr>
<tr class="even">
<td>GQ</td>
<td>the Phred-scaled confidence for the genotype</td>
</tr>
</tbody>
</table>
<p>For more information on VCF files visit The Broad Institute’s <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360035531692-VCF-Variant-Call-Format" class="external-link">VCF
guide</a>.</p>
</section><section id="references"><h2 class="section-heading">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<hr class="half-width">
<p>Tenaillon O, Barrick JE, Ribeck N, Deatherage DE, Blanchard JL,
Dasgupta A, Wu GC, Wielgoss S, Cruveiller S, Médigue C, Schneider D,
Lenski RE. Tempo and mode of genome evolution in a 50,000-generation
experiment (2016) Nature. 536(7615): 165–170. <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4988878/" class="external-link">Paper</a>,
<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4988878/#" class="external-link">Supplemental
materials</a> Data on NCBI SRA: <a href="https://trace.ncbi.nlm.nih.gov/Traces/sra/?study=SRP064605" class="external-link">https://trace.ncbi.nlm.nih.gov/Traces/sra/?study=SRP064605</a>
Data on EMBL-EBI ENA: <a href="https://www.ebi.ac.uk/ena/data/view/PRJNA295606" class="external-link">https://www.ebi.ac.uk/ena/data/view/PRJNA295606</a></p>
<p>This episode was adapted from the Data Carpentry Genomic lessons:</p>
<ul>
<li><a href="https://datacarpentry.org/organization-genomics/data/" class="external-link">Project
organization and management for Genomics</a></li>
<li><a href="https://datacarpentry.org/wrangling-genomics/04-variant_calling/index.html" class="external-link">Data
wrangling and processing for genomics</a></li>
</ul>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Keypoints<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>The dataset comes from a real world experiment in E. coli.</li>
<li>Publicly available FASTQ files can be downloaded from NCBI SRA.</li>
<li>Several steps are taken outside of R/RStudio to create VCF files
from FASTQ files.</li>
<li>VCF files store variant calls in a special format.</li>
</ul>
</div>
</div>
</div>
</section></section><section id="aio-03-basics-factors-dataframes"><p>Content from <a href="03-basics-factors-dataframes.html">R Basics continued - factors and data frames</a></p>
<hr>
<p> Last updated on 2023-05-01 | 
        
        <a href="https://github.com/datacarpentry/genomics-r-intro/edit/main/episodes/03-basics-factors-dataframes.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How do I get started with tabular data (e.g. spreadsheets) in
R?</li>
<li>What are some best practices for reading data into R?</li>
<li>How do I save tabular data generated in R?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Explain the basic principle of tidy datasets</li>
<li>Be able to load a tabular dataset using base R functions</li>
<li>Be able to determine the structure of a data frame including its
dimensions and the datatypes of variables</li>
<li>Be able to subset/retrieve values from a data frame</li>
<li>Understand how R may coerce data into different modes</li>
<li>Be able to change the mode of an object</li>
<li>Understand that R uses factors to store and manipulate categorical
data</li>
<li>Be able to manipulate a factor, including subsetting and
reordering</li>
<li>Be able to apply an arithmetic function to a data frame</li>
<li>Be able to coerce the class of an object (including variables in a
data frame)</li>
<li>Be able to import data from Excel</li>
<li>Be able to save a data frame as a delimited file</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section id="working-with-spreadsheets-tabular-data"><h2 class="section-heading">Working with spreadsheets (tabular data)<a class="anchor" aria-label="anchor" href="#working-with-spreadsheets-tabular-data"></a>
</h2>
<hr class="half-width">
<p>A substantial amount of the data we work with in genomics will be
tabular data, this is data arranged in rows and columns - also known as
spreadsheets. We could write a whole lesson on how to work with
spreadsheets effectively (<a href="https://datacarpentry.org/organization-genomics/" class="external-link">actually we
did</a>). For our purposes, we want to remind you of a few principles
before we work with our first set of example data:</p>
<p><strong>1) Keep raw data separate from analyzed data</strong></p>
<p>This is principle number one because if you can’t tell which files
are the original raw data, you risk making some serious mistakes
(e.g. drawing conclusion from data which have been manipulated in some
unknown way).</p>
<p><strong>2) Keep spreadsheet data Tidy</strong></p>
<p>The simplest principle of <strong>Tidy data</strong> is that we have
one row in our spreadsheet for each observation or sample, and one
column for every variable that we measure or report on. As simple as
this sounds, it’s very easily violated. Most data scientists agree that
significant amounts of their time is spent tidying data for analysis.
Read more about data organization in <a href="https://datacarpentry.org/organization-genomics/" class="external-link">our lesson</a>
and in <a href="https://www.jstatsoft.org/article/view/v059i10" class="external-link">this
paper</a>.</p>
<p><strong>3) Trust but verify</strong></p>
<p>Finally, while you don’t need to be paranoid about data, you should
have a plan for how you will prepare it for analysis. <strong>This a
focus of this lesson.</strong> You probably already have a lot of
intuition, expectations, assumptions about your data - the range of
values you expect, how many values should have been recorded, etc. Of
course, as the data get larger our human ability to keep track will
start to fail (and yes, it can fail for small data sets too). R will
help you to examine your data so that you can have greater confidence in
your analysis, and its reproducibility.</p>
<div id="tip-keeping-you-raw-data-separate" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="tip-keeping-you-raw-data-separate" class="callout-inner">
<h3 class="callout-title">Tip: Keeping you raw data separate<a class="anchor" aria-label="anchor" href="#tip-keeping-you-raw-data-separate"></a>
</h3>
<div class="callout-content">
<p>When you work with data in R, you are not changing the original file
you loaded that data from. This is different than (for example) working
with a spreadsheet program where changing the value of the cell leaves
you one “save”-click away from overwriting the original file. You have
to purposely use a writing function (e.g. <code>write.csv()</code>) to
save data loaded into R. In that case, be sure to save the manipulated
data into a new file. More on this later in the lesson.</p>
</div>
</div>
</div>
</section><section id="importing-tabular-data-into-r"><h2 class="section-heading">Importing tabular data into R<a class="anchor" aria-label="anchor" href="#importing-tabular-data-into-r"></a>
</h2>
<hr class="half-width">
<p>There are several ways to import data into R. For our purpose here,
we will focus on using the tools every R installation comes with (so
called “base” R) to import a comma-delimited file containing the results
of our variant calling workflow. We will need to load the sheet using a
function called <code>read.csv()</code>.</p>
<div id="exercise-review-the-arguments-of-the-read.csv-function" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-review-the-arguments-of-the-read.csv-function" class="callout-inner">
<h3 class="callout-title">Exercise: Review the arguments of the
<code>read.csv()</code> function<a class="anchor" aria-label="anchor" href="#exercise-review-the-arguments-of-the-read.csv-function"></a>
</h3>
<div class="callout-content">
<p><strong>Before using the <code>read.csv()</code> function, use R’s
help feature to answer the following questions</strong>.</p>
<p><em>Hint</em>: Entering ‘?’ before the function name and then running
that line will bring up the help documentation. Also, when reading this
particular help be careful to pay attention to the ‘read.csv’ expression
under the ‘Usage’ heading. Other answers will be in the ‘Arguments’
heading.</p>
<ol style="list-style-type: upper-alpha">
<li><p>What is the default parameter for ‘header’ in the
<code>read.csv()</code> function?</p></li>
<li><p>What argument would you have to change to read a file that was
delimited by semicolons (;) rather than commas?</p></li>
<li><p>What argument would you have to change to read file in which
numbers used commas for decimal separation (i.e. 1,00)?</p></li>
<li><p>What argument would you have to change to read in only the first
10,000 rows of a very large file?</p></li>
</ol>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<ol style="list-style-type: upper-alpha">
<li><p>The <code>read.csv()</code> function has the argument ‘header’
set to TRUE by default, this means the function always assumes the first
row is header information, (i.e. column names)</p></li>
<li><p>The <code>read.csv()</code> function has the argument ‘sep’ set
to “,”. This means the function assumes commas are used as delimiters,
as you would expect. Changing this parameter (e.g. <code>sep=";"</code>)
would now interpret semicolons as delimiters.</p></li>
<li><p>Although it is not listed in the <code>read.csv()</code> usage,
<code>read.csv()</code> is a “version” of the function
<code>read.table()</code> and accepts all its arguments. If you set
<code>dec=","</code> you could change the decimal operator. We’d
probably assume the delimiter is some other character.</p></li>
<li><p>You can set <code>nrow</code> to a numeric value
(e.g. <code>nrow=10000</code>) to choose how many rows of a file you
read in. This may be useful for very large files where not all the data
is needed to test some data cleaning steps you are applying.</p></li>
</ol>
<p>Hopefully, this exercise gets you thinking about using the provided
help documentation in R. There are many arguments that exist, but which
we wont have time to cover. Look here to get familiar with functions you
use frequently, you may be surprised at what you find they can do.</p>
</div>
</div>
</div>
</div>
<p>Now, let’s read in the file <code>combined_tidy_vcf.csv</code> which
will be located in <code>/home/dcuser/r_data/</code>. Call this data
<code>variants</code>. The first argument to pass to our
<code>read.csv()</code> function is the file path for our data. The file
path must be in quotes and now is a good time to remember to use tab
autocompletion. <strong>If you use tab autocompletion you avoid typos
and errors in file paths.</strong> Use it!</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## read in a CSV file and save it as 'variants'</span></span>
<span></span>
<span><span class="va">variants</span> <span class="op">&lt;-</span> <span class="fu">read.csv</span><span class="op">(</span><span class="st">"/home/dcuser/r_data/combined_tidy_vcf.csv"</span><span class="op">)</span></span></code></pre>
</div>
<p>One of the first things you should notice is that in the Environment
window, you have the <code>variants</code> object, listed as 801 obs.
(observations/rows) of 29 variables (columns). Double-clicking on the
name of the object will open a view of the data in a new tab.</p>
<figure><img src="fig/rstudio_dataframeview.png" alt="rstudio data frame view" style="width: 1000px;" class="figure mx-auto d-block"></figure></section><section id="summarizing-subsetting-and-determining-the-structure-of-a-data-frame."><h2 class="section-heading">Summarizing, subsetting, and determining the structure of a data
frame.<a class="anchor" aria-label="anchor" href="#summarizing-subsetting-and-determining-the-structure-of-a-data-frame."></a>
</h2>
<hr class="half-width">
<p>A <strong>data frame is the standard way in R to store tabular
data</strong>. A data fame could also be thought of as a collection of
vectors, all of which have the same length. Using only two functions, we
can learn a lot about out data frame including some summary statistics
as well as well as the “structure” of the data frame. Let’s examine what
each of these functions can tell us:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## get summary statistics on a data frame</span></span>
<span></span>
<span><span class="fu">summary</span><span class="op">(</span><span class="va">variants</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>  sample_id            CHROM                POS             ID         
 Length:801         Length:801         Min.   :   1521   Mode:logical  
 Class :character   Class :character   1st Qu.:1115970   NA's:801      
 Mode  :character   Mode  :character   Median :2290361                 
                                       Mean   :2243682                 
                                       3rd Qu.:3317082                 
                                       Max.   :4629225                 
                                                                       
     REF                ALT                 QUAL          FILTER       
 Length:801         Length:801         Min.   :  4.385   Mode:logical  
 Class :character   Class :character   1st Qu.:139.000   NA's:801      
 Mode  :character   Mode  :character   Median :195.000                 
                                       Mean   :172.276                 
                                       3rd Qu.:225.000                 
                                       Max.   :228.000                 
                                                                       
   INDEL              IDV              IMF               DP       
 Mode :logical   Min.   : 2.000   Min.   :0.5714   Min.   : 2.00  
 FALSE:700       1st Qu.: 7.000   1st Qu.:0.8824   1st Qu.: 7.00  
 TRUE :101       Median : 9.000   Median :1.0000   Median :10.00  
                 Mean   : 9.396   Mean   :0.9219   Mean   :10.57  
                 3rd Qu.:11.000   3rd Qu.:1.0000   3rd Qu.:13.00  
                 Max.   :20.000   Max.   :1.0000   Max.   :79.00  
                 NA's   :700      NA's   :700                     
      VDB                 RPB              MQB              BQB        
 Min.   :0.0005387   Min.   :0.0000   Min.   :0.0000   Min.   :0.1153  
 1st Qu.:0.2180410   1st Qu.:0.3776   1st Qu.:0.1070   1st Qu.:0.6963  
 Median :0.4827410   Median :0.8663   Median :0.2872   Median :0.8615  
 Mean   :0.4926291   Mean   :0.6970   Mean   :0.5330   Mean   :0.7784  
 3rd Qu.:0.7598940   3rd Qu.:1.0000   3rd Qu.:1.0000   3rd Qu.:1.0000  
 Max.   :0.9997130   Max.   :1.0000   Max.   :1.0000   Max.   :1.0000  
                     NA's   :773      NA's   :773      NA's   :773     
      MQSB              SGB               MQ0F           ICB         
 Min.   :0.01348   Min.   :-0.6931   Min.   :0.00000   Mode:logical  
 1st Qu.:0.95494   1st Qu.:-0.6762   1st Qu.:0.00000   NA's:801      
 Median :1.00000   Median :-0.6620   Median :0.00000                 
 Mean   :0.96428   Mean   :-0.6444   Mean   :0.01127                 
 3rd Qu.:1.00000   3rd Qu.:-0.6364   3rd Qu.:0.00000                 
 Max.   :1.01283   Max.   :-0.4536   Max.   :0.66667                 
 NA's   :48                                                          
   HOB                AC          AN        DP4                  MQ       
 Mode:logical   Min.   :1   Min.   :1   Length:801         Min.   :10.00  
 NA's:801       1st Qu.:1   1st Qu.:1   Class :character   1st Qu.:60.00  
                Median :1   Median :1   Mode  :character   Median :60.00  
                Mean   :1   Mean   :1                      Mean   :58.19  
                3rd Qu.:1   3rd Qu.:1                      3rd Qu.:60.00  
                Max.   :1   Max.   :1                      Max.   :60.00  
                                                                          
    Indiv              gt_PL               gt_GT   gt_GT_alleles     
 Length:801         Length:801         Min.   :1   Length:801        
 Class :character   Class :character   1st Qu.:1   Class :character  
 Mode  :character   Mode  :character   Median :1   Mode  :character  
                                       Mean   :1                     
                                       3rd Qu.:1                     
                                       Max.   :1                     
                                                                     </code></pre>
</div>
<p>Our data frame had 29 variables, so we get 29 fields that summarize
the data. The <code>QUAL</code>, <code>IMF</code>, and <code>VDB</code>
variables (and several others) are numerical data and so you get summary
statistics on the min and max values for these columns, as well as mean,
median, and interquartile ranges. Many of the other variables
(e.g. <code>sample_id</code>) are treated as characters data (more on
this in a bit).</p>
<p>There is a lot to work with, so we will subset the first three
columns into a new data frame using the <code>data.frame()</code>
function.</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## put the first three columns of variants into a new data frame called subset</span></span>
<span></span>
<span><span class="va">subset</span><span class="op">&lt;-</span><span class="fu">data.frame</span><span class="op">(</span><span class="va">variants</span><span class="op">[</span>,<span class="fu">c</span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>,<span class="fl">6</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span></code></pre>
</div>
<p>Now, let’s use the <code>str()</code> (structure) function to look a
little more closely at how data frames work:</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## get the structure of a data frame</span></span>
<span></span>
<span><span class="fu">str</span><span class="op">(</span><span class="va">subset</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>'data.frame':	801 obs. of  4 variables:
 $ sample_id: chr  "SRR2584863" "SRR2584863" "SRR2584863" "SRR2584863" ...
 $ CHROM    : chr  "CP000819.1" "CP000819.1" "CP000819.1" "CP000819.1" ...
 $ POS      : int  9972 263235 281923 433359 473901 648692 1331794 1733343 2103887 2333538 ...
 $ ALT      : chr  "G" "T" "T" "CTTTTTTTT" ...</code></pre>
</div>
<p>Ok, thats a lot up unpack! Some things to notice.</p>
<ul>
<li>the object type <code>data.frame</code> is displayed in the first
row along with its dimensions, in this case 801 observations (rows) and
4 variables (columns)</li>
<li>Each variable (column) has a name (e.g. <code>sample_id</code>).
This is followed by the object mode (e.g. chr, int, etc.). Notice that
before each variable name there is a <code>$</code> - this will be
important later.</li>
</ul></section><section id="introducing-factors"><h2 class="section-heading">Introducing Factors<a class="anchor" aria-label="anchor" href="#introducing-factors"></a>
</h2>
<hr class="half-width">
<p>Factors are the final major data structure we will introduce in our R
genomics lessons. Factors can be thought of as vectors which are
specialized for categorical data. Given R’s specialization for
statistics, this make sense since categorial and continuous variables
are usually treated differently. Sometimes you may want to have data
treated as a factor, but in other cases, this may be undesirable.</p>
<p>Let’s see the value of treating some of which are categorical in
nature as factors. Let’s take a look at just the alternate alleles</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## extract the "ALT" column to a new object</span></span>
<span></span>
<span><span class="va">alt_alleles</span> <span class="op">&lt;-</span> <span class="va">subset</span><span class="op">$</span><span class="va">ALT</span></span></code></pre>
</div>
<p>Let’s look at the first few items in our factor using
<code>head()</code>:</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="va">alt_alleles</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "G"         "T"         "T"         "CTTTTTTTT" "CCGCGC"    "T"        </code></pre>
</div>
<p>There are 801 alleles (one for each row). To simplify, lets look at
just the single-nuleotide alleles (SNPs). We can use some of the vector
indexing skills from the last episode.</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">snps</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="va">alt_alleles</span><span class="op">[</span><span class="va">alt_alleles</span><span class="op">==</span><span class="st">"A"</span><span class="op">]</span>,</span>
<span>  <span class="va">alt_alleles</span><span class="op">[</span><span class="va">alt_alleles</span><span class="op">==</span><span class="st">"T"</span><span class="op">]</span>,</span>
<span>  <span class="va">alt_alleles</span><span class="op">[</span><span class="va">alt_alleles</span><span class="op">==</span><span class="st">"G"</span><span class="op">]</span>,</span>
<span>  <span class="va">alt_alleles</span><span class="op">[</span><span class="va">alt_alleles</span><span class="op">==</span><span class="st">"C"</span><span class="op">]</span><span class="op">)</span></span></code></pre>
</div>
<p>This leaves us with a vector of the 701 alternative alleles which
were single nucleotides. Right now, they are being treated a characters,
but we could treat them as categories of SNP. Doing this will enable
some nice features. For example, we can try to generate a plot of this
character vector as it is right now:</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot</span><span class="op">(</span><span class="va">snps</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">WARNING<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="warning" tabindex="0"><code>Warning in xy.coords(x, y, xlabel, ylabel, log): NAs introduced by coercion</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">WARNING<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="warning" tabindex="0"><code>Warning in min(x): no non-missing arguments to min; returning Inf</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">WARNING<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="warning" tabindex="0"><code>Warning in max(x): no non-missing arguments to max; returning -Inf</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in plot.window(...): need finite 'ylim' values</code></pre>
</div>
<p>Whoops! Though the <code>plot()</code> function will do its best to
give us a quick plot, it is unable to do so here. One way to fix this it
to tell R to treat the SNPs as categories (i.e. a factor vector); we
will create a new object to avoid confusion using the
<code>factor()</code> function:</p>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">factor_snps</span> <span class="op">&lt;-</span> <span class="fu">factor</span><span class="op">(</span><span class="va">snps</span><span class="op">)</span></span></code></pre>
</div>
<p>Let’s learn a little more about this new type of vector:</p>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">str</span><span class="op">(</span><span class="va">factor_snps</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> Factor w/ 4 levels "A","C","G","T": 1 1 1 1 1 1 1 1 1 1 ...</code></pre>
</div>
<p>What we get back are the categories (“A”,“C”,“G”,“T”) in our factor;
these are called “Levels”. <strong>Levels are the different categories
contained in a factor</strong>. By default, R will organize the levels
in a factor in alphabetical order. So the first level in this factor is
“A”.</p>
<p>For the sake of efficiency, R stores the content of a factor as a
vector of integers, which an integer is assigned to each of the possible
levels. Recall levels are assigned in alphabetical order. In this case,
the first item in our <code>factor_snps</code> object is “A”, which
happens to be the 1st level of our factor, ordered alphabetically. This
explains the sequence of “1”s (“Factor w/ 4 levels”A”,“C”,“G”,“T”: 1 1 1
1 1 1 1 1 1 1 …“), since”A” is the first level, and the first few items
in our factor are all “A”s.</p>
<p>We can see how many items in our vector fall into each category:</p>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">summary</span><span class="op">(</span><span class="va">factor_snps</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>  A   C   G   T 
211 139 154 203 </code></pre>
</div>
<p>As you can imagine, this is already useful when you want to generate
a tally.</p>
<div id="tip-treating-objects-as-categories-without-changing-their-mode" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="tip-treating-objects-as-categories-without-changing-their-mode" class="callout-inner">
<h3 class="callout-title">Tip: treating objects as categories without
changing their mode<a class="anchor" aria-label="anchor" href="#tip-treating-objects-as-categories-without-changing-their-mode"></a>
</h3>
<div class="callout-content">
<p>You don’t have to make an object a factor to get the benefits of
treating an object as a factor. See what happens when you use the
<code>as.factor()</code> function on <code>factor_snps</code>. To
generate a tally, you can sometimes also use the <code>table()</code>
function; though sometimes you may need to combine both (i.e.
<code>table(as.factor(object))</code>)</p>
</div>
</div>
</div>
</section><section id="plotting-and-ordering-factors"><h2 class="section-heading">Plotting and ordering factors<a class="anchor" aria-label="anchor" href="#plotting-and-ordering-factors"></a>
</h2>
<hr class="half-width">
<p>One of the most common uses for factors will be when you plot
categorical values. For example, suppose we want to know how many of our
variants had each possible SNP we could generate a plot:</p>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot</span><span class="op">(</span><span class="va">factor_snps</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/03-basics-factors-dataframes-rendered-unnamed-chunk-14-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>This isn’t a particularly pretty example of a plot but it works.
We’ll be learning much more about creating nice, publication-quality
graphics later in this lesson.</p>
<p>If you recall, factors are ordered alphabetically. That might make
sense, but categories (e.g., “red”, “blue”, “green”) often do not have
an intrinsic order. What if we wanted to order our plot according to the
numerical value (i.e., in descending order of SNP frequency)? We can
enforce an order on our factors:</p>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ordered_factor_snps</span> <span class="op">&lt;-</span> <span class="fu">factor</span><span class="op">(</span><span class="va">factor_snps</span>, levels <span class="op">=</span> <span class="fu">names</span><span class="op">(</span><span class="fu">sort</span><span class="op">(</span><span class="fu">table</span><span class="op">(</span><span class="va">factor_snps</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<p>Let’s deconstruct this from the inside out (you can try each of these
commands to see why this works):</p>
<ol style="list-style-type: decimal">
<li>We create a table of <code>factor_snps</code> to get the frequency
of each SNP: <code>table(factor_snps)</code>
</li>
<li>We sort this table: <code>sort(table(factor_snps))</code>; use the
<code>decreasing =</code> parameter for this function if you wanted to
change from the default of FALSE</li>
<li>Using the <code>names</code> function gives us just the character
names of the table sorted by
frequencies:<code>names(sort(table(factor_snps)))</code>
</li>
<li>The <code>factor</code> function is what allows us to create a
factor. We give it the <code>factor_snps</code> object as input, and use
the <code>levels=</code> parameter to enforce the ordering of the
levels.</li>
</ol>
<p>Now we see our plot has be reordered:</p>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot</span><span class="op">(</span><span class="va">ordered_factor_snps</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/03-basics-factors-dataframes-rendered-unnamed-chunk-16-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Factors come in handy in many places when using R. Even using more
sophisticated plotting packages such as ggplot2 will sometimes require
you to understand how to manipulate factors.</p>
<div id="tip-packages-in-r-what-are-they-and-why-do-we-use-them" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="tip-packages-in-r-what-are-they-and-why-do-we-use-them" class="callout-inner">
<h3 class="callout-title">Tip: Packages in R – what are they and why do
we use them?<a class="anchor" aria-label="anchor" href="#tip-packages-in-r-what-are-they-and-why-do-we-use-them"></a>
</h3>
<div class="callout-content">
<p>Packages are simply collections of functions and/or data that can be
used to extend the capabilities of R beyond the core functionality that
comes with it by default. There are useful R packages available that
span all types of statistical analysis, data visualization, and more.
The main place that R packages are installed from is a website called <a href="https://cran.r-project.org/" class="external-link">CRAN</a> (the Comprehensive R Archive
Network). Many thousands of R packages are available there, and when you
use the built-in R function <code>install.packages()</code>, it will
look for a CRAN repository to install from. So, for example, to install
<a href="https://www.tidyverse.org" class="external-link">tidyverse</a> packages such as
<code>dplyr</code> and <code>ggplot2</code> (which you’ll do in the next
few lessons), you would use the following command:</p>
<div class="codewrapper sourceCode" id="cb24">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install a package from CRAN</span></span>
<span><span class="fu">install.packages</span><span class="op">(</span><span class="st">"ggplot2"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Installing ggplot2 [3.4.2] ...
	OK [linked cache in 0.25 milliseconds]
* Installed 1 package in 1.4 seconds.</code></pre>
</div>
</div>
</div>
</div>
</section><section id="subsetting-data-frames"><h2 class="section-heading">Subsetting data frames<a class="anchor" aria-label="anchor" href="#subsetting-data-frames"></a>
</h2>
<hr class="half-width">
<p>Next, we are going to talk about how you can get specific values from
data frames, and where necessary, change the mode of a column of
values.</p>
<p>The first thing to remember is that a data frame is two-dimensional
(rows and columns). Therefore, to select a specific value we will will
once again use <code>[]</code> (bracket) notation, but we will specify
more than one value (except in some cases where we are taking a
range).</p>
<div id="exercise-subsetting-a-data-frame" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-subsetting-a-data-frame" class="callout-inner">
<h3 class="callout-title">Exercise: Subsetting a data frame<a class="anchor" aria-label="anchor" href="#exercise-subsetting-a-data-frame"></a>
</h3>
<div class="callout-content">
<p><strong>Try the following indices and functions and try to figure out
what they return</strong></p>
<ol style="list-style-type: lower-alpha">
<li><p><code>variants[1,1]</code></p></li>
<li><p><code>variants[2,4]</code></p></li>
<li><p><code>variants[801,29]</code></p></li>
<li><p><code>variants[2, ]</code></p></li>
<li><p><code>variants[-1, ]</code></p></li>
<li><p><code>variants[1:4,1]</code></p></li>
<li><p><code>variants[1:10,c("REF","ALT")]</code></p></li>
<li><p><code>variants[,c("sample_id")]</code></p></li>
<li><p><code>head(variants)</code></p></li>
<li><p><code>tail(variants)</code></p></li>
<li><p><code>variants$sample_id</code></p></li>
<li><p><code>variants[variants$REF == "A",]</code></p></li>
</ol>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<ol style="list-style-type: lower-alpha"><li>
</li></ol>
<div class="codewrapper sourceCode" id="cb26">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">variants</span><span class="op">[</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "SRR2584863"</code></pre>
</div>
<ol start="2" style="list-style-type: lower-alpha"><li>
</li></ol>
<div class="codewrapper sourceCode" id="cb28">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">variants</span><span class="op">[</span><span class="fl">2</span>,<span class="fl">4</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] NA</code></pre>
</div>
<ol start="3" style="list-style-type: lower-alpha"><li>
</li></ol>
<div class="codewrapper sourceCode" id="cb30">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">variants</span><span class="op">[</span><span class="fl">801</span>,<span class="fl">29</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "T"</code></pre>
</div>
<ol start="4" style="list-style-type: lower-alpha"><li>
</li></ol>
<div class="codewrapper sourceCode" id="cb32">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">variants</span><span class="op">[</span><span class="fl">2</span>, <span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>   sample_id      CHROM    POS ID REF ALT QUAL FILTER INDEL IDV IMF DP      VDB
2 SRR2584863 CP000819.1 263235 NA   G   T   85     NA FALSE  NA  NA  6 0.096133
  RPB MQB BQB MQSB       SGB     MQ0F ICB HOB AC AN     DP4 MQ
2   1   1   1   NA -0.590765 0.166667  NA  NA  1  1 0,1,0,5 33
                                                               Indiv gt_PL
2 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam 112,0
  gt_GT gt_GT_alleles
2     1             T</code></pre>
</div>
<ol start="5" style="list-style-type: lower-alpha"><li>
</li></ol>
<div class="codewrapper sourceCode" id="cb34">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">variants</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span>, <span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>   sample_id      CHROM     POS ID      REF       ALT QUAL FILTER INDEL IDV IMF
2 SRR2584863 CP000819.1  263235 NA        G         T   85     NA FALSE  NA  NA
3 SRR2584863 CP000819.1  281923 NA        G         T  217     NA FALSE  NA  NA
4 SRR2584863 CP000819.1  433359 NA CTTTTTTT CTTTTTTTT   64     NA  TRUE  12 1.0
5 SRR2584863 CP000819.1  473901 NA     CCGC    CCGCGC  228     NA  TRUE   9 0.9
6 SRR2584863 CP000819.1  648692 NA        C         T  210     NA FALSE  NA  NA
7 SRR2584863 CP000819.1 1331794 NA        C         A  178     NA FALSE  NA  NA
  DP      VDB RPB MQB BQB     MQSB       SGB     MQ0F ICB HOB AC AN     DP4 MQ
2  6 0.096133   1   1   1       NA -0.590765 0.166667  NA  NA  1  1 0,1,0,5 33
3 10 0.774083  NA  NA  NA 0.974597 -0.662043 0.000000  NA  NA  1  1 0,0,4,5 60
4 12 0.477704  NA  NA  NA 1.000000 -0.676189 0.000000  NA  NA  1  1 0,1,3,8 60
5 10 0.659505  NA  NA  NA 0.916482 -0.662043 0.000000  NA  NA  1  1 1,0,2,7 60
6 10 0.268014  NA  NA  NA 0.916482 -0.670168 0.000000  NA  NA  1  1 0,0,7,3 60
7  8 0.624078  NA  NA  NA 0.900802 -0.651104 0.000000  NA  NA  1  1 0,0,3,5 60
                                                               Indiv gt_PL
2 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam 112,0
3 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam 247,0
4 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam  91,0
5 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam 255,0
6 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam 240,0
7 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam 208,0
  gt_GT gt_GT_alleles
2     1             T
3     1             T
4     1     CTTTTTTTT
5     1        CCGCGC
6     1             T
7     1             A</code></pre>
</div>
<ol start="6" style="list-style-type: lower-alpha"><li>
</li></ol>
<div class="codewrapper sourceCode" id="cb36">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">variants</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span>,<span class="fl">1</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "SRR2584863" "SRR2584863" "SRR2584863" "SRR2584863"</code></pre>
</div>
<ol start="7" style="list-style-type: lower-alpha"><li>
</li></ol>
<div class="codewrapper sourceCode" id="cb38">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">variants</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,<span class="fu">c</span><span class="op">(</span><span class="st">"REF"</span>,<span class="st">"ALT"</span><span class="op">)</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>                                REF
1                                 T
2                                 G
3                                 G
4                          CTTTTTTT
5                              CCGC
6                                 C
7                                 C
8                                 G
9  ACAGCCAGCCAGCCAGCCAGCCAGCCAGCCAG
10                               AT
                                                        ALT
1                                                         G
2                                                         T
3                                                         T
4                                                 CTTTTTTTT
5                                                    CCGCGC
6                                                         T
7                                                         A
8                                                         A
9  ACAGCCAGCCAGCCAGCCAGCCAGCCAGCCAGCCAGCCAGCCAGCCAGCCAGCCAG
10                                                      ATT</code></pre>
</div>
<ol start="8" style="list-style-type: lower-alpha"><li>
</li></ol>
<div class="codewrapper sourceCode" id="cb40">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">variants</span><span class="op">[</span>,<span class="fu">c</span><span class="op">(</span><span class="st">"sample_id"</span><span class="op">)</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "SRR2584863" "SRR2584863" "SRR2584863" "SRR2584863" "SRR2584863"
[6] "SRR2584863"</code></pre>
</div>
<ol style="list-style-type: lower-roman"><li>
</li></ol>
<div class="codewrapper sourceCode" id="cb42">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="va">variants</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>   sample_id      CHROM    POS ID      REF       ALT QUAL FILTER INDEL IDV IMF
1 SRR2584863 CP000819.1   9972 NA        T         G   91     NA FALSE  NA  NA
2 SRR2584863 CP000819.1 263235 NA        G         T   85     NA FALSE  NA  NA
3 SRR2584863 CP000819.1 281923 NA        G         T  217     NA FALSE  NA  NA
4 SRR2584863 CP000819.1 433359 NA CTTTTTTT CTTTTTTTT   64     NA  TRUE  12 1.0
5 SRR2584863 CP000819.1 473901 NA     CCGC    CCGCGC  228     NA  TRUE   9 0.9
6 SRR2584863 CP000819.1 648692 NA        C         T  210     NA FALSE  NA  NA
  DP       VDB RPB MQB BQB     MQSB       SGB     MQ0F ICB HOB AC AN     DP4 MQ
1  4 0.0257451  NA  NA  NA       NA -0.556411 0.000000  NA  NA  1  1 0,0,0,4 60
2  6 0.0961330   1   1   1       NA -0.590765 0.166667  NA  NA  1  1 0,1,0,5 33
3 10 0.7740830  NA  NA  NA 0.974597 -0.662043 0.000000  NA  NA  1  1 0,0,4,5 60
4 12 0.4777040  NA  NA  NA 1.000000 -0.676189 0.000000  NA  NA  1  1 0,1,3,8 60
5 10 0.6595050  NA  NA  NA 0.916482 -0.662043 0.000000  NA  NA  1  1 1,0,2,7 60
6 10 0.2680140  NA  NA  NA 0.916482 -0.670168 0.000000  NA  NA  1  1 0,0,7,3 60
                                                               Indiv gt_PL
1 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam 121,0
2 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam 112,0
3 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam 247,0
4 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam  91,0
5 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam 255,0
6 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam 240,0
  gt_GT gt_GT_alleles
1     1             G
2     1             T
3     1             T
4     1     CTTTTTTTT
5     1        CCGCGC
6     1             T</code></pre>
</div>
<ol start="10" style="list-style-type: lower-alpha"><li>
</li></ol>
<div class="codewrapper sourceCode" id="cb44">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">tail</span><span class="op">(</span><span class="va">variants</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>     sample_id      CHROM     POS ID REF ALT QUAL FILTER INDEL IDV IMF DP
796 SRR2589044 CP000819.1 3444175 NA   G   T  184     NA FALSE  NA  NA  9
797 SRR2589044 CP000819.1 3481820 NA   A   G  225     NA FALSE  NA  NA 12
798 SRR2589044 CP000819.1 3893550 NA  AG AGG  101     NA  TRUE   4   1  4
799 SRR2589044 CP000819.1 3901455 NA   A  AC   70     NA  TRUE   3   1  3
800 SRR2589044 CP000819.1 4100183 NA   A   G  177     NA FALSE  NA  NA  8
801 SRR2589044 CP000819.1 4431393 NA TGG   T  225     NA  TRUE  10   1 10
          VDB RPB MQB BQB     MQSB       SGB MQ0F ICB HOB AC AN     DP4 MQ
796 0.4714620  NA  NA  NA 0.992367 -0.651104    0  NA  NA  1  1 0,0,4,4 60
797 0.8707240  NA  NA  NA 1.000000 -0.680642    0  NA  NA  1  1 0,0,4,8 60
798 0.9182970  NA  NA  NA 1.000000 -0.556411    0  NA  NA  1  1 0,0,3,1 52
799 0.0221621  NA  NA  NA       NA -0.511536    0  NA  NA  1  1 0,0,3,0 60
800 0.9272700  NA  NA  NA 0.900802 -0.651104    0  NA  NA  1  1 0,0,3,5 60
801 0.7488140  NA  NA  NA 1.007750 -0.670168    0  NA  NA  1  1 0,0,4,6 60
                                                                 Indiv gt_PL
796 /home/dcuser/dc_workshop/results/bam/SRR2589044.aligned.sorted.bam 214,0
797 /home/dcuser/dc_workshop/results/bam/SRR2589044.aligned.sorted.bam 255,0
798 /home/dcuser/dc_workshop/results/bam/SRR2589044.aligned.sorted.bam 131,0
799 /home/dcuser/dc_workshop/results/bam/SRR2589044.aligned.sorted.bam 100,0
800 /home/dcuser/dc_workshop/results/bam/SRR2589044.aligned.sorted.bam 207,0
801 /home/dcuser/dc_workshop/results/bam/SRR2589044.aligned.sorted.bam 255,0
    gt_GT gt_GT_alleles
796     1             T
797     1             G
798     1           AGG
799     1            AC
800     1             G
801     1             T</code></pre>
</div>
<ol start="11" style="list-style-type: lower-alpha"><li>
</li></ol>
<div class="codewrapper sourceCode" id="cb46">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">variants</span><span class="op">$</span><span class="va">sample_id</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "SRR2584863" "SRR2584863" "SRR2584863" "SRR2584863" "SRR2584863"
[6] "SRR2584863"</code></pre>
</div>
<ol start="12" style="list-style-type: lower-alpha"><li>
</li></ol>
<div class="codewrapper sourceCode" id="cb48">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">variants</span><span class="op">[</span><span class="va">variants</span><span class="op">$</span><span class="va">REF</span> <span class="op">==</span> <span class="st">"A"</span>,<span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>    sample_id      CHROM     POS ID REF ALT QUAL FILTER INDEL IDV IMF DP
11 SRR2584863 CP000819.1 2407766 NA   A   C  104     NA FALSE  NA  NA  9
12 SRR2584863 CP000819.1 2446984 NA   A   C  225     NA FALSE  NA  NA 20
14 SRR2584863 CP000819.1 2665639 NA   A   T  225     NA FALSE  NA  NA 19
16 SRR2584863 CP000819.1 3339313 NA   A   C  211     NA FALSE  NA  NA 10
18 SRR2584863 CP000819.1 3481820 NA   A   G  200     NA FALSE  NA  NA  9
19 SRR2584863 CP000819.1 3488669 NA   A   C  225     NA FALSE  NA  NA 13
         VDB      RPB      MQB      BQB     MQSB       SGB     MQ0F ICB HOB AC
11 0.0230738 0.900802 0.150134 0.750668 0.500000 -0.590765 0.333333  NA  NA  1
12 0.0714027       NA       NA       NA 1.000000 -0.689466 0.000000  NA  NA  1
14 0.9960390       NA       NA       NA 1.000000 -0.690438 0.000000  NA  NA  1
16 0.4059360       NA       NA       NA 1.007750 -0.670168 0.000000  NA  NA  1
18 0.1070810       NA       NA       NA 0.974597 -0.662043 0.000000  NA  NA  1
19 0.0162706       NA       NA       NA 1.000000 -0.680642 0.000000  NA  NA  1
   AN      DP4 MQ
11  1  3,0,3,2 25
12  1 0,0,10,6 60
14  1 0,0,12,5 60
16  1  0,0,4,6 60
18  1  0,0,4,5 60
19  1  0,0,8,4 60
                                                                Indiv gt_PL
11 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam 131,0
12 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam 255,0
14 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam 255,0
16 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam 241,0
18 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam 230,0
19 /home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam 255,0
   gt_GT gt_GT_alleles
11     1             C
12     1             C
14     1             T
16     1             C
18     1             G
19     1             C</code></pre>
</div>
</div>
</div>
</div>
</div>
<p>The subsetting notation is very similar to what we learned for
vectors. The key differences include:</p>
<ul>
<li>Typically provide two values separated by commas: data.frame[row,
column]</li>
<li>In cases where you are taking a continuous range of numbers use a
colon between the numbers (start:stop, inclusive)</li>
<li>For a non continuous set of numbers, pass a vector using
<code>c()</code>
</li>
<li>Index using the name of a column(s) by passing them as vectors using
<code>c()</code>
</li>
</ul>
<p>Finally, in all of the subsetting exercises above, we printed values
to the screen. You can create a new data frame object by assigning them
to a new object name:</p>
<div class="codewrapper sourceCode" id="cb50">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># create a new data frame containing only observations from SRR2584863</span></span>
<span></span>
<span><span class="va">SRR2584863_variants</span> <span class="op">&lt;-</span> <span class="va">variants</span><span class="op">[</span><span class="va">variants</span><span class="op">$</span><span class="va">sample_id</span> <span class="op">==</span> <span class="st">"SRR2584863"</span>,<span class="op">]</span></span>
<span></span>
<span><span class="co"># check the dimension of the data frame</span></span>
<span></span>
<span><span class="fu">dim</span><span class="op">(</span><span class="va">SRR2584863_variants</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 25 29</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb52">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># get a summary of the data frame</span></span>
<span></span>
<span><span class="fu">summary</span><span class="op">(</span><span class="va">SRR2584863_variants</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>  sample_id            CHROM                POS             ID         
 Length:25          Length:25          Min.   :   9972   Mode:logical  
 Class :character   Class :character   1st Qu.:1331794   NA's:25       
 Mode  :character   Mode  :character   Median :2618472                 
                                       Mean   :2464989                 
                                       3rd Qu.:3488669                 
                                       Max.   :4616538                 
                                                                       
     REF                ALT                 QUAL         FILTER       
 Length:25          Length:25          Min.   : 31.89   Mode:logical  
 Class :character   Class :character   1st Qu.:104.00   NA's:25       
 Mode  :character   Mode  :character   Median :211.00                 
                                       Mean   :172.97                 
                                       3rd Qu.:225.00                 
                                       Max.   :228.00                 
                                                                      
   INDEL              IDV             IMF               DP      
 Mode :logical   Min.   : 2.00   Min.   :0.6667   Min.   : 2.0  
 FALSE:19        1st Qu.: 3.25   1st Qu.:0.9250   1st Qu.: 9.0  
 TRUE :6         Median : 8.00   Median :1.0000   Median :10.0  
                 Mean   : 7.00   Mean   :0.9278   Mean   :10.4  
                 3rd Qu.: 9.75   3rd Qu.:1.0000   3rd Qu.:12.0  
                 Max.   :12.00   Max.   :1.0000   Max.   :20.0  
                 NA's   :19      NA's   :19                     
      VDB               RPB              MQB               BQB        
 Min.   :0.01627   Min.   :0.9008   Min.   :0.04979   Min.   :0.7507  
 1st Qu.:0.07140   1st Qu.:0.9275   1st Qu.:0.09996   1st Qu.:0.7627  
 Median :0.37674   Median :0.9542   Median :0.15013   Median :0.7748  
 Mean   :0.40429   Mean   :0.9517   Mean   :0.39997   Mean   :0.8418  
 3rd Qu.:0.65951   3rd Qu.:0.9771   3rd Qu.:0.57507   3rd Qu.:0.8874  
 Max.   :0.99604   Max.   :1.0000   Max.   :1.00000   Max.   :1.0000  
                   NA's   :22       NA's   :22        NA's   :22      
      MQSB             SGB               MQ0F           ICB         
 Min.   :0.5000   Min.   :-0.6904   Min.   :0.00000   Mode:logical  
 1st Qu.:0.9599   1st Qu.:-0.6762   1st Qu.:0.00000   NA's:25       
 Median :0.9962   Median :-0.6620   Median :0.00000                 
 Mean   :0.9442   Mean   :-0.6341   Mean   :0.04667                 
 3rd Qu.:1.0000   3rd Qu.:-0.6168   3rd Qu.:0.00000                 
 Max.   :1.0128   Max.   :-0.4536   Max.   :0.66667                 
 NA's   :3                                                          
   HOB                AC          AN        DP4                  MQ       
 Mode:logical   Min.   :1   Min.   :1   Length:25          Min.   :10.00  
 NA's:25        1st Qu.:1   1st Qu.:1   Class :character   1st Qu.:60.00  
                Median :1   Median :1   Mode  :character   Median :60.00  
                Mean   :1   Mean   :1                      Mean   :55.52  
                3rd Qu.:1   3rd Qu.:1                      3rd Qu.:60.00  
                Max.   :1   Max.   :1                      Max.   :60.00  
                                                                          
    Indiv              gt_PL               gt_GT   gt_GT_alleles     
 Length:25          Length:25          Min.   :1   Length:25         
 Class :character   Class :character   1st Qu.:1   Class :character  
 Mode  :character   Mode  :character   Median :1   Mode  :character  
                                       Mean   :1                     
                                       3rd Qu.:1                     
                                       Max.   :1                     
                                                                     </code></pre>
</div>
</section><section id="coercing-values-in-data-frames"><h2 class="section-heading">Coercing values in data frames<a class="anchor" aria-label="anchor" href="#coercing-values-in-data-frames"></a>
</h2>
<hr class="half-width">
<div id="tip-coercion-isnt-limited-to-data-frames" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="tip-coercion-isnt-limited-to-data-frames" class="callout-inner">
<h3 class="callout-title">Tip: coercion isn’t limited to data
frames<a class="anchor" aria-label="anchor" href="#tip-coercion-isnt-limited-to-data-frames"></a>
</h3>
<div class="callout-content">
<p>While we are going to address coercion in the context of data frames
most of these methods apply to other data structures, such as
vectors</p>
</div>
</div>
</div>
<p>Sometimes, it is possible that R will misinterpret the type of data
represented in a data frame, or store that data in a mode which prevents
you from operating on the data the way you wish. For example, a long
list of gene names isn’t usually thought of as a categorical variable,
the way that your experimental condition (e.g. control, treatment) might
be. More importantly, some R packages you use to analyze your data may
expect characters as input, not factors. At other times (such as
plotting or some statistical analyses) a factor may be more appropriate.
Ultimately, you should know how to change the mode of an object.</p>
<p>First, its very important to recognize that coercion happens in R all
the time. This can be a good thing when R gets it right, or a bad thing
when the result is not what you expect. Consider:</p>
<div class="codewrapper sourceCode" id="cb54">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">snp_chromosomes</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="st">'3'</span>, <span class="st">'11'</span>, <span class="st">'X'</span>, <span class="st">'6'</span><span class="op">)</span></span>
<span><span class="fu">typeof</span><span class="op">(</span><span class="va">snp_chromosomes</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "character"</code></pre>
</div>
<p>Although there are several numbers in our vector, they are all in
quotes, so we have explicitly told R to consider them as characters.
However, even if we removed the quotes from the numbers, R would coerce
everything into a character:</p>
<div class="codewrapper sourceCode" id="cb56">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">snp_chromosomes_2</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="fl">3</span>, <span class="fl">11</span>, <span class="st">'X'</span>, <span class="fl">6</span><span class="op">)</span></span>
<span><span class="fu">typeof</span><span class="op">(</span><span class="va">snp_chromosomes_2</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "character"</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb58">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">snp_chromosomes_2</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "3"</code></pre>
</div>
<p>We can use the <code>as.</code> functions to explicitly coerce values
from one form into another. Consider the following vector of characters,
which all happen to be valid numbers:</p>
<div class="codewrapper sourceCode" id="cb60">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">snp_positions_2</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="st">"8762685"</span>, <span class="st">"66560624"</span>, <span class="st">"67545785"</span>, <span class="st">"154039662"</span><span class="op">)</span></span>
<span><span class="fu">typeof</span><span class="op">(</span><span class="va">snp_positions_2</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "character"</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb62">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">snp_positions_2</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "8762685"</code></pre>
</div>
<p>Now we can coerce <code>snp_positions_2</code> into a numeric type
using <code>as.numeric()</code>:</p>
<div class="codewrapper sourceCode" id="cb64">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">snp_positions_2</span> <span class="op">&lt;-</span> <span class="fu">as.numeric</span><span class="op">(</span><span class="va">snp_positions_2</span><span class="op">)</span></span>
<span><span class="fu">typeof</span><span class="op">(</span><span class="va">snp_positions_2</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "double"</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb66">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">snp_positions_2</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 8762685</code></pre>
</div>
<p>Sometimes coercion is straight forward, but what would happen if we
tried using <code>as.numeric()</code> on
<code>snp_chromosomes_2</code></p>
<div class="codewrapper sourceCode" id="cb68">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">snp_chromosomes_2</span> <span class="op">&lt;-</span> <span class="fu">as.numeric</span><span class="op">(</span><span class="va">snp_chromosomes_2</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">WARNING<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="warning" tabindex="0"><code>Warning: NAs introduced by coercion</code></pre>
</div>
<p>If we check, we will see that an <code>NA</code> value (R’s default
value for missing data) has been introduced.</p>
<div class="codewrapper sourceCode" id="cb70">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">snp_chromosomes_2</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1]  3 11 NA  6</code></pre>
</div>
<p>Trouble can really start when we try to coerce a factor. For example,
when we try to coerce the <code>sample_id</code> column in our data
frame into a numeric mode look at the result:</p>
<div class="codewrapper sourceCode" id="cb72">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">as.numeric</span><span class="op">(</span><span class="va">variants</span><span class="op">$</span><span class="va">sample_id</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">WARNING<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="warning" tabindex="0"><code>Warning: NAs introduced by coercion</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>  [1] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
 [26] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
 [51] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
 [76] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
[101] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
[126] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
[151] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
[176] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
[201] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
[226] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
[251] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
[276] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
[301] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
[326] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
[351] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
[376] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
[401] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
[426] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
[451] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
[476] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
[501] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
[526] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
[551] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
[576] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
[601] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
[626] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
[651] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
[676] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
[701] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
[726] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
[751] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
[776] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
[801] NA</code></pre>
</div>
<p>Strangely, it works! Almost. Instead of giving an error message, R
returns numeric values, which in this case are the integers assigned to
the levels in this factor. This kind of behavior can lead to
hard-to-find bugs, for example when we do have numbers in a factor, and
we get numbers from a coercion. If we don’t look carefully, we may not
notice a problem.</p>
<p>If you need to coerce an entire column you can overwrite it using an
expression like this one:</p>
<div class="codewrapper sourceCode" id="cb75">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># make the 'REF' column a character type column</span></span>
<span></span>
<span><span class="va">variants</span><span class="op">$</span><span class="va">REF</span> <span class="op">&lt;-</span> <span class="fu">as.character</span><span class="op">(</span><span class="va">variants</span><span class="op">$</span><span class="va">REF</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># check the type of the column</span></span>
<span><span class="fu">typeof</span><span class="op">(</span><span class="va">variants</span><span class="op">$</span><span class="va">REF</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "character"</code></pre>
</div>
</section><section id="stringsasfactors"><h2 class="section-heading">StringsAsFactors = ?<a class="anchor" aria-label="anchor" href="#stringsasfactors"></a>
</h2>
<hr class="half-width">
<p>Lets summarize this section on coercion with a few take home
messages.</p>
<ul>
<li>When you explicitly coerce one data type into another (this is known
as <strong>explicit coercion</strong>), be careful to check the result.
Ideally, you should try to see if its possible to avoid steps in your
analysis that force you to coerce.</li>
<li>R will sometimes coerce without you asking for it. This is called
(appropriately) <strong>implicit coercion</strong>. For example when we
tried to create a vector with multiple data types, R chose one type
through implicit coercion.</li>
<li>Check the structure (<code>str()</code>) of your data frames before
working with them!</li>
</ul>
<div id="tip-coercion-isnt-limited-to-data-frames-1" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="tip-coercion-isnt-limited-to-data-frames-1" class="callout-inner">
<h3 class="callout-title">Tip: coercion isn’t limited to data
frames<a class="anchor" aria-label="anchor" href="#tip-coercion-isnt-limited-to-data-frames-1"></a>
</h3>
<div class="callout-content">
<p>Prior to R 4.0 when importing a data frame using any one of the
<code>read.table()</code> functions such as <code>read.csv()</code> ,
the argument <code>StringsAsFactors</code> was by default set to true
TRUE. Setting it to FALSE will treat any non-numeric column to a
character type. <code>read.csv()</code> documentation, you will also see
you can explicitly type your columns using the <code>colClasses</code>
argument. Other R packages (such as the Tidyverse “readr”) don’t have
this particular conversion issue, but many packages will still try to
guess a data type.</p>
</div>
</div>
</div>
</section><section id="data-frame-bonus-material-math-sorting-renaming"><h2 class="section-heading">Data frame bonus material: math, sorting, renaming<a class="anchor" aria-label="anchor" href="#data-frame-bonus-material-math-sorting-renaming"></a>
</h2>
<hr class="half-width">
<p>Here are a few operations that don’t need much explanation, but which
are good to know.</p>
<p>There are lots of arithmetic functions you may want to apply to your
data frame, covering those would be a course in itself (there is some
starting material <a href="https://swcarpentry.github.io/r-novice-inflammation/15-supp-loops-in-depth/" class="external-link">here</a>).
Our lessons will cover some additional summary statistical functions in
a subsequent lesson, but overall we will focus on data cleaning and
visualization.</p>
<p>You can use functions like <code>mean()</code>, <code>min()</code>,
<code>max()</code> on an individual column. Let’s look at the “DP” or
filtered depth. This value shows the number of filtered reads that
support each of the reported variants.</p>
<div class="codewrapper sourceCode" id="cb77">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">max</span><span class="op">(</span><span class="va">variants</span><span class="op">$</span><span class="va">DP</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 79</code></pre>
</div>
<p>You can sort a data frame using the <code>order()</code>
function:</p>
<div class="codewrapper sourceCode" id="cb79">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sorted_by_DP</span> <span class="op">&lt;-</span> <span class="va">variants</span><span class="op">[</span><span class="fu">order</span><span class="op">(</span><span class="va">variants</span><span class="op">$</span><span class="va">DP</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">sorted_by_DP</span><span class="op">$</span><span class="va">DP</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 2 2 2 2 2 2</code></pre>
</div>
<div id="exercise" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise" class="callout-inner">
<h3 class="callout-title">Exercise<a class="anchor" aria-label="anchor" href="#exercise"></a>
</h3>
<div class="callout-content">
<p>The <code>order()</code> function lists values in increasing order by
default. Look at the documentation for this function and change
<code>sorted_by_DP</code> to start with variants with the greatest
filtered depth (“DP”).</p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" data-bs-parent="#accordionSolution3" aria-labelledby="headingSolution3">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb81">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span>   <span class="va">sorted_by_DP</span> <span class="op">&lt;-</span> <span class="va">variants</span><span class="op">[</span><span class="fu">order</span><span class="op">(</span><span class="va">variants</span><span class="op">$</span><span class="va">DP</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, <span class="op">]</span></span>
<span>   <span class="fu">head</span><span class="op">(</span><span class="va">sorted_by_DP</span><span class="op">$</span><span class="va">DP</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 79 46 41 29 29 27</code></pre>
</div>
</div>
</div>
</div>
</div>
<p>You can rename columns:</p>
<div class="codewrapper sourceCode" id="cb83">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">colnames</span><span class="op">(</span><span class="va">variants</span><span class="op">)</span><span class="op">[</span><span class="fu">colnames</span><span class="op">(</span><span class="va">variants</span><span class="op">)</span> <span class="op">==</span> <span class="st">"sample_id"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"strain"</span></span>
<span></span>
<span><span class="co"># check the column name (hint names are returned as a vector)</span></span>
<span><span class="fu">colnames</span><span class="op">(</span><span class="va">variants</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> [1] "strain"        "CHROM"         "POS"           "ID"           
 [5] "REF"           "ALT"           "QUAL"          "FILTER"       
 [9] "INDEL"         "IDV"           "IMF"           "DP"           
[13] "VDB"           "RPB"           "MQB"           "BQB"          
[17] "MQSB"          "SGB"           "MQ0F"          "ICB"          
[21] "HOB"           "AC"            "AN"            "DP4"          
[25] "MQ"            "Indiv"         "gt_PL"         "gt_GT"        
[29] "gt_GT_alleles"</code></pre>
</div>
</section><section id="saving-your-data-frame-to-a-file"><h2 class="section-heading">Saving your data frame to a file<a class="anchor" aria-label="anchor" href="#saving-your-data-frame-to-a-file"></a>
</h2>
<hr class="half-width">
<p>We can save data to a file. We will save our
<code>SRR2584863_variants</code> object to a .csv file using the
<code>write.csv()</code> function:</p>
<div class="codewrapper sourceCode" id="cb85">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">write.csv</span><span class="op">(</span><span class="va">SRR2584863_variants</span>, file <span class="op">=</span> <span class="st">"data/SRR2584863_variants.csv"</span><span class="op">)</span></span></code></pre>
</div>
<p>The <code>write.csv()</code> function has some additional arguments
listed in the help, but at a minimum you need to tell it what data frame
to write to file, and give a path to a file name in quotes (if you only
provide a file name, the file will be written in the current working
directory).</p>
</section><section id="importing-data-from-excel"><h2 class="section-heading">Importing data from Excel<a class="anchor" aria-label="anchor" href="#importing-data-from-excel"></a>
</h2>
<hr class="half-width">
<p>Excel is one of the most common formats, so we need to discuss how to
make these files play nicely with R. The simplest way to import data
from Excel is to <strong>save your Excel file in .csv format</strong>*.
You can then import into R right away. Sometimes you may not be able to
do this (imagine you have data in 300 Excel files, are you going to open
and export all of them?).</p>
<p>One common R package (a set of code with features you can download
and add to your R installation) is the <a href="https://CRAN.R-project.org/package=readxl" class="external-link">readxl package</a>
which can open and import Excel files. Rather than addressing package
installation this second (we’ll discuss this soon!), we can take
advantage of RStudio’s import feature which integrates this package.
(Note: this feature is available only in the latest versions of RStudio
such as is installed on our cloud instance).</p>
<p>First, in the RStudio menu go to <strong>File</strong>, select
<strong>Import Dataset</strong>, and choose <strong>From Excel…</strong>
(notice there are several other options you can explore).</p>
<figure><img src="fig/rstudio_import_menu.png%20" alt="rstudio import menu" style="width: 600px;" class="figure mx-auto d-block"></figure><p>Next, under <strong>File/Url:</strong> click the <kbd>Browse</kbd>
button and navigate to the <strong>Ecoli_metadata.xlsx</strong> file
located at <code>/home/dcuser/dc_sample_data/R</code>. You should now
see a preview of the data to be imported:</p>
<figure><img src="fig/rstudio_import_screen.png%20" alt="rstudio import screen" style="width: 1200px;" class="figure mx-auto d-block"></figure><p>Notice that you have the option to change the data type of each
variable by clicking arrow (drop-down menu) next to each column title.
Under <strong>Import Options</strong> you may also rename the data,
choose a different sheet to import, and choose how you will handle
headers and skipped rows. Under <strong>Code Preview</strong> you can
see the code that will be used to import this file. We could have
written this code and imported the Excel file without the RStudio import
function, but now you can choose your preference.</p>
<p>In this exercise, we will leave the title of the data frame as
<strong>Ecoli_metadata</strong>, and there are no other options we need
to adjust. Click the <kbd>Import</kbd> button to import the data.</p>
<p>Finally, let’s check the first few lines of the
<code>Ecoli_metadata</code> data frame:</p>
<div class="codewrapper sourceCode" id="cb86">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="va">Ecoli_metadata</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 6 × 7
  sample   generation clade   strain cit     run       genome_size
  &lt;chr&gt;         &lt;dbl&gt; &lt;chr&gt;   &lt;chr&gt;  &lt;chr&gt;   &lt;chr&gt;           &lt;dbl&gt;
1 REL606            0 NA      REL606 unknown &lt;NA&gt;             4.62
2 REL1166A       2000 unknown REL606 unknown SRR098028        4.63
3 ZDB409         5000 unknown REL606 unknown SRR098281        4.6 
4 ZDB429        10000 UC      REL606 unknown SRR098282        4.59
5 ZDB446        15000 UC      REL606 unknown SRR098283        4.66
6 ZDB458        20000 (C1,C2) REL606 unknown SRR098284        4.63</code></pre>
</div>
<p>The type of this object is ‘tibble’, a type of data frame we will
talk more about in the ‘dplyr’ section. If you needed a true R data
frame you could coerce with <code>as.data.frame()</code>.</p>
<div id="exercise-putting-it-all-together---data-frames" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-putting-it-all-together---data-frames" class="callout-inner">
<h3 class="callout-title">Exercise: Putting it all together - data
frames<a class="anchor" aria-label="anchor" href="#exercise-putting-it-all-together---data-frames"></a>
</h3>
<div class="callout-content">
<p><strong>Using the <code>Ecoli_metadata</code> data frame created
above, answer the following questions</strong></p>
<ol style="list-style-type: upper-alpha">
<li><p>What are the dimensions (# rows, # columns) of the data
frame?</p></li>
<li><p>What are categories are there in the <code>cit</code> column?
<em>hint</em>: treat column as factor</p></li>
<li><p>How many of each of the <code>cit</code> categories are
there?</p></li>
<li><p>What is the genome size for the 7th observation in this data
set?</p></li>
<li><p>What is the median value of the variable
<code>genome_size</code></p></li>
<li><p>Rename the column <code>sample</code> to
<code>sample_id</code></p></li>
<li><p>Create a new column (name genome_size_bp) and set it equal to the
genome_size multiplied by 1,000,000</p></li>
<li><p>Save the edited Ecoli_metadata data frame as
“exercise_solution.csv” in your current working directory.</p></li>
</ol>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" data-bs-parent="#accordionSolution4" aria-labelledby="headingSolution4">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb88">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">dim</span><span class="op">(</span><span class="va">Ecoli_metadata</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 30  7</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb90">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">levels</span><span class="op">(</span><span class="fu">as.factor</span><span class="op">(</span><span class="va">Ecoli_metadata</span><span class="op">$</span><span class="va">cit</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "minus"   "plus"    "unknown"</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb92">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">table</span><span class="op">(</span><span class="fu">as.factor</span><span class="op">(</span><span class="va">Ecoli_metadata</span><span class="op">$</span><span class="va">cit</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
  minus    plus unknown 
      9       9      12 </code></pre>
</div>
<div class="codewrapper sourceCode" id="cb94">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">Ecoli_metadata</span><span class="op">[</span><span class="fl">7</span>,<span class="fl">7</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 1 × 1
  genome_size
        &lt;dbl&gt;
1        4.62</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb96">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">median</span><span class="op">(</span><span class="va">Ecoli_metadata</span><span class="op">$</span><span class="va">genome_size</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 4.625</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb98">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">colnames</span><span class="op">(</span><span class="va">Ecoli_metadata</span><span class="op">)</span><span class="op">[</span><span class="fu">colnames</span><span class="op">(</span><span class="va">Ecoli_metadata</span><span class="op">)</span> <span class="op">==</span> <span class="st">"sample"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"sample_id"</span></span>
<span><span class="va">Ecoli_metadata</span><span class="op">$</span><span class="va">genome_size_bp</span> <span class="op">&lt;-</span> <span class="va">Ecoli_metadata</span><span class="op">$</span><span class="va">genome_size</span> <span class="op">*</span> <span class="fl">1000000</span></span>
<span><span class="fu">write.csv</span><span class="op">(</span><span class="va">Ecoli_metadata</span>, file <span class="op">=</span> <span class="st">"exercise_solution.csv"</span><span class="op">)</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Keypoints<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>It is easy to import data into R from tabular formats including
Excel. However, you still need to check that R has imported and
interpreted your data correctly</li>
<li>There are best practices for organizing your data (keeping it tidy)
and R is great for this</li>
<li>Base R has many useful functions for manipulating your data, but all
of R’s capabilities are greatly enhanced by software packages developed
by the community</li>
</ul>
</div>
</div>
</div>
</section></section><section id="aio-04-bioconductor-vcfr"><p>Content from <a href="04-bioconductor-vcfr.html">Using packages from Bioconductor</a></p>
<hr>
<p> Last updated on 2023-05-01 | 
        
        <a href="https://github.com/datacarpentry/genomics-r-intro/edit/main/episodes/04-bioconductor-vcfr.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How do I use packages from the Bioconductor repository?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Describe what the Bioconductor repository is and what it is used
for</li>
<li>Describe how Bioconductor differs from CRAN</li>
<li>Search Bioconductor for relevant packages</li>
<li>Install a package from Bioconductor</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section id="installing-packages-from-somewhere-else-besides-cran"><h2 class="section-heading">Installing packages from somewhere else besides CRAN?<a class="anchor" aria-label="anchor" href="#installing-packages-from-somewhere-else-besides-cran"></a>
</h2>
<hr class="half-width">
<p>So far we have told you about using packages that are included in the
base installation of R (this is what comes with R ‘out of the box’), and
packages that you can install from <a href="https://cran.r-project.org/" class="external-link">CRAN</a> (the Comprehensive R Archive
Network), which is the primary place many people look for supplemental R
packages to install. However, not all R packages are available on CRAN.
For bioinformatics-related packages in particular, there is another
repository that has many powerful packages that you can install. It is
called <a href="https://bioconductor.org/" class="external-link">Bioconductor</a> and it is a
repository specifically focused on bioinformatics packages. <a href="https://bioconductor.org/" class="external-link">Bioconductor</a> has a mission of
“promot[ing] the statistical analysis and comprehension of current and
emerging high-throughput biological assays.” This means that many if not
all of the packages available on Bioconductor are focused on the
analysis of biological data, and that it can be a great place to look
for tools to help you analyze your -omics datasets!</p>
</section><section id="so-how-do-i-use-it"><h2 class="section-heading">So how do I use it?<a class="anchor" aria-label="anchor" href="#so-how-do-i-use-it"></a>
</h2>
<hr class="half-width">
<p>Since access to the <a href="https://bioconductor.org/" class="external-link">Bioconductor</a> repository is not
built in to base R ‘out of the box’, there are a couple steps needed to
install packages from this alternative source. We will work through the
steps (only 2!) to install a package to help with the VCF analysis we
are working on, but you can use the same approach to install any of the
many thousands of available packages.</p>
<figure><img src="fig/bioconductor_website_screenshot.jpg" alt="screenshot of bioconductor homepage" class="figure mx-auto d-block"></figure></section><section id="first-install-the-biocmanager-package"><h2 class="section-heading">First, install the <code>BiocManager</code> package<a class="anchor" aria-label="anchor" href="#first-install-the-biocmanager-package"></a>
</h2>
<hr class="half-width">
<p>The first step is to install a package that <em>is</em> on CRAN,
<code>BiocManager</code>. This package will allow us to use it to
install packages from Bioconductor. You can think of Bioconductor kind
of like an alternative app store for your phone, except instead of apps
you are installing packages, and instead of your phone it’s your local R
package library.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install the BiocManager from CRAN using the base R install.packages() function</span></span>
<span><span class="fu">install.packages</span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span></code></pre>
</div>
<p>To check if this worked (and also so you can make a note of the
version for reproducibility purposes), you can run
<code>BiocManager::version()</code> and it should give you the version
number.</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># to make sure it worked, check the version</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">version</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
</section><section id="second-install-the-vcfr-package-from-bioconductor-using-biocmanager"><h2 class="section-heading">Second, install the vcfR package from Bioconductor using
<code>BiocManager</code><a class="anchor" aria-label="anchor" href="#second-install-the-vcfr-package-from-bioconductor-using-biocmanager"></a>
</h2>
<hr class="half-width">
<div id="heads-up-installing-vcfr-may-take-a-while-due-to-numerous-dependencies" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="heads-up-installing-vcfr-may-take-a-while-due-to-numerous-dependencies" class="callout-inner">
<h3 class="callout-title">Head’s Up: Installing vcfR may take a while
due to numerous dependencies<a class="anchor" aria-label="anchor" href="#heads-up-installing-vcfr-may-take-a-while-due-to-numerous-dependencies"></a>
</h3>
<div class="callout-content">
<p>Just be aware that installing packages that have many dependencies
can take a while.</p>
</div>
</div>
</div>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install the vcfR package from bioconductor using BiocManager::install()</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="st">"vcfR"</span><span class="op">)</span></span></code></pre>
</div>
<p>Depending on your particular system, you may need to also allow it to
install some dependencies or update installed packages in order to
successfully complete the process.</p>
<div id="note-installing-packages-from-bioconductor-vs-from-cran" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="note-installing-packages-from-bioconductor-vs-from-cran" class="callout-inner">
<h3 class="callout-title">Note: Installing packages from Bioconductor vs
from CRAN<a class="anchor" aria-label="anchor" href="#note-installing-packages-from-bioconductor-vs-from-cran"></a>
</h3>
<div class="callout-content">
<p>Some packages begin by being available only on Bioconductor, and then
later move to CRAN. <code>vcfR</code> is one such package, which
originally was only available from Bioconductor, but is currently
available from CRAN. The other thing to know is that
<code>BiocManager::install()</code> will also install packages from CRAN
(it is a wrapper around <code>install.packages()</code> that adds some
extra features). There are other benefits to using
<code>BiocManager::install()</code> for Bioconductor packages, many of
which are outlined <a href="https://www.bioconductor.org/install/" class="external-link">here</a>. In short,
Bioconductor packages have a release cycle that is different from CRAN
and the <code>install()</code> function is aware of that difference, so
it helps to keep package versions in line with one another in a way that
doesn’t generally happen with the base R
<code>install.packages()</code>.</p>
</div>
</div>
</div>
</section><section id="search-for-bioconductor-packages-based-on-your-analysis-needs"><h2 class="section-heading">Search for Bioconductor packages based on your analysis needs<a class="anchor" aria-label="anchor" href="#search-for-bioconductor-packages-based-on-your-analysis-needs"></a>
</h2>
<hr class="half-width">
<p>While we are only focusing in this workshop on VCF analyses, there
are hundreds or thousands of different types of data and analyses that
bioinformaticians may want to work with. Sometimes you may get a new
dataset and not know exactly where to start with analyzing or
visualizing it. The Bioconductor package search view can be a great way
to browse through the packages that are available.</p>
<figure><img src="fig/bioconductor_search.jpg" alt="screenshot of bioconductor search" class="figure mx-auto d-block"></figure><div id="tip-searching-for-packages-on-the-bioconductor-website" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="tip-searching-for-packages-on-the-bioconductor-website" class="callout-inner">
<h3 class="callout-title">Tip: Searching for packages on the
Bioconductor website<a class="anchor" aria-label="anchor" href="#tip-searching-for-packages-on-the-bioconductor-website"></a>
</h3>
<div class="callout-content">
<p>There are several thousand packages available through the
Bioconductor website. It can be a bit of a challenge to find what you
want, but one helpful resource is the <a href="https://bioconductor.org/packages/release/BiocViews.html#___Software" class="external-link">package
search page</a>.</p>
</div>
</div>
</div>
<p>In bioinformatics, there are often many different tools that can be
used in a particular instance. The authors of <code>vcfR</code> have <a href="https://github.com/knausb/vcfR#software-that-produce-vcf-files" class="external-link">compiled
some of them</a>. One of those packages that <a href="https://bioconductor.org/packages/release/bioc/html/VariantAnnotation.html" class="external-link">is
available from Bioconductor</a> is called <code>VariantAnnotation</code>
and may also be of interest to those working with vcf files in R.</p>
<div id="challenge" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="challenge" class="callout-inner">
<h3 class="callout-title">Challenge<a class="anchor" aria-label="anchor" href="#challenge"></a>
</h3>
<div class="callout-content">
<ul>
<li>Use the <code>BiocManager::available()</code> function to see what
packages are available matching a search term.</li>
<li>Use the <a href="https://bioconductor.org/packages/release/BiocViews.html#___Software" class="external-link">biocViews</a>
interface to search for packages of interest.</li>
</ul>
<p>You may or may not want to try installing the package, since not all
dependencies always install easily. However, this will at least let you
see what is available.</p>
</div>
</div>
</div>
<div id="tip-refreshing-the-rstudio-package-view-after-installing" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="tip-refreshing-the-rstudio-package-view-after-installing" class="callout-inner">
<h3 class="callout-title">Tip: Refreshing the RStudio package view after
installing<a class="anchor" aria-label="anchor" href="#tip-refreshing-the-rstudio-package-view-after-installing"></a>
</h3>
<div class="callout-content">
<p>If you install a package from Bioconductor, you may need to refresh
the RStudio package view to see it in your list. You can do this by
clicking the “Refresh” button in the Packages pane of RStudio.</p>
</div>
</div>
</div>
</section><section id="resources"><h2 class="section-heading">Resources<a class="anchor" aria-label="anchor" href="#resources"></a>
</h2>
<hr class="half-width">
<ul>
<li><a href="https://bioconductor.org/" class="external-link">Bioconductor</a></li>
<li><a href="https://bioconductor.org/packages/release/BiocViews.html#___Software" class="external-link">Bioconductor
package search</a></li>
<li><a href="https://cran.r-project.org/" class="external-link">CRAN</a></li>
</ul>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Keypoints<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>Bioconductor is an alternative package repository for bioinformatics
packages.</li>
<li>Installing packages from Bioconductor requires a new method, since
it is not compatible with the <code>install.packages()</code> function
used for CRAN.</li>
<li>Check Bioconductor to see if there is a package relevant to your
analysis before writing code yourself.</li>
</ul>
</div>
</div>
</div>
</section></section><section id="aio-05-dplyr"><p>Content from <a href="05-dplyr.html">Data Wrangling and Analyses with Tidyverse</a></p>
<hr>
<p> Last updated on 2023-05-01 | 
        
        <a href="https://github.com/datacarpentry/genomics-r-intro/edit/main/episodes/05-dplyr.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How can I manipulate data frames without repeating myself?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Describe what the <code>dplyr</code> package in R is used for.</li>
<li>Apply common <code>dplyr</code> functions to manipulate data in
R.</li>
<li>Employ the ‘pipe’ operator to link together a sequence of
functions.</li>
<li>Employ the ‘mutate’ function to apply other chosen functions to
existing columns and create new columns of data.</li>
<li>Employ the ‘split-apply-combine’ concept to split the data into
groups, apply analysis to each group, and combine the results.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<p>Bracket subsetting is handy, but it can be cumbersome and difficult
to read, especially for complicated operations.</p>
<p>Luckily, the <a href="https://cran.r-project.org/package=dplyr" class="external-link"><code>dplyr</code></a>
package provides a number of very useful functions for manipulating data
frames in a way that will reduce repetition, reduce the probability of
making errors, and probably even save you some typing. As an added
bonus, you might even find the <code>dplyr</code> grammar easier to
read.</p>
<p>Here we’re going to cover some of the most commonly used functions as
well as using pipes (<code>%&gt;%</code>) to combine them:</p>
<ol style="list-style-type: decimal">
<li><code>glimpse()</code></li>
<li><code>select()</code></li>
<li><code>filter()</code></li>
<li><code>group_by()</code></li>
<li><code>summarize()</code></li>
<li><code>mutate()</code></li>
<li>
<code>pivot_longer</code> and <code>pivot_wider</code>
</li>
</ol>
<p>Packages in R are sets of additional functions that let you do more
stuff in R. The functions we’ve been using, like <code>str()</code>,
come built into R; packages give you access to more functions. You need
to install a package and then load it to be able to use it.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">install.packages</span><span class="op">(</span><span class="st">"dplyr"</span><span class="op">)</span> <span class="co">## installs dplyr package</span></span>
<span><span class="fu">install.packages</span><span class="op">(</span><span class="st">"tidyr"</span><span class="op">)</span> <span class="co">## installs tidyr package</span></span>
<span><span class="fu">install.packages</span><span class="op">(</span><span class="st">"ggplot2"</span><span class="op">)</span> <span class="co">## installs ggplot2 package</span></span>
<span><span class="fu">install.packages</span><span class="op">(</span><span class="st">"readr"</span><span class="op">)</span> <span class="co">## install readr package</span></span></code></pre>
</div>
<p>You might get asked to choose a CRAN mirror – this is asking you to
choose a site to download the package from. The choice doesn’t matter
too much; I’d recommend choosing the RStudio mirror.</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="st">"dplyr"</span><span class="op">)</span>          <span class="co">## loads in dplyr package to use</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="st">"tidyr"</span><span class="op">)</span>          <span class="co">## loads in tidyr package to use</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="st">"ggplot2"</span><span class="op">)</span>          <span class="co">## loads in ggplot2 package to use</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="st">"readr"</span><span class="op">)</span>          <span class="co">## load in readr package to use</span></span></code></pre>
</div>
<p>You only need to install a package once per computer, but you need to
load it every time you open a new R session and want to use that
package.</p>
<div id="tip-installing-packages" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="tip-installing-packages" class="callout-inner">
<h3 class="callout-title">Tip: Installing packages<a class="anchor" aria-label="anchor" href="#tip-installing-packages"></a>
</h3>
<div class="callout-content">
<p>It may be temping to install the <code>tidyverse</code> package, as
it contains many useful collection of packages for this lesson and
beyond. However, when teaching or following this lesson, we advise that
participants install <code>dplyr</code>, <code>readr</code>,
<code>ggplot2</code>, and <code>tidyr</code> individually as shown
above. Otherwise, a substaial amount of the lesson will be spend waiting
for the installation to complete.</p>
</div>
</div>
</div>
<section id="what-is-dplyr"><h2 class="section-heading">What is dplyr?<a class="anchor" aria-label="anchor" href="#what-is-dplyr"></a>
</h2>
<hr class="half-width">
<p>The package <code>dplyr</code> is a fairly new (2014) package that
tries to provide easy tools for the most common data manipulation tasks.
This package is also included in the <a href="https://www.tidyverse.org/" class="external-link"><code>tidyverse</code> package</a>,
which is a collection of eight different packages (<code>dplyr</code>,
<code>ggplot2</code>, <code>tibble</code>, <code>tidyr</code>,
<code>readr</code>, <code>purrr</code>, <code>stringr</code>, and
<code>forcats</code>). It is built to work directly with data frames.
The thinking behind it was largely inspired by the package
<code>plyr</code> which has been in use for some time but suffered from
being slow in some cases.<code>dplyr</code> addresses this by porting
much of the computation to C++. An additional feature is the ability to
work with data stored directly in an external database. The benefits of
doing this are that the data can be managed natively in a relational
database, queries can be conducted on that database, and only the
results of the query returned.</p>
<p>This addresses a common problem with R in that all operations are
conducted in memory and thus the amount of data you can work with is
limited by available memory. The database connections essentially remove
that limitation in that you can have a database that is over 100s of GB,
conduct queries on it directly and pull back just what you need for
analysis in R.</p>
<div class="section level3">
<h3 id="loading--csv-files-in-tidy-style">Loading .csv files in tidy style<a class="anchor" aria-label="anchor" href="#loading--csv-files-in-tidy-style"></a>
</h3>
<p>The Tidyverse’s <code>readr</code> package provides its own unique
way of loading .csv files in to R using <code>read_csv()</code>, which
is similar to <code>read.csv()</code>. <code>read_csv()</code> allows
users to load in their data faster, doesn’t create row names, and allows
you to access non-standard variable names (ie. variables that start with
numbers of contain spaces), and outputs your data on the R console in a
tidier way. In short, it’s a much friendlier way of loading in
potentially messy data.</p>
<p>Now let’s load our vcf .csv file using <code>read_csv()</code>:</p>
</div>
<div class="section level3">
<h3 id="taking-a-quick-look-at-data-frames">Taking a quick look at data frames<a class="anchor" aria-label="anchor" href="#taking-a-quick-look-at-data-frames"></a>
</h3>
<p>Similar to <code>str()</code>, which comes built into R,
<code>glimpse()</code> is a <code>dplyr</code> function that (as the
name suggests) gives a glimpse of the data frame.</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Rows: 801
Columns: 29
$ sample_id     &lt;chr&gt; "SRR2584863", "SRR2584863", "SRR2584863", "SRR2584863", …
$ CHROM         &lt;chr&gt; "CP000819.1", "CP000819.1", "CP000819.1", "CP000819.1", …
$ POS           &lt;dbl&gt; 9972, 263235, 281923, 433359, 473901, 648692, 1331794, 1…
$ ID            &lt;lgl&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, …
$ REF           &lt;chr&gt; "T", "G", "G", "CTTTTTTT", "CCGC", "C", "C", "G", "ACAGC…
$ ALT           &lt;chr&gt; "G", "T", "T", "CTTTTTTTT", "CCGCGC", "T", "A", "A", "AC…
$ QUAL          &lt;dbl&gt; 91.0000, 85.0000, 217.0000, 64.0000, 228.0000, 210.0000,…
$ FILTER        &lt;lgl&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, …
$ INDEL         &lt;lgl&gt; FALSE, FALSE, FALSE, TRUE, TRUE, FALSE, FALSE, FALSE, TR…
$ IDV           &lt;dbl&gt; NA, NA, NA, 12, 9, NA, NA, NA, 2, 7, NA, NA, NA, NA, NA,…
$ IMF           &lt;dbl&gt; NA, NA, NA, 1.000000, 0.900000, NA, NA, NA, 0.666667, 1.…
$ DP            &lt;dbl&gt; 4, 6, 10, 12, 10, 10, 8, 11, 3, 7, 9, 20, 12, 19, 15, 10…
$ VDB           &lt;dbl&gt; 0.0257451, 0.0961330, 0.7740830, 0.4777040, 0.6595050, 0…
$ RPB           &lt;dbl&gt; NA, 1.000000, NA, NA, NA, NA, NA, NA, NA, NA, 0.900802, …
$ MQB           &lt;dbl&gt; NA, 1.0000000, NA, NA, NA, NA, NA, NA, NA, NA, 0.1501340…
$ BQB           &lt;dbl&gt; NA, 1.000000, NA, NA, NA, NA, NA, NA, NA, NA, 0.750668, …
$ MQSB          &lt;dbl&gt; NA, NA, 0.974597, 1.000000, 0.916482, 0.916482, 0.900802…
$ SGB           &lt;dbl&gt; -0.556411, -0.590765, -0.662043, -0.676189, -0.662043, -…
$ MQ0F          &lt;dbl&gt; 0.000000, 0.166667, 0.000000, 0.000000, 0.000000, 0.0000…
$ ICB           &lt;lgl&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, …
$ HOB           &lt;lgl&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, …
$ AC            &lt;dbl&gt; 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,…
$ AN            &lt;dbl&gt; 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,…
$ DP4           &lt;chr&gt; "0,0,0,4", "0,1,0,5", "0,0,4,5", "0,1,3,8", "1,0,2,7", "…
$ MQ            &lt;dbl&gt; 60, 33, 60, 60, 60, 60, 60, 60, 60, 60, 25, 60, 10, 60, …
$ Indiv         &lt;chr&gt; "/home/dcuser/dc_workshop/results/bam/SRR2584863.aligned…
$ gt_PL         &lt;dbl&gt; 1210, 1120, 2470, 910, 2550, 2400, 2080, 2550, 11128, 19…
$ gt_GT         &lt;dbl&gt; 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,…
$ gt_GT_alleles &lt;chr&gt; "G", "T", "T", "CTTTTTTTT", "CCGCGC", "T", "A", "A", "AC…</code></pre>
</div>
<p>In the above output, we can already gather some information about
<code>variants</code>, such as the number of rows and columns, column
names, type of vector in the columns, and the first few entries of each
column. Although what we see is similar to outputs of
<code>str()</code>, this method gives a cleaner visual output.</p>
</div>
<div class="section level3">
<h3 id="selecting-columns-and-filtering-rows">Selecting columns and filtering rows<a class="anchor" aria-label="anchor" href="#selecting-columns-and-filtering-rows"></a>
</h3>
<p>To select columns of a data frame, use <code>select()</code>. The
first argument to this function is the data frame
(<code>variants</code>), and the subsequent arguments are the columns to
keep.</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">select</span><span class="op">(</span><span class="va">variants</span>, <span class="va">sample_id</span>, <span class="va">REF</span>, <span class="va">ALT</span>, <span class="va">DP</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 801 × 4
   sample_id  REF                              ALT                            DP
   &lt;chr&gt;      &lt;chr&gt;                            &lt;chr&gt;                       &lt;dbl&gt;
 1 SRR2584863 T                                G                               4
 2 SRR2584863 G                                T                               6
 3 SRR2584863 G                                T                              10
 4 SRR2584863 CTTTTTTT                         CTTTTTTTT                      12
 5 SRR2584863 CCGC                             CCGCGC                         10
 6 SRR2584863 C                                T                              10
 7 SRR2584863 C                                A                               8
 8 SRR2584863 G                                A                              11
 9 SRR2584863 ACAGCCAGCCAGCCAGCCAGCCAGCCAGCCAG ACAGCCAGCCAGCCAGCCAGCCAGCC…     3
10 SRR2584863 AT                               ATT                             7
# ℹ 791 more rows</code></pre>
</div>
<p>To select all columns <em>except</em> certain ones, put a “-” in
front of the variable to exclude it.</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">select</span><span class="op">(</span><span class="va">variants</span>, <span class="op">-</span><span class="va">CHROM</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 801 × 28
   sample_id      POS ID    REF      ALT    QUAL FILTER INDEL   IDV    IMF    DP
   &lt;chr&gt;        &lt;dbl&gt; &lt;lgl&gt; &lt;chr&gt;    &lt;chr&gt; &lt;dbl&gt; &lt;lgl&gt;  &lt;lgl&gt; &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt;
 1 SRR2584863    9972 NA    T        G        91 NA     FALSE    NA NA         4
 2 SRR2584863  263235 NA    G        T        85 NA     FALSE    NA NA         6
 3 SRR2584863  281923 NA    G        T       217 NA     FALSE    NA NA        10
 4 SRR2584863  433359 NA    CTTTTTTT CTTT…    64 NA     TRUE     12  1        12
 5 SRR2584863  473901 NA    CCGC     CCGC…   228 NA     TRUE      9  0.9      10
 6 SRR2584863  648692 NA    C        T       210 NA     FALSE    NA NA        10
 7 SRR2584863 1331794 NA    C        A       178 NA     FALSE    NA NA         8
 8 SRR2584863 1733343 NA    G        A       225 NA     FALSE    NA NA        11
 9 SRR2584863 2103887 NA    ACAGCCA… ACAG…    56 NA     TRUE      2  0.667     3
10 SRR2584863 2333538 NA    AT       ATT     167 NA     TRUE      7  1         7
# ℹ 791 more rows
# ℹ 17 more variables: VDB &lt;dbl&gt;, RPB &lt;dbl&gt;, MQB &lt;dbl&gt;, BQB &lt;dbl&gt;, MQSB &lt;dbl&gt;,
#   SGB &lt;dbl&gt;, MQ0F &lt;dbl&gt;, ICB &lt;lgl&gt;, HOB &lt;lgl&gt;, AC &lt;dbl&gt;, AN &lt;dbl&gt;, DP4 &lt;chr&gt;,
#   MQ &lt;dbl&gt;, Indiv &lt;chr&gt;, gt_PL &lt;dbl&gt;, gt_GT &lt;dbl&gt;, gt_GT_alleles &lt;chr&gt;</code></pre>
</div>
<p><code>dplyr</code> also provides useful functions to select columns
based on their names. For instance, <code>ends_with()</code> allows you
to select columns that ends with specific letters. For instance, if you
wanted to select columns that end with the letter “B”:</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">select</span><span class="op">(</span><span class="va">variants</span>, <span class="fu">ends_with</span><span class="op">(</span><span class="st">"B"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 801 × 8
      VDB   RPB   MQB   BQB   MQSB    SGB ICB   HOB  
    &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt; &lt;lgl&gt; &lt;lgl&gt;
 1 0.0257    NA    NA    NA NA     -0.556 NA    NA   
 2 0.0961     1     1     1 NA     -0.591 NA    NA   
 3 0.774     NA    NA    NA  0.975 -0.662 NA    NA   
 4 0.478     NA    NA    NA  1     -0.676 NA    NA   
 5 0.660     NA    NA    NA  0.916 -0.662 NA    NA   
 6 0.268     NA    NA    NA  0.916 -0.670 NA    NA   
 7 0.624     NA    NA    NA  0.901 -0.651 NA    NA   
 8 0.992     NA    NA    NA  1.01  -0.670 NA    NA   
 9 0.902     NA    NA    NA  1     -0.454 NA    NA   
10 0.568     NA    NA    NA  1.01  -0.617 NA    NA   
# ℹ 791 more rows</code></pre>
</div>
<div id="challenge" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge" class="callout-inner">
<h3 class="callout-title">Challenge<a class="anchor" aria-label="anchor" href="#challenge"></a>
</h3>
<div class="callout-content">
<p>Create a table that contains all the columns with the letter “i” and
column “POS”, without columns “Indiv” and “FILTER”. Hint: look at for a
function called <code>contains()</code>, which can be found in the help
documentation for ends with we just covered (<code>?ends_with</code>).
Note that contains() is not case sensistive.</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># First, we select "POS" and all columns with letter "i". This will contain columns Indiv and FILTER. </span></span>
<span><span class="va">variants_subset</span> <span class="op">&lt;-</span> <span class="fu">select</span><span class="op">(</span><span class="va">variants</span>, <span class="va">POS</span>, <span class="fu">contains</span><span class="op">(</span><span class="st">"i"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># Next, we remove columns Indiv and FILTER</span></span>
<span><span class="va">variants_result</span> <span class="op">&lt;-</span> <span class="fu">select</span><span class="op">(</span><span class="va">variants_subset</span>, <span class="op">-</span><span class="va">Indiv</span>, <span class="op">-</span><span class="va">FILTER</span><span class="op">)</span></span>
<span><span class="va">variants_result</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 801 × 7
       POS sample_id  ID    INDEL   IDV    IMF ICB  
     &lt;dbl&gt; &lt;chr&gt;      &lt;lgl&gt; &lt;lgl&gt; &lt;dbl&gt;  &lt;dbl&gt; &lt;lgl&gt;
 1    9972 SRR2584863 NA    FALSE    NA NA     NA   
 2  263235 SRR2584863 NA    FALSE    NA NA     NA   
 3  281923 SRR2584863 NA    FALSE    NA NA     NA   
 4  433359 SRR2584863 NA    TRUE     12  1     NA   
 5  473901 SRR2584863 NA    TRUE      9  0.9   NA   
 6  648692 SRR2584863 NA    FALSE    NA NA     NA   
 7 1331794 SRR2584863 NA    FALSE    NA NA     NA   
 8 1733343 SRR2584863 NA    FALSE    NA NA     NA   
 9 2103887 SRR2584863 NA    TRUE      2  0.667 NA   
10 2333538 SRR2584863 NA    TRUE      7  1     NA   
# ℹ 791 more rows</code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="challenge" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge" class="callout-inner">
<h3 class="callout-title">Challenge<em> (continued)</em><a class="anchor" aria-label="anchor" href="#challenge"></a>
</h3>
<div class="callout-content">
<p>We can also get to <code>variants_result</code> in one line of
code:</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2">
  Alternative solution
  </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">variants_result</span> <span class="op">&lt;-</span> <span class="fu">select</span><span class="op">(</span><span class="va">variants</span>, <span class="va">POS</span>, <span class="fu">contains</span><span class="op">(</span><span class="st">"i"</span><span class="op">)</span>, <span class="op">-</span><span class="va">Indiv</span>, <span class="op">-</span><span class="va">FILTER</span><span class="op">)</span></span>
<span><span class="va">variants_result</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 801 × 7
       POS sample_id  ID    INDEL   IDV    IMF ICB  
     &lt;dbl&gt; &lt;chr&gt;      &lt;lgl&gt; &lt;lgl&gt; &lt;dbl&gt;  &lt;dbl&gt; &lt;lgl&gt;
 1    9972 SRR2584863 NA    FALSE    NA NA     NA   
 2  263235 SRR2584863 NA    FALSE    NA NA     NA   
 3  281923 SRR2584863 NA    FALSE    NA NA     NA   
 4  433359 SRR2584863 NA    TRUE     12  1     NA   
 5  473901 SRR2584863 NA    TRUE      9  0.9   NA   
 6  648692 SRR2584863 NA    FALSE    NA NA     NA   
 7 1331794 SRR2584863 NA    FALSE    NA NA     NA   
 8 1733343 SRR2584863 NA    FALSE    NA NA     NA   
 9 2103887 SRR2584863 NA    TRUE      2  0.667 NA   
10 2333538 SRR2584863 NA    TRUE      7  1     NA   
# ℹ 791 more rows</code></pre>
</div>
</div>
</div>
</div>
</div>
<p>To choose rows, use <code>filter()</code>:</p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">filter</span><span class="op">(</span><span class="va">variants</span>, <span class="va">sample_id</span> <span class="op">==</span> <span class="st">"SRR2584863"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 25 × 29
   sample_id  CHROM        POS ID    REF   ALT    QUAL FILTER INDEL   IDV    IMF
   &lt;chr&gt;      &lt;chr&gt;      &lt;dbl&gt; &lt;lgl&gt; &lt;chr&gt; &lt;chr&gt; &lt;dbl&gt; &lt;lgl&gt;  &lt;lgl&gt; &lt;dbl&gt;  &lt;dbl&gt;
 1 SRR2584863 CP000819… 9.97e3 NA    T     G        91 NA     FALSE    NA NA    
 2 SRR2584863 CP000819… 2.63e5 NA    G     T        85 NA     FALSE    NA NA    
 3 SRR2584863 CP000819… 2.82e5 NA    G     T       217 NA     FALSE    NA NA    
 4 SRR2584863 CP000819… 4.33e5 NA    CTTT… CTTT…    64 NA     TRUE     12  1    
 5 SRR2584863 CP000819… 4.74e5 NA    CCGC  CCGC…   228 NA     TRUE      9  0.9  
 6 SRR2584863 CP000819… 6.49e5 NA    C     T       210 NA     FALSE    NA NA    
 7 SRR2584863 CP000819… 1.33e6 NA    C     A       178 NA     FALSE    NA NA    
 8 SRR2584863 CP000819… 1.73e6 NA    G     A       225 NA     FALSE    NA NA    
 9 SRR2584863 CP000819… 2.10e6 NA    ACAG… ACAG…    56 NA     TRUE      2  0.667
10 SRR2584863 CP000819… 2.33e6 NA    AT    ATT     167 NA     TRUE      7  1    
# ℹ 15 more rows
# ℹ 18 more variables: DP &lt;dbl&gt;, VDB &lt;dbl&gt;, RPB &lt;dbl&gt;, MQB &lt;dbl&gt;, BQB &lt;dbl&gt;,
#   MQSB &lt;dbl&gt;, SGB &lt;dbl&gt;, MQ0F &lt;dbl&gt;, ICB &lt;lgl&gt;, HOB &lt;lgl&gt;, AC &lt;dbl&gt;,
#   AN &lt;dbl&gt;, DP4 &lt;chr&gt;, MQ &lt;dbl&gt;, Indiv &lt;chr&gt;, gt_PL &lt;dbl&gt;, gt_GT &lt;dbl&gt;,
#   gt_GT_alleles &lt;chr&gt;</code></pre>
</div>
<p><code>filter()</code> will keep all the rows that match the
conditions that are provided. Here are a few examples:</p>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># rows for which the reference genome has T or G</span></span>
<span><span class="fu">filter</span><span class="op">(</span><span class="va">variants</span>, <span class="va">REF</span> <span class="op">%in%</span> <span class="fu">c</span><span class="op">(</span><span class="st">"T"</span>, <span class="st">"G"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 340 × 29
   sample_id CHROM    POS ID    REF   ALT    QUAL FILTER INDEL   IDV   IMF    DP
   &lt;chr&gt;     &lt;chr&gt;  &lt;dbl&gt; &lt;lgl&gt; &lt;chr&gt; &lt;chr&gt; &lt;dbl&gt; &lt;lgl&gt;  &lt;lgl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
 1 SRR25848… CP00… 9.97e3 NA    T     G      91   NA     FALSE    NA    NA     4
 2 SRR25848… CP00… 2.63e5 NA    G     T      85   NA     FALSE    NA    NA     6
 3 SRR25848… CP00… 2.82e5 NA    G     T     217   NA     FALSE    NA    NA    10
 4 SRR25848… CP00… 1.73e6 NA    G     A     225   NA     FALSE    NA    NA    11
 5 SRR25848… CP00… 2.62e6 NA    G     T      31.9 NA     FALSE    NA    NA    12
 6 SRR25848… CP00… 3.00e6 NA    G     A     225   NA     FALSE    NA    NA    15
 7 SRR25848… CP00… 3.91e6 NA    G     T     225   NA     FALSE    NA    NA    10
 8 SRR25848… CP00… 9.97e3 NA    T     G     214   NA     FALSE    NA    NA    10
 9 SRR25848… CP00… 1.06e4 NA    G     A     225   NA     FALSE    NA    NA    11
10 SRR25848… CP00… 6.40e4 NA    G     A     225   NA     FALSE    NA    NA    18
# ℹ 330 more rows
# ℹ 17 more variables: VDB &lt;dbl&gt;, RPB &lt;dbl&gt;, MQB &lt;dbl&gt;, BQB &lt;dbl&gt;, MQSB &lt;dbl&gt;,
#   SGB &lt;dbl&gt;, MQ0F &lt;dbl&gt;, ICB &lt;lgl&gt;, HOB &lt;lgl&gt;, AC &lt;dbl&gt;, AN &lt;dbl&gt;, DP4 &lt;chr&gt;,
#   MQ &lt;dbl&gt;, Indiv &lt;chr&gt;, gt_PL &lt;dbl&gt;, gt_GT &lt;dbl&gt;, gt_GT_alleles &lt;chr&gt;</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># rows that have TRUE in the column INDEL</span></span>
<span><span class="fu">filter</span><span class="op">(</span><span class="va">variants</span>, <span class="va">INDEL</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 101 × 29
   sample_id CHROM    POS ID    REF   ALT    QUAL FILTER INDEL   IDV   IMF    DP
   &lt;chr&gt;     &lt;chr&gt;  &lt;dbl&gt; &lt;lgl&gt; &lt;chr&gt; &lt;chr&gt; &lt;dbl&gt; &lt;lgl&gt;  &lt;lgl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
 1 SRR25848… CP00… 4.33e5 NA    CTTT… CTTT…  64   NA     TRUE     12 1        12
 2 SRR25848… CP00… 4.74e5 NA    CCGC  CCGC… 228   NA     TRUE      9 0.9      10
 3 SRR25848… CP00… 2.10e6 NA    ACAG… ACAG…  56   NA     TRUE      2 0.667     3
 4 SRR25848… CP00… 2.33e6 NA    AT    ATT   167   NA     TRUE      7 1         7
 5 SRR25848… CP00… 3.90e6 NA    A     AC     43.4 NA     TRUE      2 1         2
 6 SRR25848… CP00… 4.43e6 NA    TGG   T     228   NA     TRUE     10 1        10
 7 SRR25848… CP00… 1.48e5 NA    AGGGG AGGG… 122   NA     TRUE      8 1         8
 8 SRR25848… CP00… 1.58e5 NA    GTTT… GTTT…  19.5 NA     TRUE      6 1         6
 9 SRR25848… CP00… 1.73e5 NA    CAA   CA    180   NA     TRUE     11 1        11
10 SRR25848… CP00… 1.75e5 NA    GAA   GA    194   NA     TRUE     10 1        10
# ℹ 91 more rows
# ℹ 17 more variables: VDB &lt;dbl&gt;, RPB &lt;dbl&gt;, MQB &lt;dbl&gt;, BQB &lt;dbl&gt;, MQSB &lt;dbl&gt;,
#   SGB &lt;dbl&gt;, MQ0F &lt;dbl&gt;, ICB &lt;lgl&gt;, HOB &lt;lgl&gt;, AC &lt;dbl&gt;, AN &lt;dbl&gt;, DP4 &lt;chr&gt;,
#   MQ &lt;dbl&gt;, Indiv &lt;chr&gt;, gt_PL &lt;dbl&gt;, gt_GT &lt;dbl&gt;, gt_GT_alleles &lt;chr&gt;</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># rows that don't have missing data in the IDV column</span></span>
<span><span class="fu">filter</span><span class="op">(</span><span class="va">variants</span>, <span class="op">!</span><span class="fu">is.na</span><span class="op">(</span><span class="va">IDV</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 101 × 29
   sample_id CHROM    POS ID    REF   ALT    QUAL FILTER INDEL   IDV   IMF    DP
   &lt;chr&gt;     &lt;chr&gt;  &lt;dbl&gt; &lt;lgl&gt; &lt;chr&gt; &lt;chr&gt; &lt;dbl&gt; &lt;lgl&gt;  &lt;lgl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
 1 SRR25848… CP00… 4.33e5 NA    CTTT… CTTT…  64   NA     TRUE     12 1        12
 2 SRR25848… CP00… 4.74e5 NA    CCGC  CCGC… 228   NA     TRUE      9 0.9      10
 3 SRR25848… CP00… 2.10e6 NA    ACAG… ACAG…  56   NA     TRUE      2 0.667     3
 4 SRR25848… CP00… 2.33e6 NA    AT    ATT   167   NA     TRUE      7 1         7
 5 SRR25848… CP00… 3.90e6 NA    A     AC     43.4 NA     TRUE      2 1         2
 6 SRR25848… CP00… 4.43e6 NA    TGG   T     228   NA     TRUE     10 1        10
 7 SRR25848… CP00… 1.48e5 NA    AGGGG AGGG… 122   NA     TRUE      8 1         8
 8 SRR25848… CP00… 1.58e5 NA    GTTT… GTTT…  19.5 NA     TRUE      6 1         6
 9 SRR25848… CP00… 1.73e5 NA    CAA   CA    180   NA     TRUE     11 1        11
10 SRR25848… CP00… 1.75e5 NA    GAA   GA    194   NA     TRUE     10 1        10
# ℹ 91 more rows
# ℹ 17 more variables: VDB &lt;dbl&gt;, RPB &lt;dbl&gt;, MQB &lt;dbl&gt;, BQB &lt;dbl&gt;, MQSB &lt;dbl&gt;,
#   SGB &lt;dbl&gt;, MQ0F &lt;dbl&gt;, ICB &lt;lgl&gt;, HOB &lt;lgl&gt;, AC &lt;dbl&gt;, AN &lt;dbl&gt;, DP4 &lt;chr&gt;,
#   MQ &lt;dbl&gt;, Indiv &lt;chr&gt;, gt_PL &lt;dbl&gt;, gt_GT &lt;dbl&gt;, gt_GT_alleles &lt;chr&gt;</code></pre>
</div>
<p>We have a column titled “QUAL”. This is a Phred-scaled confidence
score that a polymorphism exists at this position given the sequencing
data. Lower QUAL scores indicate low probability of a polymorphism
existing at that site. <code>filter()</code> can be useful for selecting
mutations that have a QUAL score above a certain threshold:</p>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># rows with QUAL values greater than or equal to 100</span></span>
<span><span class="fu">filter</span><span class="op">(</span><span class="va">variants</span>, <span class="va">QUAL</span> <span class="op">&gt;=</span> <span class="fl">100</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 666 × 29
   sample_id CHROM    POS ID    REF   ALT    QUAL FILTER INDEL   IDV   IMF    DP
   &lt;chr&gt;     &lt;chr&gt;  &lt;dbl&gt; &lt;lgl&gt; &lt;chr&gt; &lt;chr&gt; &lt;dbl&gt; &lt;lgl&gt;  &lt;lgl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
 1 SRR25848… CP00… 2.82e5 NA    G     T       217 NA     FALSE    NA  NA      10
 2 SRR25848… CP00… 4.74e5 NA    CCGC  CCGC…   228 NA     TRUE      9   0.9    10
 3 SRR25848… CP00… 6.49e5 NA    C     T       210 NA     FALSE    NA  NA      10
 4 SRR25848… CP00… 1.33e6 NA    C     A       178 NA     FALSE    NA  NA       8
 5 SRR25848… CP00… 1.73e6 NA    G     A       225 NA     FALSE    NA  NA      11
 6 SRR25848… CP00… 2.33e6 NA    AT    ATT     167 NA     TRUE      7   1       7
 7 SRR25848… CP00… 2.41e6 NA    A     C       104 NA     FALSE    NA  NA       9
 8 SRR25848… CP00… 2.45e6 NA    A     C       225 NA     FALSE    NA  NA      20
 9 SRR25848… CP00… 2.67e6 NA    A     T       225 NA     FALSE    NA  NA      19
10 SRR25848… CP00… 3.00e6 NA    G     A       225 NA     FALSE    NA  NA      15
# ℹ 656 more rows
# ℹ 17 more variables: VDB &lt;dbl&gt;, RPB &lt;dbl&gt;, MQB &lt;dbl&gt;, BQB &lt;dbl&gt;, MQSB &lt;dbl&gt;,
#   SGB &lt;dbl&gt;, MQ0F &lt;dbl&gt;, ICB &lt;lgl&gt;, HOB &lt;lgl&gt;, AC &lt;dbl&gt;, AN &lt;dbl&gt;, DP4 &lt;chr&gt;,
#   MQ &lt;dbl&gt;, Indiv &lt;chr&gt;, gt_PL &lt;dbl&gt;, gt_GT &lt;dbl&gt;, gt_GT_alleles &lt;chr&gt;</code></pre>
</div>
<p><code>filter()</code> allows you to combine multiple conditions. You
can separate them using a <code>,</code> as arguments to the function,
they will be combined using the <code>&amp;</code> (AND) logical
operator. If you need to use the <code>|</code> (OR) logical operator,
you can specify it explicitly:</p>
<div class="codewrapper sourceCode" id="cb24">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># this is equivalent to:</span></span>
<span><span class="co">#   filter(variants, sample_id == "SRR2584863" &amp; QUAL &gt;= 100)</span></span>
<span><span class="fu">filter</span><span class="op">(</span><span class="va">variants</span>, <span class="va">sample_id</span> <span class="op">==</span> <span class="st">"SRR2584863"</span>, <span class="va">QUAL</span> <span class="op">&gt;=</span> <span class="fl">100</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 19 × 29
   sample_id CHROM    POS ID    REF   ALT    QUAL FILTER INDEL   IDV   IMF    DP
   &lt;chr&gt;     &lt;chr&gt;  &lt;dbl&gt; &lt;lgl&gt; &lt;chr&gt; &lt;chr&gt; &lt;dbl&gt; &lt;lgl&gt;  &lt;lgl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
 1 SRR25848… CP00… 2.82e5 NA    G     T       217 NA     FALSE    NA  NA      10
 2 SRR25848… CP00… 4.74e5 NA    CCGC  CCGC…   228 NA     TRUE      9   0.9    10
 3 SRR25848… CP00… 6.49e5 NA    C     T       210 NA     FALSE    NA  NA      10
 4 SRR25848… CP00… 1.33e6 NA    C     A       178 NA     FALSE    NA  NA       8
 5 SRR25848… CP00… 1.73e6 NA    G     A       225 NA     FALSE    NA  NA      11
 6 SRR25848… CP00… 2.33e6 NA    AT    ATT     167 NA     TRUE      7   1       7
 7 SRR25848… CP00… 2.41e6 NA    A     C       104 NA     FALSE    NA  NA       9
 8 SRR25848… CP00… 2.45e6 NA    A     C       225 NA     FALSE    NA  NA      20
 9 SRR25848… CP00… 2.67e6 NA    A     T       225 NA     FALSE    NA  NA      19
10 SRR25848… CP00… 3.00e6 NA    G     A       225 NA     FALSE    NA  NA      15
11 SRR25848… CP00… 3.34e6 NA    A     C       211 NA     FALSE    NA  NA      10
12 SRR25848… CP00… 3.40e6 NA    C     A       225 NA     FALSE    NA  NA      14
13 SRR25848… CP00… 3.48e6 NA    A     G       200 NA     FALSE    NA  NA       9
14 SRR25848… CP00… 3.49e6 NA    A     C       225 NA     FALSE    NA  NA      13
15 SRR25848… CP00… 3.91e6 NA    G     T       225 NA     FALSE    NA  NA      10
16 SRR25848… CP00… 4.10e6 NA    A     G       225 NA     FALSE    NA  NA      16
17 SRR25848… CP00… 4.20e6 NA    A     C       225 NA     FALSE    NA  NA      11
18 SRR25848… CP00… 4.43e6 NA    TGG   T       228 NA     TRUE     10   1      10
19 SRR25848… CP00… 4.62e6 NA    A     C       185 NA     FALSE    NA  NA       9
# ℹ 17 more variables: VDB &lt;dbl&gt;, RPB &lt;dbl&gt;, MQB &lt;dbl&gt;, BQB &lt;dbl&gt;, MQSB &lt;dbl&gt;,
#   SGB &lt;dbl&gt;, MQ0F &lt;dbl&gt;, ICB &lt;lgl&gt;, HOB &lt;lgl&gt;, AC &lt;dbl&gt;, AN &lt;dbl&gt;, DP4 &lt;chr&gt;,
#   MQ &lt;dbl&gt;, Indiv &lt;chr&gt;, gt_PL &lt;dbl&gt;, gt_GT &lt;dbl&gt;, gt_GT_alleles &lt;chr&gt;</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb26">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># using `|` logical operator</span></span>
<span><span class="fu">filter</span><span class="op">(</span><span class="va">variants</span>, <span class="va">sample_id</span> <span class="op">==</span> <span class="st">"SRR2584863"</span>, <span class="op">(</span><span class="va">MQ</span> <span class="op">&gt;=</span> <span class="fl">50</span> <span class="op">|</span> <span class="va">QUAL</span> <span class="op">&gt;=</span> <span class="fl">100</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 23 × 29
   sample_id  CHROM        POS ID    REF   ALT    QUAL FILTER INDEL   IDV    IMF
   &lt;chr&gt;      &lt;chr&gt;      &lt;dbl&gt; &lt;lgl&gt; &lt;chr&gt; &lt;chr&gt; &lt;dbl&gt; &lt;lgl&gt;  &lt;lgl&gt; &lt;dbl&gt;  &lt;dbl&gt;
 1 SRR2584863 CP000819… 9.97e3 NA    T     G        91 NA     FALSE    NA NA    
 2 SRR2584863 CP000819… 2.82e5 NA    G     T       217 NA     FALSE    NA NA    
 3 SRR2584863 CP000819… 4.33e5 NA    CTTT… CTTT…    64 NA     TRUE     12  1    
 4 SRR2584863 CP000819… 4.74e5 NA    CCGC  CCGC…   228 NA     TRUE      9  0.9  
 5 SRR2584863 CP000819… 6.49e5 NA    C     T       210 NA     FALSE    NA NA    
 6 SRR2584863 CP000819… 1.33e6 NA    C     A       178 NA     FALSE    NA NA    
 7 SRR2584863 CP000819… 1.73e6 NA    G     A       225 NA     FALSE    NA NA    
 8 SRR2584863 CP000819… 2.10e6 NA    ACAG… ACAG…    56 NA     TRUE      2  0.667
 9 SRR2584863 CP000819… 2.33e6 NA    AT    ATT     167 NA     TRUE      7  1    
10 SRR2584863 CP000819… 2.41e6 NA    A     C       104 NA     FALSE    NA NA    
# ℹ 13 more rows
# ℹ 18 more variables: DP &lt;dbl&gt;, VDB &lt;dbl&gt;, RPB &lt;dbl&gt;, MQB &lt;dbl&gt;, BQB &lt;dbl&gt;,
#   MQSB &lt;dbl&gt;, SGB &lt;dbl&gt;, MQ0F &lt;dbl&gt;, ICB &lt;lgl&gt;, HOB &lt;lgl&gt;, AC &lt;dbl&gt;,
#   AN &lt;dbl&gt;, DP4 &lt;chr&gt;, MQ &lt;dbl&gt;, Indiv &lt;chr&gt;, gt_PL &lt;dbl&gt;, gt_GT &lt;dbl&gt;,
#   gt_GT_alleles &lt;chr&gt;</code></pre>
</div>
<div id="challenge-1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-1" class="callout-inner">
<h3 class="callout-title">Challenge<a class="anchor" aria-label="anchor" href="#challenge-1"></a>
</h3>
<div class="callout-content">
<p>Select all the mutations that occurred between the positions 1e6 (one
million) and 2e6 (inclusive) that have a QUAL greater than 200, and
exclude INDEL mutations. Hint: to flip logical values such as TRUE to a
FALSE, we can use to negation symbol “!”. (eg. !TRUE == FALSE).</p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" data-bs-parent="#accordionSolution3" aria-labelledby="headingSolution3">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb28">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">filter</span><span class="op">(</span><span class="va">variants</span>, <span class="va">POS</span> <span class="op">&gt;=</span> <span class="fl">1e6</span> <span class="op">&amp;</span> <span class="va">POS</span> <span class="op">&lt;=</span> <span class="fl">2e6</span>, <span class="va">QUAL</span> <span class="op">&gt;</span> <span class="fl">200</span>, <span class="op">!</span><span class="va">INDEL</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 77 × 29
   sample_id CHROM    POS ID    REF   ALT    QUAL FILTER INDEL   IDV   IMF    DP
   &lt;chr&gt;     &lt;chr&gt;  &lt;dbl&gt; &lt;lgl&gt; &lt;chr&gt; &lt;chr&gt; &lt;dbl&gt; &lt;lgl&gt;  &lt;lgl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
 1 SRR25848… CP00… 1.73e6 NA    G     A       225 NA     FALSE    NA    NA    11
 2 SRR25848… CP00… 1.00e6 NA    A     G       225 NA     FALSE    NA    NA    15
 3 SRR25848… CP00… 1.02e6 NA    A     G       225 NA     FALSE    NA    NA    12
 4 SRR25848… CP00… 1.06e6 NA    C     T       225 NA     FALSE    NA    NA    17
 5 SRR25848… CP00… 1.06e6 NA    A     G       206 NA     FALSE    NA    NA     9
 6 SRR25848… CP00… 1.07e6 NA    G     T       225 NA     FALSE    NA    NA    11
 7 SRR25848… CP00… 1.07e6 NA    T     C       225 NA     FALSE    NA    NA    12
 8 SRR25848… CP00… 1.10e6 NA    C     T       225 NA     FALSE    NA    NA    15
 9 SRR25848… CP00… 1.11e6 NA    C     T       212 NA     FALSE    NA    NA     9
10 SRR25848… CP00… 1.11e6 NA    A     G       225 NA     FALSE    NA    NA    14
# ℹ 67 more rows
# ℹ 17 more variables: VDB &lt;dbl&gt;, RPB &lt;dbl&gt;, MQB &lt;dbl&gt;, BQB &lt;dbl&gt;, MQSB &lt;dbl&gt;,
#   SGB &lt;dbl&gt;, MQ0F &lt;dbl&gt;, ICB &lt;lgl&gt;, HOB &lt;lgl&gt;, AC &lt;dbl&gt;, AN &lt;dbl&gt;, DP4 &lt;chr&gt;,
#   MQ &lt;dbl&gt;, Indiv &lt;chr&gt;, gt_PL &lt;dbl&gt;, gt_GT &lt;dbl&gt;, gt_GT_alleles &lt;chr&gt;</code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="pipes">Pipes<a class="anchor" aria-label="anchor" href="#pipes"></a>
</h3>
<p>But what if you wanted to select and filter? We can do this with
pipes. Pipes, are a fairly recent addition to R. Pipes let you take the
output of one function and send it directly to the next, which is useful
when you need to many things to the same data set. It was possible to do
this before pipes were added to R, but it was much messier and more
difficult. Pipes in R look like <code>%&gt;%</code> and are made
available via the <code>magrittr</code> package, which is installed as
part of <code>dplyr</code>. If you use RStudio, you can type the pipe
with <kbd>Ctrl</kbd> + <kbd>Shift</kbd> + <kbd>M</kbd> if you’re using a
PC, or <kbd>Cmd</kbd> + <kbd>Shift</kbd> + <kbd>M</kbd> if you’re using
a Mac.</p>
<div class="codewrapper sourceCode" id="cb30">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">variants</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">filter</span><span class="op">(</span><span class="va">sample_id</span> <span class="op">==</span> <span class="st">"SRR2584863"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">REF</span>, <span class="va">ALT</span>, <span class="va">DP</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 25 × 3
   REF                              ALT                                       DP
   &lt;chr&gt;                            &lt;chr&gt;                                  &lt;dbl&gt;
 1 T                                G                                          4
 2 G                                T                                          6
 3 G                                T                                         10
 4 CTTTTTTT                         CTTTTTTTT                                 12
 5 CCGC                             CCGCGC                                    10
 6 C                                T                                         10
 7 C                                A                                          8
 8 G                                A                                         11
 9 ACAGCCAGCCAGCCAGCCAGCCAGCCAGCCAG ACAGCCAGCCAGCCAGCCAGCCAGCCAGCCAGCCAGC…     3
10 AT                               ATT                                        7
# ℹ 15 more rows</code></pre>
</div>
<p>In the above code, we use the pipe to send the <code>variants</code>
data set first through <code>filter()</code>, to keep rows where
<code>sample_id</code> matches a particular sample, and then through
<code>select()</code> to keep only the <code>REF</code>,
<code>ALT</code>, and <code>DP</code> columns. Since <code>%&gt;%</code>
takes the object on its left and passes it as the first argument to the
function on its right, we don’t need to explicitly include the data
frame as an argument to the <code>filter()</code> and
<code>select()</code> functions any more.</p>
<p>Some may find it helpful to read the pipe like the word “then”. For
instance, in the above example, we took the data frame
<code>variants</code>, <em>then</em> we <code>filter</code>ed for rows
where <code>sample_id</code> was SRR2584863, <em>then</em> we
<code>select</code>ed the <code>REF</code>, <code>ALT</code>, and
<code>DP</code> columns, <em>then</em> we showed only the first six
rows. The <strong><code>dplyr</code></strong> functions by themselves
are somewhat simple, but by combining them into linear workflows with
the pipe, we can accomplish more complex manipulations of data
frames.</p>
<p>If we want to create a new object with this smaller version of the
data we can do so by assigning it a new name:</p>
<div class="codewrapper sourceCode" id="cb32">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">SRR2584863_variants</span> <span class="op">&lt;-</span> <span class="va">variants</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">filter</span><span class="op">(</span><span class="va">sample_id</span> <span class="op">==</span> <span class="st">"SRR2584863"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">REF</span>, <span class="va">ALT</span>, <span class="va">DP</span><span class="op">)</span></span></code></pre>
</div>
<p>This new object includes all of the data from this sample. Let’s look
at just the first six rows to confirm it’s what we want:</p>
<div class="codewrapper sourceCode" id="cb33">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">SRR2584863_variants</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 25 × 3
   REF                              ALT                                       DP
   &lt;chr&gt;                            &lt;chr&gt;                                  &lt;dbl&gt;
 1 T                                G                                          4
 2 G                                T                                          6
 3 G                                T                                         10
 4 CTTTTTTT                         CTTTTTTTT                                 12
 5 CCGC                             CCGCGC                                    10
 6 C                                T                                         10
 7 C                                A                                          8
 8 G                                A                                         11
 9 ACAGCCAGCCAGCCAGCCAGCCAGCCAGCCAG ACAGCCAGCCAGCCAGCCAGCCAGCCAGCCAGCCAGC…     3
10 AT                               ATT                                        7
# ℹ 15 more rows</code></pre>
</div>
<p>Similar to <code>head()</code> and <code>tail()</code> functions, we
can also look at the first or last six rows using tidyverse function
<code>slice()</code>. Slice is a more versatile function that allows
users to specify a range to view:</p>
<div class="codewrapper sourceCode" id="cb35">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">SRR2584863_variants</span> <span class="op">%&gt;%</span> <span class="fu">slice</span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 6 × 3
  REF      ALT          DP
  &lt;chr&gt;    &lt;chr&gt;     &lt;dbl&gt;
1 T        G             4
2 G        T             6
3 G        T            10
4 CTTTTTTT CTTTTTTTT    12
5 CCGC     CCGCGC       10
6 C        T            10</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb37">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">SRR2584863_variants</span> <span class="op">%&gt;%</span> <span class="fu">slice</span><span class="op">(</span><span class="fl">10</span><span class="op">:</span><span class="fl">25</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 16 × 3
   REF   ALT      DP
   &lt;chr&gt; &lt;chr&gt; &lt;dbl&gt;
 1 AT    ATT       7
 2 A     C         9
 3 A     C        20
 4 G     T        12
 5 A     T        19
 6 G     A        15
 7 A     C        10
 8 C     A        14
 9 A     G         9
10 A     C        13
11 A     AC        2
12 G     T        10
13 A     G        16
14 A     C        11
15 TGG   T        10
16 A     C         9</code></pre>
</div>
<div id="exercise-pipe-and-filter" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-pipe-and-filter" class="callout-inner">
<h3 class="callout-title">Exercise: Pipe and filter<a class="anchor" aria-label="anchor" href="#exercise-pipe-and-filter"></a>
</h3>
<div class="callout-content">
<p>Starting with the <code>variants</code> data frame, use pipes to
subset the data to include only observations from SRR2584863 sample,
where the filtered depth (DP) is at least 10. Showing only 5th through
11th rows of columns <code>REF</code>, <code>ALT</code>, and
<code>POS</code>.</p>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" data-bs-parent="#accordionSolution4" aria-labelledby="headingSolution4">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb39">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span> <span class="va">variants</span> <span class="op">%&gt;%</span></span>
<span> <span class="fu">filter</span><span class="op">(</span><span class="va">sample_id</span> <span class="op">==</span> <span class="st">"SRR2584863"</span> <span class="op">&amp;</span> <span class="va">DP</span> <span class="op">&gt;=</span> <span class="fl">10</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span> <span class="fu">slice</span><span class="op">(</span><span class="fl">5</span><span class="op">:</span><span class="fl">11</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span> <span class="fu">select</span><span class="op">(</span><span class="va">sample_id</span>, <span class="va">DP</span>, <span class="va">REF</span>, <span class="va">ALT</span>, <span class="va">POS</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 7 × 5
  sample_id     DP REF   ALT       POS
  &lt;chr&gt;      &lt;dbl&gt; &lt;chr&gt; &lt;chr&gt;   &lt;dbl&gt;
1 SRR2584863    11 G     A     1733343
2 SRR2584863    20 A     C     2446984
3 SRR2584863    12 G     T     2618472
4 SRR2584863    19 A     T     2665639
5 SRR2584863    15 G     A     2999330
6 SRR2584863    10 A     C     3339313
7 SRR2584863    14 C     A     3401754</code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="mutate">Mutate<a class="anchor" aria-label="anchor" href="#mutate"></a>
</h3>
<p>Frequently you’ll want to create new columns based on the values in
existing columns, for example to do unit conversions or find the ratio
of values in two columns. For this we’ll use the <code>dplyr</code>
function <code>mutate()</code>.</p>
<p>For example, we can convert the polymorphism confidence value QUAL to
a probability value according to the formula:</p>
<p>Probability = 1- 10 ^ -(QUAL/10)</p>
<p>We can use <code>mutate</code> to add a column (<code>POLPROB</code>)
to our <code>variants</code> data frame that shows the probability of a
polymorphism at that site given the data.</p>
<div class="codewrapper sourceCode" id="cb41">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">variants</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">mutate</span><span class="op">(</span>POLPROB <span class="op">=</span> <span class="fl">1</span> <span class="op">-</span> <span class="op">(</span><span class="fl">10</span> <span class="op">^</span> <span class="op">-</span><span class="op">(</span><span class="va">QUAL</span><span class="op">/</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 801 × 30
   sample_id  CHROM        POS ID    REF   ALT    QUAL FILTER INDEL   IDV    IMF
   &lt;chr&gt;      &lt;chr&gt;      &lt;dbl&gt; &lt;lgl&gt; &lt;chr&gt; &lt;chr&gt; &lt;dbl&gt; &lt;lgl&gt;  &lt;lgl&gt; &lt;dbl&gt;  &lt;dbl&gt;
 1 SRR2584863 CP000819… 9.97e3 NA    T     G        91 NA     FALSE    NA NA    
 2 SRR2584863 CP000819… 2.63e5 NA    G     T        85 NA     FALSE    NA NA    
 3 SRR2584863 CP000819… 2.82e5 NA    G     T       217 NA     FALSE    NA NA    
 4 SRR2584863 CP000819… 4.33e5 NA    CTTT… CTTT…    64 NA     TRUE     12  1    
 5 SRR2584863 CP000819… 4.74e5 NA    CCGC  CCGC…   228 NA     TRUE      9  0.9  
 6 SRR2584863 CP000819… 6.49e5 NA    C     T       210 NA     FALSE    NA NA    
 7 SRR2584863 CP000819… 1.33e6 NA    C     A       178 NA     FALSE    NA NA    
 8 SRR2584863 CP000819… 1.73e6 NA    G     A       225 NA     FALSE    NA NA    
 9 SRR2584863 CP000819… 2.10e6 NA    ACAG… ACAG…    56 NA     TRUE      2  0.667
10 SRR2584863 CP000819… 2.33e6 NA    AT    ATT     167 NA     TRUE      7  1    
# ℹ 791 more rows
# ℹ 19 more variables: DP &lt;dbl&gt;, VDB &lt;dbl&gt;, RPB &lt;dbl&gt;, MQB &lt;dbl&gt;, BQB &lt;dbl&gt;,
#   MQSB &lt;dbl&gt;, SGB &lt;dbl&gt;, MQ0F &lt;dbl&gt;, ICB &lt;lgl&gt;, HOB &lt;lgl&gt;, AC &lt;dbl&gt;,
#   AN &lt;dbl&gt;, DP4 &lt;chr&gt;, MQ &lt;dbl&gt;, Indiv &lt;chr&gt;, gt_PL &lt;dbl&gt;, gt_GT &lt;dbl&gt;,
#   gt_GT_alleles &lt;chr&gt;, POLPROB &lt;dbl&gt;</code></pre>
</div>
<div id="exercise" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise" class="callout-inner">
<h3 class="callout-title">Exercise<a class="anchor" aria-label="anchor" href="#exercise"></a>
</h3>
<div class="callout-content">
<p>There are a lot of columns in our data set, so let’s just look at the
<code>sample_id</code>, <code>POS</code>, <code>QUAL</code>, and
<code>POLPROB</code> columns for now. Add a line to the above code to
only show those columns.</p>
</div>
</div>
</div>
<div id="accordionSolution5" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution5" aria-expanded="false" aria-controls="collapseSolution5">
  <h4 class="accordion-header" id="headingSolution5">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution5" class="accordion-collapse collapse" data-bs-parent="#accordionSolution5" aria-labelledby="headingSolution5">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb43">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">variants</span> <span class="op">%&gt;%</span></span>
<span> <span class="fu">mutate</span><span class="op">(</span>POLPROB <span class="op">=</span> <span class="fl">1</span> <span class="op">-</span> <span class="fl">10</span> <span class="op">^</span> <span class="op">-</span><span class="op">(</span><span class="va">QUAL</span><span class="op">/</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span> <span class="fu">select</span><span class="op">(</span><span class="va">sample_id</span>, <span class="va">POS</span>, <span class="va">QUAL</span>, <span class="va">POLPROB</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 801 × 4
   sample_id      POS  QUAL POLPROB
   &lt;chr&gt;        &lt;dbl&gt; &lt;dbl&gt;   &lt;dbl&gt;
 1 SRR2584863    9972    91    1.00
 2 SRR2584863  263235    85    1.00
 3 SRR2584863  281923   217    1   
 4 SRR2584863  433359    64    1.00
 5 SRR2584863  473901   228    1   
 6 SRR2584863  648692   210    1   
 7 SRR2584863 1331794   178    1   
 8 SRR2584863 1733343   225    1   
 9 SRR2584863 2103887    56    1.00
10 SRR2584863 2333538   167    1   
# ℹ 791 more rows</code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="group_by-and-summarize-functions">group_by() and summarize() functions<a class="anchor" aria-label="anchor" href="#group_by-and-summarize-functions"></a>
</h3>
<p>Many data analysis tasks can be approached using the
“split-apply-combine” paradigm: split the data into groups, apply some
analysis to each group, and then combine the results. <code>dplyr</code>
makes this very easy through the use of the <code>group_by()</code>
function, which splits the data into groups.</p>
<p>We can use <code>group_by()</code> to tally the number of mutations
detected in each sample using the function <code>tally()</code>:</p>
<div class="codewrapper sourceCode" id="cb45">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">variants</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">sample_id</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">tally</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 3 × 2
  sample_id      n
  &lt;chr&gt;      &lt;int&gt;
1 SRR2584863    25
2 SRR2584866   766
3 SRR2589044    10</code></pre>
</div>
<p>Since counting or tallying values is a common use case for
<code>group_by()</code>, an alternative function was created to bypasses
<code>group_by()</code> using the function <code>count()</code>:</p>
<div class="codewrapper sourceCode" id="cb47">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">variants</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">count</span><span class="op">(</span><span class="va">sample_id</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 3 × 2
  sample_id      n
  &lt;chr&gt;      &lt;int&gt;
1 SRR2584863    25
2 SRR2584866   766
3 SRR2589044    10</code></pre>
</div>
<div id="challenge-2" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-2" class="callout-inner">
<h3 class="callout-title">Challenge<a class="anchor" aria-label="anchor" href="#challenge-2"></a>
</h3>
<div class="callout-content">
<ul>
<li>How many mutations are INDELs?</li>
</ul>
</div>
</div>
</div>
<div id="accordionSolution6" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution6" aria-expanded="false" aria-controls="collapseSolution6">
  <h4 class="accordion-header" id="headingSolution6">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution6" class="accordion-collapse collapse" data-bs-parent="#accordionSolution6" aria-labelledby="headingSolution6">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb49">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">variants</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">count</span><span class="op">(</span><span class="va">INDEL</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 2 × 2
  INDEL     n
  &lt;lgl&gt; &lt;int&gt;
1 FALSE   700
2 TRUE    101</code></pre>
</div>
</div>
</div>
</div>
</div>
<p>When the data is grouped, <code>summarize()</code> can be used to
collapse each group into a single-row summary. <code>summarize()</code>
does this by applying an aggregating or summary function to each
group.</p>
<p>It can be a bit tricky at first, but we can imagine physically
splitting the data frame by groups and applying a certain function to
summarize the data.</p>
<center>
<img src="fig/split_apply_combine.png" alt="rstudio default session" style="width: 500px;" class="figure">
</center>
<p><a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a></p>
<p>We can also apply many other functions to individual columns to get
other summary statistics. For example,we can use built-in functions like
<code>mean()</code>, <code>median()</code>, <code>min()</code>, and
<code>max()</code>. These are called “built-in functions” because they
come with R and don’t require that you install any additional packages.
By default, all <strong>R functions operating on vectors that contains
missing data will return NA</strong>. It’s a way to make sure that users
know they have missing data, and make a conscious decision on how to
deal with it. When dealing with simple statistics like the mean, the
easiest way to ignore <code>NA</code> (the missing data) is to use
<code>na.rm = TRUE</code> (<code>rm</code> stands for remove).</p>
<p>So to view the mean, median, maximum, and minimum filtered depth
(<code>DP</code>) for each sample:</p>
<div class="codewrapper sourceCode" id="cb51">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">variants</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">sample_id</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">summarize</span><span class="op">(</span></span>
<span>    mean_DP <span class="op">=</span> <span class="fu">mean</span><span class="op">(</span><span class="va">DP</span><span class="op">)</span>,</span>
<span>    median_DP <span class="op">=</span> <span class="fu">median</span><span class="op">(</span><span class="va">DP</span><span class="op">)</span>,</span>
<span>    min_DP <span class="op">=</span> <span class="fu">min</span><span class="op">(</span><span class="va">DP</span><span class="op">)</span>,</span>
<span>    max_DP <span class="op">=</span> <span class="fu">max</span><span class="op">(</span><span class="va">DP</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 3 × 5
  sample_id  mean_DP median_DP min_DP max_DP
  &lt;chr&gt;        &lt;dbl&gt;     &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;
1 SRR2584863    10.4      10        2     20
2 SRR2584866    10.6      10        2     79
3 SRR2589044     9.3       9.5      3     16</code></pre>
</div>
</div>
<div class="section level3">
<h3 id="reshaping-data-frames">Reshaping data frames<a class="anchor" aria-label="anchor" href="#reshaping-data-frames"></a>
</h3>
<p>It can sometimes be useful to transform the “long” tidy format, into
the wide format. This transformation can be done with the
<code>pivot_wider()</code> function provided by the <code>tidyr</code>
package (also part of the <code>tidyverse</code>).</p>
<p><code>pivot_wider()</code> takes a data frame as the first argument,
and two arguments: the column name that will become the columns and the
column name that will become the cells in the wide data.</p>
<div class="codewrapper sourceCode" id="cb53">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">variants_wide</span> <span class="op">&lt;-</span> <span class="va">variants</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">sample_id</span>, <span class="va">CHROM</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">summarize</span><span class="op">(</span>mean_DP <span class="op">=</span> <span class="fu">mean</span><span class="op">(</span><span class="va">DP</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">pivot_wider</span><span class="op">(</span>names_from <span class="op">=</span> <span class="va">sample_id</span>, values_from <span class="op">=</span> <span class="va">mean_DP</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>`summarise()` has grouped output by 'sample_id'. You can override using the
`.groups` argument.</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb55">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">variants_wide</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 1 × 4
  CHROM      SRR2584863 SRR2584866 SRR2589044
  &lt;chr&gt;           &lt;dbl&gt;      &lt;dbl&gt;      &lt;dbl&gt;
1 CP000819.1       10.4       10.6        9.3</code></pre>
</div>
<p>The opposite operation of <code>pivot_wider()</code> is taken care by
<code>pivot_longer()</code>. We specify the names of the new columns,
and here add <code>-CHROM</code> as this column shouldn’t be affected by
the reshaping:</p>
<div class="codewrapper sourceCode" id="cb57">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">variants_wide</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">pivot_longer</span><span class="op">(</span><span class="op">-</span><span class="va">CHROM</span>, names_to <span class="op">=</span> <span class="st">"sample_id"</span>, values_to <span class="op">=</span> <span class="st">"mean_DP"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 3 × 3
  CHROM      sample_id  mean_DP
  &lt;chr&gt;      &lt;chr&gt;        &lt;dbl&gt;
1 CP000819.1 SRR2584863    10.4
2 CP000819.1 SRR2584866    10.6
3 CP000819.1 SRR2589044     9.3</code></pre>
</div>
</div>
<div class="section level3">
<h3 id="resources">Resources<a class="anchor" aria-label="anchor" href="#resources"></a>
</h3>
<ul>
<li><p><a href="https://github.com/rstudio/cheatsheets/raw/master/data-transformation.pdf" class="external-link">Handy
dplyr cheatsheet</a></p></li>
<li><p><a href="https://usepa.github.io/introR/2015/01/14/03-Clean/" class="external-link">Much of this
lesson was copied or adapted from Jeff Hollister’s
materials</a></p></li>
</ul>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Keypoints<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>Use the <code>dplyr</code> package to manipulate data frames.</li>
<li>Use <code>glimpse()</code> to quickly look at your data frame.</li>
<li>Use <code>select()</code> to choose variables from a data
frame.</li>
<li>Use <code>filter()</code> to choose data based on values.</li>
<li>Use <code>mutate()</code> to create new variables.</li>
<li>Use <code>group_by()</code> and <code>summarize()</code> to work
with subsets of data.</li>
</ul>
</div>
</div>
</div>
</div>
</section><div class="footnotes footnotes-end-of-document">
<hr>
<ol>
<li id="fn1"><p>The figure was adapted from the Software Carpentry
lesson, <a href="https://swcarpentry.github.io/r-novice-gapminder/13-dplyr/" class="external-link">R for
Reproducible Scientific Analysis</a><a href="#fnref1" class="footnote-back">↩︎</a></p></li>
</ol>
</div></section><section id="aio-06-data-visualization"><p>Content from <a href="06-data-visualization.html">Data Visualization with ggplot2</a></p>
<hr>
<p> Last updated on 2023-05-01 | 
        
        <a href="https://github.com/datacarpentry/genomics-r-intro/edit/main/episodes/06-data-visualization.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What is ggplot2?</li>
<li>What is mapping, and what is aesthetics?</li>
<li>What is the process of creating a publication-quality plots with
ggplot in R?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Describe the role of data, aesthetics, and geoms in ggplot
functions.</li>
<li>Choose the correct aesthetics and alter the geom parameters for a
scatter plot, histogram, or box plot.</li>
<li>Layer multiple geometries in a single plot.</li>
<li>Customize plot scales, titles, themes, and fonts.</li>
<li>Apply a facet to a plot.</li>
<li>Apply additional ggplot2-compatible plotting libraries.</li>
<li>Save a ggplot to a file.</li>
<li>List several resources for getting help with ggplot.</li>
<li>List several resources for creating informative scientific
plots.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section id="introduction-to-ggplot2"><h2 class="section-heading">Introduction to <strong><code>ggplot2</code></strong><a class="anchor" aria-label="anchor" href="#introduction-to-ggplot2"></a>
</h2>
<hr class="half-width">
<figure><img src="https://ggplot2.tidyverse.org/logo.png" align="right" alt="Line plot enclosed in hexagon shape with ggplot2 typed beneath and www.rstudio.com at the bottom." class="figure mx-auto d-block"></figure><p><strong><code>ggplot2</code></strong> is a plotting package that
makes it simple to create complex plots from data in a data frame. It
provides a more programmatic interface for specifying what variables to
plot, how they are displayed, and general visual properties. Therefore,
we only need minimal changes if the underlying data change or if we
decide to change from a bar plot to a scatter plot. This helps in
creating publication-quality plots with minimal amounts of adjustments
and tweaking.</p>
<p>The <strong>gg</strong> in “<strong>ggplot</strong>” stands for
“<strong>G</strong>rammar of <strong>G</strong>raphics,” which is an
elegant yet powerful way to describe the making of scientific plots. In
short, the grammar of graphics breaks down every plot into a few
components, namely, a dataset, a set of geoms (visual marks that
represent the data points), and a coordinate system. You can imagine
this is a grammar that gives unique names to each component appearing in
a plot and conveys specific information about data. With
<strong>ggplot</strong>, graphics are built step by step by adding new
elements.</p>
<p>The idea of <strong>mapping</strong> is crucial in
<strong>ggplot</strong>. One familiar example is to <em>map</em> the
value of one variable in a dataset to <span class="math inline">\(x\)</span> and the other to <span class="math inline">\(y\)</span>. However, we often encounter datasets
that include multiple (more than two) variables. In this case,
<strong>ggplot</strong> allows you to <em>map</em> those other variables
to visual marks such as <strong>color</strong> and
<strong>shape</strong> (<strong>aesthetics</strong> or
<code>aes</code>). One thing you may want to remember is the difference
between <strong>discrete</strong> and <strong>continuous</strong>
variables. Some aesthetics, such as the shape of dots, do not accept
continuous variables. If forced to do so, R will give an error. This is
easy to understand; we cannot create a continuum of shapes for a
variable, unlike, say, color.</p>
<p><strong>Tip:</strong> when having doubts about whether a variable is
<a href="https://en.wikipedia.org/wiki/Continuous_or_discrete_variable" class="external-link">continuous
or discrete</a>, a quick way to check is to use the <a href="https://www.geeksforgeeks.org/get-summary-of-results-produced-by-functions-in-r-programming-summary-function/" class="external-link"><code>summary()</code></a>
function. Continuous variables have descriptive statistics but not the
discrete variables.</p>
</section><section id="installing-tidyverse"><h2 class="section-heading">Installing <code>tidyverse</code><a class="anchor" aria-label="anchor" href="#installing-tidyverse"></a>
</h2>
<hr class="half-width">
<p><strong><code>ggplot2</code></strong> belongs to the <a href="https://www.tidyverse.org/" class="external-link"><strong><code>tidyverse</code></strong>
framework</a>. Therefore, we will start with loading the package
<strong><code>tidyverse</code></strong>. If
<strong><code>tidyverse</code></strong> is not already installed, then
we need to install first. If it is already installed, then we can skip
the following step:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">install.packages</span><span class="op">(</span><span class="st">"tidyverse"</span><span class="op">)</span> <span class="co"># Installing tidyverse package, includes ggplot2 and other packages such as dplyr, readr, tidyr</span></span></code></pre>
</div>
<p>Now, let’s load the <code>tidyverse</code> package:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">tidyverse</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──
✔ dplyr     1.1.2     ✔ readr     2.1.4
✔ forcats   1.0.0     ✔ stringr   1.5.0
✔ ggplot2   3.4.2     ✔ tibble    3.2.1
✔ lubridate 1.9.2     ✔ tidyr     1.3.0
✔ purrr     1.0.1     
── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()
ℹ Use the conflicted package (&lt;http://conflicted.r-lib.org/&gt;) to force all conflicts to become errors</code></pre>
</div>
<p>As we can see from above output <strong><code>ggplot2</code></strong>
has been already loaded along with other packages as part of the
<strong><code>tidyverse</code></strong> framework.</p>
</section><section id="loading-the-dataset"><h2 class="section-heading">Loading the dataset<a class="anchor" aria-label="anchor" href="#loading-the-dataset"></a>
</h2>
<hr class="half-width">
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">variants</span> <span class="op">=</span> <span class="fu">read_csv</span><span class="op">(</span><span class="st">"https://raw.githubusercontent.com/naupaka/vcfr-for-data-carpentry-draft/main/output/combined_tidy_vcf.csv"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Rows: 801 Columns: 29
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr  (7): sample_id, CHROM, REF, ALT, DP4, Indiv, gt_GT_alleles
dbl (16): POS, QUAL, IDV, IMF, DP, VDB, RPB, MQB, BQB, MQSB, SGB, MQ0F, AC, ...
num  (1): gt_PL
lgl  (5): ID, FILTER, INDEL, ICB, HOB

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
</div>
<p>Explore the <em>structure</em> (types of columns and number of rows)
of the dataset using <a href="https://dplyr.tidyverse.org/index.html" class="external-link">dplyr</a>’s <a href="https://dplyr.tidyverse.org/reference/glimpse.html" class="external-link"><code>glimpse()</code></a>
(for more info, see the <a href="https://datacarpentry.org/genomics-r-intro/05-dplyr/" class="external-link">Data
Wrangling and Analyses with Tidyverse</a> episode)</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">glimpse</span><span class="op">(</span><span class="va">variants</span><span class="op">)</span> <span class="co"># Show a snapshot of the rows and columns</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Rows: 801
Columns: 29
$ sample_id     &lt;chr&gt; "SRR2584863", "SRR2584863", "SRR2584863", "SRR2584863", …
$ CHROM         &lt;chr&gt; "CP000819.1", "CP000819.1", "CP000819.1", "CP000819.1", …
$ POS           &lt;dbl&gt; 9972, 263235, 281923, 433359, 473901, 648692, 1331794, 1…
$ ID            &lt;lgl&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, …
$ REF           &lt;chr&gt; "T", "G", "G", "CTTTTTTT", "CCGC", "C", "C", "G", "ACAGC…
$ ALT           &lt;chr&gt; "G", "T", "T", "CTTTTTTTT", "CCGCGC", "T", "A", "A", "AC…
$ QUAL          &lt;dbl&gt; 91.0000, 85.0000, 217.0000, 64.0000, 228.0000, 210.0000,…
$ FILTER        &lt;lgl&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, …
$ INDEL         &lt;lgl&gt; FALSE, FALSE, FALSE, TRUE, TRUE, FALSE, FALSE, FALSE, TR…
$ IDV           &lt;dbl&gt; NA, NA, NA, 12, 9, NA, NA, NA, 2, 7, NA, NA, NA, NA, NA,…
$ IMF           &lt;dbl&gt; NA, NA, NA, 1.000000, 0.900000, NA, NA, NA, 0.666667, 1.…
$ DP            &lt;dbl&gt; 4, 6, 10, 12, 10, 10, 8, 11, 3, 7, 9, 20, 12, 19, 15, 10…
$ VDB           &lt;dbl&gt; 0.0257451, 0.0961330, 0.7740830, 0.4777040, 0.6595050, 0…
$ RPB           &lt;dbl&gt; NA, 1.000000, NA, NA, NA, NA, NA, NA, NA, NA, 0.900802, …
$ MQB           &lt;dbl&gt; NA, 1.0000000, NA, NA, NA, NA, NA, NA, NA, NA, 0.1501340…
$ BQB           &lt;dbl&gt; NA, 1.000000, NA, NA, NA, NA, NA, NA, NA, NA, 0.750668, …
$ MQSB          &lt;dbl&gt; NA, NA, 0.974597, 1.000000, 0.916482, 0.916482, 0.900802…
$ SGB           &lt;dbl&gt; -0.556411, -0.590765, -0.662043, -0.676189, -0.662043, -…
$ MQ0F          &lt;dbl&gt; 0.000000, 0.166667, 0.000000, 0.000000, 0.000000, 0.0000…
$ ICB           &lt;lgl&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, …
$ HOB           &lt;lgl&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, …
$ AC            &lt;dbl&gt; 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,…
$ AN            &lt;dbl&gt; 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,…
$ DP4           &lt;chr&gt; "0,0,0,4", "0,1,0,5", "0,0,4,5", "0,1,3,8", "1,0,2,7", "…
$ MQ            &lt;dbl&gt; 60, 33, 60, 60, 60, 60, 60, 60, 60, 60, 25, 60, 10, 60, …
$ Indiv         &lt;chr&gt; "/home/dcuser/dc_workshop/results/bam/SRR2584863.aligned…
$ gt_PL         &lt;dbl&gt; 1210, 1120, 2470, 910, 2550, 2400, 2080, 2550, 11128, 19…
$ gt_GT         &lt;dbl&gt; 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,…
$ gt_GT_alleles &lt;chr&gt; "G", "T", "T", "CTTTTTTTT", "CCGCGC", "T", "A", "A", "AC…</code></pre>
</div>
<p>Alternatively, we can display the first a few rows (vertically) of
the table using <a href="https://www.geeksforgeeks.org/get-the-first-parts-of-a-data-set-in-r-programming-head-function/" class="external-link"><code>head()</code></a>:</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="va">variants</span><span class="op">)</span></span></code></pre>
</div>
<table style="width:100%;" class="table">
<colgroup>
<col width="4%">
<col width="4%">
<col width="2%">
<col width="1%">
<col width="3%">
<col width="4%">
<col width="2%">
<col width="2%">
<col width="2%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="4%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="3%">
<col width="4%">
<col width="3%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="3%">
<col width="1%">
<col width="27%">
<col width="2%">
<col width="2%">
<col width="5%">
</colgroup>
<thead><tr class="header">
<th align="left">sample_id</th>
<th align="left">CHROM</th>
<th align="right">POS</th>
<th align="left">ID</th>
<th align="left">REF</th>
<th align="left">ALT</th>
<th align="right">QUAL</th>
<th align="left">FILTER</th>
<th align="left">INDEL</th>
<th align="right">IDV</th>
<th align="right">IMF</th>
<th align="right">DP</th>
<th align="right">VDB</th>
<th align="right">RPB</th>
<th align="right">MQB</th>
<th align="right">BQB</th>
<th align="right">MQSB</th>
<th align="right">SGB</th>
<th align="right">MQ0F</th>
<th align="left">ICB</th>
<th align="left">HOB</th>
<th align="right">AC</th>
<th align="right">AN</th>
<th align="left">DP4</th>
<th align="right">MQ</th>
<th align="left">Indiv</th>
<th align="right">gt_PL</th>
<th align="right">gt_GT</th>
<th align="left">gt_GT_alleles</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">SRR2584863</td>
<td align="left">CP000819.1</td>
<td align="right">9972</td>
<td align="left">NA</td>
<td align="left">T</td>
<td align="left">G</td>
<td align="right">91</td>
<td align="left">NA</td>
<td align="left">FALSE</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">4</td>
<td align="right">0.0257451</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">-0.556411</td>
<td align="right">0.000000</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="left">0,0,0,4</td>
<td align="right">60</td>
<td align="left">/home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam</td>
<td align="right">1210</td>
<td align="right">1</td>
<td align="left">G</td>
</tr>
<tr class="even">
<td align="left">SRR2584863</td>
<td align="left">CP000819.1</td>
<td align="right">263235</td>
<td align="left">NA</td>
<td align="left">G</td>
<td align="left">T</td>
<td align="right">85</td>
<td align="left">NA</td>
<td align="left">FALSE</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">6</td>
<td align="right">0.0961330</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">NA</td>
<td align="right">-0.590765</td>
<td align="right">0.166667</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="left">0,1,0,5</td>
<td align="right">33</td>
<td align="left">/home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam</td>
<td align="right">1120</td>
<td align="right">1</td>
<td align="left">T</td>
</tr>
<tr class="odd">
<td align="left">SRR2584863</td>
<td align="left">CP000819.1</td>
<td align="right">281923</td>
<td align="left">NA</td>
<td align="left">G</td>
<td align="left">T</td>
<td align="right">217</td>
<td align="left">NA</td>
<td align="left">FALSE</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">10</td>
<td align="right">0.7740830</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">0.974597</td>
<td align="right">-0.662043</td>
<td align="right">0.000000</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="left">0,0,4,5</td>
<td align="right">60</td>
<td align="left">/home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam</td>
<td align="right">2470</td>
<td align="right">1</td>
<td align="left">T</td>
</tr>
<tr class="even">
<td align="left">SRR2584863</td>
<td align="left">CP000819.1</td>
<td align="right">433359</td>
<td align="left">NA</td>
<td align="left">CTTTTTTT</td>
<td align="left">CTTTTTTTT</td>
<td align="right">64</td>
<td align="left">NA</td>
<td align="left">TRUE</td>
<td align="right">12</td>
<td align="right">1.0</td>
<td align="right">12</td>
<td align="right">0.4777040</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">1.000000</td>
<td align="right">-0.676189</td>
<td align="right">0.000000</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="left">0,1,3,8</td>
<td align="right">60</td>
<td align="left">/home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam</td>
<td align="right">910</td>
<td align="right">1</td>
<td align="left">CTTTTTTTT</td>
</tr>
<tr class="odd">
<td align="left">SRR2584863</td>
<td align="left">CP000819.1</td>
<td align="right">473901</td>
<td align="left">NA</td>
<td align="left">CCGC</td>
<td align="left">CCGCGC</td>
<td align="right">228</td>
<td align="left">NA</td>
<td align="left">TRUE</td>
<td align="right">9</td>
<td align="right">0.9</td>
<td align="right">10</td>
<td align="right">0.6595050</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">0.916482</td>
<td align="right">-0.662043</td>
<td align="right">0.000000</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="left">1,0,2,7</td>
<td align="right">60</td>
<td align="left">/home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam</td>
<td align="right">2550</td>
<td align="right">1</td>
<td align="left">CCGCGC</td>
</tr>
<tr class="even">
<td align="left">SRR2584863</td>
<td align="left">CP000819.1</td>
<td align="right">648692</td>
<td align="left">NA</td>
<td align="left">C</td>
<td align="left">T</td>
<td align="right">210</td>
<td align="left">NA</td>
<td align="left">FALSE</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">10</td>
<td align="right">0.2680140</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">0.916482</td>
<td align="right">-0.670168</td>
<td align="right">0.000000</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="left">0,0,7,3</td>
<td align="right">60</td>
<td align="left">/home/dcuser/dc_workshop/results/bam/SRR2584863.aligned.sorted.bam</td>
<td align="right">2400</td>
<td align="right">1</td>
<td align="left">T</td>
</tr>
</tbody>
</table>
<p><strong><code>ggplot2</code></strong> functions like data in the
<strong>long</strong> format, i.e., a column for every dimension
(variable), and a row for every observation. Well-structured data will
save you time when making figures with
<strong><code>ggplot2</code></strong></p>
<p><strong><code>ggplot2</code></strong> graphics are built step-by-step
by adding new elements. Adding layers in this fashion allows for
extensive flexibility and customization of plots, and more equally
important the readability of the code.</p>
<p>To build a ggplot, we will use the following basic template that can
be used for different types of plots:</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> <span class="sc">&lt;</span>DATA<span class="sc">&gt;</span>, <span class="at">mapping =</span> <span class="fu">aes</span>(<span class="sc">&lt;</span>MAPPINGS<span class="sc">&gt;</span>)) <span class="sc">+</span>  <span class="er">&lt;</span>GEOM_FUNCTION<span class="sc">&gt;</span>()</span></code></pre>
</div>
<ul>
<li>use the <code>ggplot()</code> function and bind the plot to a
specific data frame using the <code>data</code> argument</li>
</ul>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">variants</span><span class="op">)</span></span></code></pre>
</div>
<ul>
<li>define a mapping (using the aesthetic (<code>aes</code>) function),
by selecting the variables to be plotted and specifying how to present
them in the graph, e.g. as x and y positions or characteristics such as
size, shape, color, etc.</li>
</ul>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">variants</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">POS</span>, y <span class="op">=</span> <span class="va">DP</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<ul>
<li>add ‘geoms’ – graphical representations of the data in the plot
(points, lines, bars). <strong><code>ggplot2</code></strong> offers many
different geoms; we will use some common ones today, including:
<ul>
<li>
<a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link"><code>geom_point()</code></a>
for scatter plots, dot plots, etc.</li>
<li>
<a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link"><code>geom_boxplot()</code></a>
for, well, boxplots!</li>
<li>
<a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link"><code>geom_line()</code></a>
for trend lines, time series, etc.</li>
</ul>
</li>
</ul>
<p>To add a geom to the plot use the <code>+</code> operator. Because we
have two continuous variables, let’s use <a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link"><code>geom_point()</code></a>
(i.e., a scatter plot) first:</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">variants</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">POS</span>, y <span class="op">=</span> <span class="va">DP</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/06-data-visualization-rendered-first-ggplot-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>The <code>+</code> in the <strong><code>ggplot2</code></strong>
package is particularly useful because it allows you to modify existing
<code>ggplot</code> objects. This means you can easily set up plot
templates and conveniently explore different types of plots, so the
above plot can also be generated with code like this:</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Assign plot to a variable</span></span>
<span><span class="va">coverage_plot</span> <span class="op">&lt;-</span> <span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">variants</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">POS</span>, y <span class="op">=</span> <span class="va">DP</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Draw the plot</span></span>
<span><span class="va">coverage_plot</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<p><strong>Notes</strong></p>
<ul>
<li>Anything you put in the <a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link"><code>ggplot()</code></a>
function can be seen by any geom layers that you add (i.e., these are
universal plot settings). This includes the x- and y-axis mapping you
set up in <a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link"><code>aes()</code></a>.</li>
<li>You can also specify mappings for a given geom independently of the
mappings defined globally in the <a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link"><code>ggplot()</code></a>
function.</li>
<li>The <code>+</code> sign used to add new layers must be placed at the
end of the line containing the <em>previous</em> layer. If, instead, the
<code>+</code> sign is added at the beginning of the line containing the
new layer, <strong><code>ggplot2</code></strong> will not add the new
layer and will return an error message.</li>
</ul>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># This is the correct syntax for adding layers</span></span>
<span><span class="va">coverage_plot</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># This will not add the new layer and will return an error message</span></span>
<span><span class="va">coverage_plot</span></span>
<span>  <span class="op">+</span> <span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
</section><section id="building-your-plots-iteratively"><h2 class="section-heading">Building your plots iteratively<a class="anchor" aria-label="anchor" href="#building-your-plots-iteratively"></a>
</h2>
<hr class="half-width">
<p>Building plots with <strong><code>ggplot2</code></strong> is
typically an iterative process. We start by defining the dataset we’ll
use, lay out the axes, and choose a geom:</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">variants</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">POS</span>, y <span class="op">=</span> <span class="va">DP</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/06-data-visualization-rendered-create-ggplot-object-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Then, we start modifying this plot to extract more information from
it. For instance, we can add transparency (<code>alpha</code>) to avoid
over-plotting:</p>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">variants</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">POS</span>, y <span class="op">=</span> <span class="va">DP</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_point</span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/06-data-visualization-rendered-adding-transparency-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>We can also add colors for all the points:</p>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">variants</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">POS</span>, y <span class="op">=</span> <span class="va">DP</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_point</span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.5</span>, color <span class="op">=</span> <span class="st">"blue"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/06-data-visualization-rendered-adding-colors-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Or to color each species in the plot differently, you could use a
vector as an input to the argument <strong>color</strong>.
<strong><code>ggplot2</code></strong> will provide a different color
corresponding to different values in the vector. Here is an example
where we color with <strong><code>sample_id</code></strong>:</p>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">variants</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">POS</span>, y <span class="op">=</span> <span class="va">DP</span>, color <span class="op">=</span> <span class="va">sample_id</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_point</span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/06-data-visualization-rendered-color-by-sample-1-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Notice that we can change the geom layer and colors will be still
determined by <strong><code>sample_id</code></strong></p>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">variants</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">POS</span>, y <span class="op">=</span> <span class="va">DP</span>, color <span class="op">=</span> <span class="va">sample_id</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_point</span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/06-data-visualization-rendered-color-by-sample-2-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>To make our plot more readable, we can add axis labels:</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">variants</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">POS</span>, y <span class="op">=</span> <span class="va">DP</span>, color <span class="op">=</span> <span class="va">sample_id</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_point</span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">labs</span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Base Pair Position"</span>,</span>
<span>       y <span class="op">=</span> <span class="st">"Read Depth (DP)"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/06-data-visualization-rendered-add-axis-labels-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>To add a <em>main</em> title to the plot, we use <a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link"><code>ggtitle()</code></a>:</p>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">variants</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">POS</span>, y <span class="op">=</span> <span class="va">DP</span>, color <span class="op">=</span> <span class="va">sample_id</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_point</span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">labs</span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Base Pair Position"</span>,</span>
<span>       y <span class="op">=</span> <span class="st">"Read Depth (DP)"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"Read Depth vs. Position"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/06-data-visualization-rendered-add-main-title-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Now the figure is complete and ready to be exported and saved to a
file. This can be achieved easily using <a href="https://ggplot2.tidyverse.org/reference/ggsave.html" class="external-link"><code>ggsave()</code></a>,
which can write, by default, the most recent generated figure into
different formats (e.g., <code>jpeg</code>, <code>png</code>,
<code>pdf</code>) according to the file extension. So, for example, to
create a pdf version of the above figure with a dimension of <span class="math inline">\(6\times4\)</span> inches:</p>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggsave</span> <span class="op">(</span><span class="st">"depth.pdf"</span>, width <span class="op">=</span> <span class="fl">6</span>, height <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span></code></pre>
</div>
<p>If we check the <em>current working directory</em>, there should be a
newly created file called <code>depth.pdf</code> with the above
plot.</p>
<div id="challenge" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge" class="callout-inner">
<h3 class="callout-title">Challenge<a class="anchor" aria-label="anchor" href="#challenge"></a>
</h3>
<div class="callout-content">
<p>Use what you just learned to create a scatter plot of mapping quality
(<code>MQ</code>) over position (<code>POS</code>) with the samples
showing in different colors. Make sure to give your plot relevant axis
labels.</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span> <span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">variants</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">POS</span>, y <span class="op">=</span> <span class="va">MQ</span>, color <span class="op">=</span> <span class="va">sample_id</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">labs</span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Base Pair Position"</span>,</span>
<span>       y <span class="op">=</span> <span class="st">"Mapping Quality (MQ)"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/06-data-visualization-rendered-scatter-challenge-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure>
</div>
</div>
</div>
</div>
<p>To further customize the plot, we can change the default font
format:</p>
<div class="codewrapper sourceCode" id="cb24">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">variants</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">POS</span>, y <span class="op">=</span> <span class="va">DP</span>, color <span class="op">=</span> <span class="va">sample_id</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_point</span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">labs</span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Base Pair Position"</span>,</span>
<span>       y <span class="op">=</span> <span class="st">"Read Depth (DP)"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"Read Depth vs. Position"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span>text <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>family <span class="op">=</span> <span class="st">"Bookman"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/06-data-visualization-rendered-change-font-family-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure></section><section id="faceting"><h2 class="section-heading">Faceting<a class="anchor" aria-label="anchor" href="#faceting"></a>
</h2>
<hr class="half-width">
<p><strong><code>ggplot2</code></strong> has a special technique called
<em>faceting</em> that allows the user to split one plot into multiple
plots (panels) based on a factor (variable) included in the dataset. We
will use it to split our mapping quality plot into three panels, one for
each sample.</p>
<div class="codewrapper sourceCode" id="cb25">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">variants</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">POS</span>, y <span class="op">=</span> <span class="va">MQ</span>, color <span class="op">=</span> <span class="va">sample_id</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span> <span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span> <span class="fu">labs</span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Base Pair Position"</span>,</span>
<span>      y <span class="op">=</span> <span class="st">"Mapping Quality (MQ)"</span><span class="op">)</span> <span class="op">+</span></span>
<span> <span class="fu">facet_grid</span><span class="op">(</span><span class="va">.</span> <span class="op">~</span> <span class="va">sample_id</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/06-data-visualization-rendered-first-facet-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>This looks okay, but it would be easier to read if the plot facets
were stacked vertically rather than horizontally. The
<code>facet_grid</code> geometry allows you to explicitly specify how
you want your plots to be arranged via formula notation
(<code>rows ~ columns</code>; the dot (<code>.</code>) indicates every
other variable in the data i.e., no faceting on that side of the
formula).</p>
<div class="codewrapper sourceCode" id="cb26">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">variants</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">POS</span>, y <span class="op">=</span> <span class="va">MQ</span>, color <span class="op">=</span> <span class="va">sample_id</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span> <span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span> <span class="fu">labs</span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Base Pair Position"</span>,</span>
<span>      y <span class="op">=</span> <span class="st">"Mapping Quality (MQ)"</span><span class="op">)</span> <span class="op">+</span></span>
<span> <span class="fu">facet_grid</span><span class="op">(</span><span class="va">sample_id</span> <span class="op">~</span> <span class="va">.</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/06-data-visualization-rendered-second-facet-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Usually plots with white background look more readable when printed.
We can set the background to white using the function <a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link"><code>theme_bw()</code></a>.
Additionally, you can remove the grid:</p>
<div class="codewrapper sourceCode" id="cb27">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">variants</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">POS</span>, y <span class="op">=</span> <span class="va">MQ</span>, color <span class="op">=</span> <span class="va">sample_id</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">labs</span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Base Pair Position"</span>,</span>
<span>       y <span class="op">=</span> <span class="st">"Mapping Quality (MQ)"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">facet_grid</span><span class="op">(</span><span class="va">sample_id</span> <span class="op">~</span> <span class="va">.</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span>panel.grid <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/06-data-visualization-rendered-facet-plot-white-bg-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div id="challenge-1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-1" class="callout-inner">
<h3 class="callout-title">Challenge<a class="anchor" aria-label="anchor" href="#challenge-1"></a>
</h3>
<div class="callout-content">
<p>Use what you just learned to create a scatter plot of PHRED scaled
quality (<code>QUAL</code>) over position (<code>POS</code>) with the
samples showing in different colors. Make sure to give your plot
relevant axis labels.</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb28">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span> <span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">variants</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">POS</span>, y <span class="op">=</span> <span class="va">QUAL</span>, color <span class="op">=</span> <span class="va">sample_id</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">labs</span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Base Pair Position"</span>,</span>
<span>       y <span class="op">=</span> <span class="st">"PHRED-sacled Quality (QUAL)"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">facet_grid</span><span class="op">(</span><span class="va">sample_id</span> <span class="op">~</span> <span class="va">.</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/06-data-visualization-rendered-scatter-challenge-2-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure>
</div>
</div>
</div>
</div>
</section><section id="barplots"><h2 class="section-heading">Barplots<a class="anchor" aria-label="anchor" href="#barplots"></a>
</h2>
<hr class="half-width">
<p>We can create barplots using the <a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link"><code>geom_bar</code></a>
geom. Let’s make a barplot showing the number of variants for each
sample that are indels.</p>
<div class="codewrapper sourceCode" id="cb29">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">variants</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">INDEL</span>, fill <span class="op">=</span> <span class="va">sample_id</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_bar</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">facet_grid</span><span class="op">(</span><span class="va">sample_id</span> <span class="op">~</span> <span class="va">.</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/06-data-visualization-rendered-barplot-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div id="challenge-2" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-2" class="callout-inner">
<h3 class="callout-title">Challenge<a class="anchor" aria-label="anchor" href="#challenge-2"></a>
</h3>
<div class="callout-content">
<p>Since we already have the sample_id labels on the individual plot
facets, we don’t need the legend. Use the help file for
<code>geom_bar</code> and any other online resources you want to use to
remove the legend from the plot.</p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" data-bs-parent="#accordionSolution3" aria-labelledby="headingSolution3">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb30">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">variants</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">INDEL</span>, color <span class="op">=</span> <span class="va">sample_id</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>   <span class="fu">geom_bar</span><span class="op">(</span>show.legend <span class="op">=</span> <span class="cn">F</span><span class="op">)</span> <span class="op">+</span></span>
<span>   <span class="fu">facet_grid</span><span class="op">(</span><span class="va">sample_id</span> <span class="op">~</span> <span class="va">.</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/06-data-visualization-rendered-barplot-challenge-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure>
</div>
</div>
</div>
</div>
</section><section id="density"><h2 class="section-heading">Density<a class="anchor" aria-label="anchor" href="#density"></a>
</h2>
<hr class="half-width">
<p>We can create density plots using the <a href="https://ggplot2.tidyverse.org/reference/geom_density.html" class="external-link"><code>geom_density</code></a>
geom that shows the distribution of of a variable in the dataset. Let’s
plot the distribution of <code>DP</code></p>
<div class="codewrapper sourceCode" id="cb31">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">variants</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">DP</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_density</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/06-data-visualization-rendered-density-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>This plot tells us that the most of frequent <code>DP</code> (read
depth) for the variants is about 10 reads.</p>
<div id="challenge-3" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-3" class="callout-inner">
<h3 class="callout-title">Challenge<a class="anchor" aria-label="anchor" href="#challenge-3"></a>
</h3>
<div class="callout-content">
<p>Use <a href="https://ggplot2.tidyverse.org/reference/geom_density.html" class="external-link"><code>geom_density</code></a>
to plot the distribution of <code>DP</code> with a different fill for
each sample. Use a white background for the plot.</p>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" data-bs-parent="#accordionSolution4" aria-labelledby="headingSolution4">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb32">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">variants</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">DP</span>, fill <span class="op">=</span> <span class="va">sample_id</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>   <span class="fu">geom_density</span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>   <span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/06-data-visualization-rendered-density-challenge-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure>
</div>
</div>
</div>
</div>
</section><section id="ggplot2-themes"><h2 class="section-heading">
<strong><code>ggplot2</code></strong> themes<a class="anchor" aria-label="anchor" href="#ggplot2-themes"></a>
</h2>
<hr class="half-width">
<p>In addition to <a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link"><code>theme_bw()</code></a>,
which changes the plot background to white,
<strong><code>ggplot2</code></strong> comes with several other themes
which can be useful to quickly change the look of your visualization.
The complete list of themes is available at <a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">https://ggplot2.tidyverse.org/reference/ggtheme.html</a>.
<code>theme_minimal()</code> and <code>theme_light()</code> are popular,
and <code>theme_void()</code> can be useful as a starting point to
create a new hand-crafted theme.</p>
<p>The <a href="https://jrnold.github.io/ggthemes/reference/index.html" class="external-link">ggthemes</a>
package provides a wide variety of options (including Microsoft Excel,
<a href="https://jrnold.github.io/ggthemes/reference/theme_excel.html" class="external-link">old</a>
and <a href="https://jrnold.github.io/ggthemes/reference/theme_excel_new.html" class="external-link">new</a>).
The <a href="https://exts.ggplot2.tidyverse.org/" class="external-link"><strong><code>ggplot2</code></strong>
extensions website</a> provides a list of packages that extend the
capabilities of <strong><code>ggplot2</code></strong>, including
additional themes.</p>
<div id="challenge-4" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="challenge-4" class="callout-inner">
<h3 class="callout-title">Challenge<a class="anchor" aria-label="anchor" href="#challenge-4"></a>
</h3>
<div class="callout-content">
<p>With all of this information in hand, please take another five
minutes to either improve one of the plots generated in this exercise or
create a beautiful graph of your own. Use the RStudio <a href="https://raw.githubusercontent.com/rstudio/cheatsheets/main/data-visualization.pdf" class="external-link"><strong><code>ggplot2</code></strong>
cheat sheet</a> for inspiration. Here are some ideas:</p>
<ul>
<li>See if you can change the size or shape of the plotting symbol.</li>
<li>Can you find a way to change the name of the legend? What about its
labels?</li>
<li>Try using a different color palette (see the <a href="https://www.cookbook-r.com/Graphs/Colors_(ggplot2)/" class="external-link">Cookbook for
R</a>.</li>
</ul>
</div>
</div>
</div>
</section><section id="more-ggplot2-plots"><h2 class="section-heading">More <strong><code>ggplot2</code></strong> Plots<a class="anchor" aria-label="anchor" href="#more-ggplot2-plots"></a>
</h2>
<hr class="half-width">
<p><strong><code>ggplot2</code></strong> offers many more informative
and beautiful plots (<code>geoms</code>) of interest for biologists
(although not covered in this lesson) that are worth exploring, such
as</p>
<ul>
<li>
<a href="https://ggplot2.tidyverse.org/reference/geom_tile.html" class="external-link"><code>geom_tile()</code></a>,
for heatmaps,</li>
<li>
<a href="https://ggplot2.tidyverse.org/reference/geom_jitter.html" class="external-link"><code>geom_jitter()</code></a>,
for strip charts, and</li>
<li>
<a href="https://ggplot2.tidyverse.org/reference/geom_violin.html" class="external-link"><code>geom_violin()</code></a>,
for violin plots</li>
</ul></section><section id="resources"><h2 class="section-heading">Resources<a class="anchor" aria-label="anchor" href="#resources"></a>
</h2>
<hr class="half-width">
<ul>
<li>
<a href="https://www.amazon.com/gp/product/331924275X/" class="external-link">ggplot2:
Elegant Graphics for Data Analysis</a> (<a href="https://ggplot2-book.org/" class="external-link">online version</a>)</li>
<li><a href="https://www.amazon.com/Grammar-Graphics-Statistics-Computing/dp/0387245448/" class="external-link">The
Grammar of Graphics (Statistics and Computing)</a></li>
<li>
<a href="https://www.amazon.com/gp/product/0691181624/" class="external-link">Data
Visualization: A Practical Introduction</a> (<a href="https://socviz.co/" class="external-link">online version</a>)</li>
<li>
<a href="https://r-graph-gallery.com/" class="external-link">The R Graph Gallery</a> (<a href="https://bookdown.org/content/b298e479-b1ab-49fa-b83d-a57c2b034d49/" class="external-link">the
book</a>)</li>
</ul>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Keypoints<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>ggplot2 is a powerful tool for high-quality plots</li>
<li>ggplot2 provides a flexiable and readable grammar to build
plots</li>
</ul>
</div>
</div>
</div>
</section></section><section id="aio-07-r-help"><p>Content from <a href="07-r-help.html">Getting help with R</a></p>
<hr>
<p> Last updated on 2023-05-10 | 
        
        <a href="https://github.com/datacarpentry/genomics-r-intro/edit/main/episodes/07-r-help.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How do I get help using R and RStudio?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Locate help for an R function using <code>?</code>, <code>??</code>,
and <code>args()</code>
</li>
<li>Check the version of R</li>
<li>Be able to ask effective questions when searching for help on forums
or using web searches</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section id="getting-help-with-r"><h2 class="section-heading">Getting help with R<a class="anchor" aria-label="anchor" href="#getting-help-with-r"></a>
</h2>
<hr class="half-width">
<figure><img src="fig/oreilly_book_covers.png" alt="rstudio default session" style="width: 600px;" class="figure mx-auto d-block"></figure><p>No matter how much experience you have with R, you will find yourself
needing help. There is no shame in researching how to do something in R,
and most people will find themselves looking up how to do the same
things that they “should know how to do” over and over again. Here are
some tips to make this process as helpful and efficient as possible.</p>
<blockquote>
<p>“Never memorize something that you can look up” -- A. Einstein</p>
</blockquote>
</section><section id="finding-help-on-stackoverflow-and-biostars"><h2 class="section-heading">Finding help on Stackoverflow and Biostars<a class="anchor" aria-label="anchor" href="#finding-help-on-stackoverflow-and-biostars"></a>
</h2>
<hr class="half-width">
<p>Two popular websites will be of great help with many R problems. For
<strong>general</strong> <strong>R questions</strong>, <a href="https://stackoverflow.com/" class="external-link">Stack Overflow</a> is probably the
most popular online community for developers. If you start your question
“How to do X in R” results from Stack Overflow are usually near the top
of the list. For <strong>bioinformatics specific questions</strong>, <a href="https://www.biostars.org/" class="external-link">Biostars</a> is a popular online
forum.</p>
<div id="tip-asking-for-help-using-online-forums" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="tip-asking-for-help-using-online-forums" class="callout-inner">
<h3 class="callout-title">Tip: Asking for help using online forums:<a class="anchor" aria-label="anchor" href="#tip-asking-for-help-using-online-forums"></a>
</h3>
<div class="callout-content">
<ul>
<li>When searching for R help, look for answers with the <a href="https://stackoverflow.com/questions/tagged/r" class="external-link">r</a> tag.</li>
<li>Get an account; not required to view answers but to required to
post</li>
<li>Put in effort to check thoroughly before you post a question; folks
get annoyed if you ask a very common question that has been answered
multiple times</li>
<li>Be careful. While forums are very helpful, you can’t know for sure
if the advice you are getting is correct</li>
<li>See the <a href="https://blog.revolutionanalytics.com/2014/01/how-to-ask-for-r-help.html" class="external-link">How
to ask for R help</a> blog post for more useful tips</li>
</ul>
</div>
</div>
</div>
</section><section id="help-people-help-you"><h2 class="section-heading">Help people help you<a class="anchor" aria-label="anchor" href="#help-people-help-you"></a>
</h2>
<hr class="half-width">
<p>Often, in order to duplicate the issue you are having, someone may
need to see the data you are working with or verify the versions of R or
R packages you are using. The following R functions will help with
this:</p>
<p>You can <strong>check the version of R</strong> you are working with
using the <code>sessionInfo()</code> function. Actually, it is good to
save this information as part of your notes on any analysis you are
doing. When you run the same script that has worked fine a dozen times
before, looking back at these notes will remind you that you upgraded R
and forget to check your script.</p>
<pre><code><span><span class="fu">sessionInfo</span><span class="op">(</span><span class="op">)</span></span></code></pre>
<pre><code>R version 3.2.3 (2015-12-10)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 14.04.3 LTS

locale:
[1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8
[4] LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8
[7] LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C
[10] LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base

loaded via a namespace (and not attached):
[1] tools_3.2.3     packrat_0.4.9-1</code></pre>
<p>Many times, there may be some issues with your data and the way it is
formatted. In that case, you may want to share that data with someone
else. However, you may not need to share the whole dataset; looking at a
subset of your 50,000 row, 10,000 column dataframe may be TMI (too much
information)! You can take an object you have in memory such as
dataframe (if you don’t know what this means yet, we will get to it!)
and save it to a file. In our example we will use the
<code>dput()</code> function on the <code>iris</code> dataframe which is
an example dataset that is installed in R:</p>
<pre><code><span><span class="fu">dput</span><span class="op">(</span><span class="fu">head</span><span class="op">(</span><span class="va">iris</span><span class="op">)</span><span class="op">)</span> <span class="co"># iris is an example data.frame that comes with R</span></span>
<span>                 <span class="co"># the `head()` function just takes the first 6 lines of the iris dataset</span></span></code></pre>
<p>This generates some output (below) which you will be better able to
interpret after covering the other R lessons. This info would be helpful
in understanding how the data is formatted and possibly revealing
problematic issues.</p>
<pre><code><span><span class="fu">structure</span><span class="op">(</span><span class="fu">list</span><span class="op">(</span>Sepal.Length <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">5.1</span>, <span class="fl">4.9</span>, <span class="fl">4.7</span>, <span class="fl">4.6</span>, <span class="fl">5</span>, <span class="fl">5.4</span><span class="op">)</span>,</span>
<span>    Sepal.Width <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">3.5</span>, <span class="fl">3</span>, <span class="fl">3.2</span>, <span class="fl">3.1</span>, <span class="fl">3.6</span>, <span class="fl">3.9</span><span class="op">)</span>, Petal.Length <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">1.4</span>,</span>
<span>    <span class="fl">1.4</span>, <span class="fl">1.3</span>, <span class="fl">1.5</span>, <span class="fl">1.4</span>, <span class="fl">1.7</span><span class="op">)</span>, Petal.Width <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">0.2</span>, <span class="fl">0.2</span>, <span class="fl">0.2</span>,</span>
<span>    <span class="fl">0.2</span>, <span class="fl">0.2</span>, <span class="fl">0.4</span><span class="op">)</span>, Species <span class="op">=</span> <span class="fu">structure</span><span class="op">(</span><span class="fu">c</span><span class="op">(</span><span class="fl">1L</span>, <span class="fl">1L</span>, <span class="fl">1L</span>, <span class="fl">1L</span>, <span class="fl">1L</span>,</span>
<span>    <span class="fl">1L</span><span class="op">)</span>, .Label <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"setosa"</span>, <span class="st">"versicolor"</span>, <span class="st">"virginica"</span><span class="op">)</span>, class <span class="op">=</span> <span class="st">"factor"</span><span class="op">)</span><span class="op">)</span>, .Names <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"Sepal.Length"</span>,</span>
<span><span class="st">"Sepal.Width"</span>, <span class="st">"Petal.Length"</span>, <span class="st">"Petal.Width"</span>, <span class="st">"Species"</span><span class="op">)</span>, row.names <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="cn">NA</span>,</span>
<span><span class="fl">6L</span><span class="op">)</span>, class <span class="op">=</span> <span class="st">"data.frame"</span><span class="op">)</span></span></code></pre>
<p>Alternatively, you can also save objects in R memory to a file by
specifying the name of the object, in this case the <code>iris</code>
data frame, and passing a filename to the <code>file=</code>
argument.</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">saveRDS</span><span class="op">(</span><span class="va">iris</span>, file<span class="op">=</span><span class="st">"iris.rds"</span><span class="op">)</span> <span class="co"># By convention, we use the .rds file extension</span></span></code></pre>
</div>
</section><section id="final-faqs-on-r"><h2 class="section-heading">Final FAQs on R<a class="anchor" aria-label="anchor" href="#final-faqs-on-r"></a>
</h2>
<hr class="half-width">
<p>Finally, here are a few pieces of introductory R knowledge that are
too good to pass up. While we won’t return to them in this course, we
put them here because they come up commonly:</p>
<p><strong>Do I need to click Run every time I want to run a
script?</strong></p>
<ul>
<li>
<p>No. In fact, the most common shortcut key allows you to run a
command (or any lines of the script that are highlighted):</p>
<ul>
<li>Windows execution shortcut: <kbd>Ctrl</kbd>+<kbd>Enter</kbd>
</li>
<li>Mac execution shortcut: <kbd>Cmd(⌘)</kbd>+<kbd>Enter</kbd>
</li>
</ul>
<p>To see a complete list of shortcuts, click on the <kbd>Tools</kbd>
menu and select <kbd>Keyboard Shortcuts Help</kbd></p>
</li>
</ul>
<p><strong>What’s with the brackets in R console output?</strong></p>
<ul>
<li>R returns an index with your result. When your result contains
multiple values, the number tells you what ordinal number begins the
line, for example:</li>
</ul>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fl">1</span><span class="op">:</span><span class="fl">101</span> <span class="co"># generates the sequence of numbers from 1 to 101</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>  [1]   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17  18
 [19]  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35  36
 [37]  37  38  39  40  41  42  43  44  45  46  47  48  49  50  51  52  53  54
 [55]  55  56  57  58  59  60  61  62  63  64  65  66  67  68  69  70  71  72
 [73]  73  74  75  76  77  78  79  80  81  82  83  84  85  86  87  88  89  90
 [91]  91  92  93  94  95  96  97  98  99 100 101</code></pre>
</div>
<p>In the output above, <code>[81]</code> indicates that the first value
on that line is the 81st item in your result</p>
<p><strong>Can I run my R script without RStudio?</strong></p>
<ul>
<li>Yes, remember - RStudio is running R. You get to use lots of the
enhancements RStudio provides, but R works independent of RStudio. See
<a href="https://support.rstudio.com/hc/en-us/articles/218012917-How-to-run-R-scripts-from-the-command-line" class="external-link">these
tips</a> for running your commands at the command line</li>
</ul>
<p><strong>Where else can I learn about RStudio?</strong></p>
<ul>
<li>Check out the <kbd>Help</kbd> menu, especially “Cheatsheets”
section</li>
</ul>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Keypoints<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>R provides thousands of functions for analyzing data, and provides
several way to get help</li>
<li>Using R will mean searching for online help, and there are tips and
resources on how to search effectively</li>
</ul>
</div>
</div>
</div>
</section></section>
</div>
    </main>
</div>
<!-- END  : inst/pkgdown/templates/content-extra.html -->

      </div>
<!--/div.row-->
      		<footer class="row footer mx-md-3"><hr>
<div class="col-md-6">
				<p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>
        
        <a href="https://github.com/datacarpentry/genomics-r-intro/edit/main/README.md" class="external-link">Edit on GitHub</a>
        
	
        | <a href="https://github.com/datacarpentry/genomics-r-intro/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a> 
        | <a href="https://github.com/datacarpentry/genomics-r-intro/" class="external-link">Source</a></p>
				<p><a href="https://github.com/datacarpentry/genomics-r-intro/blob/main/CITATION" class="external-link">Cite</a> | <a href="mailto:team@carpentries.org">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">
        
        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>
        
        <p><a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">Template licensed under CC-BY 4.0</a> by <a href="https://carpentries.org" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.12.2" class="external-link">sandpaper (0.12.2)</a>,
        <a href="https://github.com/carpentries/pegboard" class="external-link">pegboard (0.5.3)</a>,
      and <a href="https://github.com/carpentries/varnish/tree/0.2.17" class="external-link">varnish (0.2.17)</a>.</p>
			</div>
		</footer>
</div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
			<i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back to top"></i><br><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "TrainingMaterial",
  "@id": "https://datacarpentry.github.io/genomics-r-intro/aio.html",
  "dct:conformsTo": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, The Carpentries",
  "name": "All in One View",
  "creativeWorkStatus": "active",
  "url": "https://datacarpentry.github.io/genomics-r-intro/aio.html",
  "identifier": "https://datacarpentry.github.io/genomics-r-intro/aio.html",
  "dateCreated": "2018-03-12",
  "dateModified": "2023-05-30",
  "datePublished": "2023-05-30"
}

  </script><script>
		feather.replace();
	</script><!-- Matomo
    2022-11-07: we have gotten a notification that we have an overage for our
    tracking and I'm pretty sure this has to do with Workbench usage.
    Considering that I am not _currently_ using this tracking because I do not
    yet know how to access the data, I am turning this off for now.
  <script>
    var _paq = window._paq = window._paq || [];
    /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
    _paq.push(["setDocumentTitle", document.domain + "/" + document.title]);
    _paq.push(["setDomains", ["*.preview.carpentries.org","*.datacarpentry.github.io","*.datacarpentry.org","*.librarycarpentry.github.io","*.librarycarpentry.org","*.swcarpentry.github.io", "*.carpentries.github.io"]]);
    _paq.push(["setDoNotTrack", true]);
    _paq.push(["disableCookies"]);
    _paq.push(['trackPageView']);
    _paq.push(['enableLinkTracking']);
    (function() {
          var u="https://carpentries.matomo.cloud/";
          _paq.push(['setTrackerUrl', u+'matomo.php']);
          _paq.push(['setSiteId', '1']);
          var d=document, g=d.createElement('script'), s=d.getElementsByTagName('script')[0];
          g.async=true; g.src='https://cdn.matomo.cloud/carpentries.matomo.cloud/matomo.js'; s.parentNode.insertBefore(g,s);
        })();
  </script>
  End Matomo Code -->
</body>
</html><!-- END:   inst/pkgdown/templates/layout.html-->

