<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><title>Intro to R and RStudio for Genomics: Introduction to the example dataset and file type</title><meta name="viewport" content="width=device-width, initial-scale=1"><link rel="stylesheet" type="text/css" href="../assets/styles.css"><script src="../assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="manifest" href="../site.webmanifest"><link rel="mask-icon" href="../safari-pinned-tab.svg" color="#5bbad5"><meta name="msapplication-TileColor" content="#da532c"><meta name="theme-color" content="#ffffff"></head><body>
    <header id="top" class="navbar navbar-expand-md navbar-light bg-white top-nav data"><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-6">
      <div class="large-logo">
        <img alt="Data Carpentry" src="../assets/images/data-logo.svg"><abbr class="badge badge-light" title="This lesson is in the beta phase, which means that it is ready for teaching by instructors outside of the original author team." style="background-color: #001483; border-radius: 5px">
          <a href="https://cdh.carpentries.org/the-lesson-life-cycle.html#polishing-beta-stage" class="external-link alert-link" style="color: #FFF7F1">
            <i aria-hidden="true" class="icon" data-feather="alert-circle" style="border-radius: 5px"></i>
            Beta
          </a>
          <span class="visually-hidden">This lesson is in the beta phase, which means that it is ready for teaching by instructors outside of the original author team.</span>
        </abbr>
        
      </div>
    </div>
    <div class="selector-container">
      
      
      <div class="dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1"><li><button class="dropdown-item" type="button" onclick="window.location.href='../02-data-prelude.html';">Learner View</button></li>
        </ul></div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl navbar-light bg-white bottom-nav data" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Data Carpentry" src="../assets/images/data-logo-sm.svg"></div>
    <div class="lesson-title-md">
      Intro to R and RStudio for Genomics
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="search button" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0"><li class="nav-item">
          <span class="lesson-title">
            Intro to R and RStudio for Genomics
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/instructor-notes.html">Instructor Notes</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/images.html">Extract All Images</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"><hr><li><a class="dropdown-item" href="discuss.html">Discussion</a></li><li><a class="dropdown-item" href="reference.html">Glossary</a></li>
          </ul></li>
      </ul></div>
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled><input class="form-control me-2 searchbox" type="search" placeholder="Search" aria-label="Search"><button class="btn btn-outline-success tablet-search-button" type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="search button"></i>
        </button>
      </fieldset></form>
  </div><!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Intro to R and RStudio for Genomics
</div>

<aside class="col-md-12 lesson-progress"><div style="width: 31%" class="percentage">
    31%
  </div>
  <div class="progress data">
    <div class="progress-bar data" role="progressbar" style="width: 31%" aria-valuenow="31" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="../02-data-prelude.html">Learner View</a>
                      </div>
                    </div>
                  </div><!--/div.accordion-item-->
                </div><!--/div.accordion-flush-->
              </div><!--div.sidenav-view-selector -->
            </div><!--/div.col -->
      
            <hr></div><!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Schedule</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="00-introduction.html">1. Introducing R and RStudio IDE</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="01-r-basics.html">2. R Basics</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlushcurrent">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-headingcurrent">
      <button class="accordion-button" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapsecurrent" aria-expanded="true" aria-controls="flush-collapsecurrent">
        <span class="visually-hidden">Current Chapter</span>
        <span class="current-chapter">
        3. Introduction to the example dataset and file type
        </span>
      </button>
    </div><!--/div.accordion-header-->
        
    <div id="flush-collapsecurrent" class="accordion-collapse collapse show" aria-labelledby="flush-headingcurrent" data-bs-parent="#accordionFlushcurrent">
      <div class="accordion-body">
        <ul><li><a href="#preface">Preface</a></li>
<li><a href="#introduction-to-the-dataset">Introduction to the dataset</a></li>
<li><a href="#how-vcf-files-are-generated">How VCF files are generated</a></li>
<li><a href="#how-variant-calls-are-stored-in-vcf-files">How variant calls are stored in VCF files</a></li>
<li><a href="#references">References</a></li>
        </ul></div><!--/div.accordion-body-->
    </div><!--/div.accordion-collapse-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="03-basics-factors-dataframes.html">4. R Basics continued - factors and data frames</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="04-bioconductor-vcfr.html">5. Using packages from Bioconductor</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="05-dplyr.html">6. Data Wrangling and Analyses with Tidyverse</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush8">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading8">
        <a href="06-data-visualization.html">7. Data Visualization with ggplot2</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush9">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading9">
        <a href="07-r-help.html">8. Getting help with R</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width"><div class="accordion accordion-flush resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul><li>
                        <a href="../instructor/key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="../instructor/instructor-notes.html">Instructor Notes</a>
                      </li>
                      <li>
                        <a href="../instructor/images.html">Extract All Images</a>
                      </li>
                      <hr><li><a class="dropdown-item" href="discuss.html">Discussion</a></li><li><a class="dropdown-item" href="reference.html">Glossary</a></li>
                    </ul></div>
                </div>
              </div>
            </div>
            <hr class="half-width resources"><a href="../instructor/aio.html">See all in one page</a>
            

            <hr class="d-none d-sm-block d-md-none"><div class="d-grid gap-1">
            
            </div>
          </div><!-- /div.accordion -->
        </div><!-- /div.sidebar-inner -->
      </nav></div><!-- /div.sidebar -->
  </div><!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-instructor.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <nav class="lesson-content mx-md-4" aria-label="Previous and Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="../instructor/01-r-basics.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="../instructor/03-basics-factors-dataframes.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="../instructor/01-r-basics.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: R Basics
        </a>
        <a class="chapter-link float-end" href="../instructor/03-basics-factors-dataframes.html" rel="next">
          Next: R Basics continued -... 
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
      <hr></nav><main id="main-content" class="main-content"><div class="container lesson-content">
        <h1>Introduction to the example dataset and file type</h1>
        <p>Last updated on 2024-03-12 |
        
        <a href="https://github.com/datacarpentry/genomics-r-intro/edit/main/episodes/02-data-prelude.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
        
        
        
        <p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 15 minutes</p>
        
        <div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>

        

<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul><li>What data are we using in the lesson?</li>
<li>What are VCF files?</li>
</ul></div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul><li>Know what the example dataset represents</li>
<li>Know the concepts of how VCF files are generated</li>
</ul></div>
</div>
</div>
</div>
</div>
<section id="preface"><h2 class="section-heading">Preface<a class="anchor" aria-label="anchor" href="#preface"></a>
</h2>
<hr class="half-width"><p>The Intro to R and RStudio for Genomics is a part of the Genomics
Data Carpentry lessons. In this lesson we will learn the necessary skill
sets for R and RStudio and apply them directly to a real next-generation
sequencing (NGS) data in the variant calling format (VCF) file type.
Previous Genomics Data Carpentry lessons teach learners how to generate
a VCF file from FASTQ files downloaded from NCBI Sequence Read Archive
(SRA), so we won’t cover that here. Instead, in this episode we will
give a brief overview of the data and a what VCF file types are for
those who wish to teach the Intro to R and RStudio for Genomics lesson
independently of the Genomics Data Carpentry lessons.</p>
<p>This dataset was selected for several reasons, including:</p>
<ul><li>Simple, but iconic NGS-problem: Examine a population where we want
to characterize changes in sequence <em>a priori</em>
</li>
<li>Dataset publicly available - in this case through the NCBI SRA (<a href="https://www.ncbi.nlm.nih.gov/sra" class="external-link">http://www.ncbi.nlm.nih.gov/sra</a>)</li>
</ul></section><section id="introduction-to-the-dataset"><h2 class="section-heading">Introduction to the dataset<a class="anchor" aria-label="anchor" href="#introduction-to-the-dataset"></a>
</h2>
<hr class="half-width"><p>Microbes are ideal organisms for exploring ‘Long-term Evolution
Experiments’ (LTEEs) - thousands of generations can be generated and
stored in a way that would be virtually impossible for more complex
eukaryotic systems. In <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4988878/" class="external-link">Tenaillon
et al 2016</a>, 12 populations of <em>Escherichia coli</em> were
propagated for more than 50,000 generations in a glucose-limited minimal
medium. This medium was supplemented with citrate which <em>E. coli</em>
cannot metabolize in the aerobic conditions of the experiment.
Sequencing of the populations at regular time points reveals that
spontaneous citrate-using mutants (Cit+) appeared in a population of
<em>E.coli</em> (designated Ara-3) at around 31,000 generations. It
should be noted that spontaneous Cit+ mutants are extraordinarily rare -
inability to metabolize citrate is one of the defining characters of the
<em>E. coli</em> species. Eventually, Cit+ mutants became the dominant
population as the experimental growth medium contained a high
concentration of citrate relative to glucose. Around the same time that
this mutation emerged, another phenotype become prominent in the Ara-3
population. Many <em>E. coli</em> began to develop excessive numbers of
mutations, meaning they became hypermutable.</p>
<p>Strains from generation 0 to generation 50,000 were sequenced,
including ones that were both Cit+ and Cit- and hypermutable in later
generations.</p>
<p>For the purposes of this workshop we’re going to be working with 3 of
the sequence reads from this experiment.</p>
<table class="table"><colgroup><col width="3%"><col width="82%"><col width="2%"><col width="1%"><col width="3%"><col width="2%"><col width="4%"></colgroup><thead><tr class="header"><th>SRA Run Number</th>
<th>Clone</th>
<th>Generation</th>
<th>Cit</th>
<th>Hypermutable</th>
<th>Read Length</th>
<th>Sequencing Depth</th>
</tr></thead><tbody><tr class="odd"><td>SRR2589044</td>
<td>REL2181A</td>
<td>5,000</td>
<td>Unknown</td>
<td>None</td>
<td>150</td>
<td>60.2</td>
</tr><tr class="even"><td>SRR2584863</td>
<td>REL7179B</td>
<td>15,000</td>
<td>Unknown</td>
<td>None</td>
<td>150</td>
<td>88</td>
</tr><tr class="odd"><td>SRR2584866</td>
<td>REL11365</td>
<td>50,000</td>
<td>Cit+</td>
<td>plus</td>
<td>150</td>
<td>138.3</td>
</tr></tbody></table><p>We want to be able to look at differences in mutation rates between
hypermutable and non-hypermutable strains. We also want to analyze the
sequences to figure out what changes occurred in genomes to make the
strains Cit+. Ultimately, we will use R to answer the questions:</p>
<ul><li>How many base pair changes are there between the Cit+ and Cit-
strains?</li>
<li>What are the base pair changes between strains?</li>
</ul></section><section id="how-vcf-files-are-generated"><h2 class="section-heading">How VCF files are generated<a class="anchor" aria-label="anchor" href="#how-vcf-files-are-generated"></a>
</h2>
<hr class="half-width"><p>Publicly accessible sequencing files in FASTQ formats can be
downloaded from NCBI SRA. However, at FASTQ files contain unaligned
sequences of varying quality, and requires clean up and alignment steps
for variants to be called from the reference genome.</p>
<p>Five steps are taken to transform FASTQ files to variant calls
contained in VCF files and at each step, specialized non-R based
bioinformatics tools that are used:</p>
<center>
<img src="../fig/variant_calling_workflow.png" alt="variant calling workflow. Sequence reads (FASTQ files), Quality control (FASTQ files), Alignment to Genome (SAM/BAM files), Alignment cleanup (BAM file ready for variant calling), Variant Calling (VCF file)" style="width: 250px;" class="figure"></center>
</section><section id="how-variant-calls-are-stored-in-vcf-files"><h2 class="section-heading">How variant calls are stored in VCF files<a class="anchor" aria-label="anchor" href="#how-variant-calls-are-stored-in-vcf-files"></a>
</h2>
<hr class="half-width"><p>VCF files contain variants that were called against a reference
genome. These files are slightly more complicated than regular tables
you can open using programs like Excel and contain two sections: header
and records.</p>
<p>Below you will see the header (which describes the format), the time
and date the file was created, the version of bcftools that was used,
the command line parameters used, and some additional information:</p>
<pre><code><span><span class="co">##fileformat=VCFv4.2</span></span>
<span><span class="co">##FILTER=&lt;ID=PASS,Description="All filters passed"&gt;</span></span>
<span><span class="co">##bcftoolsVersion=1.8+htslib-1.8</span></span>
<span><span class="co">##bcftoolsCommand=mpileup -O b -o results/bcf/SRR2584866_raw.bcf -f data/ref_genome/ecoli_rel606.fasta results/bam/SRR2584866.aligned.sorted.bam</span></span>
<span><span class="co">##reference=file://data/ref_genome/ecoli_rel606.fasta</span></span>
<span><span class="co">##contig=&lt;ID=CP000819.1,length=4629812&gt;</span></span>
<span><span class="co">##ALT=&lt;ID=*,Description="Represents allele(s) other than observed."&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=INDEL,Number=0,Type=Flag,Description="Indicates that the variant is an INDEL."&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=IDV,Number=1,Type=Integer,Description="Maximum number of reads supporting an indel"&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=IMF,Number=1,Type=Float,Description="Maximum fraction of reads supporting an indel"&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=DP,Number=1,Type=Integer,Description="Raw read depth"&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=VDB,Number=1,Type=Float,Description="Variant Distance Bias for filtering splice-site artefacts in RNA-seq data (bigger is better)",Version=</span></span>
<span><span class="co">##INFO=&lt;ID=RPB,Number=1,Type=Float,Description="Mann-Whitney U test of Read Position Bias (bigger is better)"&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=MQB,Number=1,Type=Float,Description="Mann-Whitney U test of Mapping Quality Bias (bigger is better)"&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=BQB,Number=1,Type=Float,Description="Mann-Whitney U test of Base Quality Bias (bigger is better)"&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=MQSB,Number=1,Type=Float,Description="Mann-Whitney U test of Mapping Quality vs Strand Bias (bigger is better)"&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=SGB,Number=1,Type=Float,Description="Segregation based metric."&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=MQ0F,Number=1,Type=Float,Description="Fraction of MQ0 reads (smaller is better)"&gt;</span></span>
<span><span class="co">##FORMAT=&lt;ID=PL,Number=G,Type=Integer,Description="List of Phred-scaled genotype likelihoods"&gt;</span></span>
<span><span class="co">##FORMAT=&lt;ID=GT,Number=1,Type=String,Description="Genotype"&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=ICB,Number=1,Type=Float,Description="Inbreeding Coefficient Binomial test (bigger is better)"&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=HOB,Number=1,Type=Float,Description="Bias in the number of HOMs number (smaller is better)"&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=AC,Number=A,Type=Integer,Description="Allele count in genotypes for each ALT allele, in the same order as listed"&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=AN,Number=1,Type=Integer,Description="Total number of alleles in called genotypes"&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=DP4,Number=4,Type=Integer,Description="Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases"&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=MQ,Number=1,Type=Integer,Description="Average mapping quality"&gt;</span></span>
<span><span class="co">##bcftools_callVersion=1.8+htslib-1.8</span></span>
<span><span class="co">##bcftools_callCommand=call --ploidy 1 -m -v -o results/bcf/SRR2584866_variants.vcf results/bcf/SRR2584866_raw.bcf; Date=Tue Oct  9 18:48:10 2018</span></span></code></pre>
<p>Followed by information on each of the variations observed:</p>
<pre><code>#CHROM  POS     ID      REF     ALT     QUAL    FILTER  INFO    FORMAT  results/bam/SRR2584866.aligned.sorted.bam
CP000819.1      1521    .       C       T       207     .       DP=9;VDB=0.993024;SGB=-0.662043;MQSB=0.974597;MQ0F=0;AC=1;AN=1;DP4=0,0,4,5;MQ=60
CP000819.1      1612    .       A       G       225     .       DP=13;VDB=0.52194;SGB=-0.676189;MQSB=0.950952;MQ0F=0;AC=1;AN=1;DP4=0,0,6,5;MQ=60
CP000819.1      9092    .       A       G       225     .       DP=14;VDB=0.717543;SGB=-0.670168;MQSB=0.916482;MQ0F=0;AC=1;AN=1;DP4=0,0,7,3;MQ=60
CP000819.1      9972    .       T       G       214     .       DP=10;VDB=0.022095;SGB=-0.670168;MQSB=1;MQ0F=0;AC=1;AN=1;DP4=0,0,2,8;MQ=60      GT:PL
CP000819.1      10563   .       G       A       225     .       DP=11;VDB=0.958658;SGB=-0.670168;MQSB=0.952347;MQ0F=0;AC=1;AN=1;DP4=0,0,5,5;MQ=60
CP000819.1      22257   .       C       T       127     .       DP=5;VDB=0.0765947;SGB=-0.590765;MQSB=1;MQ0F=0;AC=1;AN=1;DP4=0,0,2,3;MQ=60      GT:PL
CP000819.1      38971   .       A       G       225     .       DP=14;VDB=0.872139;SGB=-0.680642;MQSB=1;MQ0F=0;AC=1;AN=1;DP4=0,0,4,8;MQ=60      GT:PL
CP000819.1      42306   .       A       G       225     .       DP=15;VDB=0.969686;SGB=-0.686358;MQSB=1;MQ0F=0;AC=1;AN=1;DP4=0,0,5,9;MQ=60      GT:PL
CP000819.1      45277   .       A       G       225     .       DP=15;VDB=0.470998;SGB=-0.680642;MQSB=0.95494;MQ0F=0;AC=1;AN=1;DP4=0,0,7,5;MQ=60
CP000819.1      56613   .       C       G       183     .       DP=12;VDB=0.879703;SGB=-0.676189;MQSB=1;MQ0F=0;AC=1;AN=1;DP4=0,0,8,3;MQ=60      GT:PL
CP000819.1      62118   .       A       G       225     .       DP=19;VDB=0.414981;SGB=-0.691153;MQSB=0.906029;MQ0F=0;AC=1;AN=1;DP4=0,0,8,10;MQ=59
CP000819.1      64042   .       G       A       225     .       DP=18;VDB=0.451328;SGB=-0.689466;MQSB=1;MQ0F=0;AC=1;AN=1;DP4=0,0,7,9;MQ=60      GT:PL</code></pre>
<p>The first few columns represent the information we have about a
predicted variation.</p>
<table class="table"><colgroup><col width="4%"><col width="95%"></colgroup><thead><tr class="header"><th>column</th>
<th>info</th>
</tr></thead><tbody><tr class="odd"><td>CHROM</td>
<td>contig location where the variation occurs</td>
</tr><tr class="even"><td>POS</td>
<td>position within the contig where the variation occurs</td>
</tr><tr class="odd"><td>ID</td>
<td>a <code>.</code> until we add annotation information</td>
</tr><tr class="even"><td>REF</td>
<td>reference genotype (forward strand)</td>
</tr><tr class="odd"><td>ALT</td>
<td>sample genotype (forward strand)</td>
</tr><tr class="even"><td>QUAL</td>
<td>Phred-scaled probability that the observed variant exists at this
site (higher is better)</td>
</tr><tr class="odd"><td>FILTER</td>
<td>a <code>.</code> if no quality filters have been applied, PASS if a
filter is passed, or the name of the filters this variant failed</td>
</tr></tbody></table><p>In an ideal world, the information in the <code>QUAL</code> column
would be all we needed to filter out bad variant calls. However, in
reality we need to filter on multiple other metrics.</p>
<p>The last two columns contain the genotypes and can be tricky to
decode.</p>
<table class="table"><colgroup><col width="4%"><col width="95%"></colgroup><thead><tr class="header"><th>column</th>
<th>info</th>
</tr></thead><tbody><tr class="odd"><td>FORMAT</td>
<td>lists in order the metrics presented in the final column</td>
</tr><tr class="even"><td>results</td>
<td>lists the values associated with those metrics in order</td>
</tr></tbody></table><p>For our file, the metrics presented are GT:PL:GQ.</p>
<table class="table"><colgroup><col width="4%"><col width="95%"></colgroup><thead><tr class="header"><th>metric</th>
<th>definition</th>
</tr></thead><tbody><tr class="odd"><td>AD, DP</td>
<td>the depth per allele by sample and coverage</td>
</tr><tr class="even"><td>GT</td>
<td>the genotype for the sample at this loci. For a diploid organism,
the GT field indicates the two alleles carried by the sample, encoded by
a 0 for the REF allele, 1 for the first ALT allele, 2 for the second ALT
allele, etc. A 0/0 means homozygous reference, 0/1 is heterozygous, and
1/1 is homozygous for the alternate allele.</td>
</tr><tr class="odd"><td>PL</td>
<td>the likelihoods of the given genotypes</td>
</tr><tr class="even"><td>GQ</td>
<td>the Phred-scaled confidence for the genotype</td>
</tr></tbody></table><p>For more information on VCF files visit The Broad Institute’s <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360035531692-VCF-Variant-Call-Format" class="external-link">VCF
guide</a>.</p>
</section><section id="references"><h2 class="section-heading">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<hr class="half-width"><p>Tenaillon O, Barrick JE, Ribeck N, Deatherage DE, Blanchard JL,
Dasgupta A, Wu GC, Wielgoss S, Cruveiller S, Médigue C, Schneider D,
Lenski RE. Tempo and mode of genome evolution in a 50,000-generation
experiment (2016) Nature. 536(7615): 165–170. <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4988878/" class="external-link">Paper</a>,
<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4988878/#" class="external-link">Supplemental
materials</a> Data on NCBI SRA: <a href="https://trace.ncbi.nlm.nih.gov/Traces/sra/?study=SRP064605" class="external-link">https://trace.ncbi.nlm.nih.gov/Traces/sra/?study=SRP064605</a>
Data on EMBL-EBI ENA: <a href="https://www.ebi.ac.uk/ena/data/view/PRJNA295606" class="external-link">https://www.ebi.ac.uk/ena/data/view/PRJNA295606</a></p>
<p>This episode was adapted from the Data Carpentry Genomic lessons:</p>
<ul><li><a href="https://datacarpentry.org/organization-genomics/data/" class="external-link">Project
organization and management for Genomics</a></li>
<li><a href="https://datacarpentry.org/wrangling-genomics/04-variant_calling/index.html" class="external-link">Data
wrangling and processing for genomics</a></li>
</ul><div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul><li>The dataset comes from a real world experiment in E. coli.</li>
<li>Publicly available FASTQ files can be downloaded from NCBI SRA.</li>
<li>Several steps are taken outside of R/RStudio to create VCF files
from FASTQ files.</li>
<li>VCF files store variant calls in a special format.</li>
</ul></div>
</div>
</div>
</section></div> <!-- / div.lesson-content -->
    </main><!-- / main#main-content.main-content --><nav class="bottom-pagination mx-md-4" aria-label="Previous and Next Chapter"><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="../instructor/01-r-basics.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="../instructor/03-basics-factors-dataframes.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="../instructor/01-r-basics.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: R Basics
        </a>
        <a class="chapter-link float-end" href="../instructor/03-basics-factors-dataframes.html" rel="next">
          Next: R Basics continued -... 
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
    </nav></div> <!-- / div.primary-content.col-xs-12 -->
<!-- END:   inst/pkgdown/templates/content-instructor.html-->

      </div><!--/div.row-->
      		<footer class="row footer mx-md-3"><hr><div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>
        
        <a href="https://github.com/datacarpentry/genomics-r-intro/edit/main/episodes/02-data-prelude.Rmd" class="external-link">Edit on GitHub</a>
        
	
        | <a href="https://github.com/datacarpentry/genomics-r-intro/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/datacarpentry/genomics-r-intro/" class="external-link">Source</a></p>
				<p><a href="https://github.com/datacarpentry/genomics-r-intro/blob/main/CITATION" class="external-link">Cite</a> | <a href="mailto:team@carpentries.org">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">
        
        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>
        
        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.16.3" class="external-link">sandpaper (0.16.3)</a>, <a href="https://github.com/carpentries/pegboard/tree/0.7.4" class="external-link">pegboard (0.7.4)</a>, and <a href="https://github.com/carpentries/varnish/tree/1.0.1" class="external-link">varnish (1.0.1)</a></p>
			</div>
		</footer></div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "TrainingMaterial",
  "@id": "https://datacarpentry.github.io/genomics-r-intro/instructor/02-data-prelude.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, The Carpentries",
  "name": "Introduction to the example dataset and file type",
  "creativeWorkStatus": "active",
  "url": "https://datacarpentry.github.io/genomics-r-intro/instructor/02-data-prelude.html",
  "identifier": "https://datacarpentry.github.io/genomics-r-intro/instructor/02-data-prelude.html",
  "dateCreated": "2018-03-12",
  "dateModified": "2024-03-12",
  "datePublished": "2024-03-13"
}

  </script><script>
		feather.replace();
	</script><!-- Matomo
    2022-11-07: we have gotten a notification that we have an overage for our
    tracking and I'm pretty sure this has to do with Workbench usage.
    Considering that I am not _currently_ using this tracking because I do not
    yet know how to access the data, I am turning this off for now.
  <script>
    var _paq = window._paq = window._paq || [];
    /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
    _paq.push(["setDocumentTitle", document.domain + "/" + document.title]);
    _paq.push(["setDomains", ["*.preview.carpentries.org","*.datacarpentry.github.io","*.datacarpentry.org","*.librarycarpentry.github.io","*.librarycarpentry.org","*.swcarpentry.github.io", "*.carpentries.github.io"]]);
    _paq.push(["setDoNotTrack", true]);
    _paq.push(["disableCookies"]);
    _paq.push(['trackPageView']);
    _paq.push(['enableLinkTracking']);
    (function() {
          var u="https://carpentries.matomo.cloud/";
          _paq.push(['setTrackerUrl', u+'matomo.php']);
          _paq.push(['setSiteId', '1']);
          var d=document, g=d.createElement('script'), s=d.getElementsByTagName('script')[0];
          g.async=true; g.src='https://cdn.matomo.cloud/carpentries.matomo.cloud/matomo.js'; s.parentNode.insertBefore(g,s);
        })();
  </script>
  End Matomo Code --></body></html><!-- END:   inst/pkgdown/templates/layout.html-->

