<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><title>Intro to R and RStudio for Genomics: Summary and Schedule</title><meta name="viewport" content="width=device-width, initial-scale=1"><link rel="stylesheet" type="text/css" href="../assets/styles.css"><script src="../assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [ ['$','$'], ['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="manifest" href="../site.webmanifest"><link rel="mask-icon" href="../safari-pinned-tab.svg" color="#5bbad5"><meta name="msapplication-TileColor" content="#da532c"><meta name="theme-color" content="#ffffff"></head><body>
    <header id="top" class="navbar navbar-expand-md navbar-light bg-white top-nav data"><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-6">
      <div class="large-logo">
        <img alt="Data Carpentry" src="../assets/images/data-logo.svg"><abbr class="icon" title="This lesson is in the alpha phase, which means that it has been taught once and lesson authors are iterating on feedback." style="text-decoration: unset">
           
          <a href="https://cdh.carpentries.org/the-lesson-life-cycle.html#field-testing-alpha-stage" class="external-link alert-link" style="color: #383838">Alpha
            <i aria-hidden="true" class="icon" data-feather="alert-triangle" style="background: #FFC700; border-radius: 5px"></i>
          </a>
          <span class="visually-hidden">This lesson is in the alpha phase, which means that it has been taught once and lesson authors are iterating on feedback.</span>
        </abbr>
        
      </div>
    </div>
    <div class="selector-container ">
      
      
      <div class="dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1"><li><button class="dropdown-item" type="button" onclick="window.location.href='../index.html';">Learner View</button></li>
        </ul></div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl navbar-light bg-white bottom-nav data" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Data Carpentry" src="../assets/images/data-logo-sm.svg"></div>
    <div class="lesson-title-md">
      Intro to R and RStudio for Genomics
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
      <i role="img" aria-label="search button" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0"><li class="nav-item">
          <span class="lesson-title">
            Intro to R and RStudio for Genomics
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/instructor-notes.html">Instructor Notes</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/images.html">Extract All Images</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"></ul></li>
      </ul></div>
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled><input class="form-control me-2 searchbox" type="search" placeholder="Search" aria-label="Search"><button class="btn btn-outline-success tablet-search-button" type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="search button"></i>
        </button>
      </fieldset></form>
  </div><!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Intro to R and RStudio for Genomics
</div>

<aside class="col-md-12 lesson-progress"><div style="width: %" class="percentage">
    %
  </div>
  <div class="progress data">
    <div class="progress-bar data" role="progressbar" style="width: %" aria-valuenow="" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
  <div id="sidebar-col" class="col-lg-4">
    <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle"><i class="search-icon" data-feather="x" role="img"></i></button>
        <div class="sidebar-inner">
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="../index.html">Learner View</a>
                      </div>
                    </div>
                  </div><!--/div.accordion-item-->
                </div><!--/div.accordion-flush-->
              </div><!--div.sidenav-view-selector -->
            </div><!--/div.col -->
      
            <hr></div><!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlushcurrent">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-headingcurrent">
        <span class="visually-hidden">Current Chapter</span>
        <span class="current-chapter">
        Summary and Schedule
        </span>
      
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="00-introduction.html">1. Introducing R and RStudio IDE</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="01-r-basics.html">2. R Basics</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="02-data-prelude.html">3. Introduction to the example dataset and file type</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="03-basics-factors-dataframes.html">4. R Basics continued - factors and data frames</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="04-bioconductor-vcfr.html">5. Using packages from Bioconductor</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="05-dplyr.html">6. Data Wrangling and Analyses with Tidyverse</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush8">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading8">
        <a href="06-data-visualization.html">7. Data Visualization with ggplot2</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush9">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading9">
        <a href="07-r-help.html">8. Getting help with R</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width"><div class="accordion accordion-flush resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul><li>
                        <a href="../instructor/key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="../instructor/instructor-notes.html">Instructor Notes</a>
                      </li>
                      <li>
                        <a href="../instructor/images.html">Extract All Images</a>
                      </li>
                      
                    </ul></div>
                </div>
              </div>
            </div>
            <hr class="half-width resources"><a href="../instructor/aio.html">See all in one page</a>

            <hr class="d-none d-sm-block d-md-none"><div class="d-grid gap-1">
            
            </div>
          </div><!-- /div.accordion -->
        </div><!-- /div.sidebar-inner -->
      </nav></div><!-- /div.sidebar -->
  </div><!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-syllabus.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <nav class="lesson-content mx-md-4" aria-label="Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link float-end" href="../instructor/00-introduction.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link float-end" href="../instructor/00-introduction.html" rel="next">Next: Introducing R and... <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <hr></nav><main id="main-content" class="main-content"><h1 class="schedule-heading">Summary and Schedule</h1>
      

<p><strong>Welcome to R!</strong> Working with a programming language
(especially if it’s your first time) often feels intimidating, but the
rewards outweigh any frustrations. An important secret of coding is that
even experienced programmers find it difficult and frustrating at times
– so if even the best feel that way, why let intimidation stop you?
Given time and practice* you will soon find it easier and easier to
accomplish what you want.</p>
<p>Why learn to code? Bioinformatics – like biology – is messy.
Different organisms, different systems, different conditions, all behave
differently. Experiments at the bench require a variety of approaches –
from tested protocols to trial-and-error. Bioinformatics is also an
experimental science, otherwise we could use the same software and same
parameters for every genome assembly. Learning to code opens up the full
possibilities of computing, especially given that most bioinformatics
tools exist only at the command line. Think of it this way: if you could
only do molecular biology using a kit, you could probably accomplish a
fair amount. However, if you don’t understand the biochemistry of the
kit, how would you troubleshoot? How would you do experiments for which
there are no kits?</p>
<p>R is one of the most widely-used and powerful programming languages
in bioinformatics. R especially shines where a variety of statistical
tools are required (e.g. RNA-Seq, population genomics, etc.) and in the
generation of publication-quality graphs and figures. Rather than get
into an R vs. Python debate (both are useful), keep in mind that many of
the concepts you will learn apply to Python and other programming
languages.</p>
<p>Finally, we won’t lie; R is not the easiest-to-learn programming
language ever created. So, don’t get discouraged! The truth is that even
with the modest amount of R we will cover today, you can start using
some sophisticated R software packages, and have a general sense of how
to interpret an R script. Get through these lessons, and you are on your
way to being an accomplished R user!</p>
<p>* We very intentionally used the word practice. One of the other
“secrets” of programming is that you can only learn so much by reading
about it. Do the exercises in class, re-do them on your own, and then
work on your own problems.</p>
<div id="prerequisites" class="callout prereq">
<div class="callout-square">
<i class="callout-icon" data-feather="check"></i>
</div>
<div id="prerequisites" class="callout-inner">
<h3 class="callout-title">Prerequisites<a class="anchor" aria-label="anchor" href="#prerequisites"></a>
</h3>
<div class="callout-content">
<ul><li>
<strong>Experimenter’s Mindset</strong>: We define the
“Experimenter’s mindset” as an approach to bioinformatics that treats it
like any other experiment. There are probably a variety of metaphors we
could employ (data are our reagents, scripts are our protocols, etc.),
but the most important idea of the mindset is to remind you that as a
researcher, you need to employ all of your training in the bench or
field to working with analyses. Evaluate results critically, and don’t
expect that things will always work the first time, or that they will
always work in the same way.</li>
<li>
<strong>Genomics Data Carpentry Instance</strong>: This lesson
assumes you are using a Genomics Data Carpentry instance as described on
the <a href="https://www.datacarpentry.org/genomics-workshop/setup.html" class="external-link">Genomics
Workshop setup page</a>
</li>
</ul></div>
</div>
</div>


      <section id="schedule"><table class="table table schedule table-striped" role="presentation"><tbody><tr><td></td><td><a href="#setup">Setup Instructions</a></td><td> Download files required for the lesson</td>
          </tr><tr><td>
<span class="visually-hidden">Duration: </span>00h 00m
</td>
<td>
 1. <a href="00-introduction.html">Introducing R and RStudio IDE</a>
</td>
<td>
Why use R?<br>Why use RStudio and how does it differ from R?
</td>
</tr><tr><td>
<span class="visually-hidden">Duration: </span>00h 45m
</td>
<td>
 2. <a href="01-r-basics.html">R Basics</a>
</td>
<td>
What will these lessons not cover?<br>What are the basic features of
the R language?<br>What are the most common objects in R?
</td>
</tr><tr><td>
<span class="visually-hidden">Duration: </span>02h 05m
</td>
<td>
 3. <a href="02-data-prelude.html">Introduction to the example dataset
and file type</a>
</td>
<td>
What data are we using in the lesson?<br>What are VCF files?
</td>
</tr><tr><td>
<span class="visually-hidden">Duration: </span>02h 20m
</td>
<td>
 4. <a href="03-basics-factors-dataframes.html">R Basics continued -
factors and data frames</a>
</td>
<td>
How do I get started with tabular data (e.g. spreadsheets) in
R?<br>What are some best practices for reading data into R?<br>How do
I save tabular data generated in R?
</td>
</tr><tr><td>
<span class="visually-hidden">Duration: </span>03h 50m
</td>
<td>
 5. <a href="04-bioconductor-vcfr.html">Using packages from
Bioconductor</a>
</td>
<td>
How do I use packages from the Bioconductor repository?
</td>
</tr><tr><td>
<span class="visually-hidden">Duration: </span>04h 03m
</td>
<td>
 6. <a href="05-dplyr.html">Data Wrangling and Analyses with
Tidyverse</a>
</td>
<td>
How can I manipulate data frames without repeating myself?
</td>
</tr><tr><td>
<span class="visually-hidden">Duration: </span>04h 58m
</td>
<td>
 7. <a href="06-data-visualization.html">Data Visualization with
ggplot2</a>
</td>
<td>
What is ggplot2?<br>What is mapping, and what is aesthetics?<br>What
is the process of creating a publication-quality plots with ggplot in R?
</td>
</tr><tr><td>
<span class="visually-hidden">Duration: </span>06h 28m
</td>
<td>
 8. <a href="07-r-help.html">Getting help with R</a>
</td>
<td>
How do I get help using R and RStudio?
</td>
</tr><tr><td>
<span class="visually-hidden">Duration: </span>06h 43m
</td>
<td>
Finish
</td>
<td>
</td>
</tr></tbody></table><p>
      The actual schedule may vary slightly depending on the topics and exercises chosen by the instructor.
      </p>
      </section><section id="setup"><p>This lesson is an additional lesson to <a href="https://datacarpentry.org/genomics-workshop/" class="external-link">the genomics
workshop</a>. Below, is a detailed setup instructions for the main
workshop which can also be found on <a href="https://datacarpentry.org/genomics-workshop/setup.html" class="external-link">the main
setup page</a>. If you are only here for the Intro to R and RStudio for
Genomics lesson, and do not wish to work on the cloud, you can go for
option B below where you will only need to <a href="https://figshare.com/articles/Data_Carpentry_Genomics_beta_2_0/7726454" class="external-link">download
the data files</a> to your local working directory where you will create
the r-project in.</p>
<div class="section level1">
<h1 id="genomics-workshop-setup-directions">Genomics workshop setup directions<a class="anchor" aria-label="anchor" href="#genomics-workshop-setup-directions"></a></h1>
</div>
<div class="section level1">
<h1 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a></h1>
<p>This workshop is designed to be run on pre-imaged Amazon Web Services
(AWS) instances. With the exception of a spreadsheet program, all of the
software and data used in the workshop are hosted on an Amazon Machine
Image (AMI). Please follow the instructions below to prepare your
computer for the workshop:</p>
<ul><li>Required additional software + Option A<br><strong>OR</strong>
</li>
<li>Required additional software + Option B</li>
</ul><div class="section level2">
<h2 id="required-additional-software">Required additional software<a class="anchor" aria-label="anchor" href="#required-additional-software"></a></h2>
<p>This lesson requires a working spreadsheet program. If you don’t have
a spreadsheet program already, you can use LibreOffice. It’s a free,
open source spreadsheet program. Directions to install are included for
each Windows, Mac OS X, and Linux systems below. For Windows, you will
also need to install Git Bash, PuTTY, or the Ubuntu Subsystem.</p>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">
  Windows
  </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<ul><li>Install LibreOffice by going to <a href="https://www.libreoffice.org/download/libreoffice-fresh/" class="external-link">the
installation page</a>. The version for Windows should automatically be
selected. Click Download Version X.X.X (whichever is the most recent
version). You will go to a page that asks about a donation, but you
don’t need to make one. Your download should begin automatically.</li>
<li>Once the installer is downloaded, double click on it and LibreOffice
should install.</li>
<li>Download the <a href="https://git-for-windows.github.io/" class="external-link">Git for
Windows installer</a>. Run the installer and follow the steps below:
<ul><li>Click on “Next” four times (two times if you’ve previously installed
Git). You don’t need to change anything in the Information, location,
components, and start menu screens.</li>
<li><strong>From the dropdown menu select “Use the Nano editor by
default” (NOTE: you will need to scroll up to find it) and click on
“Next”.</strong></li>
<li>On the page that says “Adjusting the name of the initial branch in
new repositories”, ensure that “Let Git decide” is selected. This will
ensure the highest level of compatibility for our lessons.</li>
<li>Ensure that “Git from the command line and also from 3rd-party
software” is selected and click on “Next”. (If you don’t do this Git
Bash will not work properly, requiring you to remove the Git Bash
installation, re-run the installer and to select the “Git from the
command line and also from 3rd-party software” option.)</li>
<li>Ensure that “Use the native Windows Secure Channel Library” is
selected and click on “Next”.</li>
<li>Ensure that “Checkout Windows-style, commit Unix-style line endings”
is selected and click on “Next”.</li>
<li><strong>Ensure that “Use Windows’ default console window” is
selected and click on “Next”.</strong></li>
<li>Ensure that “Default (fast-forward or merge) is selected and
click”Next”</li>
<li>Ensure that “Git Credential Manager Core” is selected and click on
“Next”.</li>
<li>Ensure that “Enable file system caching” is selected and click on
“Next”.</li>
<li>Click on “Install”.</li>
<li>Click on “Finish”.</li>
<li>Check the settings for you your “HOME” environment variable.
<ul><li>If your “HOME” environment variable is not set (or you don’t know
what this is):</li>
<li>Open command prompt (Open Start Menu then type <code>cmd</code> and
press [Enter])</li>
<li>Type the following line into the command prompt window exactly as
shown: <code>setx HOME "%USERPROFILE%"</code>
</li>
<li>Press [Enter], you should see
<code>SUCCESS: Specified value was saved.</code>
</li>
<li>Quit command prompt by typing <code>exit</code> then pressing
[Enter]</li>
</ul></li>
</ul></li>
<li>An <strong>alternative option</strong> is to install PuTTY by going
to the <a href="https://www.chiark.greenend.org.uk/~sgtatham/putty/latest.html" class="external-link">the
installation page</a>. For most newer computers, click on
putty-64bit-X.XX-installer.msi to download the 64-bit version. If you
have an older laptop, you may need to get the 32-bit version
putty-X.XX-installer.msi. If you aren’t sure whether you need the 64 or
32 bit version, you can check your laptop version by following <a href="https://support.microsoft.com/en-us/help/15056/windows-32-64-bit-faq" class="external-link">the
instructions here</a>. Once the installer is downloaded, double click on
it, and PuTTY should install.</li>
<li>
<strong>Another alternative option</strong> is to use the Ubuntu
Subsystem for Windows. This option is only available for Windows 10 -
detailed <a href="https://docs.microsoft.com/en-us/windows/wsl/install-win10" class="external-link">instructions
are available here</a>.</li>
</ul><div class="section level2">
<h2 id="mac-os-x">Mac OS X<a class="anchor" aria-label="anchor" href="#mac-os-x"></a></h2>
<ul><li>Install LibreOffice by going to <a href="https://www.libreoffice.org/download/libreoffice-fresh/" class="external-link">the
installation page</a>. The version for Mac should automatically be
selected. Click Download Version X.X.X (whichever is the most recent
version). You will go to a page that asks about a donation, but you
don’t need to make one. Your download should begin automatically.</li>
<li>Once the installer is downloaded, double click on it and LibreOffice
should install.</li>
</ul></div>
<div class="section level2">
<h2 id="linux">Linux<a class="anchor" aria-label="anchor" href="#linux"></a></h2>
<ul><li>Install LibreOffice by going to <a href="https://www.libreoffice.org/download/libreoffice-fresh/" class="external-link">the
installation page</a>. The version for Linux should automatically be
selected. Click Download Version X.X.X (whichever is the most recent
version). You will go to a page that asks about a donation, but you
don’t need to make one. Your download should begin automatically.</li>
<li>Once the installer is downloaded, double click on it and LibreOffice
should install.</li>
</ul></div>
</div>
</div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="option-a-recommended-using-the-lessons-with-amazon-web-services-aws">Option A (<strong>Recommended</strong>): Using the lessons with
Amazon Web Services (AWS)<a class="anchor" aria-label="anchor" href="#option-a-recommended-using-the-lessons-with-amazon-web-services-aws"></a></h2>
<p>If you are signed up to take a Genomics Data Carpentry workshop, you
do <em>not</em> need to worry about setting up an AMI instance. The
Carpentries staff will create an instance for you and this will be
provided to you at no cost. This is true for both self-organized and
centrally-organized workshops. Your Instructor will provide instructions
for connecting to the AMI instance at the workshop.</p>
<p>If you would like to work through these lessons independently,
outside of a workshop, you will need to start your own AMI instance.
Follow these <a href="https://datacarpentry.org/genomics-workshop/AMI-setup/" class="external-link">instructions
on creating an Amazon instance</a>. Use the AMI
<code>ami-04b3bc83255f918b0</code> (Data Carpentry Genomics with R 4.0)
listed on the Community AMIs page. Please note that you must set your
location as <code>N. Virginia</code> in order to access this community
AMI. You can change your location in the upper right corner of the main
AWS menu bar. The cost of using this AMI for a few days, with the
t2.medium instance type is very low (about USD $1.50 per user, per day).
Data Carpentry has <em>no</em> control over AWS pricing structure and
provides this cost estimate with no guarantees. Please read AWS
documentation on pricing for up-to-date information.</p>
<p>If you’re an Instructor or Maintainer or want to contribute to these
lessons, please get in touch with us <a href="../mailto:team@carpentries.org">team@carpentries.org</a> and we will
start instances for you.</p>
</div>
<div class="section level2">
<h2 id="option-b-using-the-lessons-on-your-local-machine">Option B: Using the lessons on your local machine<a class="anchor" aria-label="anchor" href="#option-b-using-the-lessons-on-your-local-machine"></a></h2>
<p>While not recommended, it is possible to work through the lessons on
your local machine (i.e. without using AWS). To do this, you will need
to install all of the software used in the workshop and obtain a copy of
the dataset. Instructions for doing this are below.</p>
<div class="section level3">
<h3 id="data">Data<a class="anchor" aria-label="anchor" href="#data"></a></h3>
<p>The data used in this workshop is available on FigShare. Because this
workshop works with real data, be aware that file sizes for the data are
large. Please read the FigShare page linked below for information about
the data and access to the data files.</p>
<p><a href="https://figshare.com/articles/Data_Carpentry_Genomics_beta_2_0/7726454" class="external-link">FigShare
Data Carpentry Genomics Beta 2.0</a></p>
<p>More information about these data will be presented in the <a href="https://www.datacarpentry.org/organization-genomics/data/" class="external-link">first
lesson of the workshop</a>.</p>
</div>
<div class="section level3">
<h3 id="software">Software<a class="anchor" aria-label="anchor" href="#software"></a></h3>
<table class="table"><colgroup><col width="7%"><col width="6%"><col width="5%"><col width="18%"><col width="62%"></colgroup><thead><tr class="header"><th>Software</th>
<th>Version</th>
<th>Manual</th>
<th>Available for</th>
<th>Description</th>
</tr></thead><tbody><tr class="odd"><td><a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/" class="external-link">FastQC</a></td>
<td>0.11.7</td>
<td><a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/" class="external-link">Link</a></td>
<td>Linux, MacOS, Windows</td>
<td>Quality control tool for high throughput sequence data.</td>
</tr><tr class="even"><td><a href="https://www.usadellab.org/cms/?page=trimmomatic" class="external-link">Trimmomatic</a></td>
<td>0.38</td>
<td><a href="https://www.usadellab.org/cms/uploads/supplementary/Trimmomatic/TrimmomaticManual_V0.32.pdf" class="external-link">Link</a></td>
<td>Linux, MacOS, Windows</td>
<td>A flexible read trimming tool for Illumina NGS data.</td>
</tr><tr class="odd"><td><a href="https://bio-bwa.sourceforge.net/" class="external-link">BWA</a></td>
<td>0.7.17</td>
<td><a href="https://bio-bwa.sourceforge.net/bwa.shtml" class="external-link">Link</a></td>
<td>Linux, MacOS</td>
<td>Mapping DNA sequences against reference genome.</td>
</tr><tr class="even"><td><a href="https://samtools.sourceforge.net/" class="external-link">SAMtools</a></td>
<td>1.9</td>
<td><a href="https://www.htslib.org/doc/samtools.html" class="external-link">Link</a></td>
<td>Linux, MacOS</td>
<td>Utilities for manipulating alignments in the SAM format.</td>
</tr><tr class="odd"><td><a href="https://samtools.github.io/bcftools/" class="external-link">BCFtools</a></td>
<td>1.8</td>
<td><a href="https://samtools.github.io/bcftools/bcftools.html" class="external-link">Link</a></td>
<td>Linux, MacOS</td>
<td>Utilities for variant calling and manipulating VCFs and BCFs.</td>
</tr><tr class="even"><td><a href="https://software.broadinstitute.org/software/igv/home" class="external-link">IGV</a></td>
<td><a href="https://software.broadinstitute.org/software/igv/download" class="external-link">Link</a></td>
<td><a href="https://software.broadinstitute.org/software/igv/UserGuide" class="external-link">Link</a></td>
<td>Linux, MacOS, Windows</td>
<td>Visualization and interactive exploration of large genomics
datasets.</td>
</tr></tbody></table></div>
<div class="section level3">
<h3 id="quickstart-software-installation-instructions">QuickStart Software Installation Instructions<a class="anchor" aria-label="anchor" href="#quickstart-software-installation-instructions"></a></h3>
<p>These are the QuickStart installation instructions. They assume
familiarity with the command line and with installation in general. As
there are different operating systems and many different versions of
operating systems and environments, these may not work on your computer.
If an installation doesn’t work for you, please refer to the user guide
for the tool, listed in the table above.</p>
<p>We have installed software using <a href="https://docs.conda.io/en/latest/miniconda.html" class="external-link">miniconda</a>.
Miniconda is a package manager that simplifies the installation process.
Please first install miniconda3 (installation instructions below), and
then proceed to the installation of individual tools.</p>
</div>
<div class="section level3">
<h3 id="miniconda3">Miniconda3<a class="anchor" aria-label="anchor" href="#miniconda3"></a></h3>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2">
  MacOS
  </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<p>To install miniconda3, type:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> curl <span class="at">-O</span> https://repo.anaconda.com/miniconda/Miniconda3-latest-MacOSX-x86_64.sh</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> bash Miniconda3-latest-MacOSX-x86_64.sh</span></code></pre>
</div>
<p>Then, follow the instructions that you are prompted with on the
screen to install Miniconda3.</p>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="fastqc">FastQC<a class="anchor" aria-label="anchor" href="#fastqc"></a></h3>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3">
  MacOS
  </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" data-bs-parent="#accordionSolution3" aria-labelledby="headingSolution3">
<div class="accordion-body">
<p>To install FastQC, type:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> conda install <span class="at">-c</span> bioconda fastqc=0.11.7=5</span></code></pre>
</div>
<div class="section level2">
<h2 id="fastqc-source-code-installation">FastQC Source Code Installation<a class="anchor" aria-label="anchor" href="#fastqc-source-code-installation"></a></h2>
<p>If you prefer to install from source, follow the directions
below:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> cd ~/src</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> curl <span class="at">-O</span> http://www.bioinformatics.babraham.ac.uk/projects/fastqc/fastqc_v0.11.7.zip</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> unzip fastqc_v0.11.7.zip</span></code></pre>
</div>
<p>Link the fastqc executable to the ~/bin folder that you have already
added to the path.</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> ln <span class="at">-sf</span> ~/src/FastQC/fastqc ~/bin/fastqc</span></code></pre>
</div>
<p>Due to what seems a packaging error the executable flag on the fastqc
program is not set. We need to set it ourselves.</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> chmod +x ~/bin/fastqc</span></code></pre>
</div>
<p><strong>Test your installation by running:</strong></p>
</div>
</div>
</div>
</div>
</div>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> fastqc <span class="at">-h</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="trimmomatic">Trimmomatic<a class="anchor" aria-label="anchor" href="#trimmomatic"></a></h3>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4">
  MacOS
  </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" data-bs-parent="#accordionSolution4" aria-labelledby="headingSolution4">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> install <span class="at">-c</span> bioconda trimmomatic=0.38=0</span></code></pre>
</div>
<div class="section level2">
<h2 id="trimmomatic-source-code-installation">Trimmomatic Source Code Installation<a class="anchor" aria-label="anchor" href="#trimmomatic-source-code-installation"></a></h2>
<p>If you prefer to install from source, follow the directions
below:</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> cd ~/src</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> curl <span class="at">-O</span> http://www.usadellab.org/cms/uploads/supplementary/Trimmomatic/Trimmomatic-0.38.zip</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> unzip Trimmomatic-0.38.zip</span></code></pre>
</div>
<p>The program can be invoked via:</p>
<pre><code>$ java -jar ~/src/Trimmomatic-0.38/trimmomatic-0.38.jar</code></pre>
<p>The ~/src/Trimmomatic-0.38/adapters/ directory contains Illumina
specific adapter sequences.</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> ls ~/src/Trimmomatic-0.38/adapters/</span></code></pre>
</div>
<p><strong>Test your installation by running:</strong> (assuming things
are installed in ~/src)</p>
</div>
</div>
</div>
</div>
</div>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> java <span class="at">-jar</span> ~/src/Trimmomatic-0.38/trimmomatic-0.38.jar</span></code></pre>
</div>
<div id="accordionSolution5" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution5" aria-expanded="false" aria-controls="collapseSolution5">
  <h4 class="accordion-header" id="headingSolution5">
  Simplify the Invocation, or to Test your installation if you installed with miniconda3:
  </h4>
</button>
<div id="collapseSolution5" class="accordion-collapse collapse" data-bs-parent="#accordionSolution5" aria-labelledby="headingSolution5">
<div class="accordion-body">
<p>To simplify the invocation you could also create a script in the
~/bin folder:</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> echo <span class="st">'#!/bin/bash'</span> <span class="op">&gt;</span> ~/bin/trimmomatic</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> echo <span class="st">'java -jar ~/src/Trimmomatic-0.36/trimmomatic-0.36.jar $@'</span> <span class="op">&gt;&gt;</span> ~/bin/trimmomatic</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> chmod +x ~/bin/trimmomatic</span></code></pre>
</div>
<p>Test your script by running:</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> trimmomatic</span></code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="bwa">BWA<a class="anchor" aria-label="anchor" href="#bwa"></a></h3>
<div id="accordionSolution6" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution6" aria-expanded="false" aria-controls="collapseSolution6">
  <h4 class="accordion-header" id="headingSolution6">
  MacOS
  </h4>
</button>
<div id="collapseSolution6" class="accordion-collapse collapse" data-bs-parent="#accordionSolution6" aria-labelledby="headingSolution6">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> install <span class="at">-c</span> bioconda bwa=0.7.17=ha92aebf_3</span></code></pre>
</div>
<div class="section level2">
<h2 id="bwa-source-code-installation">BWA Source Code Installation<a class="anchor" aria-label="anchor" href="#bwa-source-code-installation"></a></h2>
<p>If you prefer to install from source, follow the instructions
below:</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> cd ~/src</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> curl <span class="at">-OL</span> http://sourceforge.net/projects/bio-bwa/files/bwa-0.7.17.tar.bz2</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> tar jxvf bwa-0.7.17.tar.bz2</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> cd bwa-0.7.17</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> make</span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> export PATH=~/src/bwa-0.7.17:<span class="va">$PATH</span></span></code></pre>
</div>
<p><strong>Test your installation by running:</strong></p>
</div>
</div>
</div>
</div>
</div>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> bwa</span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="samtools">SAMtools<a class="anchor" aria-label="anchor" href="#samtools"></a></h3>
<div id="accordionSolution7" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution7" aria-expanded="false" aria-controls="collapseSolution7">
  <h4 class="accordion-header" id="headingSolution7">
  MacOS
  </h4>
</button>
<div id="collapseSolution7" class="accordion-collapse collapse" data-bs-parent="#accordionSolution7" aria-labelledby="headingSolution7">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> conda install <span class="at">-c</span> bioconda samtools=1.9=h8ee4bcc_1</span></code></pre>
</div>
<div class="section level2">
<h2 id="samtools-versions">SAMtools Versions<a class="anchor" aria-label="anchor" href="#samtools-versions"></a></h2>
<p>SAMtools has changed the command line invocation (for the better).
But this means that most of the tutorials on the web indicate an older
and obsolete usage.</p>
<p>Using SAMtools version 1.9 is important to work with the commands we
present in these lessons.</p>
</div>
<div class="section level2">
<h2 id="samtools-source-code-installation">SAMtools Source Code Installation<a class="anchor" aria-label="anchor" href="#samtools-source-code-installation"></a></h2>
<p>If you prefer to install from source, follow the instructions
below:</p>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> cd ~/src</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> curl <span class="at">-OkL</span> https://github.com/samtools/samtools/releases/download/1.9/samtools-1.9.tar.bz2</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> tar jxvf samtools-1.9.tar.bz2</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> cd samtools-1.9</span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> make</span></code></pre>
</div>
<p>Add directory to the path if necessary:</p>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> echo export <span class="kw">`</span><span class="va">PATH</span><span class="op">=</span>~/src/samtools-1.9:<span class="va">$PATH</span><span class="kw">`</span> <span class="op">&gt;&gt;</span> ~/.bashrc</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> source ~/.bashrc</span></code></pre>
</div>
<p><strong>Test your installation by running:</strong></p>
</div>
</div>
</div>
</div>
</div>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> samtools</span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="bcftools">BCFtools<a class="anchor" aria-label="anchor" href="#bcftools"></a></h3>
<div id="accordionSolution8" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution8" aria-expanded="false" aria-controls="collapseSolution8">
  <h4 class="accordion-header" id="headingSolution8">
  MacOS
  </h4>
</button>
<div id="collapseSolution8" class="accordion-collapse collapse" data-bs-parent="#accordionSolution8" aria-labelledby="headingSolution8">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> conda install <span class="at">-c</span> bioconda bcftools=1.8=h4da6232_3 </span></code></pre>
</div>
<div class="section level2">
<h2 id="bcf-tools-source-code-installation">BCF tools Source Code Installation<a class="anchor" aria-label="anchor" href="#bcf-tools-source-code-installation"></a></h2>
<p>If you prefer to install from source, follow the instructions
below:</p>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> cd ~/src</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> curl <span class="at">-OkL</span> https://github.com/samtools/bcftools/releases/download/1.8/bcftools-1.8.tar.bz2</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> tar jxvf bcftools-1.8.tar.bz2</span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> cd bcftools-1.8</span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> make</span></code></pre>
</div>
<p>Add directory to the path if necessary:</p>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> echo export <span class="kw">`</span><span class="va">PATH</span><span class="op">=</span>~/src/bcftools-1.8:<span class="va">$PATH</span><span class="kw">`</span> <span class="op">&gt;&gt;</span> ~/.bashrc</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> source ~/.bashrc</span></code></pre>
</div>
<p><strong>Test your installation by running:</strong></p>
</div>
</div>
</div>
</div>
</div>
<div class="codewrapper sourceCode" id="cb24">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> bcftools</span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="igv">IGV<a class="anchor" aria-label="anchor" href="#igv"></a></h3>
<ul><li><a href="https://software.broadinstitute.org/software/igv/download" class="external-link">Download
the IGV installation files</a></li>
<li>Install and run IGV using the <a href="https://software.broadinstitute.org/software/igv/download" class="external-link">instructions
for your operating system</a>.</li>
</ul></div>
</div>
</div>


      </section></main><nav class="bottom-pagination mx-md-4" aria-label="Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link float-end" href="../instructor/00-introduction.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link float-end" href="../instructor/00-introduction.html" rel="next">Next: Introducing R and... <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
    </nav></div>
<!-- END  : inst/pkgdown/templates/content-syllabus.html -->

      </div><!--/div.row-->
      		<footer class="row footer mx-md-3"><hr><div class="col-md-6">
				<p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>
        
        <a href="https://github.com/datacarpentry/genomics-r-intro/edit/main/index.md" class="external-link">Edit on GitHub</a>
        
	
        | <a href="https://github.com/datacarpentry/genomics-r-intro/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a> 
        | <a href="https://github.com/datacarpentry/genomics-r-intro/" class="external-link">Source</a></p>
				<p><a href="https://github.com/datacarpentry/genomics-r-intro/blob/main/CITATION" class="external-link">Cite</a> | <a href="mailto:team@carpentries.org">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">
        
        <p>Materials licensed under <a href="../LICENSE.html">CC-BY 4.0</a> by the authors</p>
        
        <p><a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">Template licensed under CC-BY 4.0</a> by <a href="https://carpentries.org" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.12.4" class="external-link">sandpaper (0.12.4)</a>,
        <a href="https://github.com/carpentries/pegboard" class="external-link">pegboard (0.5.3)</a>,
      and <a href="https://github.com/carpentries/varnish/tree/0.2.17" class="external-link">varnish (0.2.17)</a>.</p>
			</div>
		</footer></div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
			<i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back to top"></i><br><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "TrainingMaterial",
  "@id": "https://datacarpentry.github.io/genomics-r-intro/instructor/index.html",
  "dct:conformsTo": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, The Carpentries",
  "name": "Intro to R and RStudio for Genomics",
  "creativeWorkStatus": "active",
  "url": "https://datacarpentry.github.io/genomics-r-intro/instructor/index.html",
  "identifier": "https://datacarpentry.github.io/genomics-r-intro/instructor/index.html",
  "dateCreated": "2018-03-12",
  "dateModified": "2023-06-27",
  "datePublished": "2023-06-27"
}

  </script><script>
		feather.replace();
	</script><!-- Matomo
    2022-11-07: we have gotten a notification that we have an overage for our
    tracking and I'm pretty sure this has to do with Workbench usage.
    Considering that I am not _currently_ using this tracking because I do not
    yet know how to access the data, I am turning this off for now.
  <script>
    var _paq = window._paq = window._paq || [];
    /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
    _paq.push(["setDocumentTitle", document.domain + "/" + document.title]);
    _paq.push(["setDomains", ["*.preview.carpentries.org","*.datacarpentry.github.io","*.datacarpentry.org","*.librarycarpentry.github.io","*.librarycarpentry.org","*.swcarpentry.github.io", "*.carpentries.github.io"]]);
    _paq.push(["setDoNotTrack", true]);
    _paq.push(["disableCookies"]);
    _paq.push(['trackPageView']);
    _paq.push(['enableLinkTracking']);
    (function() {
          var u="https://carpentries.matomo.cloud/";
          _paq.push(['setTrackerUrl', u+'matomo.php']);
          _paq.push(['setSiteId', '1']);
          var d=document, g=d.createElement('script'), s=d.getElementsByTagName('script')[0];
          g.async=true; g.src='https://cdn.matomo.cloud/carpentries.matomo.cloud/matomo.js'; s.parentNode.insertBefore(g,s);
        })();
  </script>
  End Matomo Code --></body></html><!-- END:   inst/pkgdown/templates/layout.html-->

